// pioneer-platform@1.0.0/platform/src/store/coinInfo/coins.js
export default [
  {
    displayName: 'Bitcoin',
    symbol: 'BTC',
    network: 'BTC',
    keepkeyName: 'Bitcoin',
    color: '#FF9800',
    secondaryColor: '#F57C00',
    precision: 8,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/btc.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/btc.png',
    slip44: 2147483648,
    curveName: 'secp256k1',
    gasLimit: '',
    watchtowerType: 'xpub',
    scriptTypes: ['p2sh-p2wpkh', 'p2pkh', 'p2wpkh']
  },
  // Missing CoinCap Support
  // {
  //   icon: '/static/skelefox.svg',
  //   slip44: 2147483649,
  //   curveName: 'secp256k1',
  //   gasLimit: '',
  //   symbol: 'TEST',
  //   displayName: 'Testnet',
  //   color: '#03A9F4',
  //   precision: 8,
  //   watchtowerType: 'xpub',
  //   isToken: false
  // },
  {
    displayName: 'Bitcoin Cash',
    symbol: 'BCH',
    network: 'BCH',
    keepkeyName: 'BitcoinCash',
    color: '#8BC34A',
    secondaryColor: '#7CB342',
    precision: 8,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/bch.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/bch.png',
    slip44: 2147483793,
    curveName: 'secp256k1',
    gasLimit: '',
    watchtowerType: 'xpub',
    scriptTypes: ['p2pkh']
  },
  {
    displayName: 'Litecoin',
    symbol: 'LTC',
    network: 'LTC',
    keepkeyName: 'Litecoin',
    color: '#E0E0E0',
    secondaryColor: '#BDBDBD',
    precision: 8,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/ltc.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/ltc.png',
    slip44: 2147483650,
    curveName: 'secp256k1',
    gasLimit: '',
    watchtowerType: 'xpub',
    scriptTypes: ['p2pkh']
  },
  {
    displayName: 'Dogecoin',
    symbol: 'DOGE',
    network: 'DOGE',
    keepkeyName: 'Dogecoin',
    color: '#FFC107',
    secondaryColor: '#FFB300',
    precision: 8,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/doge.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/doge.png',
    slip44: 2147483651,
    curveName: 'secp256k1',
    gasLimit: '',
    watchtowerType: 'xpub',
    scriptTypes: ['p2pkh']
  },
  {
    displayName: 'Dash',
    symbol: 'DASH',
    network: 'DASH',
    keepkeyName: 'Dash',
    color: '#1976d2',
    secondaryColor: '#1565C0',
    precision: 8,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/dash.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/dash.png',
    slip44: 2147483653,
    curveName: 'secp256k1',
    gasLimit: '',
    watchtowerType: 'xpub',
    scriptTypes: ['p2pkh']
  },
  {
    displayName: 'DigiByte',
    symbol: 'DGB',
    network: 'DGB',
    keepkeyName: 'DigiByte',
    color: '#1E88E5',
    secondaryColor: '#1976D2',
    precision: 8,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/dgb.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/dgb.png',
    slip44: 2147483668,
    curveName: 'secp256k1',
    gasLimit: '',
    watchtowerType: 'xpub',
    scriptTypes: ['p2pkh']
  },
  {
    displayName: 'Ethereum',
    symbol: 'ETH',
    network: 'ETH',
    keepkeyName: 'Ethereum',
    color: '#5C6BC0',
    secondaryColor: '#3F51B5',
    precision: 18,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/eth.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/eth.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: ''
  },
  {
    displayName: 'Cosmos',
    symbol: 'ATOM',
    network: 'ATOM',
    keepkeyName: 'Atom',
    color: '#303f9f',
    secondaryColor: '#283593',
    precision: 6,
    isToken: false,
    icon: 'https://static.coincap.io/assets/icons/svg/atom.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/atom.png',
    slip44: 0x80000000 + 118,
    curveName: 'secp256k1',
    gasLimit: '',
    watchtowerType: 'address',
    scriptTypes: ['cosmos']
  },
  {
    displayName: 'FOX',
    symbol: 'FOX',
    network: 'ETH',
    color: '#386FF9',
    secondaryColor: '#2F5DD2',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/fox.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/fox.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xc770EEfAd204B5180dF6a14Ee197D99d808ee52d' // required for eth tokens
  },
  // Missing Watchtower Support
  // {
  //   icon: 'https://static.coincap.io/assets/icons/svg/etc.svg',
  //   slip44: 2147483709,
  //   curveName: 'secp256k1',
  //   gasLimit: '',
  //   symbol: 'ETC',
  //   network: 'ETC',
  //   displayName: 'Ethereum Classic',
  //   color: '#2E7D32',
  //   watchtowerType: 'xpub',
  //   isToken: false
  // },
  // Missing Watchtower Support
  // {
  //   icon: 'https://static.coincap.io/assets/icons/svg/btg.svg',
  //   slip44: 2147483804,
  //   curveName: 'secp256k1',
  //   gasLimit: '',
  //   symbol: 'BTG',
  //   network: 'BTG',
  //   displayName: 'BitcoinGold',
  //   color: '#F9A825',
  //   watchtowerType: 'xpub',
  //   isToken: false
  // },
  // Missing Watchtower Support
  // {
  //   icon: 'https://static.coincap.io/assets/icons/svg/zec.svg',
  //   slip44: 2147483781,
  //   curveName: 'secp256k1',
  //   gasLimit: '',
  //   symbol: 'ZEC',
  //   network: 'ZEC',
  //   displayName: 'ZCash',
  //   color: '#FBC02D',
  //   watchtowerType: 'xpub',
  //   isToken: false
  // },
  // Missing CoinCap Support
  // {
  //   icon: '/static/skelefox.svg',
  //   slip44: 2147483781,
  //   curveName: 'secp256k1',
  //   gasLimit: '',
  //   symbol: 'TAZ',
  //   network: 'TAZ',
  //   displayName: 'Zcash Testnet',
  //   color: '#4527A0',
  //   watchtowerType: 'xpub',
  //   isToken: false
  // },
  {
    displayName: 'Golem',
    symbol: 'GNT',
    network: 'ETH',
    color: '#2B8ADF',
    secondaryColor: '#547B8D',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/gnt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/gnt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xa74476443119A942dE498590Fe1f2454d7D4aC0d' // required for eth tokens
  },
  {
    displayName: 'Gnosis',
    symbol: 'GNO',
    network: 'ETH',
    color: '#4A768B',
    secondaryColor: '#1561AD',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/gno.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/gno.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x6810e776880c02933d47db1b9fc05908e5386b96' // required for eth tokens
  },
  {
    displayName: 'Singular DTV',
    symbol: 'SNGLS',
    network: 'ETH',
    color: '#B71C1C',
    secondaryColor: '#7A1515',
    precision: 0,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/sngls.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/sngls.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xaec2e87e0a235266d9c5adc9deb4b2e29b54d009' // required for eth tokens
  },
  {
    displayName: 'Iconomi',
    symbol: 'ICN',
    network: 'ETH',
    color: '#456F9F',
    secondaryColor: '#3C6477',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/icn.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/icn.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x888666CA69E0f178DED6D75b5726Cee99A87D698' // required for eth tokens
  },
  {
    displayName: 'Digix DAO',
    symbol: 'DGD',
    network: 'ETH',
    color: '#E0AA55',
    secondaryColor: '#BB9251',
    precision: 9,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/dgd.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/dgd.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xe0b7927c4af23765cb51314a0e0521a9645f0e2a' // required for eth tokens
  },
  {
    displayName: 'Melonport',
    symbol: 'MLN',
    network: 'ETH',
    color: '#EEEEEE',
    secondaryColor: '#E0E0E0',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/mln.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/mln.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xec67005c4E498Ec7f55E092bd1d35cbC47C91892' // required for eth tokens
  },
  {
    displayName: 'Augur',
    symbol: 'REP',
    network: 'ETH',
    color: '#AD1457',
    secondaryColor: '#880E4F',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/rep.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/rep.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x1985365e9f78359a9B6AD760e32412f4a445E862' // required for eth tokens
  },
  {
    displayName: 'Swarm City',
    symbol: 'SWT',
    network: 'ETH',
    color: '#F8EEA8',
    secondaryColor: '#EFD96F',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/swt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/swt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xb9e7f8568e08d5659f5d29c4997173d84cdf2607' // required for eth tokens
  },
  {
    displayName: 'Wings',
    symbol: 'WINGS',
    network: 'ETH',
    color: '#B3E5FC',
    secondaryColor: '#81D4FA',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/wings.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/wings.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x667088b212ce3d06a1b553a7221E1fD19000d9aF' // required for eth tokens
  },
  {
    displayName: 'WeTrust',
    symbol: 'TRST',
    network: 'ETH',
    color: '#80CBC4',
    secondaryColor: '#4DB6AC',
    precision: 6,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/trst.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/trst.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xcb94be6f13a1182e4a4b6140cb7bf2025d28e41b' // required for eth tokens
  },
  {
    displayName: 'iExec RLC',
    symbol: 'RLC',
    network: 'ETH',
    color: '#FDD835',
    secondaryColor: '#FFEB3B',
    precision: 9,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/rlc.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/rlc.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x607F4C5BB672230e8672085532f7e901544a7375' // required for eth tokens
  },
  // Broken for unknown reasons; Disabled on 1/28/19
  // {
  //   displayName: 'Guppy',
  //   symbol: 'GUP',
  //  network: 'ETH',
  //   color: '#80CBC4',
  //   secondaryColor: '#26A69A',
  //   precision: 3,
  //   isToken: true,
  //   icon: 'https://static.coincap.io/assets/icons/svg/gup.svg',
  //   slip44: 2147483708,
  //   curveName: 'secp256k1',
  //   watchtowerType: 'xpub',
  //   gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
  //   contractAddress: '0xf7b098298f7c69fc14610bf71d5e02c60792894c' // required for eth tokens
  // },
  {
    displayName: 'Aragon',
    symbol: 'ANT',
    network: 'ETH',
    color: '#80DEEA',
    secondaryColor: '#26C6DA',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/ant.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/ant.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x960b236A07cf122663c4303350609A66A7B288C0' // required for eth tokens
  },
  {
    displayName: 'Basic Attention Token',
    symbol: 'BAT',
    network: 'ETH',
    color: '#FF5722',
    secondaryColor: '#E64A19',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/bat.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/bat.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x0d8775f648430679a709e98d2b0cb6250d2887ef' // required for eth tokens
  },
  {
    displayName: 'Bancor',
    symbol: 'BNT',
    network: 'ETH',
    color: '#456F9F',
    secondaryColor: '#3C6477',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/bnt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/bnt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x1f573d6fb3f13d689ff844b4ce37794d79a7ff1c' // required for eth tokens
  },
  {
    displayName: 'Status',
    symbol: 'SNT',
    network: 'ETH',
    color: '#3949AB',
    secondaryColor: '#303F9F',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/snt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/snt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x744d70fdbe2ba4cf95131626614a1763df805b9e' // required for eth tokens
  },
  {
    displayName: 'Numeraire',
    symbol: 'NMR',
    network: 'ETH',
    color: '#616161',
    secondaryColor: '#424242',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/nmr.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/nmr.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x1776e1f26f98b1a5df9cd347953a26dd3cb46671' // required for eth tokens
  },
  {
    displayName: 'Edgeless',
    symbol: 'EDG',
    network: 'ETH',
    color: '#7E57C2',
    secondaryColor: '#673AB7',
    precision: 0,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/edg.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/edg.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x08711d3b02c8758f2fb3ab4e80228418a7f8e39c' // required for eth tokens
  },
  {
    displayName: 'Civic',
    symbol: 'CVC',
    network: 'ETH',
    color: '#43A047',
    secondaryColor: '#388E3C',
    precision: 8,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/cvc.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/cvc.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x41e5560054824ea6b0732e656e3ad64e20e94e45' // required for eth tokens
  },
  {
    displayName: 'Metal',
    symbol: 'MTL',
    network: 'ETH',
    color: '#616161',
    secondaryColor: '#424242',
    precision: 8,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/mtl.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/mtl.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xF433089366899D83a9f26A773D59ec7eCF30355e' // required for eth tokens
  },
  {
    displayName: 'TenX',
    symbol: 'PAY',
    network: 'ETH',
    color: '#616161',
    secondaryColor: '#424242',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/pay.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/pay.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xB97048628DB6B661D4C2aA833e95Dbe1A905B280' // required for eth tokens
  },
  {
    displayName: 'FunFair',
    symbol: 'FUN',
    network: 'ETH',
    color: '#EC407A',
    secondaryColor: '#D81B60',
    precision: 8,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/fun.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/fun.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x419d0d8bdd9af5e606ae2232ed285aff190e711b' // required for eth tokens
  },
  // Went Mainnet
  // {
  //   icon: 'https://static.coincap.io/assets/icons/svg/qtum.svg',
  //   slip44: 2147483708,
  //   curveName: 'secp256k1',
  //   gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
  //   symbol: 'QTUM',
  //   network: 'ETH',
  //   displayName: 'Qtum',
  //   color: '#2196F3',
  //   secondaryColor: '#1E88E5',
  //   precision: 18,
  //   isToken: true,
  //   watchtowerType: 'xpub',
  //   contractAddress: '0x9a642d6b3368ddc662CA244bAdf32cDA716005BC' // required for eth tokens
  // },
  {
    displayName: 'District 0x',
    symbol: 'DNT',
    network: 'ETH',
    color: '#01579B',
    secondaryColor: '#0D47A1',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/dnt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/dnt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x0abdace70d3790235af448c88547603b945604ea' // required for eth tokens
  },
  {
    displayName: '0x',
    symbol: 'ZRX',
    network: 'ETH',
    color: '#616161',
    secondaryColor: '#424242',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/zrx.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/zrx.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xe41d2489571d322189246dafa5ebde1f4699f498' // required for eth tokens
  },
  {
    displayName: 'FirstBlood',
    symbol: '1ST',
    network: 'ETH',
    color: '#B71C1C',
    secondaryColor: '#7A1515',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/1st.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/1st.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xaf30d2a7e90d7dc361c8c4585e9bb7d2f6f15bc7' // required for eth tokens
  },
  {
    displayName: 'OmiseGO',
    symbol: 'OMG',
    network: 'ETH',
    color: '#536DFE',
    secondaryColor: '#304FFE',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/omg.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/omg.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xd26114cd6EE289AccF82350c8d8487fedB8A0C07' // required for eth tokens
  },
  {
    displayName: 'Salt',
    symbol: 'SALT',
    network: 'ETH',
    color: '#81D4FA',
    secondaryColor: '#29B6F6',
    precision: 8,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/salt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/salt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0x4156D3342D5c385a87D264F90653733592000581' // required for eth tokens
  },
  {
    displayName: 'Ripio',
    symbol: 'RCN',
    network: 'ETH',
    color: '#3D5AFE',
    secondaryColor: '#304FFE',
    precision: 18,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/rcn.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/rcn.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xf970b8e36e23f7fc3fd752eea86f8be8d83375a6' // required for eth tokens
  },
  {
    displayName: 'Storj',
    symbol: 'STORJ',
    network: 'ETH',
    color: '#2196F3',
    secondaryColor: '#1976D2',
    precision: 8,
    isToken: true,
    icon: 'https://static.coincap.io/assets/icons/svg/storj.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/storj.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    watchtowerType: 'xpub',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    contractAddress: '0xb64ef51c888972c908cfacf59b47c1afbc0ab8ac' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/bnb.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/bnb.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'BNB',
    network: 'ETH',
    displayName: 'Binance Coin (ERC-20)',
    color: '#FFC107',
    secondaryColor: '#FFAB00',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xB8c77482e45F1F44dE1745F52C74426C631bDD52' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/usdt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/usdt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'USDT',
    network: 'ETH',
    displayName: 'Tether',
    color: '#4DB6AC',
    secondaryColor: '#009688',
    precision: 6,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xdac17f958d2ee523a2206206994597c13d831ec7' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/poly.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/poly.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'POLY',
    network: 'ETH',
    displayName: 'Polymath',
    color: '#5C6BC0',
    secondaryColor: '#3949AB',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x9992ec3cf6a55b00978cddf2b27bc6882d88d1ec' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/zil.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/zil.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'ZIL',
    network: 'ETH',
    displayName: 'Zilliqa',
    color: '#00ACC1',
    secondaryColor: '#009688',
    precision: 12,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x05f4a42e251f2d52b8ed15e9fedaacfcef1fad27' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/mana.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/mana.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'MANA',
    network: 'ETH',
    displayName: 'Decentraland',
    color: '#EFEBE9',
    secondaryColor: '#D7CCC8',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x0f5d2fb29fb7d3cfee444a200298f468908cc942' // required for eth tokens
  },
  // Missing CoinCap Support
  // {
  //   icon: '/static/skelefox.svg',
  //   slip44: 2147483708,
  //   curveName: 'secp256k1',
  //   gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
  //   symbol: '0xBTC',
  //   network: 'ETH',
  //   displayName: '0xBitcoin',
  //   color: '#FF6F00',
  //   watchtowerType: 'xpub',
  //   isToken: true
  // },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/gto.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/gto.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'GTO',
    network: 'ETH',
    displayName: 'Gifto',
    color: '#7C4DFF',
    secondaryColor: '#673AB7',
    precision: 5,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xc5bbae50781be1669306b9e001eff57a2957b09d' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/iost.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/iost.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'IOST',
    network: 'ETH',
    displayName: 'IOSToken',
    color: '#616161',
    secondaryColor: '#424242',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xfa1a856cfa3409cfa145fa4e20eb270df3eb21ab' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/elf.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/elf.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'ELF',
    network: 'ETH',
    displayName: 'Aelf',
    color: '#1976D2',
    secondaryColor: '#0D47A1',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xbf2179859fc6d5bee9bf9158632dc51678a4100e' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/tusd.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/tusd.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'TUSD',
    network: 'ETH',
    displayName: 'TrueUSD',
    color: '#5C6BC0',
    secondaryColor: '#3949AB',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x0000000000085d4780B73119b644AE5ecd22b376' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/ae.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/ae.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'AE',
    network: 'ETH',
    displayName: 'Aeternity',
    color: '#F06292',
    secondaryColor: '#D81B60',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x5ca9a71b1d01849c0a95490cc00559717fcf0d1d' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/mkr.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/mkr.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'MKR',
    network: 'ETH',
    displayName: 'Maker',
    color: '#80CBC4',
    secondaryColor: '#26A69A',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x9f8f72aa9304c8b593d555f12ef6589cc3a579a2' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/dai.svg', // FIXME: rename this on CoinCap
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/dai.png', // FIXME: rename this on CoinCap
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'SAI',
    network: 'ETH',
    displayName: 'Sai (Old Dai)',
    color: '#FBC02D',
    secondaryColor: '#FDD835',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359' // required for eth tokens
  },
  // {
  //   icon: 'https://static.coincap.io/assets/icons/svg/dai.svg', // FIXME: use new DAI logo
  //   transparentIcon: 'https://static.coincap.io/assets/icons/transparent/dai.png', // FIXME: use new DAI logo
  //   slip44: 2147483708,
  //   curveName: 'secp256k1',
  //   gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
  //   symbol: 'DAI',
  //   network: 'ETH',
  //   displayName: 'Dai',
  //   color: '#FBC02D',
  //   secondaryColor: '#FDD835',
  //   precision: 18,
  //   isToken: true,
  //   watchtowerType: 'xpub',
  //   contractAddress: '0x6B175474E89094C44Da98b954EedeAC495271d0F' // required for eth tokens
  // },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/spank.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/spank.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'SPANK',
    network: 'ETH',
    displayName: 'SpankChain',
    color: '#F06292',
    secondaryColor: '#E91E63',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x42d6622dece394b54999fbd73d108123806f6a18' // required for eth tokens
  },
  // Went Mainnet
  // {
  //   icon: 'https://static.coincap.io/assets/icons/svg/cmt.svg',
  //   slip44: 2147483708,
  //   curveName: 'secp256k1',
  //   gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
  //   symbol: 'CMT',
  //   network: 'ETH',
  //   displayName: 'CyberMiles',
  //   color: '#CEAF6C',
  //   secondaryColor: '#B38534',
  //   precision: 18,
  //   isToken: true,
  //   watchtowerType: 'xpub',
  //   contractAddress: '0xf85feea2fdd81d51177f6b8f35f0e6734ce45f5f' // required for eth tokens
  // },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/mco.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/mco.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'MCO',
    network: 'ETH',
    displayName: 'Crypto.com',
    color: '#4B99BF',
    secondaryColor: '#1561AD',
    precision: 8,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xb63b606ac810a52cca15e44bb630fd42d8d1d83d' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/ppt.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/ppt.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'PPT',
    network: 'ETH',
    displayName: 'Populous',
    color: '#4F6B98',
    secondaryColor: '#203556',
    precision: 8,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xd4fa1460f537bb9085d22c7bccb5dd450ef28e3a' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/pax.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/pax.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'PAX',
    network: 'ETH',
    displayName: 'Paxos Standard Token',
    color: '#1B5E20',
    secondaryColor: '#F4FF81',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x8e870d67f660d95d5be530380d0ec0bd388289e1' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/link.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/link.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'LINK',
    network: 'ETH',
    displayName: 'Chainlink',
    color: '#1A237E',
    secondaryColor: '#1565C0',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x514910771af9ca656af840dff83e8264ecf986ca' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/knc.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/knc.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'KNC',
    network: 'ETH',
    displayName: 'Kyber Network',
    color: '#00BFA5',
    secondaryColor: '#009688',
    precision: 18,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0xdd974d5c2e2928dea5f71b9825b8b646686bd200' // required for eth tokens
  },
  {
    icon: 'https://static.coincap.io/assets/icons/svg/kcs.svg',
    transparentIcon: 'https://static.coincap.io/assets/icons/transparent/kcs.png',
    slip44: 2147483708,
    curveName: 'secp256k1',
    gasLimit: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB6Eg=',
    symbol: 'KCS',
    network: 'ETH',
    displayName: 'Kucoin Shares',
    color: '#07CBF4',
    secondaryColor: '#0091EA',
    precision: 6,
    isToken: true,
    watchtowerType: 'xpub',
    contractAddress: '0x039b5649a59967e3e936d7471f9c3700100ee1ab' // required for eth tokens
  }
]
