// @pioneer-sdk/sdk-v2@0.0.3/lib/connect.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeWallets = void 0;
const TAG = " | connectWallets | ";
const log = require("@pioneer-platform/loggerdog")();
const types_1 = require("@pioneer-platform/types");
async function initializeWallets() {
    let tag = TAG + " | initializeWallets | ";
    const wallets = [];
    // Importing wallets
    const { keepkeyWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/keepkey')));
    const { evmWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/evm-web3-wallets')));
    const { keplrWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/keplr')));
    const { keystoreWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/keystore')));
    const { metamaskWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/metamask')));
    const { ledgerWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/ledger')));
    const { okxWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/okx')));
    const { trezorWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/trezor')));
    const { walletconnectWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/walletconnect')));
    const { xdefiWallet } = await Promise.resolve().then(() => __importStar(require('@pioneer-platform/xdefi')));
    log.info("walletKeepKey: ", keepkeyWallet);
    // Initialize and push each wallet into the wallets array
    let walletKeepKey = {
        type: types_1.WalletOption.KEEPKEY,
        icon: 'https://pioneers.dev/coins/keepkey.png',
        wallet: keepkeyWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletKeepKey);
    let walletMetaMask = {
        type: types_1.WalletOption.METAMASK,
        icon: 'https://pioneers.dev/coins/metamask.png',
        wallet: metamaskWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletMetaMask);
    let walletEVM = {
        type: 'EVM',
        icon: 'https://pioneers.dev/coins/evm.png',
        wallet: evmWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletEVM);
    let walletKeplr = {
        type: types_1.WalletOption.KEPLR,
        icon: 'https://pioneers.dev/coins/keplr.png',
        wallet: keplrWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletKeplr);
    let walletKeystore = {
        type: types_1.WalletOption.KEYSTORE,
        icon: 'https://pioneers.dev/coins/keystore.png',
        wallet: keystoreWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletKeystore);
    let walletLedger = {
        type: types_1.WalletOption.LEDGER,
        icon: 'https://pioneers.dev/coins/ledger.png',
        wallet: ledgerWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletLedger);
    let walletOKX = {
        type: types_1.WalletOption.OKX,
        icon: 'https://pioneers.dev/coins/okx.png',
        wallet: okxWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletOKX);
    let walletTrezor = {
        type: types_1.WalletOption.TREZOR,
        icon: 'https://pioneers.dev/coins/trezor.png',
        wallet: trezorWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletTrezor);
    let walletWalletConnect = {
        type: types_1.WalletOption.WALLETCONNECT,
        icon: 'https://pioneers.dev/coins/walletconnect.png',
        wallet: walletconnectWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletWalletConnect);
    let walletXDefi = {
        type: types_1.WalletOption.XDEFI,
        icon: 'https://pioneers.dev/coins/xdefi.png',
        wallet: xdefiWallet,
        status: 'offline',
        isConnected: false
    };
    wallets.push(walletXDefi);
    //TODO test each for detection
    return wallets;
}
exports.initializeWallets = initializeWallets;
