// @metamask/snaps-utils@1.0.2/dist/caveats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapCaveatType = void 0;
var SnapCaveatType;
(function (SnapCaveatType) {
    /**
     * Permitted derivation paths, used by `snap_getBip32Entropy`.
     */
    SnapCaveatType["PermittedDerivationPaths"] = "permittedDerivationPaths";
    /**
     * Permitted coin types, used by `snap_getBip44Entropy`.
     */
    SnapCaveatType["PermittedCoinTypes"] = "permittedCoinTypes";
    /**
     * Permission to use the Snap keyring API.
     */
    SnapCaveatType["SnapKeyring"] = "snapKeyring";
    /**
     * Caveat specifying a snap cronjob.
     */
    SnapCaveatType["SnapCronjob"] = "snapCronjob";
    /**
     * Caveat specifying access to the transaction origin, used by `endowment:transaction-insight`.
     */
    SnapCaveatType["TransactionOrigin"] = "transactionOrigin";
    /**
     * The origins that a Snap can receive JSON-RPC messages from.
     */
    SnapCaveatType["RpcOrigin"] = "rpcOrigin";
    /**
     * Caveat specifying the snap IDs that can be interacted with.
     */
    SnapCaveatType["SnapIds"] = "snapIds";
})(SnapCaveatType = exports.SnapCaveatType || (exports.SnapCaveatType = {}));
