// @trezor/utxo-lib@1.0.9/lib/crypto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hmacSHA512 = exports.hash256 = exports.hash160blake256 = exports.hash160 = exports.blake256 = exports.sha256 = exports.sha1 = exports.ripemd160 = void 0;
const blakeHash = require("blake-hash");
const createHash = require("create-hash");
const createHmac = require("create-hmac");
function ripemd160(buffer) {
    try {
        return createHash('rmd160').update(buffer).digest();
    }
    catch (err) {
        return createHash('ripemd160').update(buffer).digest();
    }
}
exports.ripemd160 = ripemd160;
function sha1(buffer) {
    return createHash('sha1').update(buffer).digest();
}
exports.sha1 = sha1;
function sha256(buffer) {
    return createHash('sha256').update(buffer).digest();
}
exports.sha256 = sha256;
function blake256(buffer) {
    return blakeHash('blake256').update(buffer).digest();
}
exports.blake256 = blake256;
function hash160(buffer) {
    return ripemd160(sha256(buffer));
}
exports.hash160 = hash160;
function hash160blake256(buffer) {
    return ripemd160(blake256(buffer));
}
exports.hash160blake256 = hash160blake256;
function hash256(buffer) {
    return sha256(sha256(buffer));
}
exports.hash256 = hash256;
function hmacSHA512(key, data) {
    return createHmac('sha512', key).update(data).digest();
}
exports.hmacSHA512 = hmacSHA512;
