// @trezor/utxo-lib@1.0.9/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.anyOf = exports.Nil = exports.BufferN = exports.Function = exports.UInt32 = exports.UInt16 = exports.UInt8 = exports.tuple = exports.maybe = exports.Hex = exports.Buffer = exports.String = exports.Boolean = exports.Array = exports.Number = exports.Hash256bit = exports.Hash160bit = exports.Buffer256bit = exports.Satoshi = void 0;
const typeforce = require("typeforce");
const SATOSHI_MAX = 21 * 1e14;
function Satoshi(value) {
    return typeforce.UInt53(value) && value <= SATOSHI_MAX;
}
exports.Satoshi = Satoshi;
exports.Buffer256bit = typeforce.BufferN(32);
exports.Hash160bit = typeforce.BufferN(20);
exports.Hash256bit = typeforce.BufferN(32);
exports.Number = typeforce.Number, exports.Array = typeforce.Array, exports.Boolean = typeforce.Boolean, exports.String = typeforce.String, exports.Buffer = typeforce.Buffer, exports.Hex = typeforce.Hex, exports.maybe = typeforce.maybe, exports.tuple = typeforce.tuple, exports.UInt8 = typeforce.UInt8, exports.UInt16 = typeforce.UInt16, exports.UInt32 = typeforce.UInt32, exports.Function = typeforce.Function, exports.BufferN = typeforce.BufferN, exports.Nil = typeforce.Nil, exports.anyOf = typeforce.anyOf;
