// osmojs@0.37.0/module/codegen/osmosis/streamswap/v1/genesis.js
import { Sale, UserPosition } from "./state";
import { Params } from "./params";
import * as _m0 from "protobufjs/minimal";
import { Long } from "../../../helpers";
/** GenesisState defines the streamswap module's genesis state. */

function createBaseGenesisState() {
  return {
    sales: [],
    userPositions: [],
    nextSaleId: Long.UZERO,
    params: undefined
  };
}

export const GenesisState = {
  encode(message, writer = _m0.Writer.create()) {
    for (const v of message.sales) {
      Sale.encode(v, writer.uint32(10).fork()).ldelim();
    }

    for (const v of message.userPositions) {
      UserPositionKV.encode(v, writer.uint32(18).fork()).ldelim();
    }

    if (!message.nextSaleId.isZero()) {
      writer.uint32(24).uint64(message.nextSaleId);
    }

    if (message.params !== undefined) {
      Params.encode(message.params, writer.uint32(34).fork()).ldelim();
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseGenesisState();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.sales.push(Sale.decode(reader, reader.uint32()));
          break;

        case 2:
          message.userPositions.push(UserPositionKV.decode(reader, reader.uint32()));
          break;

        case 3:
          message.nextSaleId = reader.uint64();
          break;

        case 4:
          message.params = Params.decode(reader, reader.uint32());
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromPartial(object) {
    var _object$sales, _object$userPositions;

    const message = createBaseGenesisState();
    message.sales = ((_object$sales = object.sales) === null || _object$sales === void 0 ? void 0 : _object$sales.map(e => Sale.fromPartial(e))) || [];
    message.userPositions = ((_object$userPositions = object.userPositions) === null || _object$userPositions === void 0 ? void 0 : _object$userPositions.map(e => UserPositionKV.fromPartial(e))) || [];
    message.nextSaleId = object.nextSaleId !== undefined && object.nextSaleId !== null ? Long.fromValue(object.nextSaleId) : Long.UZERO;
    message.params = object.params !== undefined && object.params !== null ? Params.fromPartial(object.params) : undefined;
    return message;
  }

};

function createBaseUserPositionKV() {
  return {
    accAddress: "",
    saleId: Long.UZERO,
    userPosition: undefined
  };
}

export const UserPositionKV = {
  encode(message, writer = _m0.Writer.create()) {
    if (message.accAddress !== "") {
      writer.uint32(10).string(message.accAddress);
    }

    if (!message.saleId.isZero()) {
      writer.uint32(16).uint64(message.saleId);
    }

    if (message.userPosition !== undefined) {
      UserPosition.encode(message.userPosition, writer.uint32(26).fork()).ldelim();
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseUserPositionKV();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.accAddress = reader.string();
          break;

        case 2:
          message.saleId = reader.uint64();
          break;

        case 3:
          message.userPosition = UserPosition.decode(reader, reader.uint32());
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromPartial(object) {
    var _object$accAddress;

    const message = createBaseUserPositionKV();
    message.accAddress = (_object$accAddress = object.accAddress) !== null && _object$accAddress !== void 0 ? _object$accAddress : "";
    message.saleId = object.saleId !== undefined && object.saleId !== null ? Long.fromValue(object.saleId) : Long.UZERO;
    message.userPosition = object.userPosition !== undefined && object.userPosition !== null ? UserPosition.fromPartial(object.userPosition) : undefined;
    return message;
  }

};
