// osmojs@0.37.0/module/codegen/osmosis/streamswap/v1/query.rpc.Query.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as _m0 from "protobufjs/minimal";
import { createProtobufRpcClient } from "@cosmjs/stargate";
import { QuerySales, QuerySalesResponse, QuerySale, QuerySaleResponse, QueryUserPosition, QueryUserPositionResponse } from "./query";
/** Query defines the gRPC querier service. */

export class QueryClientImpl {
  constructor(rpc) {
    _defineProperty(this, "rpc", void 0);

    this.rpc = rpc;
    this.sales = this.sales.bind(this);
    this.sale = this.sale.bind(this);
    this.userPosition = this.userPosition.bind(this);
  }

  sales(request = {
    pagination: undefined
  }) {
    const data = QuerySales.encode(request).finish();
    const promise = this.rpc.request("osmosis.streamswap.v1.Query", "Sales", data);
    return promise.then(data => QuerySalesResponse.decode(new _m0.Reader(data)));
  }

  sale(request) {
    const data = QuerySale.encode(request).finish();
    const promise = this.rpc.request("osmosis.streamswap.v1.Query", "Sale", data);
    return promise.then(data => QuerySaleResponse.decode(new _m0.Reader(data)));
  }

  userPosition(request) {
    const data = QueryUserPosition.encode(request).finish();
    const promise = this.rpc.request("osmosis.streamswap.v1.Query", "UserPosition", data);
    return promise.then(data => QueryUserPositionResponse.decode(new _m0.Reader(data)));
  }

}
export const createRpcQueryExtension = base => {
  const rpc = createProtobufRpcClient(base);
  const queryService = new QueryClientImpl(rpc);
  return {
    sales(request) {
      return queryService.sales(request);
    },

    sale(request) {
      return queryService.sale(request);
    },

    userPosition(request) {
      return queryService.userPosition(request);
    }

  };
};
