// osmojs@0.37.0/module/codegen/osmosis/streamswap/v1/tx.registry.js
//@ts-nocheck
import { MsgCreateSale, MsgSubscribe, MsgWithdraw, MsgExitSale, MsgFinalizeSale } from "./tx";
export const registry = [["/osmosis.streamswap.v1.MsgCreateSale", MsgCreateSale], ["/osmosis.streamswap.v1.MsgSubscribe", MsgSubscribe], ["/osmosis.streamswap.v1.MsgWithdraw", MsgWithdraw], ["/osmosis.streamswap.v1.MsgExitSale", MsgExitSale], ["/osmosis.streamswap.v1.MsgFinalizeSale", MsgFinalizeSale]];
export const load = protoRegistry => {
  registry.forEach(([typeUrl, mod]) => {
    protoRegistry.register(typeUrl, mod);
  });
};
export const MessageComposer = {
  encoded: {
    createSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgCreateSale",
        value: MsgCreateSale.encode(value).finish()
      };
    },

    subscribe(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgSubscribe",
        value: MsgSubscribe.encode(value).finish()
      };
    },

    withdraw(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgWithdraw",
        value: MsgWithdraw.encode(value).finish()
      };
    },

    exitSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgExitSale",
        value: MsgExitSale.encode(value).finish()
      };
    },

    finalizeSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgFinalizeSale",
        value: MsgFinalizeSale.encode(value).finish()
      };
    }

  },
  withTypeUrl: {
    createSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgCreateSale",
        value
      };
    },

    subscribe(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgSubscribe",
        value
      };
    },

    withdraw(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgWithdraw",
        value
      };
    },

    exitSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgExitSale",
        value
      };
    },

    finalizeSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgFinalizeSale",
        value
      };
    }

  },
  fromPartial: {
    createSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgCreateSale",
        value: MsgCreateSale.fromPartial(value)
      };
    },

    subscribe(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgSubscribe",
        value: MsgSubscribe.fromPartial(value)
      };
    },

    withdraw(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgWithdraw",
        value: MsgWithdraw.fromPartial(value)
      };
    },

    exitSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgExitSale",
        value: MsgExitSale.fromPartial(value)
      };
    },

    finalizeSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgFinalizeSale",
        value: MsgFinalizeSale.fromPartial(value)
      };
    }

  }
};
