// pioneer-platform@1.0.0/services/pioneer-server/dist/controllers/x-apps.controller.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.XAppsController = exports.ApiError = void 0;
/*

    Pioneer REST endpoints



 */
var TAG = ' | API | ';
var pjson = require('../../package.json');
var log = require('@pioneer-platform/loggerdog')();
var _a = require('@pioneer-platform/default-redis'), subscriber = _a.subscriber, publisher = _a.publisher, redis = _a.redis;
//TODO if no mongo use nedb?
//https://github.com/louischatriot/nedb
var connection = require("@pioneer-platform/default-mongo");
var usersDB = connection.get('users');
var txsDB = connection.get('transactions');
var txsRawDB = connection.get('transactions-raw');
var appsDB = connection.get('apps');
txsDB.createIndex({ txid: 1 }, { unique: true });
txsRawDB.createIndex({ txhash: 1 }, { unique: true });
//appsDB.createIndex({appId: 1}, {unique: true})
//globals
//rest-ts
var tsoa_1 = require("tsoa");
var ApiError = /** @class */ (function (_super) {
    __extends(ApiError, _super);
    function ApiError(name, statusCode, message) {
        var _this = _super.call(this, message) || this;
        _this.name = name;
        _this.statusCode = statusCode;
        return _this;
    }
    return ApiError;
}(Error));
exports.ApiError = ApiError;
//route
var XAppsController = /** @class */ (function (_super) {
    __extends(XAppsController, _super);
    function XAppsController() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /*
        Create

     */
    XAppsController.prototype.createApp = function (authorization, body) {
        return __awaiter(this, void 0, void 0, function () {
            var tag, success, errorResp;
            return __generator(this, function (_a) {
                tag = TAG + " | transactions | ";
                try {
                    success = appsDB.insert(body);
                    return [2 /*return*/, (success)];
                }
                catch (e) {
                    errorResp = {
                        success: false,
                        tag: tag,
                        e: e
                    };
                    log.error(tag, "e: ", { errorResp: errorResp });
                    throw new ApiError("error", 503, "error: " + e.toString());
                }
                return [2 /*return*/];
            });
        });
    };
    /*
        read


    */
    XAppsController.prototype.listApps = function () {
        return __awaiter(this, void 0, void 0, function () {
            var tag, apps, errorResp;
            return __generator(this, function (_a) {
                tag = TAG + " | health | ";
                try {
                    apps = appsDB.find();
                    return [2 /*return*/, (apps)];
                }
                catch (e) {
                    errorResp = {
                        success: false,
                        tag: tag,
                        e: e
                    };
                    log.error(tag, "e: ", { errorResp: errorResp });
                    throw new ApiError("error", 503, "error: " + e.toString());
                }
                return [2 /*return*/];
            });
        });
    };
    __decorate([
        tsoa_1.Post('/create'),
        __param(0, tsoa_1.Header('Authorization')), __param(1, tsoa_1.Body())
    ], XAppsController.prototype, "createApp");
    __decorate([
        tsoa_1.Get('/apps')
    ], XAppsController.prototype, "listApps");
    XAppsController = __decorate([
        tsoa_1.Tags('App Store Endpoints'),
        tsoa_1.Route('')
    ], XAppsController);
    return XAppsController;
}(tsoa_1.Controller));
exports.XAppsController = XAppsController;
