// @shapeshiftoss/unchained-client@10.14.0/dist/generated/optimism/models/OptimismGasFees.js
"use strict";
// @ts-nocheck
/* tslint:disable */
/* eslint-disable */
/**
 * @shapeshiftoss/optimism-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 8.3.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptimismGasFeesToJSON = exports.OptimismGasFeesFromJSONTyped = exports.OptimismGasFeesFromJSON = exports.instanceOfOptimismGasFees = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the OptimismGasFees interface.
 */
function instanceOfOptimismGasFees(value) {
    let isInstance = true;
    isInstance = isInstance && "gasPrice" in value;
    isInstance = isInstance && "l1GasPrice" in value;
    return isInstance;
}
exports.instanceOfOptimismGasFees = instanceOfOptimismGasFees;
function OptimismGasFeesFromJSON(json) {
    return OptimismGasFeesFromJSONTyped(json, false);
}
exports.OptimismGasFeesFromJSON = OptimismGasFeesFromJSON;
function OptimismGasFeesFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'gasPrice': json['gasPrice'],
        'maxFeePerGas': !(0, runtime_1.exists)(json, 'maxFeePerGas') ? undefined : json['maxFeePerGas'],
        'maxPriorityFeePerGas': !(0, runtime_1.exists)(json, 'maxPriorityFeePerGas') ? undefined : json['maxPriorityFeePerGas'],
        'l1GasPrice': json['l1GasPrice'],
    };
}
exports.OptimismGasFeesFromJSONTyped = OptimismGasFeesFromJSONTyped;
function OptimismGasFeesToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'gasPrice': value.gasPrice,
        'maxFeePerGas': value.maxFeePerGas,
        'maxPriorityFeePerGas': value.maxPriorityFeePerGas,
        'l1GasPrice': value.l1GasPrice,
    };
}
exports.OptimismGasFeesToJSON = OptimismGasFeesToJSON;
