// @pioneer-platform/metamask@3.0.1/dist/MsgCompiled-f025c504-4442524e.js
import { i as ne, j as oe, k as re, n as ie, o as se, u as ue, p as ce, c as A, q as le } from "./index-2b41ecf2.js";
import "assert";
import "fs";
import "path";
import "buffer";
import "events";
import "util";
import "string_decoder";
var K = {}, U = {}, D, M;
function ae() {
  if (M)
    return D;
  M = 1, D = o;
  var i = j();
  function o(r, s) {
    this.lo = r >>> 0, this.hi = s >>> 0;
  }
  var e = o.zero = new o(0, 0);
  e.toNumber = function() {
    return 0;
  }, e.zzEncode = e.zzDecode = function() {
    return this;
  }, e.length = function() {
    return 1;
  };
  var n = o.zeroHash = "\0\0\0\0\0\0\0\0";
  o.fromNumber = function(r) {
    if (r === 0)
      return e;
    var s = r < 0;
    s && (r = -r);
    var u = r >>> 0, f = (r - u) / 4294967296 >>> 0;
    return s && (f = ~f >>> 0, u = ~u >>> 0, ++u > 4294967295 && (u = 0, ++f > 4294967295 && (f = 0))), new o(u, f);
  }, o.from = function(r) {
    if (typeof r == "number")
      return o.fromNumber(r);
    if (i.isString(r))
      if (i.Long)
        r = i.Long.fromString(r);
      else
        return o.fromNumber(parseInt(r, 10));
    return r.low || r.high ? new o(r.low >>> 0, r.high >>> 0) : e;
  }, o.prototype.toNumber = function(r) {
    if (!r && this.hi >>> 31) {
      var s = ~this.lo + 1 >>> 0, u = ~this.hi >>> 0;
      return s || (u = u + 1 >>> 0), -(s + u * 4294967296);
    }
    return this.lo + this.hi * 4294967296;
  }, o.prototype.toLong = function(r) {
    return i.Long ? new i.Long(this.lo | 0, this.hi | 0, !!r) : { low: this.lo | 0, high: this.hi | 0, unsigned: !!r };
  };
  var t = String.prototype.charCodeAt;
  return o.fromHash = function(r) {
    return r === n ? e : new o(
      (t.call(r, 0) | t.call(r, 1) << 8 | t.call(r, 2) << 16 | t.call(r, 3) << 24) >>> 0,
      (t.call(r, 4) | t.call(r, 5) << 8 | t.call(r, 6) << 16 | t.call(r, 7) << 24) >>> 0
    );
  }, o.prototype.toHash = function() {
    return String.fromCharCode(
      this.lo & 255,
      this.lo >>> 8 & 255,
      this.lo >>> 16 & 255,
      this.lo >>> 24,
      this.hi & 255,
      this.hi >>> 8 & 255,
      this.hi >>> 16 & 255,
      this.hi >>> 24
    );
  }, o.prototype.zzEncode = function() {
    var r = this.hi >> 31;
    return this.hi = ((this.hi << 1 | this.lo >>> 31) ^ r) >>> 0, this.lo = (this.lo << 1 ^ r) >>> 0, this;
  }, o.prototype.zzDecode = function() {
    var r = -(this.lo & 1);
    return this.lo = ((this.lo >>> 1 | this.hi << 31) ^ r) >>> 0, this.hi = (this.hi >>> 1 ^ r) >>> 0, this;
  }, o.prototype.length = function() {
    var r = this.lo, s = (this.lo >>> 28 | this.hi << 4) >>> 0, u = this.hi >>> 24;
    return u === 0 ? s === 0 ? r < 16384 ? r < 128 ? 1 : 2 : r < 2097152 ? 3 : 4 : s < 16384 ? s < 128 ? 5 : 6 : s < 2097152 ? 7 : 8 : u < 128 ? 9 : 10;
  }, D;
}
var F;
function j() {
  return F || (F = 1, function(i) {
    var o = i;
    o.asPromise = ne, o.base64 = oe, o.EventEmitter = re, o.float = ie, o.inquire = se, o.utf8 = ue, o.pool = ce, o.LongBits = ae(), o.isNode = !!(typeof A < "u" && A && A.process && A.process.versions && A.process.versions.node), o.global = o.isNode && A || typeof window < "u" && window || typeof self < "u" && self || A, o.emptyArray = Object.freeze ? Object.freeze([]) : (
      /* istanbul ignore next */
      []
    ), o.emptyObject = Object.freeze ? Object.freeze({}) : (
      /* istanbul ignore next */
      {}
    ), o.isInteger = Number.isInteger || /* istanbul ignore next */
    function(t) {
      return typeof t == "number" && isFinite(t) && Math.floor(t) === t;
    }, o.isString = function(t) {
      return typeof t == "string" || t instanceof String;
    }, o.isObject = function(t) {
      return t && typeof t == "object";
    }, o.isset = /**
    * Checks if a property on a message is considered to be present.
    * @param {Object} obj Plain object or message instance
    * @param {string} prop Property name
    * @returns {boolean} `true` if considered to be present, otherwise `false`
    */
    o.isSet = function(t, r) {
      var s = t[r];
      return s != null && t.hasOwnProperty(r) ? typeof s != "object" || (Array.isArray(s) ? s.length : Object.keys(s).length) > 0 : !1;
    }, o.Buffer = function() {
      try {
        var t = o.inquire("buffer").Buffer;
        return t.prototype.utf8Write ? t : (
          /* istanbul ignore next */
          null
        );
      } catch {
        return null;
      }
    }(), o._Buffer_from = null, o._Buffer_allocUnsafe = null, o.newBuffer = function(t) {
      return typeof t == "number" ? o.Buffer ? o._Buffer_allocUnsafe(t) : new o.Array(t) : o.Buffer ? o._Buffer_from(t) : typeof Uint8Array > "u" ? t : new Uint8Array(t);
    }, o.Array = typeof Uint8Array < "u" ? Uint8Array : Array, o.Long = /* istanbul ignore next */
    o.global.dcodeIO && /* istanbul ignore next */
    o.global.dcodeIO.Long || /* istanbul ignore next */
    o.global.Long || o.inquire("long"), o.key2Re = /^true|false|0|1$/, o.key32Re = /^-?(?:0|[1-9][0-9]*)$/, o.key64Re = /^(?:[\\x00-\\xff]{8}|-?(?:0|[1-9][0-9]*))$/, o.longToHash = function(t) {
      return t ? o.LongBits.from(t).toHash() : o.LongBits.zeroHash;
    }, o.longFromHash = function(t, r) {
      var s = o.LongBits.fromHash(t);
      return o.Long ? o.Long.fromBits(s.lo, s.hi, r) : s.toNumber(!!r);
    };
    function e(t, r, s) {
      for (var u = Object.keys(r), f = 0; f < u.length; ++f)
        (t[u[f]] === void 0 || !s) && (t[u[f]] = r[u[f]]);
      return t;
    }
    o.merge = e, o.lcFirst = function(t) {
      return t.charAt(0).toLowerCase() + t.substring(1);
    };
    function n(t) {
      function r(s, u) {
        if (!(this instanceof r))
          return new r(s, u);
        Object.defineProperty(this, "message", { get: function() {
          return s;
        } }), Error.captureStackTrace ? Error.captureStackTrace(this, r) : Object.defineProperty(this, "stack", { value: new Error().stack || "" }), u && e(this, u);
      }
      return r.prototype = Object.create(Error.prototype, {
        constructor: {
          value: r,
          writable: !0,
          enumerable: !1,
          configurable: !0
        },
        name: {
          get: function() {
            return t;
          },
          set: void 0,
          enumerable: !1,
          // configurable: false would accurately preserve the behavior of
          // the original, but I'm guessing that was not intentional.
          // For an actual error subclass, this property would
          // be configurable.
          configurable: !0
        },
        toString: {
          value: function() {
            return this.name + ": " + this.message;
          },
          writable: !0,
          enumerable: !1,
          configurable: !0
        }
      }), r;
    }
    o.newError = n, o.ProtocolError = n("ProtocolError"), o.oneOfGetter = function(t) {
      for (var r = {}, s = 0; s < t.length; ++s)
        r[t[s]] = 1;
      return function() {
        for (var u = Object.keys(this), f = u.length - 1; f > -1; --f)
          if (r[u[f]] === 1 && this[u[f]] !== void 0 && this[u[f]] !== null)
            return u[f];
      };
    }, o.oneOfSetter = function(t) {
      return function(r) {
        for (var s = 0; s < t.length; ++s)
          t[s] !== r && delete this[t[s]];
      };
    }, o.toJSONOptions = {
      longs: String,
      enums: String,
      bytes: String,
      json: !0
    }, o._configure = function() {
      var t = o.Buffer;
      if (!t) {
        o._Buffer_from = o._Buffer_allocUnsafe = null;
        return;
      }
      o._Buffer_from = t.from !== Uint8Array.from && t.from || /* istanbul ignore next */
      function(r, s) {
        return new t(r, s);
      }, o._Buffer_allocUnsafe = t.allocUnsafe || /* istanbul ignore next */
      function(r) {
        return new t(r);
      };
    };
  }(U)), U;
}
var G = d, y = j(), _, x = y.LongBits, R = y.base64, q = y.utf8;
function S(i, o, e) {
  this.fn = i, this.len = o, this.next = void 0, this.val = e;
}
function E() {
}
function fe(i) {
  this.head = i.head, this.tail = i.tail, this.len = i.len, this.next = i.states;
}
function d() {
  this.len = 0, this.head = new S(E, 0, 0), this.tail = this.head, this.states = null;
}
var V = function() {
  return y.Buffer ? function() {
    return (d.create = function() {
      return new _();
    })();
  } : function() {
    return new d();
  };
};
d.create = V();
d.alloc = function(i) {
  return new y.Array(i);
};
y.Array !== Array && (d.alloc = y.pool(d.alloc, y.Array.prototype.subarray));
d.prototype._push = function(i, o, e) {
  return this.tail = this.tail.next = new S(i, o, e), this.len += o, this;
};
function T(i, o, e) {
  o[e] = i & 255;
}
function de(i, o, e) {
  for (; i > 127; )
    o[e++] = i & 127 | 128, i >>>= 7;
  o[e] = i;
}
function L(i, o) {
  this.len = i, this.next = void 0, this.val = o;
}
L.prototype = Object.create(S.prototype);
L.prototype.fn = de;
d.prototype.uint32 = function(i) {
  return this.len += (this.tail = this.tail.next = new L(
    (i = i >>> 0) < 128 ? 1 : i < 16384 ? 2 : i < 2097152 ? 3 : i < 268435456 ? 4 : 5,
    i
  )).len, this;
};
d.prototype.int32 = function(i) {
  return i < 0 ? this._push(J, 10, x.fromNumber(i)) : this.uint32(i);
};
d.prototype.sint32 = function(i) {
  return this.uint32((i << 1 ^ i >> 31) >>> 0);
};
function J(i, o, e) {
  for (; i.hi; )
    o[e++] = i.lo & 127 | 128, i.lo = (i.lo >>> 7 | i.hi << 25) >>> 0, i.hi >>>= 7;
  for (; i.lo > 127; )
    o[e++] = i.lo & 127 | 128, i.lo = i.lo >>> 7;
  o[e++] = i.lo;
}
d.prototype.uint64 = function(i) {
  var o = x.from(i);
  return this._push(J, o.length(), o);
};
d.prototype.int64 = d.prototype.uint64;
d.prototype.sint64 = function(i) {
  var o = x.from(i).zzEncode();
  return this._push(J, o.length(), o);
};
d.prototype.bool = function(i) {
  return this._push(T, 1, i ? 1 : 0);
};
function C(i, o, e) {
  o[e] = i & 255, o[e + 1] = i >>> 8 & 255, o[e + 2] = i >>> 16 & 255, o[e + 3] = i >>> 24;
}
d.prototype.fixed32 = function(i) {
  return this._push(C, 4, i >>> 0);
};
d.prototype.sfixed32 = d.prototype.fixed32;
d.prototype.fixed64 = function(i) {
  var o = x.from(i);
  return this._push(C, 4, o.lo)._push(C, 4, o.hi);
};
d.prototype.sfixed64 = d.prototype.fixed64;
d.prototype.float = function(i) {
  return this._push(y.float.writeFloatLE, 4, i);
};
d.prototype.double = function(i) {
  return this._push(y.float.writeDoubleLE, 8, i);
};
var he = y.Array.prototype.set ? function(i, o, e) {
  o.set(i, e);
} : function(i, o, e) {
  for (var n = 0; n < i.length; ++n)
    o[e + n] = i[n];
};
d.prototype.bytes = function(i) {
  var o = i.length >>> 0;
  if (!o)
    return this._push(T, 1, 0);
  if (y.isString(i)) {
    var e = d.alloc(o = R.length(i));
    R.decode(i, e, 0), i = e;
  }
  return this.uint32(o)._push(he, o, i);
};
d.prototype.string = function(i) {
  var o = q.length(i);
  return o ? this.uint32(o)._push(q.write, o, i) : this._push(T, 1, 0);
};
d.prototype.fork = function() {
  return this.states = new fe(this), this.head = this.tail = new S(E, 0, 0), this.len = 0, this;
};
d.prototype.reset = function() {
  return this.states ? (this.head = this.states.head, this.tail = this.states.tail, this.len = this.states.len, this.states = this.states.next) : (this.head = this.tail = new S(E, 0, 0), this.len = 0), this;
};
d.prototype.ldelim = function() {
  var i = this.head, o = this.tail, e = this.len;
  return this.reset().uint32(e), e && (this.tail.next = i.next, this.tail = o, this.len += e), this;
};
d.prototype.finish = function() {
  for (var i = this.head.next, o = this.constructor.alloc(this.len), e = 0; i; )
    i.fn(i.val, o, e), e += i.len, i = i.next;
  return o;
};
d._configure = function(i) {
  _ = i, d.create = V(), _._configure();
};
var pe = w, Q = G;
(w.prototype = Object.create(Q.prototype)).constructor = w;
var O = j();
function w() {
  Q.call(this);
}
w._configure = function() {
  w.alloc = O._Buffer_allocUnsafe, w.writeBytesBuffer = O.Buffer && O.Buffer.prototype instanceof Uint8Array && O.Buffer.prototype.set.name === "set" ? function(i, o, e) {
    o.set(i, e);
  } : function(i, o, e) {
    if (i.copy)
      i.copy(o, e, 0, i.length);
    else
      for (var n = 0; n < i.length; )
        o[e++] = i[n++];
  };
};
w.prototype.bytes = function(i) {
  O.isString(i) && (i = O._Buffer_from(i, "base64"));
  var o = i.length >>> 0;
  return this.uint32(o), o && this._push(w.writeBytesBuffer, o, i), this;
};
function me(i, o, e) {
  i.length < 40 ? O.utf8.write(i, o, e) : o.utf8Write ? o.utf8Write(i, e) : o.write(i, e);
}
w.prototype.string = function(i) {
  var o = O.Buffer.byteLength(i);
  return this.uint32(o), o && this._push(me, o, i), this;
};
w._configure();
var X = h, b = j(), N, Y = b.LongBits, ye = b.utf8;
function g(i, o) {
  return RangeError("index out of range: " + i.pos + " + " + (o || 1) + " > " + i.len);
}
function h(i) {
  this.buf = i, this.pos = 0, this.len = i.length;
}
var H = typeof Uint8Array < "u" ? function(i) {
  if (i instanceof Uint8Array || Array.isArray(i))
    return new h(i);
  throw Error("illegal buffer");
} : function(i) {
  if (Array.isArray(i))
    return new h(i);
  throw Error("illegal buffer");
}, Z = function() {
  return b.Buffer ? function(i) {
    return (h.create = function(o) {
      return b.Buffer.isBuffer(o) ? new N(o) : H(o);
    })(i);
  } : H;
};
h.create = Z();
h.prototype._slice = b.Array.prototype.subarray || /* istanbul ignore next */
b.Array.prototype.slice;
h.prototype.uint32 = function() {
  var i = 4294967295;
  return function() {
    if (i = (this.buf[this.pos] & 127) >>> 0, this.buf[this.pos++] < 128 || (i = (i | (this.buf[this.pos] & 127) << 7) >>> 0, this.buf[this.pos++] < 128) || (i = (i | (this.buf[this.pos] & 127) << 14) >>> 0, this.buf[this.pos++] < 128) || (i = (i | (this.buf[this.pos] & 127) << 21) >>> 0, this.buf[this.pos++] < 128) || (i = (i | (this.buf[this.pos] & 15) << 28) >>> 0, this.buf[this.pos++] < 128))
      return i;
    if ((this.pos += 5) > this.len)
      throw this.pos = this.len, g(this, 10);
    return i;
  };
}();
h.prototype.int32 = function() {
  return this.uint32() | 0;
};
h.prototype.sint32 = function() {
  var i = this.uint32();
  return i >>> 1 ^ -(i & 1) | 0;
};
function B() {
  var i = new Y(0, 0), o = 0;
  if (this.len - this.pos > 4) {
    for (; o < 4; ++o)
      if (i.lo = (i.lo | (this.buf[this.pos] & 127) << o * 7) >>> 0, this.buf[this.pos++] < 128)
        return i;
    if (i.lo = (i.lo | (this.buf[this.pos] & 127) << 28) >>> 0, i.hi = (i.hi | (this.buf[this.pos] & 127) >> 4) >>> 0, this.buf[this.pos++] < 128)
      return i;
    o = 0;
  } else {
    for (; o < 3; ++o) {
      if (this.pos >= this.len)
        throw g(this);
      if (i.lo = (i.lo | (this.buf[this.pos] & 127) << o * 7) >>> 0, this.buf[this.pos++] < 128)
        return i;
    }
    return i.lo = (i.lo | (this.buf[this.pos++] & 127) << o * 7) >>> 0, i;
  }
  if (this.len - this.pos > 4) {
    for (; o < 5; ++o)
      if (i.hi = (i.hi | (this.buf[this.pos] & 127) << o * 7 + 3) >>> 0, this.buf[this.pos++] < 128)
        return i;
  } else
    for (; o < 5; ++o) {
      if (this.pos >= this.len)
        throw g(this);
      if (i.hi = (i.hi | (this.buf[this.pos] & 127) << o * 7 + 3) >>> 0, this.buf[this.pos++] < 128)
        return i;
    }
  throw Error("invalid varint encoding");
}
h.prototype.bool = function() {
  return this.uint32() !== 0;
};
function P(i, o) {
  return (i[o - 4] | i[o - 3] << 8 | i[o - 2] << 16 | i[o - 1] << 24) >>> 0;
}
h.prototype.fixed32 = function() {
  if (this.pos + 4 > this.len)
    throw g(this, 4);
  return P(this.buf, this.pos += 4);
};
h.prototype.sfixed32 = function() {
  if (this.pos + 4 > this.len)
    throw g(this, 4);
  return P(this.buf, this.pos += 4) | 0;
};
function W() {
  if (this.pos + 8 > this.len)
    throw g(this, 8);
  return new Y(P(this.buf, this.pos += 4), P(this.buf, this.pos += 4));
}
h.prototype.float = function() {
  if (this.pos + 4 > this.len)
    throw g(this, 4);
  var i = b.float.readFloatLE(this.buf, this.pos);
  return this.pos += 4, i;
};
h.prototype.double = function() {
  if (this.pos + 8 > this.len)
    throw g(this, 4);
  var i = b.float.readDoubleLE(this.buf, this.pos);
  return this.pos += 8, i;
};
h.prototype.bytes = function() {
  var i = this.uint32(), o = this.pos, e = this.pos + i;
  if (e > this.len)
    throw g(this, i);
  if (this.pos += i, Array.isArray(this.buf))
    return this.buf.slice(o, e);
  if (o === e) {
    var n = b.Buffer;
    return n ? n.alloc(0) : new this.buf.constructor(0);
  }
  return this._slice.call(this.buf, o, e);
};
h.prototype.string = function() {
  var i = this.bytes();
  return ye.read(i, 0, i.length);
};
h.prototype.skip = function(i) {
  if (typeof i == "number") {
    if (this.pos + i > this.len)
      throw g(this, i);
    this.pos += i;
  } else
    do
      if (this.pos >= this.len)
        throw g(this);
    while (this.buf[this.pos++] & 128);
  return this;
};
h.prototype.skipType = function(i) {
  switch (i) {
    case 0:
      this.skip();
      break;
    case 1:
      this.skip(8);
      break;
    case 2:
      this.skip(this.uint32());
      break;
    case 3:
      for (; (i = this.uint32() & 7) !== 4; )
        this.skipType(i);
      break;
    case 5:
      this.skip(4);
      break;
    default:
      throw Error("invalid wire type " + i + " at offset " + this.pos);
  }
  return this;
};
h._configure = function(i) {
  N = i, h.create = Z(), N._configure();
  var o = b.Long ? "toLong" : (
    /* istanbul ignore next */
    "toNumber"
  );
  b.merge(h.prototype, {
    int64: function() {
      return B.call(this)[o](!1);
    },
    uint64: function() {
      return B.call(this)[o](!0);
    },
    sint64: function() {
      return B.call(this).zzDecode()[o](!1);
    },
    fixed64: function() {
      return W.call(this)[o](!0);
    },
    sfixed64: function() {
      return W.call(this)[o](!1);
    }
  });
};
var be = v, ee = X;
(v.prototype = Object.create(ee.prototype)).constructor = v;
var $ = j();
function v(i) {
  ee.call(this, i);
}
v._configure = function() {
  $.Buffer && (v.prototype._slice = $.Buffer.prototype.slice);
};
v.prototype.string = function() {
  var i = this.uint32();
  return this.buf.utf8Slice ? this.buf.utf8Slice(this.pos, this.pos = Math.min(this.pos + i, this.len)) : this.buf.toString("utf-8", this.pos, this.pos = Math.min(this.pos + i, this.len));
};
v._configure();
var te = {}, ge = k, I = j();
(k.prototype = Object.create(I.EventEmitter.prototype)).constructor = k;
function k(i, o, e) {
  if (typeof i != "function")
    throw TypeError("rpcImpl must be a function");
  I.EventEmitter.call(this), this.rpcImpl = i, this.requestDelimited = !!o, this.responseDelimited = !!e;
}
k.prototype.rpcCall = function i(o, e, n, t, r) {
  if (!t)
    throw TypeError("request must be specified");
  var s = this;
  if (!r)
    return I.asPromise(i, s, o, e, n, t);
  if (!s.rpcImpl) {
    setTimeout(function() {
      r(Error("already ended"));
    }, 0);
    return;
  }
  try {
    return s.rpcImpl(
      o,
      e[s.requestDelimited ? "encodeDelimited" : "encode"](t).finish(),
      function(u, f) {
        if (u)
          return s.emit("error", u, o), r(u);
        if (f === null) {
          s.end(
            /* endedByRPC */
            !0
          );
          return;
        }
        if (!(f instanceof n))
          try {
            f = n[s.responseDelimited ? "decodeDelimited" : "decode"](f);
          } catch (z) {
            return s.emit("error", z, o), r(z);
          }
        return s.emit("data", f, o), r(null, f);
      }
    );
  } catch (u) {
    s.emit("error", u, o), setTimeout(function() {
      r(u);
    }, 0);
    return;
  }
};
k.prototype.end = function(i) {
  return this.rpcImpl && (i || this.rpcImpl(null, null, null), this.rpcImpl = null, this.emit("end").off()), this;
};
(function(i) {
  var o = i;
  o.Service = ge;
})(te);
var we = {};
(function(i) {
  var o = i;
  o.build = "minimal", o.Writer = G, o.BufferWriter = pe, o.Reader = X, o.BufferReader = be, o.util = j(), o.rpc = te, o.roots = we, o.configure = e;
  function e() {
    o.util._configure(), o.Writer._configure(o.BufferWriter), o.Reader._configure(o.BufferReader);
  }
  e();
})(K);
var Oe = K;
const p = /* @__PURE__ */ le(Oe), a = p.Reader, m = p.Writer, l = p.util, c = p.roots.default || (p.roots.default = {});
c.common = function() {
  let i = {};
  return i.Asset = function() {
    function o(e) {
      if (e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.chain = "", o.prototype.symbol = "", o.prototype.ticker = "", o.prototype.synth = !1, o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      return n || (n = m.create()), e.chain != null && Object.hasOwnProperty.call(e, "chain") && n.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(e.chain), e.symbol != null && Object.hasOwnProperty.call(e, "symbol") && n.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(e.symbol), e.ticker != null && Object.hasOwnProperty.call(e, "ticker") && n.uint32(
        /* id 3, wireType 2 =*/
        26
      ).string(e.ticker), e.synth != null && Object.hasOwnProperty.call(e, "synth") && n.uint32(
        /* id 4, wireType 0 =*/
        32
      ).bool(e.synth), n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      e instanceof a || (e = a.create(e));
      let t = n === void 0 ? e.len : e.pos + n, r = new c.common.Asset();
      for (; e.pos < t; ) {
        let s = e.uint32();
        switch (s >>> 3) {
          case 1:
            r.chain = e.string();
            break;
          case 2:
            r.symbol = e.string();
            break;
          case 3:
            r.ticker = e.string();
            break;
          case 4:
            r.synth = e.bool();
            break;
          default:
            e.skipType(s & 7);
            break;
        }
      }
      return r;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      return typeof e != "object" || e === null ? "object expected" : e.chain != null && e.hasOwnProperty("chain") && !l.isString(e.chain) ? "chain: string expected" : e.symbol != null && e.hasOwnProperty("symbol") && !l.isString(e.symbol) ? "symbol: string expected" : e.ticker != null && e.hasOwnProperty("ticker") && !l.isString(e.ticker) ? "ticker: string expected" : e.synth != null && e.hasOwnProperty("synth") && typeof e.synth != "boolean" ? "synth: boolean expected" : null;
    }, o.fromObject = function(e) {
      if (e instanceof c.common.Asset)
        return e;
      let n = new c.common.Asset();
      return e.chain != null && (n.chain = String(e.chain)), e.symbol != null && (n.symbol = String(e.symbol)), e.ticker != null && (n.ticker = String(e.ticker)), e.synth != null && (n.synth = !!e.synth), n;
    }, o.toObject = function(e, n) {
      n || (n = {});
      let t = {};
      return n.defaults && (t.chain = "", t.symbol = "", t.ticker = "", t.synth = !1), e.chain != null && e.hasOwnProperty("chain") && (t.chain = e.chain), e.symbol != null && e.hasOwnProperty("symbol") && (t.symbol = e.symbol), e.ticker != null && e.hasOwnProperty("ticker") && (t.ticker = e.ticker), e.synth != null && e.hasOwnProperty("synth") && (t.synth = e.synth), t;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i.Coin = function() {
    function o(e) {
      if (e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.asset = null, o.prototype.amount = "", o.prototype.decimals = l.Long ? l.Long.fromBits(0, 0, !1) : 0, o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      return n || (n = m.create()), e.asset != null && Object.hasOwnProperty.call(e, "asset") && c.common.Asset.encode(
        e.asset,
        n.uint32(
          /* id 1, wireType 2 =*/
          10
        ).fork()
      ).ldelim(), e.amount != null && Object.hasOwnProperty.call(e, "amount") && n.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(e.amount), e.decimals != null && Object.hasOwnProperty.call(e, "decimals") && n.uint32(
        /* id 3, wireType 0 =*/
        24
      ).int64(e.decimals), n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      e instanceof a || (e = a.create(e));
      let t = n === void 0 ? e.len : e.pos + n, r = new c.common.Coin();
      for (; e.pos < t; ) {
        let s = e.uint32();
        switch (s >>> 3) {
          case 1:
            r.asset = c.common.Asset.decode(e, e.uint32());
            break;
          case 2:
            r.amount = e.string();
            break;
          case 3:
            r.decimals = e.int64();
            break;
          default:
            e.skipType(s & 7);
            break;
        }
      }
      return r;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      if (typeof e != "object" || e === null)
        return "object expected";
      if (e.asset != null && e.hasOwnProperty("asset")) {
        let n = c.common.Asset.verify(e.asset);
        if (n)
          return "asset." + n;
      }
      return e.amount != null && e.hasOwnProperty("amount") && !l.isString(e.amount) ? "amount: string expected" : e.decimals != null && e.hasOwnProperty("decimals") && !l.isInteger(e.decimals) && !(e.decimals && l.isInteger(e.decimals.low) && l.isInteger(e.decimals.high)) ? "decimals: integer|Long expected" : null;
    }, o.fromObject = function(e) {
      if (e instanceof c.common.Coin)
        return e;
      let n = new c.common.Coin();
      if (e.asset != null) {
        if (typeof e.asset != "object")
          throw TypeError(".common.Coin.asset: object expected");
        n.asset = c.common.Asset.fromObject(e.asset);
      }
      return e.amount != null && (n.amount = String(e.amount)), e.decimals != null && (l.Long ? (n.decimals = l.Long.fromValue(e.decimals)).unsigned = !1 : typeof e.decimals == "string" ? n.decimals = parseInt(e.decimals, 10) : typeof e.decimals == "number" ? n.decimals = e.decimals : typeof e.decimals == "object" && (n.decimals = new l.LongBits(
        e.decimals.low >>> 0,
        e.decimals.high >>> 0
      ).toNumber())), n;
    }, o.toObject = function(e, n) {
      n || (n = {});
      let t = {};
      if (n.defaults)
        if (t.asset = null, t.amount = "", l.Long) {
          let r = new l.Long(0, 0, !1);
          t.decimals = n.longs === String ? r.toString() : n.longs === Number ? r.toNumber() : r;
        } else
          t.decimals = n.longs === String ? "0" : 0;
      return e.asset != null && e.hasOwnProperty("asset") && (t.asset = c.common.Asset.toObject(e.asset, n)), e.amount != null && e.hasOwnProperty("amount") && (t.amount = e.amount), e.decimals != null && e.hasOwnProperty("decimals") && (typeof e.decimals == "number" ? t.decimals = n.longs === String ? String(e.decimals) : e.decimals : t.decimals = n.longs === String ? l.Long.prototype.toString.call(e.decimals) : n.longs === Number ? new l.LongBits(
        e.decimals.low >>> 0,
        e.decimals.high >>> 0
      ).toNumber() : e.decimals), t;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i.PubKeySet = function() {
    function o(e) {
      if (e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.secp256k1 = "", o.prototype.ed25519 = "", o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      return n || (n = m.create()), e.secp256k1 != null && Object.hasOwnProperty.call(e, "secp256k1") && n.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(e.secp256k1), e.ed25519 != null && Object.hasOwnProperty.call(e, "ed25519") && n.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(e.ed25519), n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      e instanceof a || (e = a.create(e));
      let t = n === void 0 ? e.len : e.pos + n, r = new c.common.PubKeySet();
      for (; e.pos < t; ) {
        let s = e.uint32();
        switch (s >>> 3) {
          case 1:
            r.secp256k1 = e.string();
            break;
          case 2:
            r.ed25519 = e.string();
            break;
          default:
            e.skipType(s & 7);
            break;
        }
      }
      return r;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      return typeof e != "object" || e === null ? "object expected" : e.secp256k1 != null && e.hasOwnProperty("secp256k1") && !l.isString(e.secp256k1) ? "secp256k1: string expected" : e.ed25519 != null && e.hasOwnProperty("ed25519") && !l.isString(e.ed25519) ? "ed25519: string expected" : null;
    }, o.fromObject = function(e) {
      if (e instanceof c.common.PubKeySet)
        return e;
      let n = new c.common.PubKeySet();
      return e.secp256k1 != null && (n.secp256k1 = String(e.secp256k1)), e.ed25519 != null && (n.ed25519 = String(e.ed25519)), n;
    }, o.toObject = function(e, n) {
      n || (n = {});
      let t = {};
      return n.defaults && (t.secp256k1 = "", t.ed25519 = ""), e.secp256k1 != null && e.hasOwnProperty("secp256k1") && (t.secp256k1 = e.secp256k1), e.ed25519 != null && e.hasOwnProperty("ed25519") && (t.ed25519 = e.ed25519), t;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i.Tx = function() {
    function o(e) {
      if (this.coins = [], this.gas = [], e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.id = "", o.prototype.chain = "", o.prototype.fromAddress = "", o.prototype.toAddress = "", o.prototype.coins = l.emptyArray, o.prototype.gas = l.emptyArray, o.prototype.memo = "", o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      var t, r;
      if (n || (n = m.create()), e.id != null && Object.hasOwnProperty.call(e, "id") && n.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(e.id), e.chain != null && Object.hasOwnProperty.call(e, "chain") && n.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(e.chain), e.fromAddress != null && Object.hasOwnProperty.call(e, "fromAddress") && n.uint32(
        /* id 3, wireType 2 =*/
        26
      ).string(e.fromAddress), e.toAddress != null && Object.hasOwnProperty.call(e, "toAddress") && n.uint32(
        /* id 4, wireType 2 =*/
        34
      ).string(e.toAddress), (t = e.coins) != null && t.length)
        for (let s = 0; s < e.coins.length; ++s)
          c.common.Coin.encode(
            e.coins[s],
            n.uint32(
              /* id 5, wireType 2 =*/
              42
            ).fork()
          ).ldelim();
      if ((r = e.gas) != null && r.length)
        for (let s = 0; s < e.gas.length; ++s)
          c.common.Coin.encode(
            e.gas[s],
            n.uint32(
              /* id 6, wireType 2 =*/
              50
            ).fork()
          ).ldelim();
      return e.memo != null && Object.hasOwnProperty.call(e, "memo") && n.uint32(
        /* id 7, wireType 2 =*/
        58
      ).string(e.memo), n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      var t, r;
      e instanceof a || (e = a.create(e));
      let s = n === void 0 ? e.len : e.pos + n, u = new c.common.Tx();
      for (; e.pos < s; ) {
        let f = e.uint32();
        switch (f >>> 3) {
          case 1:
            u.id = e.string();
            break;
          case 2:
            u.chain = e.string();
            break;
          case 3:
            u.fromAddress = e.string();
            break;
          case 4:
            u.toAddress = e.string();
            break;
          case 5:
            (t = u.coins) != null && t.length || (u.coins = []), u.coins.push(c.common.Coin.decode(e, e.uint32()));
            break;
          case 6:
            (r = u.gas) != null && r.length || (u.gas = []), u.gas.push(c.common.Coin.decode(e, e.uint32()));
            break;
          case 7:
            u.memo = e.string();
            break;
          default:
            e.skipType(f & 7);
            break;
        }
      }
      return u;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      if (typeof e != "object" || e === null)
        return "object expected";
      if (e.id != null && e.hasOwnProperty("id") && !l.isString(e.id))
        return "id: string expected";
      if (e.chain != null && e.hasOwnProperty("chain") && !l.isString(e.chain))
        return "chain: string expected";
      if (e.fromAddress != null && e.hasOwnProperty("fromAddress") && !l.isString(e.fromAddress))
        return "fromAddress: string expected";
      if (e.toAddress != null && e.hasOwnProperty("toAddress") && !l.isString(e.toAddress))
        return "toAddress: string expected";
      if (e.coins != null && e.hasOwnProperty("coins")) {
        if (!Array.isArray(e.coins))
          return "coins: array expected";
        for (let n = 0; n < e.coins.length; ++n) {
          let t = c.common.Coin.verify(e.coins[n]);
          if (t)
            return "coins." + t;
        }
      }
      if (e.gas != null && e.hasOwnProperty("gas")) {
        if (!Array.isArray(e.gas))
          return "gas: array expected";
        for (let n = 0; n < e.gas.length; ++n) {
          let t = c.common.Coin.verify(e.gas[n]);
          if (t)
            return "gas." + t;
        }
      }
      return e.memo != null && e.hasOwnProperty("memo") && !l.isString(e.memo) ? "memo: string expected" : null;
    }, o.fromObject = function(e) {
      if (e instanceof c.common.Tx)
        return e;
      let n = new c.common.Tx();
      if (e.id != null && (n.id = String(e.id)), e.chain != null && (n.chain = String(e.chain)), e.fromAddress != null && (n.fromAddress = String(e.fromAddress)), e.toAddress != null && (n.toAddress = String(e.toAddress)), e.coins) {
        if (!Array.isArray(e.coins))
          throw TypeError(".common.Tx.coins: array expected");
        n.coins = [];
        for (let t = 0; t < e.coins.length; ++t) {
          if (typeof e.coins[t] != "object")
            throw TypeError(".common.Tx.coins: object expected");
          n.coins[t] = c.common.Coin.fromObject(e.coins[t]);
        }
      }
      if (e.gas) {
        if (!Array.isArray(e.gas))
          throw TypeError(".common.Tx.gas: array expected");
        n.gas = [];
        for (let t = 0; t < e.gas.length; ++t) {
          if (typeof e.gas[t] != "object")
            throw TypeError(".common.Tx.gas: object expected");
          n.gas[t] = c.common.Coin.fromObject(e.gas[t]);
        }
      }
      return e.memo != null && (n.memo = String(e.memo)), n;
    }, o.toObject = function(e, n) {
      var t, r;
      n || (n = {});
      let s = {};
      if ((n.arrays || n.defaults) && (s.coins = [], s.gas = []), n.defaults && (s.id = "", s.chain = "", s.fromAddress = "", s.toAddress = "", s.memo = ""), e.id != null && e.hasOwnProperty("id") && (s.id = e.id), e.chain != null && e.hasOwnProperty("chain") && (s.chain = e.chain), e.fromAddress != null && e.hasOwnProperty("fromAddress") && (s.fromAddress = e.fromAddress), e.toAddress != null && e.hasOwnProperty("toAddress") && (s.toAddress = e.toAddress), (t = e.coins) != null && t.length) {
        s.coins = [];
        for (let u = 0; u < e.coins.length; ++u)
          s.coins[u] = c.common.Coin.toObject(e.coins[u], n);
      }
      if ((r = e.gas) != null && r.length) {
        s.gas = [];
        for (let u = 0; u < e.gas.length; ++u)
          s.gas[u] = c.common.Coin.toObject(e.gas[u], n);
      }
      return e.memo != null && e.hasOwnProperty("memo") && (s.memo = e.memo), s;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i.Fee = function() {
    function o(e) {
      if (this.coins = [], e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.coins = l.emptyArray, o.prototype.poolDeduct = "", o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      var t;
      if (n || (n = m.create()), (t = e.coins) != null && t.length)
        for (let r = 0; r < e.coins.length; ++r)
          c.common.Coin.encode(
            e.coins[r],
            n.uint32(
              /* id 1, wireType 2 =*/
              10
            ).fork()
          ).ldelim();
      return e.poolDeduct != null && Object.hasOwnProperty.call(e, "poolDeduct") && n.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(e.poolDeduct), n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      var t;
      e instanceof a || (e = a.create(e));
      let r = n === void 0 ? e.len : e.pos + n, s = new c.common.Fee();
      for (; e.pos < r; ) {
        let u = e.uint32();
        switch (u >>> 3) {
          case 1:
            (t = s.coins) != null && t.length || (s.coins = []), s.coins.push(c.common.Coin.decode(e, e.uint32()));
            break;
          case 2:
            s.poolDeduct = e.string();
            break;
          default:
            e.skipType(u & 7);
            break;
        }
      }
      return s;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      if (typeof e != "object" || e === null)
        return "object expected";
      if (e.coins != null && e.hasOwnProperty("coins")) {
        if (!Array.isArray(e.coins))
          return "coins: array expected";
        for (let n = 0; n < e.coins.length; ++n) {
          let t = c.common.Coin.verify(e.coins[n]);
          if (t)
            return "coins." + t;
        }
      }
      return e.poolDeduct != null && e.hasOwnProperty("poolDeduct") && !l.isString(e.poolDeduct) ? "poolDeduct: string expected" : null;
    }, o.fromObject = function(e) {
      if (e instanceof c.common.Fee)
        return e;
      let n = new c.common.Fee();
      if (e.coins) {
        if (!Array.isArray(e.coins))
          throw TypeError(".common.Fee.coins: array expected");
        n.coins = [];
        for (let t = 0; t < e.coins.length; ++t) {
          if (typeof e.coins[t] != "object")
            throw TypeError(".common.Fee.coins: object expected");
          n.coins[t] = c.common.Coin.fromObject(e.coins[t]);
        }
      }
      return e.poolDeduct != null && (n.poolDeduct = String(e.poolDeduct)), n;
    }, o.toObject = function(e, n) {
      var t;
      n || (n = {});
      let r = {};
      if ((n.arrays || n.defaults) && (r.coins = []), n.defaults && (r.poolDeduct = ""), (t = e.coins) != null && t.length) {
        r.coins = [];
        for (let s = 0; s < e.coins.length; ++s)
          r.coins[s] = c.common.Coin.toObject(e.coins[s], n);
      }
      return e.poolDeduct != null && e.hasOwnProperty("poolDeduct") && (r.poolDeduct = e.poolDeduct), r;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i.ProtoUint = function() {
    function o(e) {
      if (e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.value = "", o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      return n || (n = m.create()), e.value != null && Object.hasOwnProperty.call(e, "value") && n.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(e.value), n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      e instanceof a || (e = a.create(e));
      let t = n === void 0 ? e.len : e.pos + n, r = new c.common.ProtoUint();
      for (; e.pos < t; ) {
        let s = e.uint32();
        switch (s >>> 3) {
          case 1:
            r.value = e.string();
            break;
          default:
            e.skipType(s & 7);
            break;
        }
      }
      return r;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      return typeof e != "object" || e === null ? "object expected" : e.value != null && e.hasOwnProperty("value") && !l.isString(e.value) ? "value: string expected" : null;
    }, o.fromObject = function(e) {
      if (e instanceof c.common.ProtoUint)
        return e;
      let n = new c.common.ProtoUint();
      return e.value != null && (n.value = String(e.value)), n;
    }, o.toObject = function(e, n) {
      n || (n = {});
      let t = {};
      return n.defaults && (t.value = ""), e.value != null && e.hasOwnProperty("value") && (t.value = e.value), t;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i;
}();
c.types = function() {
  let i = {};
  return i.MsgDeposit = function() {
    function o(e) {
      if (this.coins = [], e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.coins = l.emptyArray, o.prototype.memo = "", o.prototype.signer = l.newBuffer([]), o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      var t;
      if (n || (n = m.create()), (t = e.coins) != null && t.length)
        for (let r = 0; r < e.coins.length; ++r)
          c.common.Coin.encode(
            e.coins[r],
            n.uint32(
              /* id 1, wireType 2 =*/
              10
            ).fork()
          ).ldelim();
      return e.memo != null && Object.hasOwnProperty.call(e, "memo") && n.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(e.memo), e.signer != null && Object.hasOwnProperty.call(e, "signer") && n.uint32(
        /* id 3, wireType 2 =*/
        26
      ).bytes(e.signer), n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      var t;
      e instanceof a || (e = a.create(e));
      let r = n === void 0 ? e.len : e.pos + n, s = new c.types.MsgDeposit();
      for (; e.pos < r; ) {
        let u = e.uint32();
        switch (u >>> 3) {
          case 1:
            (t = s.coins) != null && t.length || (s.coins = []), s.coins.push(c.common.Coin.decode(e, e.uint32()));
            break;
          case 2:
            s.memo = e.string();
            break;
          case 3:
            s.signer = e.bytes();
            break;
          default:
            e.skipType(u & 7);
            break;
        }
      }
      return s;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      if (typeof e != "object" || e === null)
        return "object expected";
      if (e.coins != null && e.hasOwnProperty("coins")) {
        if (!Array.isArray(e.coins))
          return "coins: array expected";
        for (let n = 0; n < e.coins.length; ++n) {
          let t = c.common.Coin.verify(e.coins[n]);
          if (t)
            return "coins." + t;
        }
      }
      return e.memo != null && e.hasOwnProperty("memo") && !l.isString(e.memo) ? "memo: string expected" : e.signer != null && e.hasOwnProperty("signer") && !(e.signer && typeof e.signer.length == "number" || l.isString(e.signer)) ? "signer: buffer expected" : null;
    }, o.fromObject = function(e) {
      if (e instanceof c.types.MsgDeposit)
        return e;
      let n = new c.types.MsgDeposit();
      if (e.coins) {
        if (!Array.isArray(e.coins))
          throw TypeError(".types.MsgDeposit.coins: array expected");
        n.coins = [];
        for (let t = 0; t < e.coins.length; ++t) {
          if (typeof e.coins[t] != "object")
            throw TypeError(".types.MsgDeposit.coins: object expected");
          n.coins[t] = c.common.Coin.fromObject(e.coins[t]);
        }
      }
      return e.memo != null && (n.memo = String(e.memo)), e.signer != null && (typeof e.signer == "string" ? l.base64.decode(
        e.signer,
        n.signer = l.newBuffer(l.base64.length(e.signer)),
        0
      ) : e.signer.length && (n.signer = e.signer)), n;
    }, o.toObject = function(e, n) {
      var t;
      n || (n = {});
      let r = {};
      if ((n.arrays || n.defaults) && (r.coins = []), n.defaults && (r.memo = "", n.bytes === String ? r.signer = "" : (r.signer = [], n.bytes !== Array && (r.signer = l.newBuffer(r.signer)))), (t = e.coins) != null && t.length) {
        r.coins = [];
        for (let s = 0; s < e.coins.length; ++s)
          r.coins[s] = c.common.Coin.toObject(e.coins[s], n);
      }
      return e.memo != null && e.hasOwnProperty("memo") && (r.memo = e.memo), e.signer != null && e.hasOwnProperty("signer") && (r.signer = n.bytes === String ? l.base64.encode(e.signer, 0, e.signer.length) : n.bytes === Array ? Array.prototype.slice.call(e.signer) : e.signer), r;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i.MsgSend = function() {
    function o(e) {
      if (this.amount = [], e)
        for (let n = Object.keys(e), t = 0; t < n.length; ++t)
          e[n[t]] != null && (this[n[t]] = e[n[t]]);
    }
    return o.prototype.fromAddress = l.newBuffer([]), o.prototype.toAddress = l.newBuffer([]), o.prototype.amount = l.emptyArray, o.create = function(e) {
      return new o(e);
    }, o.encode = function(e, n) {
      var t;
      if (n || (n = m.create()), e.fromAddress != null && Object.hasOwnProperty.call(e, "fromAddress") && n.uint32(
        /* id 1, wireType 2 =*/
        10
      ).bytes(e.fromAddress), e.toAddress != null && Object.hasOwnProperty.call(e, "toAddress") && n.uint32(
        /* id 2, wireType 2 =*/
        18
      ).bytes(e.toAddress), (t = e.amount) != null && t.length)
        for (let r = 0; r < e.amount.length; ++r)
          c.cosmos.base.v1beta1.Coin.encode(
            e.amount[r],
            n.uint32(
              /* id 3, wireType 2 =*/
              26
            ).fork()
          ).ldelim();
      return n;
    }, o.encodeDelimited = function(e, n) {
      return this.encode(e, n).ldelim();
    }, o.decode = function(e, n) {
      var t;
      e instanceof a || (e = a.create(e));
      let r = n === void 0 ? e.len : e.pos + n, s = new c.types.MsgSend();
      for (; e.pos < r; ) {
        let u = e.uint32();
        switch (u >>> 3) {
          case 1:
            s.fromAddress = e.bytes();
            break;
          case 2:
            s.toAddress = e.bytes();
            break;
          case 3:
            (t = s.amount) != null && t.length || (s.amount = []), s.amount.push(c.cosmos.base.v1beta1.Coin.decode(e, e.uint32()));
            break;
          default:
            e.skipType(u & 7);
            break;
        }
      }
      return s;
    }, o.decodeDelimited = function(e) {
      return e instanceof a || (e = new a(e)), this.decode(e, e.uint32());
    }, o.verify = function(e) {
      if (typeof e != "object" || e === null)
        return "object expected";
      if (e.fromAddress != null && e.hasOwnProperty("fromAddress") && !(e.fromAddress && typeof e.fromAddress.length == "number" || l.isString(e.fromAddress)))
        return "fromAddress: buffer expected";
      if (e.toAddress != null && e.hasOwnProperty("toAddress") && !(e.toAddress && typeof e.toAddress.length == "number" || l.isString(e.toAddress)))
        return "toAddress: buffer expected";
      if (e.amount != null && e.hasOwnProperty("amount")) {
        if (!Array.isArray(e.amount))
          return "amount: array expected";
        for (let n = 0; n < e.amount.length; ++n) {
          let t = c.cosmos.base.v1beta1.Coin.verify(e.amount[n]);
          if (t)
            return "amount." + t;
        }
      }
      return null;
    }, o.fromObject = function(e) {
      if (e instanceof c.types.MsgSend)
        return e;
      let n = new c.types.MsgSend();
      if (e.fromAddress != null && (typeof e.fromAddress == "string" ? l.base64.decode(
        e.fromAddress,
        n.fromAddress = l.newBuffer(l.base64.length(e.fromAddress)),
        0
      ) : e.fromAddress.length && (n.fromAddress = e.fromAddress)), e.toAddress != null && (typeof e.toAddress == "string" ? l.base64.decode(
        e.toAddress,
        n.toAddress = l.newBuffer(l.base64.length(e.toAddress)),
        0
      ) : e.toAddress.length && (n.toAddress = e.toAddress)), e.amount) {
        if (!Array.isArray(e.amount))
          throw TypeError(".types.MsgSend.amount: array expected");
        n.amount = [];
        for (let t = 0; t < e.amount.length; ++t) {
          if (typeof e.amount[t] != "object")
            throw TypeError(".types.MsgSend.amount: object expected");
          n.amount[t] = c.cosmos.base.v1beta1.Coin.fromObject(e.amount[t]);
        }
      }
      return n;
    }, o.toObject = function(e, n) {
      var t;
      n || (n = {});
      let r = {};
      if ((n.arrays || n.defaults) && (r.amount = []), n.defaults && (n.bytes === String ? r.fromAddress = "" : (r.fromAddress = [], n.bytes !== Array && (r.fromAddress = l.newBuffer(r.fromAddress))), n.bytes === String ? r.toAddress = "" : (r.toAddress = [], n.bytes !== Array && (r.toAddress = l.newBuffer(r.toAddress)))), e.fromAddress != null && e.hasOwnProperty("fromAddress") && (r.fromAddress = n.bytes === String ? l.base64.encode(e.fromAddress, 0, e.fromAddress.length) : n.bytes === Array ? Array.prototype.slice.call(e.fromAddress) : e.fromAddress), e.toAddress != null && e.hasOwnProperty("toAddress") && (r.toAddress = n.bytes === String ? l.base64.encode(e.toAddress, 0, e.toAddress.length) : n.bytes === Array ? Array.prototype.slice.call(e.toAddress) : e.toAddress), (t = e.amount) != null && t.length) {
        r.amount = [];
        for (let s = 0; s < e.amount.length; ++s)
          r.amount[s] = c.cosmos.base.v1beta1.Coin.toObject(e.amount[s], n);
      }
      return r;
    }, o.prototype.toJSON = function() {
      return this.constructor.toObject(this, p.util.toJSONOptions);
    }, o;
  }(), i;
}();
c.cosmos = function() {
  let i = {};
  return i.base = function() {
    let o = {};
    return o.v1beta1 = function() {
      let e = {};
      return e.Coin = function() {
        function n(t) {
          if (t)
            for (let r = Object.keys(t), s = 0; s < r.length; ++s)
              t[r[s]] != null && (this[r[s]] = t[r[s]]);
        }
        return n.prototype.denom = "", n.prototype.amount = "", n.create = function(t) {
          return new n(t);
        }, n.encode = function(t, r) {
          return r || (r = m.create()), t.denom != null && Object.hasOwnProperty.call(t, "denom") && r.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(t.denom), t.amount != null && Object.hasOwnProperty.call(t, "amount") && r.uint32(
            /* id 2, wireType 2 =*/
            18
          ).string(t.amount), r;
        }, n.encodeDelimited = function(t, r) {
          return this.encode(t, r).ldelim();
        }, n.decode = function(t, r) {
          t instanceof a || (t = a.create(t));
          let s = r === void 0 ? t.len : t.pos + r, u = new c.cosmos.base.v1beta1.Coin();
          for (; t.pos < s; ) {
            let f = t.uint32();
            switch (f >>> 3) {
              case 1:
                u.denom = t.string();
                break;
              case 2:
                u.amount = t.string();
                break;
              default:
                t.skipType(f & 7);
                break;
            }
          }
          return u;
        }, n.decodeDelimited = function(t) {
          return t instanceof a || (t = new a(t)), this.decode(t, t.uint32());
        }, n.verify = function(t) {
          return typeof t != "object" || t === null ? "object expected" : t.denom != null && t.hasOwnProperty("denom") && !l.isString(t.denom) ? "denom: string expected" : t.amount != null && t.hasOwnProperty("amount") && !l.isString(t.amount) ? "amount: string expected" : null;
        }, n.fromObject = function(t) {
          if (t instanceof c.cosmos.base.v1beta1.Coin)
            return t;
          let r = new c.cosmos.base.v1beta1.Coin();
          return t.denom != null && (r.denom = String(t.denom)), t.amount != null && (r.amount = String(t.amount)), r;
        }, n.toObject = function(t, r) {
          r || (r = {});
          let s = {};
          return r.defaults && (s.denom = "", s.amount = ""), t.denom != null && t.hasOwnProperty("denom") && (s.denom = t.denom), t.amount != null && t.hasOwnProperty("amount") && (s.amount = t.amount), s;
        }, n.prototype.toJSON = function() {
          return this.constructor.toObject(this, p.util.toJSONOptions);
        }, n;
      }(), e.DecCoin = function() {
        function n(t) {
          if (t)
            for (let r = Object.keys(t), s = 0; s < r.length; ++s)
              t[r[s]] != null && (this[r[s]] = t[r[s]]);
        }
        return n.prototype.denom = "", n.prototype.amount = "", n.create = function(t) {
          return new n(t);
        }, n.encode = function(t, r) {
          return r || (r = m.create()), t.denom != null && Object.hasOwnProperty.call(t, "denom") && r.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(t.denom), t.amount != null && Object.hasOwnProperty.call(t, "amount") && r.uint32(
            /* id 2, wireType 2 =*/
            18
          ).string(t.amount), r;
        }, n.encodeDelimited = function(t, r) {
          return this.encode(t, r).ldelim();
        }, n.decode = function(t, r) {
          t instanceof a || (t = a.create(t));
          let s = r === void 0 ? t.len : t.pos + r, u = new c.cosmos.base.v1beta1.DecCoin();
          for (; t.pos < s; ) {
            let f = t.uint32();
            switch (f >>> 3) {
              case 1:
                u.denom = t.string();
                break;
              case 2:
                u.amount = t.string();
                break;
              default:
                t.skipType(f & 7);
                break;
            }
          }
          return u;
        }, n.decodeDelimited = function(t) {
          return t instanceof a || (t = new a(t)), this.decode(t, t.uint32());
        }, n.verify = function(t) {
          return typeof t != "object" || t === null ? "object expected" : t.denom != null && t.hasOwnProperty("denom") && !l.isString(t.denom) ? "denom: string expected" : t.amount != null && t.hasOwnProperty("amount") && !l.isString(t.amount) ? "amount: string expected" : null;
        }, n.fromObject = function(t) {
          if (t instanceof c.cosmos.base.v1beta1.DecCoin)
            return t;
          let r = new c.cosmos.base.v1beta1.DecCoin();
          return t.denom != null && (r.denom = String(t.denom)), t.amount != null && (r.amount = String(t.amount)), r;
        }, n.toObject = function(t, r) {
          r || (r = {});
          let s = {};
          return r.defaults && (s.denom = "", s.amount = ""), t.denom != null && t.hasOwnProperty("denom") && (s.denom = t.denom), t.amount != null && t.hasOwnProperty("amount") && (s.amount = t.amount), s;
        }, n.prototype.toJSON = function() {
          return this.constructor.toObject(this, p.util.toJSONOptions);
        }, n;
      }(), e.IntProto = function() {
        function n(t) {
          if (t)
            for (let r = Object.keys(t), s = 0; s < r.length; ++s)
              t[r[s]] != null && (this[r[s]] = t[r[s]]);
        }
        return n.prototype.int = "", n.create = function(t) {
          return new n(t);
        }, n.encode = function(t, r) {
          return r || (r = m.create()), t.int != null && Object.hasOwnProperty.call(t, "int") && r.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(t.int), r;
        }, n.encodeDelimited = function(t, r) {
          return this.encode(t, r).ldelim();
        }, n.decode = function(t, r) {
          t instanceof a || (t = a.create(t));
          let s = r === void 0 ? t.len : t.pos + r, u = new c.cosmos.base.v1beta1.IntProto();
          for (; t.pos < s; ) {
            let f = t.uint32();
            switch (f >>> 3) {
              case 1:
                u.int = t.string();
                break;
              default:
                t.skipType(f & 7);
                break;
            }
          }
          return u;
        }, n.decodeDelimited = function(t) {
          return t instanceof a || (t = new a(t)), this.decode(t, t.uint32());
        }, n.verify = function(t) {
          return typeof t != "object" || t === null ? "object expected" : t.int != null && t.hasOwnProperty("int") && !l.isString(t.int) ? "int: string expected" : null;
        }, n.fromObject = function(t) {
          if (t instanceof c.cosmos.base.v1beta1.IntProto)
            return t;
          let r = new c.cosmos.base.v1beta1.IntProto();
          return t.int != null && (r.int = String(t.int)), r;
        }, n.toObject = function(t, r) {
          r || (r = {});
          let s = {};
          return r.defaults && (s.int = ""), t.int != null && t.hasOwnProperty("int") && (s.int = t.int), s;
        }, n.prototype.toJSON = function() {
          return this.constructor.toObject(this, p.util.toJSONOptions);
        }, n;
      }(), e.DecProto = function() {
        function n(t) {
          if (t)
            for (let r = Object.keys(t), s = 0; s < r.length; ++s)
              t[r[s]] != null && (this[r[s]] = t[r[s]]);
        }
        return n.prototype.dec = "", n.create = function(t) {
          return new n(t);
        }, n.encode = function(t, r) {
          return r || (r = m.create()), t.dec != null && Object.hasOwnProperty.call(t, "dec") && r.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(t.dec), r;
        }, n.encodeDelimited = function(t, r) {
          return this.encode(t, r).ldelim();
        }, n.decode = function(t, r) {
          t instanceof a || (t = a.create(t));
          let s = r === void 0 ? t.len : t.pos + r, u = new c.cosmos.base.v1beta1.DecProto();
          for (; t.pos < s; ) {
            let f = t.uint32();
            switch (f >>> 3) {
              case 1:
                u.dec = t.string();
                break;
              default:
                t.skipType(f & 7);
                break;
            }
          }
          return u;
        }, n.decodeDelimited = function(t) {
          return t instanceof a || (t = new a(t)), this.decode(t, t.uint32());
        }, n.verify = function(t) {
          return typeof t != "object" || t === null ? "object expected" : t.dec != null && t.hasOwnProperty("dec") && !l.isString(t.dec) ? "dec: string expected" : null;
        }, n.fromObject = function(t) {
          if (t instanceof c.cosmos.base.v1beta1.DecProto)
            return t;
          let r = new c.cosmos.base.v1beta1.DecProto();
          return t.dec != null && (r.dec = String(t.dec)), r;
        }, n.toObject = function(t, r) {
          r || (r = {});
          let s = {};
          return r.defaults && (s.dec = ""), t.dec != null && t.hasOwnProperty("dec") && (s.dec = t.dec), s;
        }, n.prototype.toJSON = function() {
          return this.constructor.toObject(this, p.util.toJSONOptions);
        }, n;
      }(), e;
    }(), o;
  }(), i;
}();
export {
  c as default
};
