// osmojs@0.37.0/main/codegen/osmosis/lockup/tx.rpc.msg.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MsgClientImpl = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));

var _tx = require("./tx");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MsgClientImpl = /*#__PURE__*/function () {
  function MsgClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, MsgClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.lockTokens = this.lockTokens.bind(this);
    this.beginUnlockingAll = this.beginUnlockingAll.bind(this);
    this.beginUnlocking = this.beginUnlocking.bind(this);
    this.extendLockup = this.extendLockup.bind(this);
  }

  (0, _createClass2["default"])(MsgClientImpl, [{
    key: "lockTokens",
    value: function lockTokens(request) {
      var data = _tx.MsgLockTokens.encode(request).finish();

      var promise = this.rpc.request("osmosis.lockup.Msg", "LockTokens", data);
      return promise.then(function (data) {
        return _tx.MsgLockTokensResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "beginUnlockingAll",
    value: function beginUnlockingAll(request) {
      var data = _tx.MsgBeginUnlockingAll.encode(request).finish();

      var promise = this.rpc.request("osmosis.lockup.Msg", "BeginUnlockingAll", data);
      return promise.then(function (data) {
        return _tx.MsgBeginUnlockingAllResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "beginUnlocking",
    value: function beginUnlocking(request) {
      var data = _tx.MsgBeginUnlocking.encode(request).finish();

      var promise = this.rpc.request("osmosis.lockup.Msg", "BeginUnlocking", data);
      return promise.then(function (data) {
        return _tx.MsgBeginUnlockingResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "extendLockup",
    value: function extendLockup(request) {
      var data = _tx.MsgExtendLockup.encode(request).finish();

      var promise = this.rpc.request("osmosis.lockup.Msg", "ExtendLockup", data);
      return promise.then(function (data) {
        return _tx.MsgExtendLockupResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return MsgClientImpl;
}();

exports.MsgClientImpl = MsgClientImpl;
