// @shapeshiftoss/caip@8.15.0/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALID_ASSET_NAMESPACE = exports.VALID_CHAIN_IDS = exports.ASSET_REFERENCE = exports.ASSET_NAMESPACE = exports.CHAIN_REFERENCE = exports.CHAIN_NAMESPACE = exports.secretChainId = exports.terraChainId = exports.kavaChainId = exports.binanceChainId = exports.thorchainChainId = exports.osmosisChainId = exports.cosmosChainId = exports.bscChainId = exports.optimismChainId = exports.avalancheChainId = exports.ethChainId = exports.ltcChainId = exports.dogeChainId = exports.bchChainId = exports.btcChainId = exports.secretAssetId = exports.terraAssetId = exports.kavaAssetId = exports.binanceAssetId = exports.thorchainAssetId = exports.osmosisAssetId = exports.cosmosAssetId = exports.foxyAssetId = exports.foxAssetId = exports.bscAssetId = exports.optimismAssetId = exports.avalancheAssetId = exports.ethAssetId = exports.ltcAssetId = exports.dogeAssetId = exports.bchAssetId = exports.btcAssetId = void 0;
exports.btcAssetId = 'bip122:000000000019d6689c085ae165831e93/slip44:0';
exports.bchAssetId = 'bip122:000000000000000000651ef99cb9fcbe/slip44:145';
exports.dogeAssetId = 'bip122:00000000001a91e3dace36e2be3bf030/slip44:3';
exports.ltcAssetId = 'bip122:12a765e31ffd4059bada1e25190f6e98/slip44:2';
exports.ethAssetId = 'eip155:1/slip44:60';
exports.avalancheAssetId = 'eip155:43114/slip44:60';
exports.optimismAssetId = 'eip155:10/slip44:60';
exports.bscAssetId = 'eip155:56/slip44:60';
exports.foxAssetId = 'eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d';
exports.foxyAssetId = 'eip155:1/erc20:0xdc49108ce5c57bc3408c3a5e95f3d864ec386ed3';
exports.cosmosAssetId = 'cosmos:cosmoshub-4/slip44:118';
exports.osmosisAssetId = 'cosmos:osmosis-1/slip44:118';
exports.thorchainAssetId = 'cosmos:thorchain-mainnet-v1/slip44:931';
exports.binanceAssetId = 'cosmos:binance-chain-tigris/slip44:714';
exports.kavaAssetId = 'cosmos:kava_2222-10/slip44:459';
exports.terraAssetId = 'cosmos:phoenix-1/slip44:330';
exports.secretAssetId = 'cosmos:secret-4/slip44:529';
exports.btcChainId = 'bip122:000000000019d6689c085ae165831e93';
exports.bchChainId = 'bip122:000000000000000000651ef99cb9fcbe';
exports.dogeChainId = 'bip122:00000000001a91e3dace36e2be3bf030';
exports.ltcChainId = 'bip122:12a765e31ffd4059bada1e25190f6e98';
exports.ethChainId = 'eip155:1';
exports.avalancheChainId = 'eip155:43114';
exports.optimismChainId = 'eip155:10';
exports.bscChainId = 'eip155:56';
exports.cosmosChainId = 'cosmos:cosmoshub-4';
exports.osmosisChainId = 'cosmos:osmosis-1';
exports.thorchainChainId = 'cosmos:thorchain-mainnet-v1';
exports.binanceChainId = 'cosmos:binance-chain-tigris';
exports.kavaChainId = 'cosmos:kava_2222-10';
exports.terraChainId = 'cosmos:phoenix-1';
exports.secretChainId = 'cosmos:secret-4';
exports.CHAIN_NAMESPACE = {
    Evm: 'eip155',
    Utxo: 'bip122',
    CosmosSdk: 'cosmos',
};
exports.CHAIN_REFERENCE = {
    EthereumMainnet: '1',
    EthereumRopsten: '3',
    EthereumRinkeby: '4',
    // EthereumKovan: '42', // currently unsupported by ShapeShift
    // https://github.com/bitcoin/bips/blob/master/bip-0122.mediawiki#definition-of-chain-id
    // chainId uses max length of 32 chars of the genesis block
    BitcoinMainnet: '000000000019d6689c085ae165831e93',
    BitcoinTestnet: '000000000933ea01ad0ee984209779ba',
    BitcoinCashMainnet: '000000000000000000651ef99cb9fcbe',
    DogecoinMainnet: '00000000001a91e3dace36e2be3bf030',
    LitecoinMainnet: '12a765e31ffd4059bada1e25190f6e98',
    CosmosHubMainnet: 'cosmoshub-4',
    CosmosHubVega: 'vega-testnet',
    OsmosisMainnet: 'osmosis-1',
    OsmosisTestnet: 'osmo-testnet-1',
    ThorchainMainnet: 'thorchain-mainnet-v1',
    AvalancheCChain: '43114',
    BinanceMainnet: 'binance-chain-tigris',
    BinanceTestnet: 'binance-chain-ganges',
    KavaMainnet: 'kava_2222-10',
    KavaTestnet: 'kava_2221-16000',
    TerraMainnet: 'phoenix-1',
    TerraTestnet: 'pisco-1',
    SecretMainnet: 'secret-4',
    SecretTestnet: 'pulsar-2',
    OptimismMainnet: '10',
    BnbSmartChainMainnet: '56', // https://docs.bnbchain.org/docs/wallet_api/#chain-ids
};
exports.ASSET_NAMESPACE = {
    cw20: 'cw20',
    cw721: 'cw721',
    erc20: 'erc20',
    erc721: 'erc721',
    bep20: 'bep20',
    bep721: 'bep721',
    slip44: 'slip44',
    native: 'native',
    ibc: 'ibc',
};
exports.ASSET_REFERENCE = {
    Bitcoin: '0',
    Litecoin: '2',
    Dogecoin: '3',
    Ethereum: '60',
    Cosmos: '118',
    Osmosis: '118',
    Thorchain: '931',
    BitcoinCash: '145',
    AvalancheC: '60',
    Binance: '714',
    Kava: '459',
    Terra: '330',
    Secret: '529',
    Optimism: '60',
    BnbSmartChain: '60', // evm chain which uses ethereum derivation path as common practice
};
exports.VALID_CHAIN_IDS = Object.freeze({
    [exports.CHAIN_NAMESPACE.Utxo]: [
        exports.CHAIN_REFERENCE.BitcoinMainnet,
        exports.CHAIN_REFERENCE.BitcoinTestnet,
        exports.CHAIN_REFERENCE.BitcoinCashMainnet,
        exports.CHAIN_REFERENCE.DogecoinMainnet,
        exports.CHAIN_REFERENCE.LitecoinMainnet,
    ],
    [exports.CHAIN_NAMESPACE.Evm]: [
        exports.CHAIN_REFERENCE.EthereumMainnet,
        exports.CHAIN_REFERENCE.EthereumRopsten,
        exports.CHAIN_REFERENCE.EthereumRinkeby,
        exports.CHAIN_REFERENCE.AvalancheCChain,
        exports.CHAIN_REFERENCE.OptimismMainnet,
        exports.CHAIN_REFERENCE.BnbSmartChainMainnet,
    ],
    [exports.CHAIN_NAMESPACE.CosmosSdk]: [
        exports.CHAIN_REFERENCE.CosmosHubMainnet,
        exports.CHAIN_REFERENCE.CosmosHubVega,
        exports.CHAIN_REFERENCE.OsmosisMainnet,
        exports.CHAIN_REFERENCE.OsmosisTestnet,
        exports.CHAIN_REFERENCE.ThorchainMainnet,
        exports.CHAIN_REFERENCE.BinanceMainnet,
        exports.CHAIN_REFERENCE.KavaMainnet,
        exports.CHAIN_REFERENCE.TerraMainnet,
        exports.CHAIN_REFERENCE.SecretMainnet,
    ],
});
exports.VALID_ASSET_NAMESPACE = Object.freeze({
    [exports.CHAIN_NAMESPACE.Utxo]: [exports.ASSET_NAMESPACE.slip44],
    [exports.CHAIN_NAMESPACE.Evm]: [
        exports.ASSET_NAMESPACE.slip44,
        exports.ASSET_NAMESPACE.erc20,
        exports.ASSET_NAMESPACE.erc721,
        exports.ASSET_NAMESPACE.bep20,
        exports.ASSET_NAMESPACE.bep721,
    ],
    [exports.CHAIN_NAMESPACE.CosmosSdk]: [
        exports.ASSET_NAMESPACE.cw20,
        exports.ASSET_NAMESPACE.cw721,
        exports.ASSET_NAMESPACE.ibc,
        exports.ASSET_NAMESPACE.native,
        exports.ASSET_NAMESPACE.slip44,
    ],
});
