// @shapeshiftoss/caip@8.15.0/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.thorchainAssetMap = exports.osmosisAssetMap = exports.cosmosAssetMap = exports.litecoinAssetMap = exports.dogecoinAssetMap = exports.bitcoinCashAssetMap = exports.bitcoinAssetMap = exports.generateAssetIdFromOsmosisDenom = exports.isValidChainPartsPair = exports.accountIdToSpecifier = exports.accountIdToChainId = exports.parseAssetIdRegExp = void 0;
const accountId_1 = require("./accountId/accountId");
const assetId_1 = require("./assetId/assetId");
const constants = __importStar(require("./constants"));
// https://regex101.com/r/xqVzV2/1
exports.parseAssetIdRegExp = /([-a-z\d]{3,8}):([-a-zA-Z\d]{1,32})\/([-a-z\d]{3,8}):([-a-zA-Z/\d]+)/;
const accountIdToChainId = (accountId) => (0, accountId_1.fromAccountId)(accountId).chainId;
exports.accountIdToChainId = accountIdToChainId;
const accountIdToSpecifier = (accountId) => (0, accountId_1.fromAccountId)(accountId).account;
exports.accountIdToSpecifier = accountIdToSpecifier;
const isValidChainPartsPair = (chainNamespace, chainReference) => { var _a; return ((_a = constants.VALID_CHAIN_IDS[chainNamespace]) === null || _a === void 0 ? void 0 : _a.includes(chainReference)) || false; };
exports.isValidChainPartsPair = isValidChainPartsPair;
const generateAssetIdFromOsmosisDenom = (denom) => {
    if (denom.startsWith('u') && denom !== 'uosmo') {
        return (0, assetId_1.toAssetId)({
            assetNamespace: constants.ASSET_NAMESPACE.native,
            assetReference: denom,
            chainId: constants.osmosisChainId,
        });
    }
    if (denom.startsWith('ibc')) {
        return (0, assetId_1.toAssetId)({
            assetNamespace: constants.ASSET_NAMESPACE.ibc,
            assetReference: denom.split('/')[1],
            chainId: constants.osmosisChainId,
        });
    }
    if (denom.startsWith('gamm')) {
        return (0, assetId_1.toAssetId)({
            assetNamespace: constants.ASSET_NAMESPACE.ibc,
            assetReference: denom,
            chainId: constants.osmosisChainId,
        });
    }
    return (0, assetId_1.toAssetId)({
        assetNamespace: constants.ASSET_NAMESPACE.slip44,
        assetReference: constants.ASSET_REFERENCE.Osmosis,
        chainId: constants.osmosisChainId,
    });
};
exports.generateAssetIdFromOsmosisDenom = generateAssetIdFromOsmosisDenom;
exports.bitcoinAssetMap = { [constants.btcAssetId]: 'bitcoin' };
exports.bitcoinCashAssetMap = { [constants.bchAssetId]: 'bitcoin-cash' };
exports.dogecoinAssetMap = { [constants.dogeAssetId]: 'dogecoin' };
exports.litecoinAssetMap = { [constants.ltcAssetId]: 'litecoin' };
exports.cosmosAssetMap = { [constants.cosmosAssetId]: 'cosmos' };
exports.osmosisAssetMap = { [constants.osmosisAssetId]: 'osmosis' };
exports.thorchainAssetMap = { [constants.thorchainAssetId]: 'thorchain' };
