// @shapeshiftoss/unchained-client@10.1.1/dist/generated/osmosis/models/InfoAllOf.js
"use strict";
// @ts-nocheck
/* tslint:disable */
/* eslint-disable */
/**
 * Osmosis Unchained API
 * Provides access to osmosis chain data.
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoAllOfToJSON = exports.InfoAllOfFromJSONTyped = exports.InfoAllOfFromJSON = exports.instanceOfInfoAllOf = void 0;
/**
 * Check if a given object implements the InfoAllOf interface.
 */
function instanceOfInfoAllOf(value) {
    let isInstance = true;
    isInstance = isInstance && "apr" in value;
    return isInstance;
}
exports.instanceOfInfoAllOf = instanceOfInfoAllOf;
function InfoAllOfFromJSON(json) {
    return InfoAllOfFromJSONTyped(json, false);
}
exports.InfoAllOfFromJSON = InfoAllOfFromJSON;
function InfoAllOfFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'apr': json['apr'],
    };
}
exports.InfoAllOfFromJSONTyped = InfoAllOfFromJSONTyped;
function InfoAllOfToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'apr': value.apr,
    };
}
exports.InfoAllOfToJSON = InfoAllOfToJSON;
