// @shapeshiftoss/unchained-client@10.1.1/dist/generated/osmosis/models/Message.js
"use strict";
// @ts-nocheck
/* tslint:disable */
/* eslint-disable */
/**
 * Osmosis Unchained API
 * Provides access to osmosis chain data.
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageToJSON = exports.MessageFromJSONTyped = exports.MessageFromJSON = exports.instanceOfMessage = void 0;
const runtime_1 = require("../runtime");
const Value_1 = require("./Value");
/**
 * Check if a given object implements the Message interface.
 */
function instanceOfMessage(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfMessage = instanceOfMessage;
function MessageFromJSON(json) {
    return MessageFromJSONTyped(json, false);
}
exports.MessageFromJSON = MessageFromJSON;
function MessageFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'from': !(0, runtime_1.exists)(json, 'from') ? undefined : json['from'],
        'origin': !(0, runtime_1.exists)(json, 'origin') ? undefined : json['origin'],
        'to': !(0, runtime_1.exists)(json, 'to') ? undefined : json['to'],
        'type': json['type'],
        'value': !(0, runtime_1.exists)(json, 'value') ? undefined : (0, Value_1.ValueFromJSON)(json['value']),
    };
}
exports.MessageFromJSONTyped = MessageFromJSONTyped;
function MessageToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'from': value.from,
        'origin': value.origin,
        'to': value.to,
        'type': value.type,
        'value': (0, Value_1.ValueToJSON)(value.value),
    };
}
exports.MessageToJSON = MessageToJSON;
