// @pioneer-platform/okx@3.0.1/dist/index.es.js
import { getSignatureAssetFor as zr } from "@pioneer-platform/swapkit-entities";
import { Psbt as Yr } from "bitcoinjs-lib";
var v = /* @__PURE__ */ ((n) => (n.INVALID_INPUT_PARAMETERS = "1000", n.UNKNOWN_PROVIDERS = "1001", n.CANNOT_FIND_INBOUND_ADDRESS = "1002", n.NO_INBOUND_ADDRESSES = "1003", n.CHAIN_HALTED_OR_UNSUPPORTED = "1004", n.MISSING_INPUT_PARAMETER = "1005", n.INVALID_TYPE_GENERIC = "1100", n.INVALID_NUMBER_STRING = "1101", n.INVALID_NUMBER = "1102", n.INVALID_BOOLEAN = "1103", n.INVALID_OBJECT = "1104", n.INVALID_ARRAY = "1105", n.SELL_AMOUNT_MUST_BE_POSITIVE_INTEGER = "2000", n.SELL_BUY_ASSETS_ARE_THE_SAME = "2001", n.MISSING_SOURCE_ADDRESS_FOR_SYNTH = "2002", n.AFF_ADDRESS_AND_BPS_OR_NEITHER = "2003", n.AFF_ADDRESS_TOO_LONG = "2004", n.AFF_BPS_INTEGER_0_100 = "2005", n.SOURCE_ADDRESS_INVALID_FOR_SELL_CHAIN = "2006", n.DESTINATION_ADDRESS_INVALID_FOR_BUY_CHAIN = "2007", n.PREFERRED_PROFVIDER_NOT_SUPPORTED = "2008", n.DESTINATION_ADDRESS_SMART_CONTRACT = "2009", n.BUY_AMOUNT_MUST_BE_POSITIVE_INTEGER = "2010", n.SOURCE_ADDRESS_SMART_CONTRACT = "2011", n.INVALID_PROVIDER = "2100", n.MISSING_CROSS_CHAIN_PROVIDER = "2101", n.MISSING_AVAX_PROVIDER = "2102", n.MISSING_BSC_PROVIDER = "2103", n.MISSING_ETH_PROVIDER = "2104", n.INVALID_PROVIDER_FOR_SWAP_OUT = "2105", n.MISSING_ARB_PROVIDER = "2106", n.INVALID_CHAIN = "2200", n.INVALID_ASSET = "2201", n.INVALID_ASSET_IDENTIFIER = "2202", n.UNSUPPORTED_CHAIN = "2204", n.UNSUPPORTED_ASSET = "2203", n.UNSUPPORTED_ASSET_FOR_SWAPOUT = "2205", n.THORNODE_QUOTE_GENERIC_ERROR = "3000", n.NOT_ENOUGH_SYNTH_BALANCE = "3001", n.SYNTH_MINTING_CAP_REACHED = "3002", n.INVALID_QUOTE_MODE = "4000", n.NO_QUOTES = "4001", n.SERVICE_UNAVAILABLE_GENERIC = "5000", n.MISSING_GAS_DATA_GENERIC = "5100", n.MISSING_TOKEN_INFO_GENERIC = "5200", n.CANT_FIND_TOKEN_LIST = "5201", n.NO_PRICE = "5202", n.PRICE_IS_STALE = "5203", n.ADDRESS_NOT_WHITELISTED = "6000", n.ADDRESS_ALREADY_CLAIMED = "6001", n.TEMPORARY_ERROR = "9999", n))(v || {});
v.INVALID_INPUT_PARAMETERS + "", v.UNKNOWN_PROVIDERS + "", v.CANNOT_FIND_INBOUND_ADDRESS + "", v.NO_INBOUND_ADDRESSES + "", v.CHAIN_HALTED_OR_UNSUPPORTED + "", v.MISSING_INPUT_PARAMETER + "", v.INVALID_TYPE_GENERIC + "", v.INVALID_NUMBER_STRING + "", v.INVALID_NUMBER + "", v.INVALID_BOOLEAN + "", v.INVALID_OBJECT + "", v.INVALID_ARRAY + "", v.SELL_AMOUNT_MUST_BE_POSITIVE_INTEGER + "", v.SELL_BUY_ASSETS_ARE_THE_SAME + "", v.MISSING_SOURCE_ADDRESS_FOR_SYNTH + "", v.AFF_ADDRESS_AND_BPS_OR_NEITHER + "", v.AFF_ADDRESS_TOO_LONG + "", v.AFF_BPS_INTEGER_0_100 + "", v.SOURCE_ADDRESS_INVALID_FOR_SELL_CHAIN + "", v.DESTINATION_ADDRESS_INVALID_FOR_BUY_CHAIN + "", v.PREFERRED_PROFVIDER_NOT_SUPPORTED + "", v.DESTINATION_ADDRESS_SMART_CONTRACT + "", v.BUY_AMOUNT_MUST_BE_POSITIVE_INTEGER + "", v.INVALID_PROVIDER + "", v.MISSING_CROSS_CHAIN_PROVIDER + "", v.MISSING_AVAX_PROVIDER + "", v.MISSING_BSC_PROVIDER + "", v.MISSING_ETH_PROVIDER + "", v.MISSING_ARB_PROVIDER + "", v.INVALID_PROVIDER_FOR_SWAP_OUT + "", v.INVALID_CHAIN + "", v.INVALID_ASSET + "", v.UNSUPPORTED_CHAIN + "", v.UNSUPPORTED_ASSET + "", v.UNSUPPORTED_ASSET_FOR_SWAPOUT + "", v.THORNODE_QUOTE_GENERIC_ERROR + "", v.NOT_ENOUGH_SYNTH_BALANCE + "", v.SYNTH_MINTING_CAP_REACHED + "", v.INVALID_QUOTE_MODE + "", v.NO_QUOTES + "", v.SERVICE_UNAVAILABLE_GENERIC + "", v.MISSING_GAS_DATA_GENERIC + "", v.MISSING_TOKEN_INFO_GENERIC + "", v.CANT_FIND_TOKEN_LIST + "", v.NO_PRICE + "", v.PRICE_IS_STALE + "", v.ADDRESS_NOT_WHITELISTED + "", v.ADDRESS_ALREADY_CLAIMED + "";
var gt = /* @__PURE__ */ ((n) => (n.Arbitrum = "ARB", n.Avalanche = "AVAX", n.Binance = "BNB", n.BinanceSmartChain = "BSC", n.Bitcoin = "BTC", n.BitcoinCash = "BCH", n.Cosmos = "GAIA", n.Dogecoin = "DOGE", n.Ethereum = "ETH", n.Litecoin = "LTC", n.Optimism = "OP", n.Polygon = "MATIC", n.THORChain = "THOR", n))(gt || {}), It = /* @__PURE__ */ ((n) => (n.Arbitrum = "42161", n.ArbitrumHex = "0xa4b1", n.Avalanche = "43114", n.AvalancheHex = "0xa86a", n.Binance = "Binance-Chain-Tigris", n.BinanceHex = "", n.BinanceSmartChain = "56", n.BinanceSmartChainHex = "0x38", n.Bitcoin = "bitcoin", n.BitcoinHex = "", n.BitcoinCash = "bitcoincash", n.BitcoinCashHex = "", n.Cosmos = "cosmoshub-4", n.CosmosHex = "", n.Dogecoin = "dogecoin", n.DogecoinHex = "", n.Ethereum = "1", n.EthereumHex = "0x1", n.Litecoin = "litecoin", n.LitecoinHex = "", n.Optimism = "10", n.OptimismHex = "0xa", n.Polygon = "137", n.PolygonHex = "0x89", n.THORChain = "thorchain-mainnet-v1", n.THORChainHex = "", n.THORChainStagenet = "thorchain-stagenet-v2", n))(It || {}), Wr = /* @__PURE__ */ ((n) => (n.Arbitrum = "https://arb1.arbitrum.io/rpc", n.Avalanche = "https://node-router.thorswap.net/avalanche-c", n.Binance = "", n.BinanceSmartChain = "https://bsc-dataseed.binance.org", n.Bitcoin = "https://node-router.thorswap.net/bitcoin", n.BitcoinCash = "https://node-router.thorswap.net/bitcoin-cash", n.Cosmos = "https://node-router.thorswap.net/cosmos/rpc", n.Dogecoin = "https://node-router.thorswap.net/dogecoin", n.Ethereum = "https://node-router.thorswap.net/ethereum", n.Litecoin = "https://node-router.thorswap.net/litecoin", n.Optimism = "https://mainnet.optimism.io", n.Polygon = "https://polygon-rpc.com", n.THORChain = "https://rpc.thorswap.net", n.THORChainStagenet = "https://stagenet-rpc.ninerealms.com", n))(Wr || {});
const Ct = Object.values(gt), Xr = Object.keys(gt), lr = Ct.reduce(
  (n, d) => {
    const p = Xr.find((c) => gt[c] === d);
    return p && (n[d] = p), n;
  },
  {}
);
Ct.reduce(
  (n, d) => (n[d] = It[lr[d]], n),
  {}
);
Ct.reduce(
  (n, d) => (n[d] = Wr[lr[d]], n),
  {}
);
Ct.reduce(
  (n, d) => (n[d] = It[`${lr[d]}Hex`], n),
  {}
);
var Ut = /* @__PURE__ */ ((n) => (n.Average = "average", n.Fast = "fast", n.Fastest = "fastest", n))(Ut || {}), Kr = /* @__PURE__ */ ((n) => (n.KEYSTORE = "KEYSTORE", n.XDEFI = "XDEFI", n.METAMASK = "METAMASK", n.COINBASE_WEB = "COINBASE_WEB", n.TREZOR = "TREZOR", n.KEEPKEY = "KEEPKEY", n.TRUSTWALLET_WEB = "TRUSTWALLET_WEB", n.LEDGER = "LEDGER", n.KEPLR = "KEPLR", n.OKX = "OKX", n.BRAVE = "BRAVE", n.WALLETCONNECT = "WALLETCONNECT", n))(Kr || {}), Jr = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {};
function $r(n) {
  return n && n.__esModule && Object.prototype.hasOwnProperty.call(n, "default") ? n.default : n;
}
var ur = { exports: {} };
ur.exports;
(function(n) {
  (function(d, p) {
    function c(a, t) {
      if (!a)
        throw new Error(t || "Assertion failed");
    }
    function _(a, t) {
      a.super_ = t;
      var i = function() {
      };
      i.prototype = t.prototype, a.prototype = new i(), a.prototype.constructor = a;
    }
    function u(a, t, i) {
      if (u.isBN(a))
        return a;
      this.negative = 0, this.words = null, this.length = 0, this.red = null, a !== null && ((t === "le" || t === "be") && (i = t, t = 10), this._init(a || 0, t || 10, i || "be"));
    }
    typeof d == "object" ? d.exports = u : p.BN = u, u.BN = u, u.wordSize = 26;
    var I;
    try {
      typeof window < "u" && typeof window.Buffer < "u" ? I = window.Buffer : I = require("buffer").Buffer;
    } catch {
    }
    u.isBN = function(t) {
      return t instanceof u ? !0 : t !== null && typeof t == "object" && t.constructor.wordSize === u.wordSize && Array.isArray(t.words);
    }, u.max = function(t, i) {
      return t.cmp(i) > 0 ? t : i;
    }, u.min = function(t, i) {
      return t.cmp(i) < 0 ? t : i;
    }, u.prototype._init = function(t, i, e) {
      if (typeof t == "number")
        return this._initNumber(t, i, e);
      if (typeof t == "object")
        return this._initArray(t, i, e);
      i === "hex" && (i = 16), c(i === (i | 0) && i >= 2 && i <= 36), t = t.toString().replace(/\s+/g, "");
      var s = 0;
      t[0] === "-" && (s++, this.negative = 1), s < t.length && (i === 16 ? this._parseHex(t, s, e) : (this._parseBase(t, i, s), e === "le" && this._initArray(this.toArray(), i, e)));
    }, u.prototype._initNumber = function(t, i, e) {
      t < 0 && (this.negative = 1, t = -t), t < 67108864 ? (this.words = [t & 67108863], this.length = 1) : t < 4503599627370496 ? (this.words = [
        t & 67108863,
        t / 67108864 & 67108863
      ], this.length = 2) : (c(t < 9007199254740992), this.words = [
        t & 67108863,
        t / 67108864 & 67108863,
        1
      ], this.length = 3), e === "le" && this._initArray(this.toArray(), i, e);
    }, u.prototype._initArray = function(t, i, e) {
      if (c(typeof t.length == "number"), t.length <= 0)
        return this.words = [0], this.length = 1, this;
      this.length = Math.ceil(t.length / 3), this.words = new Array(this.length);
      for (var s = 0; s < this.length; s++)
        this.words[s] = 0;
      var f, l, m = 0;
      if (e === "be")
        for (s = t.length - 1, f = 0; s >= 0; s -= 3)
          l = t[s] | t[s - 1] << 8 | t[s - 2] << 16, this.words[f] |= l << m & 67108863, this.words[f + 1] = l >>> 26 - m & 67108863, m += 24, m >= 26 && (m -= 26, f++);
      else if (e === "le")
        for (s = 0, f = 0; s < t.length; s += 3)
          l = t[s] | t[s + 1] << 8 | t[s + 2] << 16, this.words[f] |= l << m & 67108863, this.words[f + 1] = l >>> 26 - m & 67108863, m += 24, m >= 26 && (m -= 26, f++);
      return this._strip();
    };
    function dt(a, t) {
      var i = a.charCodeAt(t);
      if (i >= 48 && i <= 57)
        return i - 48;
      if (i >= 65 && i <= 70)
        return i - 55;
      if (i >= 97 && i <= 102)
        return i - 87;
      c(!1, "Invalid character in " + a);
    }
    function A(a, t, i) {
      var e = dt(a, i);
      return i - 1 >= t && (e |= dt(a, i - 1) << 4), e;
    }
    u.prototype._parseHex = function(t, i, e) {
      this.length = Math.ceil((t.length - i) / 6), this.words = new Array(this.length);
      for (var s = 0; s < this.length; s++)
        this.words[s] = 0;
      var f = 0, l = 0, m;
      if (e === "be")
        for (s = t.length - 1; s >= i; s -= 2)
          m = A(t, i, s) << f, this.words[l] |= m & 67108863, f >= 18 ? (f -= 18, l += 1, this.words[l] |= m >>> 26) : f += 8;
      else {
        var o = t.length - i;
        for (s = o % 2 === 0 ? i + 1 : i; s < t.length; s += 2)
          m = A(t, i, s) << f, this.words[l] |= m & 67108863, f >= 18 ? (f -= 18, l += 1, this.words[l] |= m >>> 26) : f += 8;
      }
      this._strip();
    };
    function x(a, t, i, e) {
      for (var s = 0, f = 0, l = Math.min(a.length, i), m = t; m < l; m++) {
        var o = a.charCodeAt(m) - 48;
        s *= e, o >= 49 ? f = o - 49 + 10 : o >= 17 ? f = o - 17 + 10 : f = o, c(o >= 0 && f < e, "Invalid character"), s += f;
      }
      return s;
    }
    u.prototype._parseBase = function(t, i, e) {
      this.words = [0], this.length = 1;
      for (var s = 0, f = 1; f <= 67108863; f *= i)
        s++;
      s--, f = f / i | 0;
      for (var l = t.length - e, m = l % s, o = Math.min(l, l - m) + e, r = 0, h = e; h < o; h += s)
        r = x(t, h, h + s, i), this.imuln(f), this.words[0] + r < 67108864 ? this.words[0] += r : this._iaddn(r);
      if (m !== 0) {
        var M = 1;
        for (r = x(t, h, t.length, i), h = 0; h < m; h++)
          M *= i;
        this.imuln(M), this.words[0] + r < 67108864 ? this.words[0] += r : this._iaddn(r);
      }
      this._strip();
    }, u.prototype.copy = function(t) {
      t.words = new Array(this.length);
      for (var i = 0; i < this.length; i++)
        t.words[i] = this.words[i];
      t.length = this.length, t.negative = this.negative, t.red = this.red;
    };
    function mt(a, t) {
      a.words = t.words, a.length = t.length, a.negative = t.negative, a.red = t.red;
    }
    if (u.prototype._move = function(t) {
      mt(t, this);
    }, u.prototype.clone = function() {
      var t = new u(null);
      return this.copy(t), t;
    }, u.prototype._expand = function(t) {
      for (; this.length < t; )
        this.words[this.length++] = 0;
      return this;
    }, u.prototype._strip = function() {
      for (; this.length > 1 && this.words[this.length - 1] === 0; )
        this.length--;
      return this._normSign();
    }, u.prototype._normSign = function() {
      return this.length === 1 && this.words[0] === 0 && (this.negative = 0), this;
    }, typeof Symbol < "u" && typeof Symbol.for == "function")
      try {
        u.prototype[Symbol.for("nodejs.util.inspect.custom")] = ut;
      } catch {
        u.prototype.inspect = ut;
      }
    else
      u.prototype.inspect = ut;
    function ut() {
      return (this.red ? "<BN-R: " : "<BN: ") + this.toString(16) + ">";
    }
    var Mt = [
      "",
      "0",
      "00",
      "000",
      "0000",
      "00000",
      "000000",
      "0000000",
      "00000000",
      "000000000",
      "0000000000",
      "00000000000",
      "000000000000",
      "0000000000000",
      "00000000000000",
      "000000000000000",
      "0000000000000000",
      "00000000000000000",
      "000000000000000000",
      "0000000000000000000",
      "00000000000000000000",
      "000000000000000000000",
      "0000000000000000000000",
      "00000000000000000000000",
      "000000000000000000000000",
      "0000000000000000000000000"
    ], Ot = [
      0,
      0,
      25,
      16,
      12,
      11,
      10,
      9,
      8,
      8,
      7,
      7,
      7,
      7,
      6,
      6,
      6,
      6,
      6,
      6,
      6,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5,
      5
    ], bt = [
      0,
      0,
      33554432,
      43046721,
      16777216,
      48828125,
      60466176,
      40353607,
      16777216,
      43046721,
      1e7,
      19487171,
      35831808,
      62748517,
      7529536,
      11390625,
      16777216,
      24137569,
      34012224,
      47045881,
      64e6,
      4084101,
      5153632,
      6436343,
      7962624,
      9765625,
      11881376,
      14348907,
      17210368,
      20511149,
      243e5,
      28629151,
      33554432,
      39135393,
      45435424,
      52521875,
      60466176
    ];
    u.prototype.toString = function(t, i) {
      t = t || 10, i = i | 0 || 1;
      var e;
      if (t === 16 || t === "hex") {
        e = "";
        for (var s = 0, f = 0, l = 0; l < this.length; l++) {
          var m = this.words[l], o = ((m << s | f) & 16777215).toString(16);
          f = m >>> 24 - s & 16777215, s += 2, s >= 26 && (s -= 26, l--), f !== 0 || l !== this.length - 1 ? e = Mt[6 - o.length] + o + e : e = o + e;
        }
        for (f !== 0 && (e = f.toString(16) + e); e.length % i !== 0; )
          e = "0" + e;
        return this.negative !== 0 && (e = "-" + e), e;
      }
      if (t === (t | 0) && t >= 2 && t <= 36) {
        var r = Ot[t], h = bt[t];
        e = "";
        var M = this.clone();
        for (M.negative = 0; !M.isZero(); ) {
          var g = M.modrn(h).toString(t);
          M = M.idivn(h), M.isZero() ? e = g + e : e = Mt[r - g.length] + g + e;
        }
        for (this.isZero() && (e = "0" + e); e.length % i !== 0; )
          e = "0" + e;
        return this.negative !== 0 && (e = "-" + e), e;
      }
      c(!1, "Base should be between 2 and 36");
    }, u.prototype.toNumber = function() {
      var t = this.words[0];
      return this.length === 2 ? t += this.words[1] * 67108864 : this.length === 3 && this.words[2] === 1 ? t += 4503599627370496 + this.words[1] * 67108864 : this.length > 2 && c(!1, "Number can only safely store up to 53 bits"), this.negative !== 0 ? -t : t;
    }, u.prototype.toJSON = function() {
      return this.toString(16, 2);
    }, I && (u.prototype.toBuffer = function(t, i) {
      return this.toArrayLike(I, t, i);
    }), u.prototype.toArray = function(t, i) {
      return this.toArrayLike(Array, t, i);
    };
    var St = function(t, i) {
      return t.allocUnsafe ? t.allocUnsafe(i) : new t(i);
    };
    u.prototype.toArrayLike = function(t, i, e) {
      this._strip();
      var s = this.byteLength(), f = e || Math.max(1, s);
      c(s <= f, "byte array longer than desired length"), c(f > 0, "Requested array length <= 0");
      var l = St(t, f), m = i === "le" ? "LE" : "BE";
      return this["_toArrayLike" + m](l, s), l;
    }, u.prototype._toArrayLikeLE = function(t, i) {
      for (var e = 0, s = 0, f = 0, l = 0; f < this.length; f++) {
        var m = this.words[f] << l | s;
        t[e++] = m & 255, e < t.length && (t[e++] = m >> 8 & 255), e < t.length && (t[e++] = m >> 16 & 255), l === 6 ? (e < t.length && (t[e++] = m >> 24 & 255), s = 0, l = 0) : (s = m >>> 24, l += 2);
      }
      if (e < t.length)
        for (t[e++] = s; e < t.length; )
          t[e++] = 0;
    }, u.prototype._toArrayLikeBE = function(t, i) {
      for (var e = t.length - 1, s = 0, f = 0, l = 0; f < this.length; f++) {
        var m = this.words[f] << l | s;
        t[e--] = m & 255, e >= 0 && (t[e--] = m >> 8 & 255), e >= 0 && (t[e--] = m >> 16 & 255), l === 6 ? (e >= 0 && (t[e--] = m >> 24 & 255), s = 0, l = 0) : (s = m >>> 24, l += 2);
      }
      if (e >= 0)
        for (t[e--] = s; e >= 0; )
          t[e--] = 0;
    }, Math.clz32 ? u.prototype._countBits = function(t) {
      return 32 - Math.clz32(t);
    } : u.prototype._countBits = function(t) {
      var i = t, e = 0;
      return i >= 4096 && (e += 13, i >>>= 13), i >= 64 && (e += 7, i >>>= 7), i >= 8 && (e += 4, i >>>= 4), i >= 2 && (e += 2, i >>>= 2), e + i;
    }, u.prototype._zeroBits = function(t) {
      if (t === 0)
        return 26;
      var i = t, e = 0;
      return i & 8191 || (e += 13, i >>>= 13), i & 127 || (e += 7, i >>>= 7), i & 15 || (e += 4, i >>>= 4), i & 3 || (e += 2, i >>>= 2), i & 1 || e++, e;
    }, u.prototype.bitLength = function() {
      var t = this.words[this.length - 1], i = this._countBits(t);
      return (this.length - 1) * 26 + i;
    };
    function Rt(a) {
      for (var t = new Array(a.bitLength()), i = 0; i < t.length; i++) {
        var e = i / 26 | 0, s = i % 26;
        t[i] = a.words[e] >>> s & 1;
      }
      return t;
    }
    u.prototype.zeroBits = function() {
      if (this.isZero())
        return 0;
      for (var t = 0, i = 0; i < this.length; i++) {
        var e = this._zeroBits(this.words[i]);
        if (t += e, e !== 26)
          break;
      }
      return t;
    }, u.prototype.byteLength = function() {
      return Math.ceil(this.bitLength() / 8);
    }, u.prototype.toTwos = function(t) {
      return this.negative !== 0 ? this.abs().inotn(t).iaddn(1) : this.clone();
    }, u.prototype.fromTwos = function(t) {
      return this.testn(t - 1) ? this.notn(t).iaddn(1).ineg() : this.clone();
    }, u.prototype.isNeg = function() {
      return this.negative !== 0;
    }, u.prototype.neg = function() {
      return this.clone().ineg();
    }, u.prototype.ineg = function() {
      return this.isZero() || (this.negative ^= 1), this;
    }, u.prototype.iuor = function(t) {
      for (; this.length < t.length; )
        this.words[this.length++] = 0;
      for (var i = 0; i < t.length; i++)
        this.words[i] = this.words[i] | t.words[i];
      return this._strip();
    }, u.prototype.ior = function(t) {
      return c((this.negative | t.negative) === 0), this.iuor(t);
    }, u.prototype.or = function(t) {
      return this.length > t.length ? this.clone().ior(t) : t.clone().ior(this);
    }, u.prototype.uor = function(t) {
      return this.length > t.length ? this.clone().iuor(t) : t.clone().iuor(this);
    }, u.prototype.iuand = function(t) {
      var i;
      this.length > t.length ? i = t : i = this;
      for (var e = 0; e < i.length; e++)
        this.words[e] = this.words[e] & t.words[e];
      return this.length = i.length, this._strip();
    }, u.prototype.iand = function(t) {
      return c((this.negative | t.negative) === 0), this.iuand(t);
    }, u.prototype.and = function(t) {
      return this.length > t.length ? this.clone().iand(t) : t.clone().iand(this);
    }, u.prototype.uand = function(t) {
      return this.length > t.length ? this.clone().iuand(t) : t.clone().iuand(this);
    }, u.prototype.iuxor = function(t) {
      var i, e;
      this.length > t.length ? (i = this, e = t) : (i = t, e = this);
      for (var s = 0; s < e.length; s++)
        this.words[s] = i.words[s] ^ e.words[s];
      if (this !== i)
        for (; s < i.length; s++)
          this.words[s] = i.words[s];
      return this.length = i.length, this._strip();
    }, u.prototype.ixor = function(t) {
      return c((this.negative | t.negative) === 0), this.iuxor(t);
    }, u.prototype.xor = function(t) {
      return this.length > t.length ? this.clone().ixor(t) : t.clone().ixor(this);
    }, u.prototype.uxor = function(t) {
      return this.length > t.length ? this.clone().iuxor(t) : t.clone().iuxor(this);
    }, u.prototype.inotn = function(t) {
      c(typeof t == "number" && t >= 0);
      var i = Math.ceil(t / 26) | 0, e = t % 26;
      this._expand(i), e > 0 && i--;
      for (var s = 0; s < i; s++)
        this.words[s] = ~this.words[s] & 67108863;
      return e > 0 && (this.words[s] = ~this.words[s] & 67108863 >> 26 - e), this._strip();
    }, u.prototype.notn = function(t) {
      return this.clone().inotn(t);
    }, u.prototype.setn = function(t, i) {
      c(typeof t == "number" && t >= 0);
      var e = t / 26 | 0, s = t % 26;
      return this._expand(e + 1), i ? this.words[e] = this.words[e] | 1 << s : this.words[e] = this.words[e] & ~(1 << s), this._strip();
    }, u.prototype.iadd = function(t) {
      var i;
      if (this.negative !== 0 && t.negative === 0)
        return this.negative = 0, i = this.isub(t), this.negative ^= 1, this._normSign();
      if (this.negative === 0 && t.negative !== 0)
        return t.negative = 0, i = this.isub(t), t.negative = 1, i._normSign();
      var e, s;
      this.length > t.length ? (e = this, s = t) : (e = t, s = this);
      for (var f = 0, l = 0; l < s.length; l++)
        i = (e.words[l] | 0) + (s.words[l] | 0) + f, this.words[l] = i & 67108863, f = i >>> 26;
      for (; f !== 0 && l < e.length; l++)
        i = (e.words[l] | 0) + f, this.words[l] = i & 67108863, f = i >>> 26;
      if (this.length = e.length, f !== 0)
        this.words[this.length] = f, this.length++;
      else if (e !== this)
        for (; l < e.length; l++)
          this.words[l] = e.words[l];
      return this;
    }, u.prototype.add = function(t) {
      var i;
      return t.negative !== 0 && this.negative === 0 ? (t.negative = 0, i = this.sub(t), t.negative ^= 1, i) : t.negative === 0 && this.negative !== 0 ? (this.negative = 0, i = t.sub(this), this.negative = 1, i) : this.length > t.length ? this.clone().iadd(t) : t.clone().iadd(this);
    }, u.prototype.isub = function(t) {
      if (t.negative !== 0) {
        t.negative = 0;
        var i = this.iadd(t);
        return t.negative = 1, i._normSign();
      } else if (this.negative !== 0)
        return this.negative = 0, this.iadd(t), this.negative = 1, this._normSign();
      var e = this.cmp(t);
      if (e === 0)
        return this.negative = 0, this.length = 1, this.words[0] = 0, this;
      var s, f;
      e > 0 ? (s = this, f = t) : (s = t, f = this);
      for (var l = 0, m = 0; m < f.length; m++)
        i = (s.words[m] | 0) - (f.words[m] | 0) + l, l = i >> 26, this.words[m] = i & 67108863;
      for (; l !== 0 && m < s.length; m++)
        i = (s.words[m] | 0) + l, l = i >> 26, this.words[m] = i & 67108863;
      if (l === 0 && m < s.length && s !== this)
        for (; m < s.length; m++)
          this.words[m] = s.words[m];
      return this.length = Math.max(this.length, m), s !== this && (this.negative = 1), this._strip();
    }, u.prototype.sub = function(t) {
      return this.clone().isub(t);
    };
    function dr(a, t, i) {
      i.negative = t.negative ^ a.negative;
      var e = a.length + t.length | 0;
      i.length = e, e = e - 1 | 0;
      var s = a.words[0] | 0, f = t.words[0] | 0, l = s * f, m = l & 67108863, o = l / 67108864 | 0;
      i.words[0] = m;
      for (var r = 1; r < e; r++) {
        for (var h = o >>> 26, M = o & 67108863, g = Math.min(r, t.length - 1), w = Math.max(0, r - a.length + 1); w <= g; w++) {
          var ct = r - w | 0;
          s = a.words[ct] | 0, f = t.words[w] | 0, l = s * f + M, h += l / 67108864 | 0, M = l & 67108863;
        }
        i.words[r] = M | 0, o = h | 0;
      }
      return o !== 0 ? i.words[r] = o | 0 : i.length--, i._strip();
    }
    var mr = function(t, i, e) {
      var s = t.words, f = i.words, l = e.words, m = 0, o, r, h, M = s[0] | 0, g = M & 8191, w = M >>> 13, ct = s[1] | 0, y = ct & 8191, S = ct >>> 13, Mr = s[2] | 0, T = Mr & 8191, O = Mr >>> 13, wr = s[3] | 0, b = wr & 8191, D = wr >>> 13, _r = s[4] | 0, U = _r & 8191, P = _r >>> 13, Er = s[5] | 0, C = Er & 8191, B = Er >>> 13, Nr = s[6] | 0, L = Nr & 8191, F = Nr >>> 13, yr = s[7] | 0, k = yr & 8191, G = yr >>> 13, Sr = s[8] | 0, H = Sr & 8191, V = Sr >>> 13, Ar = s[9] | 0, q = Ar & 8191, W = Ar >>> 13, Ir = f[0] | 0, K = Ir & 8191, Z = Ir >>> 13, Rr = f[1] | 0, z = Rr & 8191, Y = Rr >>> 13, xr = f[2] | 0, X = xr & 8191, J = xr >>> 13, Tr = f[3] | 0, $ = Tr & 8191, Q = Tr >>> 13, Or = f[4] | 0, j = Or & 8191, tt = Or >>> 13, br = f[5] | 0, rt = br & 8191, it = br >>> 13, Dr = f[6] | 0, et = Dr & 8191, nt = Dr >>> 13, Ur = f[7] | 0, ot = Ur & 8191, st = Ur >>> 13, Pr = f[8] | 0, ht = Pr & 8191, ft = Pr >>> 13, Cr = f[9] | 0, at = Cr & 8191, lt = Cr >>> 13;
      e.negative = t.negative ^ i.negative, e.length = 19, o = Math.imul(g, K), r = Math.imul(g, Z), r = r + Math.imul(w, K) | 0, h = Math.imul(w, Z);
      var Ft = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Ft >>> 26) | 0, Ft &= 67108863, o = Math.imul(y, K), r = Math.imul(y, Z), r = r + Math.imul(S, K) | 0, h = Math.imul(S, Z), o = o + Math.imul(g, z) | 0, r = r + Math.imul(g, Y) | 0, r = r + Math.imul(w, z) | 0, h = h + Math.imul(w, Y) | 0;
      var kt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (kt >>> 26) | 0, kt &= 67108863, o = Math.imul(T, K), r = Math.imul(T, Z), r = r + Math.imul(O, K) | 0, h = Math.imul(O, Z), o = o + Math.imul(y, z) | 0, r = r + Math.imul(y, Y) | 0, r = r + Math.imul(S, z) | 0, h = h + Math.imul(S, Y) | 0, o = o + Math.imul(g, X) | 0, r = r + Math.imul(g, J) | 0, r = r + Math.imul(w, X) | 0, h = h + Math.imul(w, J) | 0;
      var Gt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Gt >>> 26) | 0, Gt &= 67108863, o = Math.imul(b, K), r = Math.imul(b, Z), r = r + Math.imul(D, K) | 0, h = Math.imul(D, Z), o = o + Math.imul(T, z) | 0, r = r + Math.imul(T, Y) | 0, r = r + Math.imul(O, z) | 0, h = h + Math.imul(O, Y) | 0, o = o + Math.imul(y, X) | 0, r = r + Math.imul(y, J) | 0, r = r + Math.imul(S, X) | 0, h = h + Math.imul(S, J) | 0, o = o + Math.imul(g, $) | 0, r = r + Math.imul(g, Q) | 0, r = r + Math.imul(w, $) | 0, h = h + Math.imul(w, Q) | 0;
      var Ht = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Ht >>> 26) | 0, Ht &= 67108863, o = Math.imul(U, K), r = Math.imul(U, Z), r = r + Math.imul(P, K) | 0, h = Math.imul(P, Z), o = o + Math.imul(b, z) | 0, r = r + Math.imul(b, Y) | 0, r = r + Math.imul(D, z) | 0, h = h + Math.imul(D, Y) | 0, o = o + Math.imul(T, X) | 0, r = r + Math.imul(T, J) | 0, r = r + Math.imul(O, X) | 0, h = h + Math.imul(O, J) | 0, o = o + Math.imul(y, $) | 0, r = r + Math.imul(y, Q) | 0, r = r + Math.imul(S, $) | 0, h = h + Math.imul(S, Q) | 0, o = o + Math.imul(g, j) | 0, r = r + Math.imul(g, tt) | 0, r = r + Math.imul(w, j) | 0, h = h + Math.imul(w, tt) | 0;
      var Vt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Vt >>> 26) | 0, Vt &= 67108863, o = Math.imul(C, K), r = Math.imul(C, Z), r = r + Math.imul(B, K) | 0, h = Math.imul(B, Z), o = o + Math.imul(U, z) | 0, r = r + Math.imul(U, Y) | 0, r = r + Math.imul(P, z) | 0, h = h + Math.imul(P, Y) | 0, o = o + Math.imul(b, X) | 0, r = r + Math.imul(b, J) | 0, r = r + Math.imul(D, X) | 0, h = h + Math.imul(D, J) | 0, o = o + Math.imul(T, $) | 0, r = r + Math.imul(T, Q) | 0, r = r + Math.imul(O, $) | 0, h = h + Math.imul(O, Q) | 0, o = o + Math.imul(y, j) | 0, r = r + Math.imul(y, tt) | 0, r = r + Math.imul(S, j) | 0, h = h + Math.imul(S, tt) | 0, o = o + Math.imul(g, rt) | 0, r = r + Math.imul(g, it) | 0, r = r + Math.imul(w, rt) | 0, h = h + Math.imul(w, it) | 0;
      var qt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (qt >>> 26) | 0, qt &= 67108863, o = Math.imul(L, K), r = Math.imul(L, Z), r = r + Math.imul(F, K) | 0, h = Math.imul(F, Z), o = o + Math.imul(C, z) | 0, r = r + Math.imul(C, Y) | 0, r = r + Math.imul(B, z) | 0, h = h + Math.imul(B, Y) | 0, o = o + Math.imul(U, X) | 0, r = r + Math.imul(U, J) | 0, r = r + Math.imul(P, X) | 0, h = h + Math.imul(P, J) | 0, o = o + Math.imul(b, $) | 0, r = r + Math.imul(b, Q) | 0, r = r + Math.imul(D, $) | 0, h = h + Math.imul(D, Q) | 0, o = o + Math.imul(T, j) | 0, r = r + Math.imul(T, tt) | 0, r = r + Math.imul(O, j) | 0, h = h + Math.imul(O, tt) | 0, o = o + Math.imul(y, rt) | 0, r = r + Math.imul(y, it) | 0, r = r + Math.imul(S, rt) | 0, h = h + Math.imul(S, it) | 0, o = o + Math.imul(g, et) | 0, r = r + Math.imul(g, nt) | 0, r = r + Math.imul(w, et) | 0, h = h + Math.imul(w, nt) | 0;
      var Wt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Wt >>> 26) | 0, Wt &= 67108863, o = Math.imul(k, K), r = Math.imul(k, Z), r = r + Math.imul(G, K) | 0, h = Math.imul(G, Z), o = o + Math.imul(L, z) | 0, r = r + Math.imul(L, Y) | 0, r = r + Math.imul(F, z) | 0, h = h + Math.imul(F, Y) | 0, o = o + Math.imul(C, X) | 0, r = r + Math.imul(C, J) | 0, r = r + Math.imul(B, X) | 0, h = h + Math.imul(B, J) | 0, o = o + Math.imul(U, $) | 0, r = r + Math.imul(U, Q) | 0, r = r + Math.imul(P, $) | 0, h = h + Math.imul(P, Q) | 0, o = o + Math.imul(b, j) | 0, r = r + Math.imul(b, tt) | 0, r = r + Math.imul(D, j) | 0, h = h + Math.imul(D, tt) | 0, o = o + Math.imul(T, rt) | 0, r = r + Math.imul(T, it) | 0, r = r + Math.imul(O, rt) | 0, h = h + Math.imul(O, it) | 0, o = o + Math.imul(y, et) | 0, r = r + Math.imul(y, nt) | 0, r = r + Math.imul(S, et) | 0, h = h + Math.imul(S, nt) | 0, o = o + Math.imul(g, ot) | 0, r = r + Math.imul(g, st) | 0, r = r + Math.imul(w, ot) | 0, h = h + Math.imul(w, st) | 0;
      var Kt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Kt >>> 26) | 0, Kt &= 67108863, o = Math.imul(H, K), r = Math.imul(H, Z), r = r + Math.imul(V, K) | 0, h = Math.imul(V, Z), o = o + Math.imul(k, z) | 0, r = r + Math.imul(k, Y) | 0, r = r + Math.imul(G, z) | 0, h = h + Math.imul(G, Y) | 0, o = o + Math.imul(L, X) | 0, r = r + Math.imul(L, J) | 0, r = r + Math.imul(F, X) | 0, h = h + Math.imul(F, J) | 0, o = o + Math.imul(C, $) | 0, r = r + Math.imul(C, Q) | 0, r = r + Math.imul(B, $) | 0, h = h + Math.imul(B, Q) | 0, o = o + Math.imul(U, j) | 0, r = r + Math.imul(U, tt) | 0, r = r + Math.imul(P, j) | 0, h = h + Math.imul(P, tt) | 0, o = o + Math.imul(b, rt) | 0, r = r + Math.imul(b, it) | 0, r = r + Math.imul(D, rt) | 0, h = h + Math.imul(D, it) | 0, o = o + Math.imul(T, et) | 0, r = r + Math.imul(T, nt) | 0, r = r + Math.imul(O, et) | 0, h = h + Math.imul(O, nt) | 0, o = o + Math.imul(y, ot) | 0, r = r + Math.imul(y, st) | 0, r = r + Math.imul(S, ot) | 0, h = h + Math.imul(S, st) | 0, o = o + Math.imul(g, ht) | 0, r = r + Math.imul(g, ft) | 0, r = r + Math.imul(w, ht) | 0, h = h + Math.imul(w, ft) | 0;
      var Zt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Zt >>> 26) | 0, Zt &= 67108863, o = Math.imul(q, K), r = Math.imul(q, Z), r = r + Math.imul(W, K) | 0, h = Math.imul(W, Z), o = o + Math.imul(H, z) | 0, r = r + Math.imul(H, Y) | 0, r = r + Math.imul(V, z) | 0, h = h + Math.imul(V, Y) | 0, o = o + Math.imul(k, X) | 0, r = r + Math.imul(k, J) | 0, r = r + Math.imul(G, X) | 0, h = h + Math.imul(G, J) | 0, o = o + Math.imul(L, $) | 0, r = r + Math.imul(L, Q) | 0, r = r + Math.imul(F, $) | 0, h = h + Math.imul(F, Q) | 0, o = o + Math.imul(C, j) | 0, r = r + Math.imul(C, tt) | 0, r = r + Math.imul(B, j) | 0, h = h + Math.imul(B, tt) | 0, o = o + Math.imul(U, rt) | 0, r = r + Math.imul(U, it) | 0, r = r + Math.imul(P, rt) | 0, h = h + Math.imul(P, it) | 0, o = o + Math.imul(b, et) | 0, r = r + Math.imul(b, nt) | 0, r = r + Math.imul(D, et) | 0, h = h + Math.imul(D, nt) | 0, o = o + Math.imul(T, ot) | 0, r = r + Math.imul(T, st) | 0, r = r + Math.imul(O, ot) | 0, h = h + Math.imul(O, st) | 0, o = o + Math.imul(y, ht) | 0, r = r + Math.imul(y, ft) | 0, r = r + Math.imul(S, ht) | 0, h = h + Math.imul(S, ft) | 0, o = o + Math.imul(g, at) | 0, r = r + Math.imul(g, lt) | 0, r = r + Math.imul(w, at) | 0, h = h + Math.imul(w, lt) | 0;
      var zt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (zt >>> 26) | 0, zt &= 67108863, o = Math.imul(q, z), r = Math.imul(q, Y), r = r + Math.imul(W, z) | 0, h = Math.imul(W, Y), o = o + Math.imul(H, X) | 0, r = r + Math.imul(H, J) | 0, r = r + Math.imul(V, X) | 0, h = h + Math.imul(V, J) | 0, o = o + Math.imul(k, $) | 0, r = r + Math.imul(k, Q) | 0, r = r + Math.imul(G, $) | 0, h = h + Math.imul(G, Q) | 0, o = o + Math.imul(L, j) | 0, r = r + Math.imul(L, tt) | 0, r = r + Math.imul(F, j) | 0, h = h + Math.imul(F, tt) | 0, o = o + Math.imul(C, rt) | 0, r = r + Math.imul(C, it) | 0, r = r + Math.imul(B, rt) | 0, h = h + Math.imul(B, it) | 0, o = o + Math.imul(U, et) | 0, r = r + Math.imul(U, nt) | 0, r = r + Math.imul(P, et) | 0, h = h + Math.imul(P, nt) | 0, o = o + Math.imul(b, ot) | 0, r = r + Math.imul(b, st) | 0, r = r + Math.imul(D, ot) | 0, h = h + Math.imul(D, st) | 0, o = o + Math.imul(T, ht) | 0, r = r + Math.imul(T, ft) | 0, r = r + Math.imul(O, ht) | 0, h = h + Math.imul(O, ft) | 0, o = o + Math.imul(y, at) | 0, r = r + Math.imul(y, lt) | 0, r = r + Math.imul(S, at) | 0, h = h + Math.imul(S, lt) | 0;
      var Yt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Yt >>> 26) | 0, Yt &= 67108863, o = Math.imul(q, X), r = Math.imul(q, J), r = r + Math.imul(W, X) | 0, h = Math.imul(W, J), o = o + Math.imul(H, $) | 0, r = r + Math.imul(H, Q) | 0, r = r + Math.imul(V, $) | 0, h = h + Math.imul(V, Q) | 0, o = o + Math.imul(k, j) | 0, r = r + Math.imul(k, tt) | 0, r = r + Math.imul(G, j) | 0, h = h + Math.imul(G, tt) | 0, o = o + Math.imul(L, rt) | 0, r = r + Math.imul(L, it) | 0, r = r + Math.imul(F, rt) | 0, h = h + Math.imul(F, it) | 0, o = o + Math.imul(C, et) | 0, r = r + Math.imul(C, nt) | 0, r = r + Math.imul(B, et) | 0, h = h + Math.imul(B, nt) | 0, o = o + Math.imul(U, ot) | 0, r = r + Math.imul(U, st) | 0, r = r + Math.imul(P, ot) | 0, h = h + Math.imul(P, st) | 0, o = o + Math.imul(b, ht) | 0, r = r + Math.imul(b, ft) | 0, r = r + Math.imul(D, ht) | 0, h = h + Math.imul(D, ft) | 0, o = o + Math.imul(T, at) | 0, r = r + Math.imul(T, lt) | 0, r = r + Math.imul(O, at) | 0, h = h + Math.imul(O, lt) | 0;
      var Xt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Xt >>> 26) | 0, Xt &= 67108863, o = Math.imul(q, $), r = Math.imul(q, Q), r = r + Math.imul(W, $) | 0, h = Math.imul(W, Q), o = o + Math.imul(H, j) | 0, r = r + Math.imul(H, tt) | 0, r = r + Math.imul(V, j) | 0, h = h + Math.imul(V, tt) | 0, o = o + Math.imul(k, rt) | 0, r = r + Math.imul(k, it) | 0, r = r + Math.imul(G, rt) | 0, h = h + Math.imul(G, it) | 0, o = o + Math.imul(L, et) | 0, r = r + Math.imul(L, nt) | 0, r = r + Math.imul(F, et) | 0, h = h + Math.imul(F, nt) | 0, o = o + Math.imul(C, ot) | 0, r = r + Math.imul(C, st) | 0, r = r + Math.imul(B, ot) | 0, h = h + Math.imul(B, st) | 0, o = o + Math.imul(U, ht) | 0, r = r + Math.imul(U, ft) | 0, r = r + Math.imul(P, ht) | 0, h = h + Math.imul(P, ft) | 0, o = o + Math.imul(b, at) | 0, r = r + Math.imul(b, lt) | 0, r = r + Math.imul(D, at) | 0, h = h + Math.imul(D, lt) | 0;
      var Jt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Jt >>> 26) | 0, Jt &= 67108863, o = Math.imul(q, j), r = Math.imul(q, tt), r = r + Math.imul(W, j) | 0, h = Math.imul(W, tt), o = o + Math.imul(H, rt) | 0, r = r + Math.imul(H, it) | 0, r = r + Math.imul(V, rt) | 0, h = h + Math.imul(V, it) | 0, o = o + Math.imul(k, et) | 0, r = r + Math.imul(k, nt) | 0, r = r + Math.imul(G, et) | 0, h = h + Math.imul(G, nt) | 0, o = o + Math.imul(L, ot) | 0, r = r + Math.imul(L, st) | 0, r = r + Math.imul(F, ot) | 0, h = h + Math.imul(F, st) | 0, o = o + Math.imul(C, ht) | 0, r = r + Math.imul(C, ft) | 0, r = r + Math.imul(B, ht) | 0, h = h + Math.imul(B, ft) | 0, o = o + Math.imul(U, at) | 0, r = r + Math.imul(U, lt) | 0, r = r + Math.imul(P, at) | 0, h = h + Math.imul(P, lt) | 0;
      var $t = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + ($t >>> 26) | 0, $t &= 67108863, o = Math.imul(q, rt), r = Math.imul(q, it), r = r + Math.imul(W, rt) | 0, h = Math.imul(W, it), o = o + Math.imul(H, et) | 0, r = r + Math.imul(H, nt) | 0, r = r + Math.imul(V, et) | 0, h = h + Math.imul(V, nt) | 0, o = o + Math.imul(k, ot) | 0, r = r + Math.imul(k, st) | 0, r = r + Math.imul(G, ot) | 0, h = h + Math.imul(G, st) | 0, o = o + Math.imul(L, ht) | 0, r = r + Math.imul(L, ft) | 0, r = r + Math.imul(F, ht) | 0, h = h + Math.imul(F, ft) | 0, o = o + Math.imul(C, at) | 0, r = r + Math.imul(C, lt) | 0, r = r + Math.imul(B, at) | 0, h = h + Math.imul(B, lt) | 0;
      var Qt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (Qt >>> 26) | 0, Qt &= 67108863, o = Math.imul(q, et), r = Math.imul(q, nt), r = r + Math.imul(W, et) | 0, h = Math.imul(W, nt), o = o + Math.imul(H, ot) | 0, r = r + Math.imul(H, st) | 0, r = r + Math.imul(V, ot) | 0, h = h + Math.imul(V, st) | 0, o = o + Math.imul(k, ht) | 0, r = r + Math.imul(k, ft) | 0, r = r + Math.imul(G, ht) | 0, h = h + Math.imul(G, ft) | 0, o = o + Math.imul(L, at) | 0, r = r + Math.imul(L, lt) | 0, r = r + Math.imul(F, at) | 0, h = h + Math.imul(F, lt) | 0;
      var jt = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (jt >>> 26) | 0, jt &= 67108863, o = Math.imul(q, ot), r = Math.imul(q, st), r = r + Math.imul(W, ot) | 0, h = Math.imul(W, st), o = o + Math.imul(H, ht) | 0, r = r + Math.imul(H, ft) | 0, r = r + Math.imul(V, ht) | 0, h = h + Math.imul(V, ft) | 0, o = o + Math.imul(k, at) | 0, r = r + Math.imul(k, lt) | 0, r = r + Math.imul(G, at) | 0, h = h + Math.imul(G, lt) | 0;
      var tr = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (tr >>> 26) | 0, tr &= 67108863, o = Math.imul(q, ht), r = Math.imul(q, ft), r = r + Math.imul(W, ht) | 0, h = Math.imul(W, ft), o = o + Math.imul(H, at) | 0, r = r + Math.imul(H, lt) | 0, r = r + Math.imul(V, at) | 0, h = h + Math.imul(V, lt) | 0;
      var rr = (m + o | 0) + ((r & 8191) << 13) | 0;
      m = (h + (r >>> 13) | 0) + (rr >>> 26) | 0, rr &= 67108863, o = Math.imul(q, at), r = Math.imul(q, lt), r = r + Math.imul(W, at) | 0, h = Math.imul(W, lt);
      var ir = (m + o | 0) + ((r & 8191) << 13) | 0;
      return m = (h + (r >>> 13) | 0) + (ir >>> 26) | 0, ir &= 67108863, l[0] = Ft, l[1] = kt, l[2] = Gt, l[3] = Ht, l[4] = Vt, l[5] = qt, l[6] = Wt, l[7] = Kt, l[8] = Zt, l[9] = zt, l[10] = Yt, l[11] = Xt, l[12] = Jt, l[13] = $t, l[14] = Qt, l[15] = jt, l[16] = tr, l[17] = rr, l[18] = ir, m !== 0 && (l[19] = m, e.length++), e;
    };
    Math.imul || (mr = dr);
    function cr(a, t, i) {
      i.negative = t.negative ^ a.negative, i.length = a.length + t.length;
      for (var e = 0, s = 0, f = 0; f < i.length - 1; f++) {
        var l = s;
        s = 0;
        for (var m = e & 67108863, o = Math.min(f, t.length - 1), r = Math.max(0, f - a.length + 1); r <= o; r++) {
          var h = f - r, M = a.words[h] | 0, g = t.words[r] | 0, w = M * g, ct = w & 67108863;
          l = l + (w / 67108864 | 0) | 0, ct = ct + m | 0, m = ct & 67108863, l = l + (ct >>> 26) | 0, s += l >>> 26, l &= 67108863;
        }
        i.words[f] = m, e = l, l = s;
      }
      return e !== 0 ? i.words[f] = e : i.length--, i._strip();
    }
    function pr(a, t, i) {
      return cr(a, t, i);
    }
    u.prototype.mulTo = function(t, i) {
      var e, s = this.length + t.length;
      return this.length === 10 && t.length === 10 ? e = mr(this, t, i) : s < 63 ? e = dr(this, t, i) : s < 1024 ? e = cr(this, t, i) : e = pr(this, t, i), e;
    }, u.prototype.mul = function(t) {
      var i = new u(null);
      return i.words = new Array(this.length + t.length), this.mulTo(t, i);
    }, u.prototype.mulf = function(t) {
      var i = new u(null);
      return i.words = new Array(this.length + t.length), pr(this, t, i);
    }, u.prototype.imul = function(t) {
      return this.clone().mulTo(t, this);
    }, u.prototype.imuln = function(t) {
      var i = t < 0;
      i && (t = -t), c(typeof t == "number"), c(t < 67108864);
      for (var e = 0, s = 0; s < this.length; s++) {
        var f = (this.words[s] | 0) * t, l = (f & 67108863) + (e & 67108863);
        e >>= 26, e += f / 67108864 | 0, e += l >>> 26, this.words[s] = l & 67108863;
      }
      return e !== 0 && (this.words[s] = e, this.length++), i ? this.ineg() : this;
    }, u.prototype.muln = function(t) {
      return this.clone().imuln(t);
    }, u.prototype.sqr = function() {
      return this.mul(this);
    }, u.prototype.isqr = function() {
      return this.imul(this.clone());
    }, u.prototype.pow = function(t) {
      var i = Rt(t);
      if (i.length === 0)
        return new u(1);
      for (var e = this, s = 0; s < i.length && i[s] === 0; s++, e = e.sqr())
        ;
      if (++s < i.length)
        for (var f = e.sqr(); s < i.length; s++, f = f.sqr())
          i[s] !== 0 && (e = e.mul(f));
      return e;
    }, u.prototype.iushln = function(t) {
      c(typeof t == "number" && t >= 0);
      var i = t % 26, e = (t - i) / 26, s = 67108863 >>> 26 - i << 26 - i, f;
      if (i !== 0) {
        var l = 0;
        for (f = 0; f < this.length; f++) {
          var m = this.words[f] & s, o = (this.words[f] | 0) - m << i;
          this.words[f] = o | l, l = m >>> 26 - i;
        }
        l && (this.words[f] = l, this.length++);
      }
      if (e !== 0) {
        for (f = this.length - 1; f >= 0; f--)
          this.words[f + e] = this.words[f];
        for (f = 0; f < e; f++)
          this.words[f] = 0;
        this.length += e;
      }
      return this._strip();
    }, u.prototype.ishln = function(t) {
      return c(this.negative === 0), this.iushln(t);
    }, u.prototype.iushrn = function(t, i, e) {
      c(typeof t == "number" && t >= 0);
      var s;
      i ? s = (i - i % 26) / 26 : s = 0;
      var f = t % 26, l = Math.min((t - f) / 26, this.length), m = 67108863 ^ 67108863 >>> f << f, o = e;
      if (s -= l, s = Math.max(0, s), o) {
        for (var r = 0; r < l; r++)
          o.words[r] = this.words[r];
        o.length = l;
      }
      if (l !== 0)
        if (this.length > l)
          for (this.length -= l, r = 0; r < this.length; r++)
            this.words[r] = this.words[r + l];
        else
          this.words[0] = 0, this.length = 1;
      var h = 0;
      for (r = this.length - 1; r >= 0 && (h !== 0 || r >= s); r--) {
        var M = this.words[r] | 0;
        this.words[r] = h << 26 - f | M >>> f, h = M & m;
      }
      return o && h !== 0 && (o.words[o.length++] = h), this.length === 0 && (this.words[0] = 0, this.length = 1), this._strip();
    }, u.prototype.ishrn = function(t, i, e) {
      return c(this.negative === 0), this.iushrn(t, i, e);
    }, u.prototype.shln = function(t) {
      return this.clone().ishln(t);
    }, u.prototype.ushln = function(t) {
      return this.clone().iushln(t);
    }, u.prototype.shrn = function(t) {
      return this.clone().ishrn(t);
    }, u.prototype.ushrn = function(t) {
      return this.clone().iushrn(t);
    }, u.prototype.testn = function(t) {
      c(typeof t == "number" && t >= 0);
      var i = t % 26, e = (t - i) / 26, s = 1 << i;
      if (this.length <= e)
        return !1;
      var f = this.words[e];
      return !!(f & s);
    }, u.prototype.imaskn = function(t) {
      c(typeof t == "number" && t >= 0);
      var i = t % 26, e = (t - i) / 26;
      if (c(this.negative === 0, "imaskn works only with positive numbers"), this.length <= e)
        return this;
      if (i !== 0 && e++, this.length = Math.min(e, this.length), i !== 0) {
        var s = 67108863 ^ 67108863 >>> i << i;
        this.words[this.length - 1] &= s;
      }
      return this._strip();
    }, u.prototype.maskn = function(t) {
      return this.clone().imaskn(t);
    }, u.prototype.iaddn = function(t) {
      return c(typeof t == "number"), c(t < 67108864), t < 0 ? this.isubn(-t) : this.negative !== 0 ? this.length === 1 && (this.words[0] | 0) <= t ? (this.words[0] = t - (this.words[0] | 0), this.negative = 0, this) : (this.negative = 0, this.isubn(t), this.negative = 1, this) : this._iaddn(t);
    }, u.prototype._iaddn = function(t) {
      this.words[0] += t;
      for (var i = 0; i < this.length && this.words[i] >= 67108864; i++)
        this.words[i] -= 67108864, i === this.length - 1 ? this.words[i + 1] = 1 : this.words[i + 1]++;
      return this.length = Math.max(this.length, i + 1), this;
    }, u.prototype.isubn = function(t) {
      if (c(typeof t == "number"), c(t < 67108864), t < 0)
        return this.iaddn(-t);
      if (this.negative !== 0)
        return this.negative = 0, this.iaddn(t), this.negative = 1, this;
      if (this.words[0] -= t, this.length === 1 && this.words[0] < 0)
        this.words[0] = -this.words[0], this.negative = 1;
      else
        for (var i = 0; i < this.length && this.words[i] < 0; i++)
          this.words[i] += 67108864, this.words[i + 1] -= 1;
      return this._strip();
    }, u.prototype.addn = function(t) {
      return this.clone().iaddn(t);
    }, u.prototype.subn = function(t) {
      return this.clone().isubn(t);
    }, u.prototype.iabs = function() {
      return this.negative = 0, this;
    }, u.prototype.abs = function() {
      return this.clone().iabs();
    }, u.prototype._ishlnsubmul = function(t, i, e) {
      var s = t.length + e, f;
      this._expand(s);
      var l, m = 0;
      for (f = 0; f < t.length; f++) {
        l = (this.words[f + e] | 0) + m;
        var o = (t.words[f] | 0) * i;
        l -= o & 67108863, m = (l >> 26) - (o / 67108864 | 0), this.words[f + e] = l & 67108863;
      }
      for (; f < this.length - e; f++)
        l = (this.words[f + e] | 0) + m, m = l >> 26, this.words[f + e] = l & 67108863;
      if (m === 0)
        return this._strip();
      for (c(m === -1), m = 0, f = 0; f < this.length; f++)
        l = -(this.words[f] | 0) + m, m = l >> 26, this.words[f] = l & 67108863;
      return this.negative = 1, this._strip();
    }, u.prototype._wordDiv = function(t, i) {
      var e = this.length - t.length, s = this.clone(), f = t, l = f.words[f.length - 1] | 0, m = this._countBits(l);
      e = 26 - m, e !== 0 && (f = f.ushln(e), s.iushln(e), l = f.words[f.length - 1] | 0);
      var o = s.length - f.length, r;
      if (i !== "mod") {
        r = new u(null), r.length = o + 1, r.words = new Array(r.length);
        for (var h = 0; h < r.length; h++)
          r.words[h] = 0;
      }
      var M = s.clone()._ishlnsubmul(f, 1, o);
      M.negative === 0 && (s = M, r && (r.words[o] = 1));
      for (var g = o - 1; g >= 0; g--) {
        var w = (s.words[f.length + g] | 0) * 67108864 + (s.words[f.length + g - 1] | 0);
        for (w = Math.min(w / l | 0, 67108863), s._ishlnsubmul(f, w, g); s.negative !== 0; )
          w--, s.negative = 0, s._ishlnsubmul(f, 1, g), s.isZero() || (s.negative ^= 1);
        r && (r.words[g] = w);
      }
      return r && r._strip(), s._strip(), i !== "div" && e !== 0 && s.iushrn(e), {
        div: r || null,
        mod: s
      };
    }, u.prototype.divmod = function(t, i, e) {
      if (c(!t.isZero()), this.isZero())
        return {
          div: new u(0),
          mod: new u(0)
        };
      var s, f, l;
      return this.negative !== 0 && t.negative === 0 ? (l = this.neg().divmod(t, i), i !== "mod" && (s = l.div.neg()), i !== "div" && (f = l.mod.neg(), e && f.negative !== 0 && f.iadd(t)), {
        div: s,
        mod: f
      }) : this.negative === 0 && t.negative !== 0 ? (l = this.divmod(t.neg(), i), i !== "mod" && (s = l.div.neg()), {
        div: s,
        mod: l.mod
      }) : this.negative & t.negative ? (l = this.neg().divmod(t.neg(), i), i !== "div" && (f = l.mod.neg(), e && f.negative !== 0 && f.isub(t)), {
        div: l.div,
        mod: f
      }) : t.length > this.length || this.cmp(t) < 0 ? {
        div: new u(0),
        mod: this
      } : t.length === 1 ? i === "div" ? {
        div: this.divn(t.words[0]),
        mod: null
      } : i === "mod" ? {
        div: null,
        mod: new u(this.modrn(t.words[0]))
      } : {
        div: this.divn(t.words[0]),
        mod: new u(this.modrn(t.words[0]))
      } : this._wordDiv(t, i);
    }, u.prototype.div = function(t) {
      return this.divmod(t, "div", !1).div;
    }, u.prototype.mod = function(t) {
      return this.divmod(t, "mod", !1).mod;
    }, u.prototype.umod = function(t) {
      return this.divmod(t, "mod", !0).mod;
    }, u.prototype.divRound = function(t) {
      var i = this.divmod(t);
      if (i.mod.isZero())
        return i.div;
      var e = i.div.negative !== 0 ? i.mod.isub(t) : i.mod, s = t.ushrn(1), f = t.andln(1), l = e.cmp(s);
      return l < 0 || f === 1 && l === 0 ? i.div : i.div.negative !== 0 ? i.div.isubn(1) : i.div.iaddn(1);
    }, u.prototype.modrn = function(t) {
      var i = t < 0;
      i && (t = -t), c(t <= 67108863);
      for (var e = (1 << 26) % t, s = 0, f = this.length - 1; f >= 0; f--)
        s = (e * s + (this.words[f] | 0)) % t;
      return i ? -s : s;
    }, u.prototype.modn = function(t) {
      return this.modrn(t);
    }, u.prototype.idivn = function(t) {
      var i = t < 0;
      i && (t = -t), c(t <= 67108863);
      for (var e = 0, s = this.length - 1; s >= 0; s--) {
        var f = (this.words[s] | 0) + e * 67108864;
        this.words[s] = f / t | 0, e = f % t;
      }
      return this._strip(), i ? this.ineg() : this;
    }, u.prototype.divn = function(t) {
      return this.clone().idivn(t);
    }, u.prototype.egcd = function(t) {
      c(t.negative === 0), c(!t.isZero());
      var i = this, e = t.clone();
      i.negative !== 0 ? i = i.umod(t) : i = i.clone();
      for (var s = new u(1), f = new u(0), l = new u(0), m = new u(1), o = 0; i.isEven() && e.isEven(); )
        i.iushrn(1), e.iushrn(1), ++o;
      for (var r = e.clone(), h = i.clone(); !i.isZero(); ) {
        for (var M = 0, g = 1; !(i.words[0] & g) && M < 26; ++M, g <<= 1)
          ;
        if (M > 0)
          for (i.iushrn(M); M-- > 0; )
            (s.isOdd() || f.isOdd()) && (s.iadd(r), f.isub(h)), s.iushrn(1), f.iushrn(1);
        for (var w = 0, ct = 1; !(e.words[0] & ct) && w < 26; ++w, ct <<= 1)
          ;
        if (w > 0)
          for (e.iushrn(w); w-- > 0; )
            (l.isOdd() || m.isOdd()) && (l.iadd(r), m.isub(h)), l.iushrn(1), m.iushrn(1);
        i.cmp(e) >= 0 ? (i.isub(e), s.isub(l), f.isub(m)) : (e.isub(i), l.isub(s), m.isub(f));
      }
      return {
        a: l,
        b: m,
        gcd: e.iushln(o)
      };
    }, u.prototype._invmp = function(t) {
      c(t.negative === 0), c(!t.isZero());
      var i = this, e = t.clone();
      i.negative !== 0 ? i = i.umod(t) : i = i.clone();
      for (var s = new u(1), f = new u(0), l = e.clone(); i.cmpn(1) > 0 && e.cmpn(1) > 0; ) {
        for (var m = 0, o = 1; !(i.words[0] & o) && m < 26; ++m, o <<= 1)
          ;
        if (m > 0)
          for (i.iushrn(m); m-- > 0; )
            s.isOdd() && s.iadd(l), s.iushrn(1);
        for (var r = 0, h = 1; !(e.words[0] & h) && r < 26; ++r, h <<= 1)
          ;
        if (r > 0)
          for (e.iushrn(r); r-- > 0; )
            f.isOdd() && f.iadd(l), f.iushrn(1);
        i.cmp(e) >= 0 ? (i.isub(e), s.isub(f)) : (e.isub(i), f.isub(s));
      }
      var M;
      return i.cmpn(1) === 0 ? M = s : M = f, M.cmpn(0) < 0 && M.iadd(t), M;
    }, u.prototype.gcd = function(t) {
      if (this.isZero())
        return t.abs();
      if (t.isZero())
        return this.abs();
      var i = this.clone(), e = t.clone();
      i.negative = 0, e.negative = 0;
      for (var s = 0; i.isEven() && e.isEven(); s++)
        i.iushrn(1), e.iushrn(1);
      do {
        for (; i.isEven(); )
          i.iushrn(1);
        for (; e.isEven(); )
          e.iushrn(1);
        var f = i.cmp(e);
        if (f < 0) {
          var l = i;
          i = e, e = l;
        } else if (f === 0 || e.cmpn(1) === 0)
          break;
        i.isub(e);
      } while (!0);
      return e.iushln(s);
    }, u.prototype.invm = function(t) {
      return this.egcd(t).a.umod(t);
    }, u.prototype.isEven = function() {
      return (this.words[0] & 1) === 0;
    }, u.prototype.isOdd = function() {
      return (this.words[0] & 1) === 1;
    }, u.prototype.andln = function(t) {
      return this.words[0] & t;
    }, u.prototype.bincn = function(t) {
      c(typeof t == "number");
      var i = t % 26, e = (t - i) / 26, s = 1 << i;
      if (this.length <= e)
        return this._expand(e + 1), this.words[e] |= s, this;
      for (var f = s, l = e; f !== 0 && l < this.length; l++) {
        var m = this.words[l] | 0;
        m += f, f = m >>> 26, m &= 67108863, this.words[l] = m;
      }
      return f !== 0 && (this.words[l] = f, this.length++), this;
    }, u.prototype.isZero = function() {
      return this.length === 1 && this.words[0] === 0;
    }, u.prototype.cmpn = function(t) {
      var i = t < 0;
      if (this.negative !== 0 && !i)
        return -1;
      if (this.negative === 0 && i)
        return 1;
      this._strip();
      var e;
      if (this.length > 1)
        e = 1;
      else {
        i && (t = -t), c(t <= 67108863, "Number is too big");
        var s = this.words[0] | 0;
        e = s === t ? 0 : s < t ? -1 : 1;
      }
      return this.negative !== 0 ? -e | 0 : e;
    }, u.prototype.cmp = function(t) {
      if (this.negative !== 0 && t.negative === 0)
        return -1;
      if (this.negative === 0 && t.negative !== 0)
        return 1;
      var i = this.ucmp(t);
      return this.negative !== 0 ? -i | 0 : i;
    }, u.prototype.ucmp = function(t) {
      if (this.length > t.length)
        return 1;
      if (this.length < t.length)
        return -1;
      for (var i = 0, e = this.length - 1; e >= 0; e--) {
        var s = this.words[e] | 0, f = t.words[e] | 0;
        if (s !== f) {
          s < f ? i = -1 : s > f && (i = 1);
          break;
        }
      }
      return i;
    }, u.prototype.gtn = function(t) {
      return this.cmpn(t) === 1;
    }, u.prototype.gt = function(t) {
      return this.cmp(t) === 1;
    }, u.prototype.gten = function(t) {
      return this.cmpn(t) >= 0;
    }, u.prototype.gte = function(t) {
      return this.cmp(t) >= 0;
    }, u.prototype.ltn = function(t) {
      return this.cmpn(t) === -1;
    }, u.prototype.lt = function(t) {
      return this.cmp(t) === -1;
    }, u.prototype.lten = function(t) {
      return this.cmpn(t) <= 0;
    }, u.prototype.lte = function(t) {
      return this.cmp(t) <= 0;
    }, u.prototype.eqn = function(t) {
      return this.cmpn(t) === 0;
    }, u.prototype.eq = function(t) {
      return this.cmp(t) === 0;
    }, u.red = function(t) {
      return new R(t);
    }, u.prototype.toRed = function(t) {
      return c(!this.red, "Already a number in reduction context"), c(this.negative === 0, "red works only with positives"), t.convertTo(this)._forceRed(t);
    }, u.prototype.fromRed = function() {
      return c(this.red, "fromRed works only with numbers in reduction context"), this.red.convertFrom(this);
    }, u.prototype._forceRed = function(t) {
      return this.red = t, this;
    }, u.prototype.forceRed = function(t) {
      return c(!this.red, "Already a number in reduction context"), this._forceRed(t);
    }, u.prototype.redAdd = function(t) {
      return c(this.red, "redAdd works only with red numbers"), this.red.add(this, t);
    }, u.prototype.redIAdd = function(t) {
      return c(this.red, "redIAdd works only with red numbers"), this.red.iadd(this, t);
    }, u.prototype.redSub = function(t) {
      return c(this.red, "redSub works only with red numbers"), this.red.sub(this, t);
    }, u.prototype.redISub = function(t) {
      return c(this.red, "redISub works only with red numbers"), this.red.isub(this, t);
    }, u.prototype.redShl = function(t) {
      return c(this.red, "redShl works only with red numbers"), this.red.shl(this, t);
    }, u.prototype.redMul = function(t) {
      return c(this.red, "redMul works only with red numbers"), this.red._verify2(this, t), this.red.mul(this, t);
    }, u.prototype.redIMul = function(t) {
      return c(this.red, "redMul works only with red numbers"), this.red._verify2(this, t), this.red.imul(this, t);
    }, u.prototype.redSqr = function() {
      return c(this.red, "redSqr works only with red numbers"), this.red._verify1(this), this.red.sqr(this);
    }, u.prototype.redISqr = function() {
      return c(this.red, "redISqr works only with red numbers"), this.red._verify1(this), this.red.isqr(this);
    }, u.prototype.redSqrt = function() {
      return c(this.red, "redSqrt works only with red numbers"), this.red._verify1(this), this.red.sqrt(this);
    }, u.prototype.redInvm = function() {
      return c(this.red, "redInvm works only with red numbers"), this.red._verify1(this), this.red.invm(this);
    }, u.prototype.redNeg = function() {
      return c(this.red, "redNeg works only with red numbers"), this.red._verify1(this), this.red.neg(this);
    }, u.prototype.redPow = function(t) {
      return c(this.red && !t.red, "redPow(normalNum)"), this.red._verify1(this), this.red.pow(this, t);
    };
    var Bt = {
      k256: null,
      p224: null,
      p192: null,
      p25519: null
    };
    function wt(a, t) {
      this.name = a, this.p = new u(t, 16), this.n = this.p.bitLength(), this.k = new u(1).iushln(this.n).isub(this.p), this.tmp = this._tmp();
    }
    wt.prototype._tmp = function() {
      var t = new u(null);
      return t.words = new Array(Math.ceil(this.n / 13)), t;
    }, wt.prototype.ireduce = function(t) {
      var i = t, e;
      do
        this.split(i, this.tmp), i = this.imulK(i), i = i.iadd(this.tmp), e = i.bitLength();
      while (e > this.n);
      var s = e < this.n ? -1 : i.ucmp(this.p);
      return s === 0 ? (i.words[0] = 0, i.length = 1) : s > 0 ? i.isub(this.p) : i.strip !== void 0 ? i.strip() : i._strip(), i;
    }, wt.prototype.split = function(t, i) {
      t.iushrn(this.n, 0, i);
    }, wt.prototype.imulK = function(t) {
      return t.imul(this.k);
    };
    function Dt() {
      wt.call(
        this,
        "k256",
        "ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff fffffffe fffffc2f"
      );
    }
    _(Dt, wt), Dt.prototype.split = function(t, i) {
      for (var e = 4194303, s = Math.min(t.length, 9), f = 0; f < s; f++)
        i.words[f] = t.words[f];
      if (i.length = s, t.length <= 9) {
        t.words[0] = 0, t.length = 1;
        return;
      }
      var l = t.words[9];
      for (i.words[i.length++] = l & e, f = 10; f < t.length; f++) {
        var m = t.words[f] | 0;
        t.words[f - 10] = (m & e) << 4 | l >>> 22, l = m;
      }
      l >>>= 22, t.words[f - 10] = l, l === 0 && t.length > 10 ? t.length -= 10 : t.length -= 9;
    }, Dt.prototype.imulK = function(t) {
      t.words[t.length] = 0, t.words[t.length + 1] = 0, t.length += 2;
      for (var i = 0, e = 0; e < t.length; e++) {
        var s = t.words[e] | 0;
        i += s * 977, t.words[e] = i & 67108863, i = s * 64 + (i / 67108864 | 0);
      }
      return t.words[t.length - 1] === 0 && (t.length--, t.words[t.length - 1] === 0 && t.length--), t;
    };
    function vr() {
      wt.call(
        this,
        "p224",
        "ffffffff ffffffff ffffffff ffffffff 00000000 00000000 00000001"
      );
    }
    _(vr, wt);
    function gr() {
      wt.call(
        this,
        "p192",
        "ffffffff ffffffff ffffffff fffffffe ffffffff ffffffff"
      );
    }
    _(gr, wt);
    function Lt() {
      wt.call(
        this,
        "25519",
        "7fffffffffffffff ffffffffffffffff ffffffffffffffff ffffffffffffffed"
      );
    }
    _(Lt, wt), Lt.prototype.imulK = function(t) {
      for (var i = 0, e = 0; e < t.length; e++) {
        var s = (t.words[e] | 0) * 19 + i, f = s & 67108863;
        s >>>= 26, t.words[e] = f, i = s;
      }
      return i !== 0 && (t.words[t.length++] = i), t;
    }, u._prime = function(t) {
      if (Bt[t])
        return Bt[t];
      var i;
      if (t === "k256")
        i = new Dt();
      else if (t === "p224")
        i = new vr();
      else if (t === "p192")
        i = new gr();
      else if (t === "p25519")
        i = new Lt();
      else
        throw new Error("Unknown prime " + t);
      return Bt[t] = i, i;
    };
    function R(a) {
      if (typeof a == "string") {
        var t = u._prime(a);
        this.m = t.p, this.prime = t;
      } else
        c(a.gtn(1), "modulus must be greater than 1"), this.m = a, this.prime = null;
    }
    R.prototype._verify1 = function(t) {
      c(t.negative === 0, "red works only with positives"), c(t.red, "red works only with red numbers");
    }, R.prototype._verify2 = function(t, i) {
      c((t.negative | i.negative) === 0, "red works only with positives"), c(
        t.red && t.red === i.red,
        "red works only with red numbers"
      );
    }, R.prototype.imod = function(t) {
      return this.prime ? this.prime.ireduce(t)._forceRed(this) : (mt(t, t.umod(this.m)._forceRed(this)), t);
    }, R.prototype.neg = function(t) {
      return t.isZero() ? t.clone() : this.m.sub(t)._forceRed(this);
    }, R.prototype.add = function(t, i) {
      this._verify2(t, i);
      var e = t.add(i);
      return e.cmp(this.m) >= 0 && e.isub(this.m), e._forceRed(this);
    }, R.prototype.iadd = function(t, i) {
      this._verify2(t, i);
      var e = t.iadd(i);
      return e.cmp(this.m) >= 0 && e.isub(this.m), e;
    }, R.prototype.sub = function(t, i) {
      this._verify2(t, i);
      var e = t.sub(i);
      return e.cmpn(0) < 0 && e.iadd(this.m), e._forceRed(this);
    }, R.prototype.isub = function(t, i) {
      this._verify2(t, i);
      var e = t.isub(i);
      return e.cmpn(0) < 0 && e.iadd(this.m), e;
    }, R.prototype.shl = function(t, i) {
      return this._verify1(t), this.imod(t.ushln(i));
    }, R.prototype.imul = function(t, i) {
      return this._verify2(t, i), this.imod(t.imul(i));
    }, R.prototype.mul = function(t, i) {
      return this._verify2(t, i), this.imod(t.mul(i));
    }, R.prototype.isqr = function(t) {
      return this.imul(t, t.clone());
    }, R.prototype.sqr = function(t) {
      return this.mul(t, t);
    }, R.prototype.sqrt = function(t) {
      if (t.isZero())
        return t.clone();
      var i = this.m.andln(3);
      if (c(i % 2 === 1), i === 3) {
        var e = this.m.add(new u(1)).iushrn(2);
        return this.pow(t, e);
      }
      for (var s = this.m.subn(1), f = 0; !s.isZero() && s.andln(1) === 0; )
        f++, s.iushrn(1);
      c(!s.isZero());
      var l = new u(1).toRed(this), m = l.redNeg(), o = this.m.subn(1).iushrn(1), r = this.m.bitLength();
      for (r = new u(2 * r * r).toRed(this); this.pow(r, o).cmp(m) !== 0; )
        r.redIAdd(m);
      for (var h = this.pow(r, s), M = this.pow(t, s.addn(1).iushrn(1)), g = this.pow(t, s), w = f; g.cmp(l) !== 0; ) {
        for (var ct = g, y = 0; ct.cmp(l) !== 0; y++)
          ct = ct.redSqr();
        c(y < w);
        var S = this.pow(h, new u(1).iushln(w - y - 1));
        M = M.redMul(S), h = S.redSqr(), g = g.redMul(h), w = y;
      }
      return M;
    }, R.prototype.invm = function(t) {
      var i = t._invmp(this.m);
      return i.negative !== 0 ? (i.negative = 0, this.imod(i).redNeg()) : this.imod(i);
    }, R.prototype.pow = function(t, i) {
      if (i.isZero())
        return new u(1).toRed(this);
      if (i.cmpn(1) === 0)
        return t.clone();
      var e = 4, s = new Array(1 << e);
      s[0] = new u(1).toRed(this), s[1] = t;
      for (var f = 2; f < s.length; f++)
        s[f] = this.mul(s[f - 1], t);
      var l = s[0], m = 0, o = 0, r = i.bitLength() % 26;
      for (r === 0 && (r = 26), f = i.length - 1; f >= 0; f--) {
        for (var h = i.words[f], M = r - 1; M >= 0; M--) {
          var g = h >> M & 1;
          if (l !== s[0] && (l = this.sqr(l)), g === 0 && m === 0) {
            o = 0;
            continue;
          }
          m <<= 1, m |= g, o++, !(o !== e && (f !== 0 || M !== 0)) && (l = this.mul(l, s[m]), o = 0, m = 0);
        }
        r = 26;
      }
      return l;
    }, R.prototype.convertTo = function(t) {
      var i = t.umod(this.m);
      return i === t ? i.clone() : i;
    }, R.prototype.convertFrom = function(t) {
      var i = t.clone();
      return i.red = null, i;
    }, u.mont = function(t) {
      return new At(t);
    };
    function At(a) {
      R.call(this, a), this.shift = this.m.bitLength(), this.shift % 26 !== 0 && (this.shift += 26 - this.shift % 26), this.r = new u(1).iushln(this.shift), this.r2 = this.imod(this.r.sqr()), this.rinv = this.r._invmp(this.m), this.minv = this.rinv.mul(this.r).isubn(1).div(this.m), this.minv = this.minv.umod(this.r), this.minv = this.r.sub(this.minv);
    }
    _(At, R), At.prototype.convertTo = function(t) {
      return this.imod(t.ushln(this.shift));
    }, At.prototype.convertFrom = function(t) {
      var i = this.imod(t.mul(this.rinv));
      return i.red = null, i;
    }, At.prototype.imul = function(t, i) {
      if (t.isZero() || i.isZero())
        return t.words[0] = 0, t.length = 1, t;
      var e = t.imul(i), s = e.maskn(this.shift).mul(this.minv).imaskn(this.shift).mul(this.m), f = e.isub(s).iushrn(this.shift), l = f;
      return f.cmp(this.m) >= 0 ? l = f.isub(this.m) : f.cmpn(0) < 0 && (l = f.iadd(this.m)), l._forceRed(this);
    }, At.prototype.mul = function(t, i) {
      if (t.isZero() || i.isZero())
        return new u(0)._forceRed(this);
      var e = t.mul(i), s = e.maskn(this.shift).mul(this.minv).imaskn(this.shift).mul(this.m), f = e.isub(s).iushrn(this.shift), l = f;
      return f.cmp(this.m) >= 0 ? l = f.isub(this.m) : f.cmpn(0) < 0 && (l = f.iadd(this.m)), l._forceRed(this);
    }, At.prototype.invm = function(t) {
      var i = this.imod(t._invmp(this.m).mul(this.r2));
      return i._forceRed(this);
    };
  })(n, Jr);
})(ur);
var Qr = ur.exports;
const jr = /* @__PURE__ */ $r(Qr), ti = "logger/5.7.0";
let Br = !1, Lr = !1;
const Pt = { debug: 1, default: 2, info: 2, warning: 3, error: 4, off: 5 };
let Fr = Pt.default, er = null;
function ri() {
  try {
    const n = [];
    if (["NFD", "NFC", "NFKD", "NFKC"].forEach((d) => {
      try {
        if ("test".normalize(d) !== "test")
          throw new Error("bad normalize");
      } catch {
        n.push(d);
      }
    }), n.length)
      throw new Error("missing " + n.join(", "));
    if (String.fromCharCode(233).normalize("NFD") !== String.fromCharCode(101, 769))
      throw new Error("broken implementation");
  } catch (n) {
    return n.message;
  }
  return null;
}
const kr = ri();
var hr;
(function(n) {
  n.DEBUG = "DEBUG", n.INFO = "INFO", n.WARNING = "WARNING", n.ERROR = "ERROR", n.OFF = "OFF";
})(hr || (hr = {}));
var Et;
(function(n) {
  n.UNKNOWN_ERROR = "UNKNOWN_ERROR", n.NOT_IMPLEMENTED = "NOT_IMPLEMENTED", n.UNSUPPORTED_OPERATION = "UNSUPPORTED_OPERATION", n.NETWORK_ERROR = "NETWORK_ERROR", n.SERVER_ERROR = "SERVER_ERROR", n.TIMEOUT = "TIMEOUT", n.BUFFER_OVERRUN = "BUFFER_OVERRUN", n.NUMERIC_FAULT = "NUMERIC_FAULT", n.MISSING_NEW = "MISSING_NEW", n.INVALID_ARGUMENT = "INVALID_ARGUMENT", n.MISSING_ARGUMENT = "MISSING_ARGUMENT", n.UNEXPECTED_ARGUMENT = "UNEXPECTED_ARGUMENT", n.CALL_EXCEPTION = "CALL_EXCEPTION", n.INSUFFICIENT_FUNDS = "INSUFFICIENT_FUNDS", n.NONCE_EXPIRED = "NONCE_EXPIRED", n.REPLACEMENT_UNDERPRICED = "REPLACEMENT_UNDERPRICED", n.UNPREDICTABLE_GAS_LIMIT = "UNPREDICTABLE_GAS_LIMIT", n.TRANSACTION_REPLACED = "TRANSACTION_REPLACED", n.ACTION_REJECTED = "ACTION_REJECTED";
})(Et || (Et = {}));
const Gr = "0123456789abcdef";
class N {
  constructor(d) {
    Object.defineProperty(this, "version", {
      enumerable: !0,
      value: d,
      writable: !1
    });
  }
  _log(d, p) {
    const c = d.toLowerCase();
    Pt[c] == null && this.throwArgumentError("invalid log level name", "logLevel", d), !(Fr > Pt[c]) && console.log.apply(console, p);
  }
  debug(...d) {
    this._log(N.levels.DEBUG, d);
  }
  info(...d) {
    this._log(N.levels.INFO, d);
  }
  warn(...d) {
    this._log(N.levels.WARNING, d);
  }
  makeError(d, p, c) {
    if (Lr)
      return this.makeError("censored error", p, {});
    p || (p = N.errors.UNKNOWN_ERROR), c || (c = {});
    const _ = [];
    Object.keys(c).forEach((A) => {
      const x = c[A];
      try {
        if (x instanceof Uint8Array) {
          let mt = "";
          for (let ut = 0; ut < x.length; ut++)
            mt += Gr[x[ut] >> 4], mt += Gr[x[ut] & 15];
          _.push(A + "=Uint8Array(0x" + mt + ")");
        } else
          _.push(A + "=" + JSON.stringify(x));
      } catch {
        _.push(A + "=" + JSON.stringify(c[A].toString()));
      }
    }), _.push(`code=${p}`), _.push(`version=${this.version}`);
    const u = d;
    let I = "";
    switch (p) {
      case Et.NUMERIC_FAULT: {
        I = "NUMERIC_FAULT";
        const A = d;
        switch (A) {
          case "overflow":
          case "underflow":
          case "division-by-zero":
            I += "-" + A;
            break;
          case "negative-power":
          case "negative-width":
            I += "-unsupported";
            break;
          case "unbound-bitwise-result":
            I += "-unbound-result";
            break;
        }
        break;
      }
      case Et.CALL_EXCEPTION:
      case Et.INSUFFICIENT_FUNDS:
      case Et.MISSING_NEW:
      case Et.NONCE_EXPIRED:
      case Et.REPLACEMENT_UNDERPRICED:
      case Et.TRANSACTION_REPLACED:
      case Et.UNPREDICTABLE_GAS_LIMIT:
        I = p;
        break;
    }
    I && (d += " [ See: https://links.ethers.org/v5-errors-" + I + " ]"), _.length && (d += " (" + _.join(", ") + ")");
    const dt = new Error(d);
    return dt.reason = u, dt.code = p, Object.keys(c).forEach(function(A) {
      dt[A] = c[A];
    }), dt;
  }
  throwError(d, p, c) {
    throw this.makeError(d, p, c);
  }
  throwArgumentError(d, p, c) {
    return this.throwError(d, N.errors.INVALID_ARGUMENT, {
      argument: p,
      value: c
    });
  }
  assert(d, p, c, _) {
    d || this.throwError(p, c, _);
  }
  assertArgument(d, p, c, _) {
    d || this.throwArgumentError(p, c, _);
  }
  checkNormalize(d) {
    kr && this.throwError("platform missing String.prototype.normalize", N.errors.UNSUPPORTED_OPERATION, {
      operation: "String.prototype.normalize",
      form: kr
    });
  }
  checkSafeUint53(d, p) {
    typeof d == "number" && (p == null && (p = "value not safe"), (d < 0 || d >= 9007199254740991) && this.throwError(p, N.errors.NUMERIC_FAULT, {
      operation: "checkSafeInteger",
      fault: "out-of-safe-range",
      value: d
    }), d % 1 && this.throwError(p, N.errors.NUMERIC_FAULT, {
      operation: "checkSafeInteger",
      fault: "non-integer",
      value: d
    }));
  }
  checkArgumentCount(d, p, c) {
    c ? c = ": " + c : c = "", d < p && this.throwError("missing argument" + c, N.errors.MISSING_ARGUMENT, {
      count: d,
      expectedCount: p
    }), d > p && this.throwError("too many arguments" + c, N.errors.UNEXPECTED_ARGUMENT, {
      count: d,
      expectedCount: p
    });
  }
  checkNew(d, p) {
    (d === Object || d == null) && this.throwError("missing new", N.errors.MISSING_NEW, { name: p.name });
  }
  checkAbstract(d, p) {
    d === p ? this.throwError("cannot instantiate abstract class " + JSON.stringify(p.name) + " directly; use a sub-class", N.errors.UNSUPPORTED_OPERATION, { name: d.name, operation: "new" }) : (d === Object || d == null) && this.throwError("missing new", N.errors.MISSING_NEW, { name: p.name });
  }
  static globalLogger() {
    return er || (er = new N(ti)), er;
  }
  static setCensorship(d, p) {
    if (!d && p && this.globalLogger().throwError("cannot permanently disable censorship", N.errors.UNSUPPORTED_OPERATION, {
      operation: "setCensorship"
    }), Br) {
      if (!d)
        return;
      this.globalLogger().throwError("error censorship permanent", N.errors.UNSUPPORTED_OPERATION, {
        operation: "setCensorship"
      });
    }
    Lr = !!d, Br = !!p;
  }
  static setLogLevel(d) {
    const p = Pt[d.toLowerCase()];
    if (p == null) {
      N.globalLogger().warn("invalid log level - " + d);
      return;
    }
    Fr = p;
  }
  static from(d) {
    return new N(d);
  }
}
N.errors = Et;
N.levels = hr;
const ii = "bytes/5.7.0", nr = new N(ii);
function ei(n) {
  return !!n.toHexString;
}
function Hr(n) {
  return typeof n == "number" && n == n && n % 1 === 0;
}
function Zr(n) {
  if (n == null)
    return !1;
  if (n.constructor === Uint8Array)
    return !0;
  if (typeof n == "string" || !Hr(n.length) || n.length < 0)
    return !1;
  for (let d = 0; d < n.length; d++) {
    const p = n[d];
    if (!Hr(p) || p < 0 || p >= 256)
      return !1;
  }
  return !0;
}
function fr(n, d) {
  return !(typeof n != "string" || !n.match(/^0x[0-9A-Fa-f]*$/) || d && n.length !== 2 + 2 * d);
}
const or = "0123456789abcdef";
function ni(n, d) {
  if (d || (d = {}), typeof n == "number") {
    nr.checkSafeUint53(n, "invalid hexlify value");
    let p = "";
    for (; n; )
      p = or[n & 15] + p, n = Math.floor(n / 16);
    return p.length ? (p.length % 2 && (p = "0" + p), "0x" + p) : "0x00";
  }
  if (typeof n == "bigint")
    return n = n.toString(16), n.length % 2 ? "0x0" + n : "0x" + n;
  if (d.allowMissingPrefix && typeof n == "string" && n.substring(0, 2) !== "0x" && (n = "0x" + n), ei(n))
    return n.toHexString();
  if (fr(n))
    return n.length % 2 && (d.hexPad === "left" ? n = "0x0" + n.substring(2) : d.hexPad === "right" ? n += "0" : nr.throwArgumentError("hex data is odd-length", "value", n)), n.toLowerCase();
  if (Zr(n)) {
    let p = "0x";
    for (let c = 0; c < n.length; c++) {
      let _ = n[c];
      p += or[(_ & 240) >> 4] + or[_ & 15];
    }
    return p;
  }
  return nr.throwArgumentError("invalid hexlify value", "value", n);
}
const oi = "bignumber/5.7.0";
var ar = jr.BN;
const yt = new N(oi), sr = {}, Vr = 9007199254740991;
let qr = !1;
class pt {
  constructor(d, p) {
    d !== sr && yt.throwError("cannot call constructor directly; use BigNumber.from", N.errors.UNSUPPORTED_OPERATION, {
      operation: "new (BigNumber)"
    }), this._hex = p, this._isBigNumber = !0, Object.freeze(this);
  }
  fromTwos(d) {
    return vt(E(this).fromTwos(d));
  }
  toTwos(d) {
    return vt(E(this).toTwos(d));
  }
  abs() {
    return this._hex[0] === "-" ? pt.from(this._hex.substring(1)) : this;
  }
  add(d) {
    return vt(E(this).add(E(d)));
  }
  sub(d) {
    return vt(E(this).sub(E(d)));
  }
  div(d) {
    return pt.from(d).isZero() && _t("division-by-zero", "div"), vt(E(this).div(E(d)));
  }
  mul(d) {
    return vt(E(this).mul(E(d)));
  }
  mod(d) {
    const p = E(d);
    return p.isNeg() && _t("division-by-zero", "mod"), vt(E(this).umod(p));
  }
  pow(d) {
    const p = E(d);
    return p.isNeg() && _t("negative-power", "pow"), vt(E(this).pow(p));
  }
  and(d) {
    const p = E(d);
    return (this.isNegative() || p.isNeg()) && _t("unbound-bitwise-result", "and"), vt(E(this).and(p));
  }
  or(d) {
    const p = E(d);
    return (this.isNegative() || p.isNeg()) && _t("unbound-bitwise-result", "or"), vt(E(this).or(p));
  }
  xor(d) {
    const p = E(d);
    return (this.isNegative() || p.isNeg()) && _t("unbound-bitwise-result", "xor"), vt(E(this).xor(p));
  }
  mask(d) {
    return (this.isNegative() || d < 0) && _t("negative-width", "mask"), vt(E(this).maskn(d));
  }
  shl(d) {
    return (this.isNegative() || d < 0) && _t("negative-width", "shl"), vt(E(this).shln(d));
  }
  shr(d) {
    return (this.isNegative() || d < 0) && _t("negative-width", "shr"), vt(E(this).shrn(d));
  }
  eq(d) {
    return E(this).eq(E(d));
  }
  lt(d) {
    return E(this).lt(E(d));
  }
  lte(d) {
    return E(this).lte(E(d));
  }
  gt(d) {
    return E(this).gt(E(d));
  }
  gte(d) {
    return E(this).gte(E(d));
  }
  isNegative() {
    return this._hex[0] === "-";
  }
  isZero() {
    return E(this).isZero();
  }
  toNumber() {
    try {
      return E(this).toNumber();
    } catch {
      _t("overflow", "toNumber", this.toString());
    }
    return null;
  }
  toBigInt() {
    try {
      return BigInt(this.toString());
    } catch {
    }
    return yt.throwError("this platform does not support BigInt", N.errors.UNSUPPORTED_OPERATION, {
      value: this.toString()
    });
  }
  toString() {
    return arguments.length > 0 && (arguments[0] === 10 ? qr || (qr = !0, yt.warn("BigNumber.toString does not accept any parameters; base-10 is assumed")) : arguments[0] === 16 ? yt.throwError("BigNumber.toString does not accept any parameters; use bigNumber.toHexString()", N.errors.UNEXPECTED_ARGUMENT, {}) : yt.throwError("BigNumber.toString does not accept parameters", N.errors.UNEXPECTED_ARGUMENT, {})), E(this).toString(10);
  }
  toHexString() {
    return this._hex;
  }
  toJSON(d) {
    return { type: "BigNumber", hex: this.toHexString() };
  }
  static from(d) {
    if (d instanceof pt)
      return d;
    if (typeof d == "string")
      return d.match(/^-?0x[0-9a-f]+$/i) ? new pt(sr, Tt(d)) : d.match(/^-?[0-9]+$/) ? new pt(sr, Tt(new ar(d))) : yt.throwArgumentError("invalid BigNumber string", "value", d);
    if (typeof d == "number")
      return d % 1 && _t("underflow", "BigNumber.from", d), (d >= Vr || d <= -Vr) && _t("overflow", "BigNumber.from", d), pt.from(String(d));
    const p = d;
    if (typeof p == "bigint")
      return pt.from(p.toString());
    if (Zr(p))
      return pt.from(ni(p));
    if (p)
      if (p.toHexString) {
        const c = p.toHexString();
        if (typeof c == "string")
          return pt.from(c);
      } else {
        let c = p._hex;
        if (c == null && p.type === "BigNumber" && (c = p.hex), typeof c == "string" && (fr(c) || c[0] === "-" && fr(c.substring(1))))
          return pt.from(c);
      }
    return yt.throwArgumentError("invalid BigNumber value", "value", d);
  }
  static isBigNumber(d) {
    return !!(d && d._isBigNumber);
  }
}
function Tt(n) {
  if (typeof n != "string")
    return Tt(n.toString(16));
  if (n[0] === "-")
    return n = n.substring(1), n[0] === "-" && yt.throwArgumentError("invalid hex", "value", n), n = Tt(n), n === "0x00" ? n : "-" + n;
  if (n.substring(0, 2) !== "0x" && (n = "0x" + n), n === "0x")
    return "0x00";
  for (n.length % 2 && (n = "0x0" + n.substring(2)); n.length > 4 && n.substring(0, 4) === "0x00"; )
    n = "0x" + n.substring(4);
  return n;
}
function vt(n) {
  return pt.from(Tt(n));
}
function E(n) {
  const d = pt.from(n).toHexString();
  return d[0] === "-" ? new ar("-" + d.substring(3), 16) : new ar(d.substring(2), 16);
}
function _t(n, d, p) {
  const c = { fault: n, operation: d };
  return p != null && (c.value = p), yt.throwError(n, N.errors.NUMERIC_FAULT, c);
}
const Nt = (n) => typeof n == "string" || typeof n == "number" || n instanceof pt, xt = (n, d = 8) => {
  const p = pt.from(n || 0);
  return {
    amount: () => p,
    plus: (c, _ = d) => xt(p.add(Nt(c) ? c : c.amount()), _),
    minus: (c, _ = d) => xt(p.sub(Nt(c) ? c : c.amount()), _),
    times: (c, _ = d) => xt(p.mul(Nt(c) ? c : c.amount()), _),
    div: (c, _ = d) => xt(p.div(Nt(c) ? c : c.amount()), _),
    lt: (c) => p.lt(Nt(c) ? c : c.amount()),
    lte: (c) => p.lte(Nt(c) ? c : c.amount()),
    gt: (c) => p.gt(Nt(c) ? c : c.amount()),
    gte: (c) => p.gte(Nt(c) ? c : c.amount()),
    eq: (c) => p.eq(Nt(c) ? c : c.amount()),
    decimal: d
  };
};
Ut.Average + "", Ut.Fast + "", Ut.Fastest + "";
const si = (n) => async ({ from: d, recipient: p, amount: c, asset: _, memo: u }) => {
  var Mt;
  const I = (Mt = window.okxwallet) == null ? void 0 : Mt.keplr, dt = I == null ? void 0 : I.getOfflineSignerOnlyAmino(It.Cosmos);
  if (!dt)
    throw new Error("No cosmos okxwallet found");
  const { createCosmJS: A } = await import("@pioneer-platform/toolbox-cosmos"), x = await A({ offlineSigner: dt, rpcUrl: n }), mt = [
    { denom: (_ == null ? void 0 : _.symbol) === "MUON" ? "umuon" : "uatom", amount: c.amount().toString() }
  ], { transactionHash: ut } = await x.sendTokens(d, p, mt, 1.6, u);
  return ut;
}, hi = async ({
  chain: n,
  ethplorerApiKey: d,
  covalentApiKey: p,
  utxoApiKey: c,
  rpcUrl: _,
  api: u
}) => {
  var I, dt, A;
  switch (n) {
    case gt.Ethereum:
    case gt.Avalanche:
    case gt.BinanceSmartChain: {
      if (!((I = window.okxwallet) != null && I.request))
        throw new Error("No okxwallet found");
      const { getWeb3WalletMethods: x, getProvider: mt } = await import("@pioneer-platform/toolbox-evm"), ut = await x({
        chain: n,
        ethplorerApiKey: d,
        covalentApiKey: p,
        ethereumWindowProvider: window.okxwallet
      }), Mt = (await window.okxwallet.request({ method: "eth_requestAccounts" }))[0];
      return { ...ut, getAddress: () => Mt, getBalance: async () => {
        const bt = await ut.getBalance(Mt), St = await mt(n).getBalance(Mt), Rt = xt(St, 18);
        return [
          { asset: zr(n), amount: Rt },
          ...bt.slice(1)
        ];
      } };
    }
    case gt.Bitcoin: {
      if (!((dt = window.okxwallet) != null && dt.bitcoin))
        throw new Error("No bitcoin okxwallet found");
      if (!c)
        throw new Error("No utxoApiKey provided");
      const { BTCToolbox: x } = await import("@pioneer-platform/toolbox-utxo"), mt = window.okxwallet.bitcoin, ut = (await mt.connect()).address, Mt = x({ rpcUrl: _, apiKey: c, apiClient: u }), Ot = async (St) => {
        const Rt = await mt.signPsbt(St.toHex(), { from: ut, type: "list" });
        return Yr.fromHex(Rt);
      };
      return { ...Mt, transfer: (St) => Mt.transfer({
        ...St,
        signTransaction: Ot
      }), getAddress: () => ut };
    }
    case gt.Cosmos: {
      if (!((A = window.okxwallet) != null && A.keplr))
        throw new Error("No cosmos okxwallet found");
      const x = window.okxwallet.keplr;
      await x.enable(It.Cosmos);
      const [{ address: mt }] = await x.getOfflineSignerOnlyAmino(It.Cosmos).getAccounts(), { GaiaToolbox: ut } = await import("@pioneer-platform/toolbox-cosmos");
      return {
        ...ut({ server: u }),
        transfer: si(_),
        getAddress: () => mt
      };
    }
    default:
      throw new Error(`No wallet for chain ${n}`);
  }
};
gt.Avalanche, gt.BinanceSmartChain, gt.Bitcoin, gt.Ethereum, gt.Cosmos;
const fi = ({
  addChain: n,
  config: { covalentApiKey: d, ethplorerApiKey: p, utxoApiKey: c }
}) => async (_) => (await _.reduce(async (I, dt) => {
  await I;
  const A = await hi({
    chain: dt,
    utxoApiKey: c,
    covalentApiKey: d,
    ethplorerApiKey: p
  }), x = await A.getAddress();
  n({
    chain: dt,
    walletMethods: { ...A, getAddress: () => x },
    wallet: { address: x, balance: [], walletType: Kr.OKX }
  });
}, Promise.resolve()), !0), ui = {
  connectMethodName: "connectOkx",
  connect: fi
};
export {
  ui as okxWallet
};
