// @shapeshiftoss/unchained-client@10.1.1/dist/generated/ethereum/models/GasFees.js
"use strict";
// @ts-nocheck
/* tslint:disable */
/* eslint-disable */
/**
 * @shapeshiftoss/ethereum-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 8.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GasFeesToJSON = exports.GasFeesFromJSONTyped = exports.GasFeesFromJSON = exports.instanceOfGasFees = void 0;
/**
 * Check if a given object implements the GasFees interface.
 */
function instanceOfGasFees(value) {
    let isInstance = true;
    isInstance = isInstance && "gasPrice" in value;
    isInstance = isInstance && "maxFeePerGas" in value;
    isInstance = isInstance && "maxPriorityFeePerGas" in value;
    return isInstance;
}
exports.instanceOfGasFees = instanceOfGasFees;
function GasFeesFromJSON(json) {
    return GasFeesFromJSONTyped(json, false);
}
exports.GasFeesFromJSON = GasFeesFromJSON;
function GasFeesFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'gasPrice': json['gasPrice'],
        'maxFeePerGas': json['maxFeePerGas'],
        'maxPriorityFeePerGas': json['maxPriorityFeePerGas'],
    };
}
exports.GasFeesFromJSONTyped = GasFeesFromJSONTyped;
function GasFeesToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'gasPrice': value.gasPrice,
        'maxFeePerGas': value.maxFeePerGas,
        'maxPriorityFeePerGas': value.maxPriorityFeePerGas,
    };
}
exports.GasFeesToJSON = GasFeesToJSON;
