// @trezor/transport@1.1.14/lib/interfaces/usb.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsbInterface = void 0;
const abstract_1 = require("./abstract");
const constants_1 = require("../constants");
const utils_1 = require("@trezor/utils");
const ERRORS = __importStar(require("../errors"));
const INTERFACE_DEVICE_DISCONNECTED = 'The device was disconnected.';
class UsbInterface extends abstract_1.AbstractInterface {
    constructor({ usbInterface, logger }) {
        super({ logger });
        this.usbInterface = usbInterface;
        if (!this.usbInterface) {
            return;
        }
        this.usbInterface.onconnect = event => {
            this.devices = [...this.devices, ...this.createDevices([event.device])];
            this.emit('transport-interface-change', this.devices);
        };
        this.usbInterface.ondisconnect = event => {
            const { device } = event;
            if (!device.serialNumber) {
                this.emit('transport-interface-error', ERRORS.DEVICE_UNREADABLE);
                this.logger.error('device does not have serial number');
                return;
            }
            const index = this.devices.findIndex(d => d.path === device.serialNumber);
            if (index > -1) {
                this.devices.splice(index, 1);
                this.emit('transport-interface-change', this.devices);
            }
            else {
                this.emit('transport-interface-error', ERRORS.DEVICE_NOT_FOUND);
                this.logger.error('device that should be removed does not exist in state');
            }
        };
    }
    enumerate() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const devices = yield this.usbInterface.getDevices();
                const [hidDevices, nonHidDevices] = this.filterDevices(devices);
                if (hidDevices.length) {
                    this.logger.error('unreadable hid device connected');
                }
                this.devices = this.createDevices(nonHidDevices);
                return this.success(this.devices.map(d => d.path));
            }
            catch (err) {
                return this.unknownError(err, []);
            }
        });
    }
    read(path) {
        return __awaiter(this, void 0, void 0, function* () {
            const device = this.findDevice(path);
            if (!device) {
                return this.error({ error: ERRORS.DEVICE_NOT_FOUND });
            }
            try {
                const res = yield device.transferIn(constants_1.ENDPOINT_ID, 64);
                if (!res.data) {
                    return this.error({ error: ERRORS.INTERFACE_DATA_TRANSFER });
                }
                if (res.data.byteLength === 0) {
                    return this.read(path);
                }
                return this.success(res.data.buffer.slice(1));
            }
            catch (err) {
                if (err.message === INTERFACE_DEVICE_DISCONNECTED) {
                    return this.error({ error: ERRORS.DEVICE_DISCONNECTED_DURING_ACTION });
                }
                return this.error({ error: ERRORS.INTERFACE_DATA_TRANSFER, message: err.message });
            }
        });
    }
    write(path, buffer) {
        return __awaiter(this, void 0, void 0, function* () {
            const device = this.findDevice(path);
            if (!device) {
                return this.error({ error: ERRORS.DEVICE_NOT_FOUND });
            }
            const newArray = new Uint8Array(64);
            newArray[0] = 63;
            newArray.set(new Uint8Array(buffer), 1);
            try {
                const result = yield device.transferOut(constants_1.ENDPOINT_ID, newArray);
                if (result.status !== 'ok') {
                    this.logger.error('transport', 'usbInterface', 'write', 'result.status', result.status);
                    throw new Error('transfer out status not ok');
                }
                return this.success(undefined);
            }
            catch (err) {
                if (err.message === INTERFACE_DEVICE_DISCONNECTED) {
                    return this.error({ error: ERRORS.DEVICE_DISCONNECTED_DURING_ACTION });
                }
                return this.error({ error: ERRORS.INTERFACE_DATA_TRANSFER, message: err.message });
            }
        });
    }
    openDevice(path, first) {
        return __awaiter(this, void 0, void 0, function* () {
            for (let i = 0; i < 5; i++) {
                const res = yield this.openInternal(path, first);
                if (res.success) {
                    return res;
                }
                yield (0, utils_1.createTimeoutPromise)(100 * i);
            }
            return this.openInternal(path, first);
        });
    }
    openInternal(path, first) {
        return __awaiter(this, void 0, void 0, function* () {
            const device = this.findDevice(path);
            if (!device) {
                return this.error({ error: ERRORS.DEVICE_NOT_FOUND });
            }
            try {
                yield device.open();
            }
            catch (err) {
                return this.error({
                    error: ERRORS.INTERFACE_UNABLE_TO_OPEN_DEVICE,
                    message: err.message,
                });
            }
            if (first) {
                try {
                    yield device.selectConfiguration(constants_1.CONFIGURATION_ID);
                    yield device.reset();
                }
                catch (err) {
                }
            }
            try {
                yield device.claimInterface(constants_1.INTERFACE_ID);
            }
            catch (err) {
                return this.error({
                    error: ERRORS.INTERFACE_UNABLE_TO_OPEN_DEVICE,
                    message: err.message,
                });
            }
            return this.success(undefined);
        });
    }
    closeDevice(path) {
        return __awaiter(this, void 0, void 0, function* () {
            const device = this.findDevice(path);
            if (!device) {
                return this.error({ error: ERRORS.DEVICE_NOT_FOUND });
            }
            if (device.opened) {
                try {
                    const interfaceId = constants_1.INTERFACE_ID;
                    yield device.releaseInterface(interfaceId);
                }
                catch (err) {
                }
            }
            if (device.opened) {
                try {
                    yield device.close();
                }
                catch (err) {
                    return this.error({
                        error: ERRORS.INTERFACE_UNABLE_TO_CLOSE_DEVICE,
                        message: err.message,
                    });
                }
            }
            return this.success(undefined);
        });
    }
    findDevice(path) {
        const device = this.devices.find(d => d.path === path);
        if (!device) {
            return;
        }
        return device.device;
    }
    createDevices(nonHidDevices) {
        let bootloaderId = 0;
        return nonHidDevices.map(device => {
            const { serialNumber } = device;
            let path = serialNumber == null || serialNumber === '' ? 'bootloader' : serialNumber;
            if (path === 'bootloader') {
                bootloaderId++;
                path += bootloaderId;
            }
            return { path, device };
        });
    }
    deviceIsHid(device) {
        return device.vendorId === constants_1.T1_HID_VENDOR;
    }
    filterDevices(devices) {
        const trezorDevices = devices.filter(dev => {
            const isTrezor = constants_1.TREZOR_USB_DESCRIPTORS.some(desc => dev.vendorId === desc.vendorId && dev.productId === desc.productId);
            return isTrezor;
        });
        const hidDevices = trezorDevices.filter(dev => this.deviceIsHid(dev));
        const nonHidDevices = trezorDevices.filter(dev => !this.deviceIsHid(dev));
        return [hidDevices, nonHidDevices];
    }
}
exports.UsbInterface = UsbInterface;
