// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/msg_tss_pool.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgTssPool = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var type_keygen_1 = require("../../../../../thorchain/v1/x/thorchain/types/type_keygen");
var type_blame_1 = require("../../../../../thorchain/v1/x/thorchain/types/type_blame");
function createBaseMsgTssPool() {
    return {
        id: "",
        poolPubKey: "",
        keygenType: 0,
        pubKeys: [],
        height: long_1.default.ZERO,
        blame: undefined,
        chains: [],
        signer: new Uint8Array(),
        keygenTime: long_1.default.ZERO,
    };
}
exports.MsgTssPool = {
    encode: function (message, writer) {
        var e_1, _a, e_2, _b;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.poolPubKey !== "") {
            writer.uint32(18).string(message.poolPubKey);
        }
        if (message.keygenType !== 0) {
            writer.uint32(24).int32(message.keygenType);
        }
        try {
            for (var _c = __values(message.pubKeys), _d = _c.next(); !_d.done; _d = _c.next()) {
                var v = _d.value;
                writer.uint32(34).string(v);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (!message.height.isZero()) {
            writer.uint32(40).int64(message.height);
        }
        if (message.blame !== undefined) {
            type_blame_1.Blame.encode(message.blame, writer.uint32(50).fork()).ldelim();
        }
        try {
            for (var _e = __values(message.chains), _f = _e.next(); !_f.done; _f = _e.next()) {
                var v = _f.value;
                writer.uint32(58).string(v);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
        if (message.signer.length !== 0) {
            writer.uint32(66).bytes(message.signer);
        }
        if (!message.keygenTime.isZero()) {
            writer.uint32(72).int64(message.keygenTime);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgTssPool();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.string();
                    break;
                case 2:
                    message.poolPubKey = reader.string();
                    break;
                case 3:
                    message.keygenType = reader.int32();
                    break;
                case 4:
                    message.pubKeys.push(reader.string());
                    break;
                case 5:
                    message.height = reader.int64();
                    break;
                case 6:
                    message.blame = type_blame_1.Blame.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.chains.push(reader.string());
                    break;
                case 8:
                    message.signer = reader.bytes();
                    break;
                case 9:
                    message.keygenTime = reader.int64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            id: isSet(object.id) ? String(object.id) : "",
            poolPubKey: isSet(object.poolPubKey) ? String(object.poolPubKey) : "",
            keygenType: isSet(object.keygenType)
                ? (0, type_keygen_1.keygenTypeFromJSON)(object.keygenType)
                : 0,
            pubKeys: Array.isArray(object === null || object === void 0 ? void 0 : object.pubKeys)
                ? object.pubKeys.map(function (e) { return String(e); })
                : [],
            height: isSet(object.height) ? long_1.default.fromString(object.height) : long_1.default.ZERO,
            blame: isSet(object.blame) ? type_blame_1.Blame.fromJSON(object.blame) : undefined,
            chains: Array.isArray(object === null || object === void 0 ? void 0 : object.chains)
                ? object.chains.map(function (e) { return String(e); })
                : [],
            signer: isSet(object.signer)
                ? bytesFromBase64(object.signer)
                : new Uint8Array(),
            keygenTime: isSet(object.keygenTime)
                ? long_1.default.fromString(object.keygenTime)
                : long_1.default.ZERO,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.id !== undefined && (obj.id = message.id);
        message.poolPubKey !== undefined && (obj.poolPubKey = message.poolPubKey);
        message.keygenType !== undefined &&
            (obj.keygenType = (0, type_keygen_1.keygenTypeToJSON)(message.keygenType));
        if (message.pubKeys) {
            obj.pubKeys = message.pubKeys.map(function (e) { return e; });
        }
        else {
            obj.pubKeys = [];
        }
        message.height !== undefined &&
            (obj.height = (message.height || long_1.default.ZERO).toString());
        message.blame !== undefined &&
            (obj.blame = message.blame ? type_blame_1.Blame.toJSON(message.blame) : undefined);
        if (message.chains) {
            obj.chains = message.chains.map(function (e) { return e; });
        }
        else {
            obj.chains = [];
        }
        message.signer !== undefined &&
            (obj.signer = base64FromBytes(message.signer !== undefined ? message.signer : new Uint8Array()));
        message.keygenTime !== undefined &&
            (obj.keygenTime = (message.keygenTime || long_1.default.ZERO).toString());
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f;
        var message = createBaseMsgTssPool();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        message.poolPubKey = (_b = object.poolPubKey) !== null && _b !== void 0 ? _b : "";
        message.keygenType = (_c = object.keygenType) !== null && _c !== void 0 ? _c : 0;
        message.pubKeys = ((_d = object.pubKeys) === null || _d === void 0 ? void 0 : _d.map(function (e) { return e; })) || [];
        message.height =
            object.height !== undefined && object.height !== null
                ? long_1.default.fromValue(object.height)
                : long_1.default.ZERO;
        message.blame =
            object.blame !== undefined && object.blame !== null
                ? type_blame_1.Blame.fromPartial(object.blame)
                : undefined;
        message.chains = ((_e = object.chains) === null || _e === void 0 ? void 0 : _e.map(function (e) { return e; })) || [];
        message.signer = (_f = object.signer) !== null && _f !== void 0 ? _f : new Uint8Array();
        message.keygenTime =
            object.keygenTime !== undefined && object.keygenTime !== null
                ? long_1.default.fromValue(object.keygenTime)
                : long_1.default.ZERO;
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, "base64").toString("binary"); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, "binary").toString("base64"); });
function base64FromBytes(arr) {
    var e_3, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_3) throw e_3.error; }
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
