// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/msg_manage_thorname.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgManageTHORName = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var common_1 = require("../../../../../thorchain/v1/common/common");
function createBaseMsgManageTHORName() {
    return {
        name: "",
        chain: "",
        address: "",
        coin: undefined,
        expireBlockHeight: long_1.default.ZERO,
        preferredAsset: undefined,
        owner: new Uint8Array(),
        signer: new Uint8Array(),
    };
}
exports.MsgManageTHORName = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.chain !== "") {
            writer.uint32(18).string(message.chain);
        }
        if (message.address !== "") {
            writer.uint32(26).string(message.address);
        }
        if (message.coin !== undefined) {
            common_1.Coin.encode(message.coin, writer.uint32(34).fork()).ldelim();
        }
        if (!message.expireBlockHeight.isZero()) {
            writer.uint32(40).int64(message.expireBlockHeight);
        }
        if (message.preferredAsset !== undefined) {
            common_1.Asset.encode(message.preferredAsset, writer.uint32(50).fork()).ldelim();
        }
        if (message.owner.length !== 0) {
            writer.uint32(58).bytes(message.owner);
        }
        if (message.signer.length !== 0) {
            writer.uint32(66).bytes(message.signer);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgManageTHORName();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.name = reader.string();
                    break;
                case 2:
                    message.chain = reader.string();
                    break;
                case 3:
                    message.address = reader.string();
                    break;
                case 4:
                    message.coin = common_1.Coin.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.expireBlockHeight = reader.int64();
                    break;
                case 6:
                    message.preferredAsset = common_1.Asset.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.owner = reader.bytes();
                    break;
                case 8:
                    message.signer = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            name: isSet(object.name) ? String(object.name) : "",
            chain: isSet(object.chain) ? String(object.chain) : "",
            address: isSet(object.address) ? String(object.address) : "",
            coin: isSet(object.coin) ? common_1.Coin.fromJSON(object.coin) : undefined,
            expireBlockHeight: isSet(object.expireBlockHeight)
                ? long_1.default.fromString(object.expireBlockHeight)
                : long_1.default.ZERO,
            preferredAsset: isSet(object.preferredAsset)
                ? common_1.Asset.fromJSON(object.preferredAsset)
                : undefined,
            owner: isSet(object.owner)
                ? bytesFromBase64(object.owner)
                : new Uint8Array(),
            signer: isSet(object.signer)
                ? bytesFromBase64(object.signer)
                : new Uint8Array(),
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.name !== undefined && (obj.name = message.name);
        message.chain !== undefined && (obj.chain = message.chain);
        message.address !== undefined && (obj.address = message.address);
        message.coin !== undefined &&
            (obj.coin = message.coin ? common_1.Coin.toJSON(message.coin) : undefined);
        message.expireBlockHeight !== undefined &&
            (obj.expireBlockHeight = (message.expireBlockHeight || long_1.default.ZERO).toString());
        message.preferredAsset !== undefined &&
            (obj.preferredAsset = message.preferredAsset
                ? common_1.Asset.toJSON(message.preferredAsset)
                : undefined);
        message.owner !== undefined &&
            (obj.owner = base64FromBytes(message.owner !== undefined ? message.owner : new Uint8Array()));
        message.signer !== undefined &&
            (obj.signer = base64FromBytes(message.signer !== undefined ? message.signer : new Uint8Array()));
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e;
        var message = createBaseMsgManageTHORName();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : "";
        message.chain = (_b = object.chain) !== null && _b !== void 0 ? _b : "";
        message.address = (_c = object.address) !== null && _c !== void 0 ? _c : "";
        message.coin =
            object.coin !== undefined && object.coin !== null
                ? common_1.Coin.fromPartial(object.coin)
                : undefined;
        message.expireBlockHeight =
            object.expireBlockHeight !== undefined &&
                object.expireBlockHeight !== null
                ? long_1.default.fromValue(object.expireBlockHeight)
                : long_1.default.ZERO;
        message.preferredAsset =
            object.preferredAsset !== undefined && object.preferredAsset !== null
                ? common_1.Asset.fromPartial(object.preferredAsset)
                : undefined;
        message.owner = (_d = object.owner) !== null && _d !== void 0 ? _d : new Uint8Array();
        message.signer = (_e = object.signer) !== null && _e !== void 0 ? _e : new Uint8Array();
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, "base64").toString("binary"); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, "binary").toString("base64"); });
function base64FromBytes(arr) {
    var e_1, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
