// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/msg_set_node_keys.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSetNodeKeys = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var common_1 = require("../../../../../thorchain/v1/common/common");
function createBaseMsgSetNodeKeys() {
    return {
        pubKeySetSet: undefined,
        validatorConsPubKey: "",
        signer: new Uint8Array(),
    };
}
exports.MsgSetNodeKeys = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pubKeySetSet !== undefined) {
            common_1.PubKeySet.encode(message.pubKeySetSet, writer.uint32(10).fork()).ldelim();
        }
        if (message.validatorConsPubKey !== "") {
            writer.uint32(18).string(message.validatorConsPubKey);
        }
        if (message.signer.length !== 0) {
            writer.uint32(26).bytes(message.signer);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgSetNodeKeys();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pubKeySetSet = common_1.PubKeySet.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.validatorConsPubKey = reader.string();
                    break;
                case 3:
                    message.signer = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pubKeySetSet: isSet(object.pubKeySetSet)
                ? common_1.PubKeySet.fromJSON(object.pubKeySetSet)
                : undefined,
            validatorConsPubKey: isSet(object.validatorConsPubKey)
                ? String(object.validatorConsPubKey)
                : "",
            signer: isSet(object.signer)
                ? bytesFromBase64(object.signer)
                : new Uint8Array(),
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.pubKeySetSet !== undefined &&
            (obj.pubKeySetSet = message.pubKeySetSet
                ? common_1.PubKeySet.toJSON(message.pubKeySetSet)
                : undefined);
        message.validatorConsPubKey !== undefined &&
            (obj.validatorConsPubKey = message.validatorConsPubKey);
        message.signer !== undefined &&
            (obj.signer = base64FromBytes(message.signer !== undefined ? message.signer : new Uint8Array()));
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseMsgSetNodeKeys();
        message.pubKeySetSet =
            object.pubKeySetSet !== undefined && object.pubKeySetSet !== null
                ? common_1.PubKeySet.fromPartial(object.pubKeySetSet)
                : undefined;
        message.validatorConsPubKey = (_a = object.validatorConsPubKey) !== null && _a !== void 0 ? _a : "";
        message.signer = (_b = object.signer) !== null && _b !== void 0 ? _b : new Uint8Array();
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, "base64").toString("binary"); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, "binary").toString("base64"); });
function base64FromBytes(arr) {
    var e_1, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
