// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/type_blame.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blame = exports.Node = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
function createBaseNode() {
    return {
        pubkey: "",
        blameData: new Uint8Array(),
        blameSignature: new Uint8Array(),
    };
}
exports.Node = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pubkey !== "") {
            writer.uint32(10).string(message.pubkey);
        }
        if (message.blameData.length !== 0) {
            writer.uint32(18).bytes(message.blameData);
        }
        if (message.blameSignature.length !== 0) {
            writer.uint32(26).bytes(message.blameSignature);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseNode();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pubkey = reader.string();
                    break;
                case 2:
                    message.blameData = reader.bytes();
                    break;
                case 3:
                    message.blameSignature = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pubkey: isSet(object.pubkey) ? String(object.pubkey) : "",
            blameData: isSet(object.blameData)
                ? bytesFromBase64(object.blameData)
                : new Uint8Array(),
            blameSignature: isSet(object.blameSignature)
                ? bytesFromBase64(object.blameSignature)
                : new Uint8Array(),
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.pubkey !== undefined && (obj.pubkey = message.pubkey);
        message.blameData !== undefined &&
            (obj.blameData = base64FromBytes(message.blameData !== undefined ? message.blameData : new Uint8Array()));
        message.blameSignature !== undefined &&
            (obj.blameSignature = base64FromBytes(message.blameSignature !== undefined
                ? message.blameSignature
                : new Uint8Array()));
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseNode();
        message.pubkey = (_a = object.pubkey) !== null && _a !== void 0 ? _a : "";
        message.blameData = (_b = object.blameData) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.blameSignature = (_c = object.blameSignature) !== null && _c !== void 0 ? _c : new Uint8Array();
        return message;
    },
};
function createBaseBlame() {
    return { failReason: "", isUnicast: false, blameNodes: [] };
}
exports.Blame = {
    encode: function (message, writer) {
        var e_1, _a;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.failReason !== "") {
            writer.uint32(10).string(message.failReason);
        }
        if (message.isUnicast === true) {
            writer.uint32(16).bool(message.isUnicast);
        }
        try {
            for (var _b = __values(message.blameNodes), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                exports.Node.encode(v, writer.uint32(26).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseBlame();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.failReason = reader.string();
                    break;
                case 2:
                    message.isUnicast = reader.bool();
                    break;
                case 3:
                    message.blameNodes.push(exports.Node.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            failReason: isSet(object.failReason) ? String(object.failReason) : "",
            isUnicast: isSet(object.isUnicast) ? Boolean(object.isUnicast) : false,
            blameNodes: Array.isArray(object === null || object === void 0 ? void 0 : object.blameNodes)
                ? object.blameNodes.map(function (e) { return exports.Node.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.failReason !== undefined && (obj.failReason = message.failReason);
        message.isUnicast !== undefined && (obj.isUnicast = message.isUnicast);
        if (message.blameNodes) {
            obj.blameNodes = message.blameNodes.map(function (e) {
                return e ? exports.Node.toJSON(e) : undefined;
            });
        }
        else {
            obj.blameNodes = [];
        }
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseBlame();
        message.failReason = (_a = object.failReason) !== null && _a !== void 0 ? _a : "";
        message.isUnicast = (_b = object.isUnicast) !== null && _b !== void 0 ? _b : false;
        message.blameNodes =
            ((_c = object.blameNodes) === null || _c === void 0 ? void 0 : _c.map(function (e) { return exports.Node.fromPartial(e); })) || [];
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, "base64").toString("binary"); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, "binary").toString("base64"); });
function base64FromBytes(arr) {
    var e_2, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
