// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/type_tss_metric.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TssKeysignMetric = exports.TssKeygenMetric = exports.NodeTssTime = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
function createBaseNodeTssTime() {
    return { address: new Uint8Array(), tssTime: long_1.default.ZERO };
}
exports.NodeTssTime = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.address.length !== 0) {
            writer.uint32(10).bytes(message.address);
        }
        if (!message.tssTime.isZero()) {
            writer.uint32(16).int64(message.tssTime);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseNodeTssTime();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.address = reader.bytes();
                    break;
                case 2:
                    message.tssTime = reader.int64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            address: isSet(object.address)
                ? bytesFromBase64(object.address)
                : new Uint8Array(),
            tssTime: isSet(object.tssTime)
                ? long_1.default.fromString(object.tssTime)
                : long_1.default.ZERO,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.address !== undefined &&
            (obj.address = base64FromBytes(message.address !== undefined ? message.address : new Uint8Array()));
        message.tssTime !== undefined &&
            (obj.tssTime = (message.tssTime || long_1.default.ZERO).toString());
        return obj;
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseNodeTssTime();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : new Uint8Array();
        message.tssTime =
            object.tssTime !== undefined && object.tssTime !== null
                ? long_1.default.fromValue(object.tssTime)
                : long_1.default.ZERO;
        return message;
    },
};
function createBaseTssKeygenMetric() {
    return { pubKey: "", nodeTssTimes: [] };
}
exports.TssKeygenMetric = {
    encode: function (message, writer) {
        var e_1, _a;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pubKey !== "") {
            writer.uint32(10).string(message.pubKey);
        }
        try {
            for (var _b = __values(message.nodeTssTimes), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                exports.NodeTssTime.encode(v, writer.uint32(18).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseTssKeygenMetric();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pubKey = reader.string();
                    break;
                case 2:
                    message.nodeTssTimes.push(exports.NodeTssTime.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pubKey: isSet(object.pubKey) ? String(object.pubKey) : "",
            nodeTssTimes: Array.isArray(object === null || object === void 0 ? void 0 : object.nodeTssTimes)
                ? object.nodeTssTimes.map(function (e) { return exports.NodeTssTime.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.pubKey !== undefined && (obj.pubKey = message.pubKey);
        if (message.nodeTssTimes) {
            obj.nodeTssTimes = message.nodeTssTimes.map(function (e) {
                return e ? exports.NodeTssTime.toJSON(e) : undefined;
            });
        }
        else {
            obj.nodeTssTimes = [];
        }
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseTssKeygenMetric();
        message.pubKey = (_a = object.pubKey) !== null && _a !== void 0 ? _a : "";
        message.nodeTssTimes =
            ((_b = object.nodeTssTimes) === null || _b === void 0 ? void 0 : _b.map(function (e) { return exports.NodeTssTime.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseTssKeysignMetric() {
    return { txId: "", nodeTssTimes: [] };
}
exports.TssKeysignMetric = {
    encode: function (message, writer) {
        var e_2, _a;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.txId !== "") {
            writer.uint32(10).string(message.txId);
        }
        try {
            for (var _b = __values(message.nodeTssTimes), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                exports.NodeTssTime.encode(v, writer.uint32(18).fork()).ldelim();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseTssKeysignMetric();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.txId = reader.string();
                    break;
                case 2:
                    message.nodeTssTimes.push(exports.NodeTssTime.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            txId: isSet(object.txId) ? String(object.txId) : "",
            nodeTssTimes: Array.isArray(object === null || object === void 0 ? void 0 : object.nodeTssTimes)
                ? object.nodeTssTimes.map(function (e) { return exports.NodeTssTime.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.txId !== undefined && (obj.txId = message.txId);
        if (message.nodeTssTimes) {
            obj.nodeTssTimes = message.nodeTssTimes.map(function (e) {
                return e ? exports.NodeTssTime.toJSON(e) : undefined;
            });
        }
        else {
            obj.nodeTssTimes = [];
        }
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseTssKeysignMetric();
        message.txId = (_a = object.txId) !== null && _a !== void 0 ? _a : "";
        message.nodeTssTimes =
            ((_b = object.nodeTssTimes) === null || _b === void 0 ? void 0 : _b.map(function (e) { return exports.NodeTssTime.fromPartial(e); })) || [];
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, "base64").toString("binary"); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, "binary").toString("base64"); });
function base64FromBytes(arr) {
    var e_3, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_3) throw e_3.error; }
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
