// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/type_solvency_voter.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolvencyVoter = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var common_1 = require("../../../../../thorchain/v1/common/common");
function createBaseSolvencyVoter() {
    return {
        id: "",
        chain: "",
        pubKey: "",
        coins: [],
        height: long_1.default.ZERO,
        consensusBlockHeight: long_1.default.ZERO,
        signers: [],
    };
}
exports.SolvencyVoter = {
    encode: function (message, writer) {
        var e_1, _a, e_2, _b;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.chain !== "") {
            writer.uint32(18).string(message.chain);
        }
        if (message.pubKey !== "") {
            writer.uint32(26).string(message.pubKey);
        }
        try {
            for (var _c = __values(message.coins), _d = _c.next(); !_d.done; _d = _c.next()) {
                var v = _d.value;
                common_1.Coin.encode(v, writer.uint32(34).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (!message.height.isZero()) {
            writer.uint32(40).int64(message.height);
        }
        if (!message.consensusBlockHeight.isZero()) {
            writer.uint32(48).int64(message.consensusBlockHeight);
        }
        try {
            for (var _e = __values(message.signers), _f = _e.next(); !_f.done; _f = _e.next()) {
                var v = _f.value;
                writer.uint32(58).string(v);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseSolvencyVoter();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.string();
                    break;
                case 2:
                    message.chain = reader.string();
                    break;
                case 3:
                    message.pubKey = reader.string();
                    break;
                case 4:
                    message.coins.push(common_1.Coin.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.height = reader.int64();
                    break;
                case 6:
                    message.consensusBlockHeight = reader.int64();
                    break;
                case 7:
                    message.signers.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            id: isSet(object.id) ? String(object.id) : "",
            chain: isSet(object.chain) ? String(object.chain) : "",
            pubKey: isSet(object.pubKey) ? String(object.pubKey) : "",
            coins: Array.isArray(object === null || object === void 0 ? void 0 : object.coins)
                ? object.coins.map(function (e) { return common_1.Coin.fromJSON(e); })
                : [],
            height: isSet(object.height) ? long_1.default.fromString(object.height) : long_1.default.ZERO,
            consensusBlockHeight: isSet(object.consensusBlockHeight)
                ? long_1.default.fromString(object.consensusBlockHeight)
                : long_1.default.ZERO,
            signers: Array.isArray(object === null || object === void 0 ? void 0 : object.signers)
                ? object.signers.map(function (e) { return String(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.id !== undefined && (obj.id = message.id);
        message.chain !== undefined && (obj.chain = message.chain);
        message.pubKey !== undefined && (obj.pubKey = message.pubKey);
        if (message.coins) {
            obj.coins = message.coins.map(function (e) { return (e ? common_1.Coin.toJSON(e) : undefined); });
        }
        else {
            obj.coins = [];
        }
        message.height !== undefined &&
            (obj.height = (message.height || long_1.default.ZERO).toString());
        message.consensusBlockHeight !== undefined &&
            (obj.consensusBlockHeight = (message.consensusBlockHeight || long_1.default.ZERO).toString());
        if (message.signers) {
            obj.signers = message.signers.map(function (e) { return e; });
        }
        else {
            obj.signers = [];
        }
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e;
        var message = createBaseSolvencyVoter();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        message.chain = (_b = object.chain) !== null && _b !== void 0 ? _b : "";
        message.pubKey = (_c = object.pubKey) !== null && _c !== void 0 ? _c : "";
        message.coins = ((_d = object.coins) === null || _d === void 0 ? void 0 : _d.map(function (e) { return common_1.Coin.fromPartial(e); })) || [];
        message.height =
            object.height !== undefined && object.height !== null
                ? long_1.default.fromValue(object.height)
                : long_1.default.ZERO;
        message.consensusBlockHeight =
            object.consensusBlockHeight !== undefined &&
                object.consensusBlockHeight !== null
                ? long_1.default.fromValue(object.consensusBlockHeight)
                : long_1.default.ZERO;
        message.signers = ((_e = object.signers) === null || _e === void 0 ? void 0 : _e.map(function (e) { return e; })) || [];
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
