// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/type_network_fee.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkFee = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
function createBaseNetworkFee() {
    return {
        chain: "",
        transactionSize: long_1.default.UZERO,
        transactionFeeRate: long_1.default.UZERO,
    };
}
exports.NetworkFee = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.chain !== "") {
            writer.uint32(10).string(message.chain);
        }
        if (!message.transactionSize.isZero()) {
            writer.uint32(16).uint64(message.transactionSize);
        }
        if (!message.transactionFeeRate.isZero()) {
            writer.uint32(24).uint64(message.transactionFeeRate);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseNetworkFee();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.chain = reader.string();
                    break;
                case 2:
                    message.transactionSize = reader.uint64();
                    break;
                case 3:
                    message.transactionFeeRate = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            chain: isSet(object.chain) ? String(object.chain) : "",
            transactionSize: isSet(object.transactionSize)
                ? long_1.default.fromString(object.transactionSize)
                : long_1.default.UZERO,
            transactionFeeRate: isSet(object.transactionFeeRate)
                ? long_1.default.fromString(object.transactionFeeRate)
                : long_1.default.UZERO,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.chain !== undefined && (obj.chain = message.chain);
        message.transactionSize !== undefined &&
            (obj.transactionSize = (message.transactionSize || long_1.default.UZERO).toString());
        message.transactionFeeRate !== undefined &&
            (obj.transactionFeeRate = (message.transactionFeeRate || long_1.default.UZERO).toString());
        return obj;
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseNetworkFee();
        message.chain = (_a = object.chain) !== null && _a !== void 0 ? _a : "";
        message.transactionSize =
            object.transactionSize !== undefined && object.transactionSize !== null
                ? long_1.default.fromValue(object.transactionSize)
                : long_1.default.UZERO;
        message.transactionFeeRate =
            object.transactionFeeRate !== undefined &&
                object.transactionFeeRate !== null
                ? long_1.default.fromValue(object.transactionFeeRate)
                : long_1.default.UZERO;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
