// @shapeshiftoss/proto-tx-builder@0.8.0/dist/proto/generated/thorchain/v1/x/thorchain/types/type_thorname.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.THORName = exports.THORNameAlias = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var common_1 = require("../../../../../thorchain/v1/common/common");
function createBaseTHORNameAlias() {
    return { chain: "", address: "" };
}
exports.THORNameAlias = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.chain !== "") {
            writer.uint32(10).string(message.chain);
        }
        if (message.address !== "") {
            writer.uint32(18).string(message.address);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseTHORNameAlias();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.chain = reader.string();
                    break;
                case 2:
                    message.address = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            chain: isSet(object.chain) ? String(object.chain) : "",
            address: isSet(object.address) ? String(object.address) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.chain !== undefined && (obj.chain = message.chain);
        message.address !== undefined && (obj.address = message.address);
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseTHORNameAlias();
        message.chain = (_a = object.chain) !== null && _a !== void 0 ? _a : "";
        message.address = (_b = object.address) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseTHORName() {
    return {
        name: "",
        expireBlockHeight: long_1.default.ZERO,
        owner: new Uint8Array(),
        preferredAsset: undefined,
        aliases: [],
    };
}
exports.THORName = {
    encode: function (message, writer) {
        var e_1, _a;
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (!message.expireBlockHeight.isZero()) {
            writer.uint32(16).int64(message.expireBlockHeight);
        }
        if (message.owner.length !== 0) {
            writer.uint32(26).bytes(message.owner);
        }
        if (message.preferredAsset !== undefined) {
            common_1.Asset.encode(message.preferredAsset, writer.uint32(34).fork()).ldelim();
        }
        try {
            for (var _b = __values(message.aliases), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                exports.THORNameAlias.encode(v, writer.uint32(42).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseTHORName();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.name = reader.string();
                    break;
                case 2:
                    message.expireBlockHeight = reader.int64();
                    break;
                case 3:
                    message.owner = reader.bytes();
                    break;
                case 4:
                    message.preferredAsset = common_1.Asset.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.aliases.push(exports.THORNameAlias.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            name: isSet(object.name) ? String(object.name) : "",
            expireBlockHeight: isSet(object.expireBlockHeight)
                ? long_1.default.fromString(object.expireBlockHeight)
                : long_1.default.ZERO,
            owner: isSet(object.owner)
                ? bytesFromBase64(object.owner)
                : new Uint8Array(),
            preferredAsset: isSet(object.preferredAsset)
                ? common_1.Asset.fromJSON(object.preferredAsset)
                : undefined,
            aliases: Array.isArray(object === null || object === void 0 ? void 0 : object.aliases)
                ? object.aliases.map(function (e) { return exports.THORNameAlias.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.name !== undefined && (obj.name = message.name);
        message.expireBlockHeight !== undefined &&
            (obj.expireBlockHeight = (message.expireBlockHeight || long_1.default.ZERO).toString());
        message.owner !== undefined &&
            (obj.owner = base64FromBytes(message.owner !== undefined ? message.owner : new Uint8Array()));
        message.preferredAsset !== undefined &&
            (obj.preferredAsset = message.preferredAsset
                ? common_1.Asset.toJSON(message.preferredAsset)
                : undefined);
        if (message.aliases) {
            obj.aliases = message.aliases.map(function (e) {
                return e ? exports.THORNameAlias.toJSON(e) : undefined;
            });
        }
        else {
            obj.aliases = [];
        }
        return obj;
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseTHORName();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : "";
        message.expireBlockHeight =
            object.expireBlockHeight !== undefined &&
                object.expireBlockHeight !== null
                ? long_1.default.fromValue(object.expireBlockHeight)
                : long_1.default.ZERO;
        message.owner = (_b = object.owner) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.preferredAsset =
            object.preferredAsset !== undefined && object.preferredAsset !== null
                ? common_1.Asset.fromPartial(object.preferredAsset)
                : undefined;
        message.aliases =
            ((_c = object.aliases) === null || _c === void 0 ? void 0 : _c.map(function (e) { return exports.THORNameAlias.fromPartial(e); })) || [];
        return message;
    },
};
var globalThis = (function () {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
var atob = globalThis.atob ||
    (function (b64) { return globalThis.Buffer.from(b64, "base64").toString("binary"); });
function bytesFromBase64(b64) {
    var bin = atob(b64);
    var arr = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
var btoa = globalThis.btoa ||
    (function (bin) { return globalThis.Buffer.from(bin, "binary").toString("base64"); });
function base64FromBytes(arr) {
    var e_2, _a;
    var bin = [];
    try {
        for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
            var byte = arr_1_1.value;
            bin.push(String.fromCharCode(byte));
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
