// @coinmasters/toolbox-evm@4.0.4/dist/index.es.js
import { aY as o, aZ as t, a_ as b, aS as l, aT as r, a$ as d, bi as i, b1 as A, aW as c, b2 as T, bb as h, b7 as n, b3 as x, ba as g, b4 as k, bg as w, aU as p, bc as M, bh as W, aV as m, b0 as C, b9 as E, bd as S, be as V, b5 as u, b8 as B, aX as N, bf as P, b6 as f } from "./index-ff937b1e.js";
import "@coinmasters/helpers";
export {
  o as ARBToolbox,
  t as AVAXToolbox,
  b as BSCToolbox,
  l as BaseEVMToolbox,
  r as EIP1193SendTransaction,
  d as ETHToolbox,
  i as EthNetwork,
  A as MATICToolbox,
  c as MAX_APPROVAL,
  T as OPToolbox,
  h as addAccountsChangedCallback,
  n as addEVMWalletNetwork,
  x as covalentApi,
  g as estimateMaxSendableAmount,
  k as ethplorerApi,
  w as getBalance,
  p as getChecksumAddressFromAsset,
  M as getETHDefaultWallet,
  W as getProvider,
  m as getTokenAddress,
  C as getToolboxByChain,
  E as getWeb3WalletMethods,
  S as isDetected,
  V as isWeb3Detected,
  u as prepareNetworkSwitch,
  B as switchEVMWalletNetwork,
  N as toChecksumAddress,
  P as toHexString,
  f as wrapMethodWithNetworkSwitch
};
