// @trezor/connect@9.1.1/lib/utils/deferred.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rejectTimeoutPromise = exports.resolveTimeoutPromise = exports.createAsync = exports.create = void 0;
function create(arg, data) {
    let localResolve = (_t) => { };
    let localReject = (_e) => { };
    let id;
    const promise = new Promise(async (resolve, reject) => {
        localResolve = resolve;
        localReject = reject;
        if (typeof arg === 'function') {
            try {
                await arg();
            }
            catch (error) {
                reject(error);
            }
        }
        if (typeof arg === 'string')
            id = arg;
    });
    return {
        id,
        data,
        resolve: localResolve,
        reject: localReject,
        promise,
    };
}
exports.create = create;
function createAsync(innerFn) {
    let localResolve = (_t) => { };
    let localReject = _e => { };
    const promise = new Promise((resolve, reject) => {
        localResolve = resolve;
        localReject = reject;
    });
    const inner = async () => {
        await innerFn();
    };
    return {
        resolve: localResolve,
        reject: localReject,
        promise,
        run: () => {
            inner();
            return promise;
        },
    };
}
exports.createAsync = createAsync;
function resolveTimeoutPromise(delay, result) {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve(result);
        }, delay);
    });
}
exports.resolveTimeoutPromise = resolveTimeoutPromise;
const rejectTimeoutPromise = (delay, error) => new Promise((_resolve, reject) => {
    setTimeout(() => {
        reject(error);
    }, delay);
});
exports.rejectTimeoutPromise = rejectTimeoutPromise;
