// @shapeshiftoss/hdwallet-metamask@1.52.14/dist/adapter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetaMaskAdapter = void 0;
const detect_provider_1 = __importDefault(require("@metamask/detect-provider"));
const onboarding_1 = __importDefault(require("@metamask/onboarding"));
const core = __importStar(require("@shapeshiftoss/hdwallet-core"));
const metamask_1 = require("./metamask");
class MetaMaskAdapter {
    constructor(keyring) {
        this.keyring = keyring;
    }
    static useKeyring(keyring) {
        return new MetaMaskAdapter(keyring);
    }
    initialize() {
        return __awaiter(this, void 0, void 0, function* () {
            return Object.keys(this.keyring.wallets).length;
        });
    }
    pairDevice() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            // TODO: remove casting, @metamask/detect-provider npm package hasn't been published with latest typings
            // https://github.com/MetaMask/detect-provider/blame/5ce916fc24779c4b36741531a41d9c8b3cbb0a17/src/index.ts#L37
            const provider = (yield (0, detect_provider_1.default)({
                mustBeMetaMask: true,
                silent: false,
                timeout: 3000,
            }));
            if (!provider) {
                const onboarding = new onboarding_1.default();
                onboarding.startOnboarding();
                console.error("Please install MetaMask!");
                throw new Error("MetaMask provider not found");
            }
            try {
                yield ((_a = provider.request) === null || _a === void 0 ? void 0 : _a.call(provider, { method: "eth_requestAccounts" }).catch(() => {
                    var _a;
                    return (_a = provider.request) === null || _a === void 0 ? void 0 : _a.call(provider, {
                        method: "wallet_requestPermissions",
                        params: [{ eth_accounts: {} }],
                    });
                }));
            }
            catch (error) {
                console.error("Could not get MetaMask accounts. ");
                throw error;
            }
            const wallet = new metamask_1.MetaMaskHDWallet(provider);
            yield wallet.initialize();
            const deviceID = yield wallet.getDeviceID();
            this.keyring.add(wallet, deviceID);
            this.keyring.emit(["MetaMask", deviceID, core.Events.CONNECT], deviceID);
            return wallet;
        });
    }
}
exports.MetaMaskAdapter = MetaMaskAdapter;
