// @trezor/connect@9.1.1/lib/backend/BlockchainLink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dispose = exports.reconnectAllBackends = exports.initBlockchain = exports.isBackendSupported = exports.setCustomBackend = exports.findBackend = exports.Blockchain = void 0;
const BackendManager_1 = require("./BackendManager");
var Blockchain_1 = require("./Blockchain");
Object.defineProperty(exports, "Blockchain", { enumerable: true, get: function () { return Blockchain_1.Blockchain; } });
const backends = new BackendManager_1.BackendManager();
const findBackend = (coin) => backends.get(coin);
exports.findBackend = findBackend;
const setCustomBackend = (coinInfo, blockchainLink) => (blockchainLink === null || blockchainLink === void 0 ? void 0 : blockchainLink.url.length)
    ? backends.setCustom(coinInfo.shortcut, blockchainLink)
    : backends.removeCustom(coinInfo.shortcut);
exports.setCustomBackend = setCustomBackend;
const isBackendSupported = (coinInfo) => backends.isSupported(coinInfo);
exports.isBackendSupported = isBackendSupported;
const initBlockchain = (coinInfo, postMessage) => backends.getOrConnect(coinInfo, postMessage);
exports.initBlockchain = initBlockchain;
const reconnectAllBackends = () => backends.reconnectAll();
exports.reconnectAllBackends = reconnectAllBackends;
const dispose = () => backends.dispose();
exports.dispose = dispose;
