// pioneer-platform@1.0.0/apps/pioneer-cli/lib/modules/config.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logLevel = exports.updateConfig = exports.setConfig = exports.getConfig = exports.getWallet = exports.checkConfigs = exports.backupWallet = exports.initWallet = exports.getApps = exports.initApps = exports.innitConfig = exports.logDir = exports.appDir = exports.backtestDir = exports.modelDir = exports.pioneerPath = exports.seedDir = exports.seedPath = exports.configPath = exports.pioneerConfig = void 0;
var path_1 = __importDefault(require("path"));
var TAG = " | Config | ";
var fs = require("fs-extra");
var homedir = require("os").homedir();
var mkdirp = require("mkdirp");
var finder = require("find-package-json");
var log = require("loggerdog-client")();
var uuid_1 = require("uuid");
exports.pioneerConfig = path_1.default.join(homedir, ".pioneer", "pioneer.json");
exports.configPath = path_1.default.join(homedir, ".pioneer", "pioneer.json");
exports.seedPath = path_1.default.join(homedir, ".pioneer", "wallet_data/wallet.seed");
exports.seedDir = path_1.default.join(homedir, ".pioneer", "wallet_data");
exports.pioneerPath = path_1.default.join(homedir, ".pioneer");
exports.modelDir = path_1.default.join(homedir, ".pioneer", "models");
exports.backtestDir = path_1.default.join(homedir, ".pioneer", "backtest");
exports.appDir = path_1.default.join(homedir, '.pioneer', 'apps');
exports.logDir = path_1.default.join(exports.pioneerPath, "log");
function innitConfig(languageSelected) {
    return __awaiter(this, void 0, void 0, function () {
        var tag, output, isCreated, isCreated2, isCreated3, queryKey, config, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tag = TAG + " | importConfig | ";
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 5, , 6]);
                    output = {};
                    return [4 /*yield*/, mkdirp(exports.pioneerPath)];
                case 2:
                    isCreated = _a.sent();
                    return [4 /*yield*/, mkdirp(exports.logDir)];
                case 3:
                    isCreated2 = _a.sent();
                    return [4 /*yield*/, mkdirp(exports.seedDir)];
                case 4:
                    isCreated3 = _a.sent();
                    queryKey = uuid_1.v4();
                    config = {};
                    config.locale = "english";
                    config.localeSelected = true;
                    config.queryKey = queryKey;
                    //config.version = finder.next().value.version;
                    config.isCli = true;
                    fs.writeFileSync(exports.pioneerConfig, JSON.stringify(config));
                    return [3 /*break*/, 6];
                case 5:
                    e_1 = _a.sent();
                    console.error(tag, "e: ", e_1);
                    return [2 /*return*/, {}];
                case 6: return [2 /*return*/];
            }
        });
    });
}
exports.innitConfig = innitConfig;
// innit Apps
function initApps() {
    return __awaiter(this, void 0, void 0, function () {
        var tag, apps, e_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tag = TAG + ' | initApps | ';
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    mkdirp.sync(exports.appDir);
                    console.log('appDir: ', exports.appDir);
                    return [4 /*yield*/, fs.readdir(exports.appDir)];
                case 2:
                    apps = _a.sent();
                    return [2 /*return*/, apps];
                case 3:
                    e_2 = _a.sent();
                    console.error(tag, 'e: ', e_2);
                    return [2 /*return*/, {}];
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.initApps = initApps;
// getApps
function getApps() {
    return __awaiter(this, void 0, void 0, function () {
        var tag, apps, e_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tag = TAG + ' | getApps | ';
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, fs.readdir(exports.appDir)];
                case 2:
                    apps = _a.sent();
                    return [2 /*return*/, apps];
                case 3:
                    e_3 = _a.sent();
                    console.error(tag, 'e: ', e_3);
                    return [2 /*return*/, {}];
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.getApps = getApps;
//innit Wallet
function initWallet(encryptedSeed, passwordHash) {
    return __awaiter(this, void 0, void 0, function () {
        var tag, isCreated, output, wallet, result, e_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tag = TAG + " | initWallet | ";
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    log.debug(tag, "seedDir: ", exports.seedDir);
                    return [4 /*yield*/, mkdirp(exports.seedDir)];
                case 2:
                    isCreated = _a.sent();
                    output = {};
                    wallet = {};
                    wallet.hash = passwordHash;
                    wallet.version = 1;
                    wallet.type = "seedwords";
                    wallet.vault = encryptedSeed;
                    result = fs.writeFileSync(exports.seedPath, JSON.stringify(wallet));
                    //console.log("result: ", result);
                    return [2 /*return*/, true];
                case 3:
                    e_4 = _a.sent();
                    console.error(tag, "e: ", e_4);
                    return [2 /*return*/, {}];
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.initWallet = initWallet;
//backup Wallet
function backupWallet(outDir, passwordHash, encryptedSeed, walletId) {
    return __awaiter(this, void 0, void 0, function () {
        var tag, isCreated, wallet, result, e_5;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tag = TAG + " | backupWallet | ";
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    log.debug(tag, "outDir: ", outDir);
                    return [4 /*yield*/, mkdirp(outDir)];
                case 2:
                    isCreated = _a.sent();
                    wallet = {};
                    wallet.hash = passwordHash;
                    wallet.version = 1;
                    wallet.type = "seedwords";
                    wallet.vault = encryptedSeed;
                    result = fs.writeFileSync(outDir + "/" + walletId + ".wallet.json", JSON.stringify(wallet));
                    //console.log("result: ", result);
                    return [2 /*return*/, result];
                case 3:
                    e_5 = _a.sent();
                    console.error(tag, "e: ", e_5);
                    return [2 /*return*/, {}];
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.backupWallet = backupWallet;
//check
function checkConfigs() {
    var output = {};
    output.isConfigured = false;
    output.isWallet = false;
    output.isRegistered = false;
    var fileFound = fs.existsSync(exports.pioneerConfig) ? true : false;
    if (fileFound) {
        output.config = JSON.parse(fs.readFileSync(exports.configPath));
        if (output.config.version)
            output.isConfigured = true;
        if (output.config.username)
            output.isRegistered = true;
    }
    if (output.config && output.config.version)
        output.isConfigured = true;
    //wallet found?
    var walletFound = fs.existsSync(exports.seedPath) ? true : false;
    if (walletFound) {
        output.isWallet = true;
    }
    return output;
}
exports.checkConfigs = checkConfigs;
function getWallet(path) {
    try {
        if (!path)
            path = exports.seedPath;
        var walletBuff = fs.readFileSync(path);
        var walletString = walletBuff.toString();
        var wallet = JSON.parse(walletString);
        return wallet;
    }
    catch (e) {
        return {};
    }
}
exports.getWallet = getWallet;
function getConfig() {
    try {
        var output = JSON.parse(fs.readFileSync(exports.configPath));
        return output;
    }
    catch (e) {
        return {};
    }
}
exports.getConfig = getConfig;
function setConfig(options) {
    return fs.writeFileSync(exports.configPath, JSON.stringify(options));
}
exports.setConfig = setConfig;
function updateConfig(options) {
    var options_ = getConfig();
    for (var key in options) {
        options_[key] = options[key];
    }
    setConfig(options_);
}
exports.updateConfig = updateConfig;
//export const logLevel = getConfig()['debug']?'debug':'info'
exports.logLevel = "debug";
