// osmojs@0.37.0/main/codegen/osmosis/gamm/v1beta1/tx.rpc.msg.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MsgClientImpl = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));

var _tx = require("./tx");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MsgClientImpl = /*#__PURE__*/function () {
  function MsgClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, MsgClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.joinPool = this.joinPool.bind(this);
    this.exitPool = this.exitPool.bind(this);
    this.swapExactAmountIn = this.swapExactAmountIn.bind(this);
    this.swapExactAmountOut = this.swapExactAmountOut.bind(this);
    this.joinSwapExternAmountIn = this.joinSwapExternAmountIn.bind(this);
    this.joinSwapShareAmountOut = this.joinSwapShareAmountOut.bind(this);
    this.exitSwapExternAmountOut = this.exitSwapExternAmountOut.bind(this);
    this.exitSwapShareAmountIn = this.exitSwapShareAmountIn.bind(this);
  }

  (0, _createClass2["default"])(MsgClientImpl, [{
    key: "joinPool",
    value: function joinPool(request) {
      var data = _tx.MsgJoinPool.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "JoinPool", data);
      return promise.then(function (data) {
        return _tx.MsgJoinPoolResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "exitPool",
    value: function exitPool(request) {
      var data = _tx.MsgExitPool.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "ExitPool", data);
      return promise.then(function (data) {
        return _tx.MsgExitPoolResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "swapExactAmountIn",
    value: function swapExactAmountIn(request) {
      var data = _tx.MsgSwapExactAmountIn.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "SwapExactAmountIn", data);
      return promise.then(function (data) {
        return _tx.MsgSwapExactAmountInResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "swapExactAmountOut",
    value: function swapExactAmountOut(request) {
      var data = _tx.MsgSwapExactAmountOut.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "SwapExactAmountOut", data);
      return promise.then(function (data) {
        return _tx.MsgSwapExactAmountOutResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "joinSwapExternAmountIn",
    value: function joinSwapExternAmountIn(request) {
      var data = _tx.MsgJoinSwapExternAmountIn.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "JoinSwapExternAmountIn", data);
      return promise.then(function (data) {
        return _tx.MsgJoinSwapExternAmountInResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "joinSwapShareAmountOut",
    value: function joinSwapShareAmountOut(request) {
      var data = _tx.MsgJoinSwapShareAmountOut.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "JoinSwapShareAmountOut", data);
      return promise.then(function (data) {
        return _tx.MsgJoinSwapShareAmountOutResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "exitSwapExternAmountOut",
    value: function exitSwapExternAmountOut(request) {
      var data = _tx.MsgExitSwapExternAmountOut.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "ExitSwapExternAmountOut", data);
      return promise.then(function (data) {
        return _tx.MsgExitSwapExternAmountOutResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "exitSwapShareAmountIn",
    value: function exitSwapShareAmountIn(request) {
      var data = _tx.MsgExitSwapShareAmountIn.encode(request).finish();

      var promise = this.rpc.request("osmosis.gamm.v1beta1.Msg", "ExitSwapShareAmountIn", data);
      return promise.then(function (data) {
        return _tx.MsgExitSwapShareAmountInResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return MsgClientImpl;
}();

exports.MsgClientImpl = MsgClientImpl;
