// osmojs@0.37.0/main/codegen/osmosis/streamswap/v1/tx.rpc.msg.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MsgClientImpl = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));

var _tx = require("./tx");

var _empty = require("../../../google/protobuf/empty");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MsgClientImpl = /*#__PURE__*/function () {
  function MsgClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, MsgClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.createSale = this.createSale.bind(this);
    this.subscribe = this.subscribe.bind(this);
    this.withdraw = this.withdraw.bind(this);
    this.exitSale = this.exitSale.bind(this);
    this.finalizeSale = this.finalizeSale.bind(this);
  }

  (0, _createClass2["default"])(MsgClientImpl, [{
    key: "createSale",
    value: function createSale(request) {
      var data = _tx.MsgCreateSale.encode(request).finish();

      var promise = this.rpc.request("osmosis.streamswap.v1.Msg", "CreateSale", data);
      return promise.then(function (data) {
        return _tx.MsgCreateSaleResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "subscribe",
    value: function subscribe(request) {
      var data = _tx.MsgSubscribe.encode(request).finish();

      var promise = this.rpc.request("osmosis.streamswap.v1.Msg", "Subscribe", data);
      return promise.then(function (data) {
        return _empty.Empty.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "withdraw",
    value: function withdraw(request) {
      var data = _tx.MsgWithdraw.encode(request).finish();

      var promise = this.rpc.request("osmosis.streamswap.v1.Msg", "Withdraw", data);
      return promise.then(function (data) {
        return _empty.Empty.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "exitSale",
    value: function exitSale(request) {
      var data = _tx.MsgExitSale.encode(request).finish();

      var promise = this.rpc.request("osmosis.streamswap.v1.Msg", "ExitSale", data);
      return promise.then(function (data) {
        return _tx.MsgExitSaleResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "finalizeSale",
    value: function finalizeSale(request) {
      var data = _tx.MsgFinalizeSale.encode(request).finish();

      var promise = this.rpc.request("osmosis.streamswap.v1.Msg", "FinalizeSale", data);
      return promise.then(function (data) {
        return _tx.MsgFinalizeSaleResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return MsgClientImpl;
}();

exports.MsgClientImpl = MsgClientImpl;
