// osmojs@0.37.0/main/codegen/osmosis/streamswap/v1/tx.registry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registry = exports.load = exports.MessageComposer = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _tx = require("./tx");

//@ts-nocheck
var registry = [["/osmosis.streamswap.v1.MsgCreateSale", _tx.MsgCreateSale], ["/osmosis.streamswap.v1.MsgSubscribe", _tx.MsgSubscribe], ["/osmosis.streamswap.v1.MsgWithdraw", _tx.MsgWithdraw], ["/osmosis.streamswap.v1.MsgExitSale", _tx.MsgExitSale], ["/osmosis.streamswap.v1.MsgFinalizeSale", _tx.MsgFinalizeSale]];
exports.registry = registry;

var load = function load(protoRegistry) {
  registry.forEach(function (_ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
        typeUrl = _ref2[0],
        mod = _ref2[1];

    protoRegistry.register(typeUrl, mod);
  });
};

exports.load = load;
var MessageComposer = {
  encoded: {
    createSale: function createSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgCreateSale",
        value: _tx.MsgCreateSale.encode(value).finish()
      };
    },
    subscribe: function subscribe(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgSubscribe",
        value: _tx.MsgSubscribe.encode(value).finish()
      };
    },
    withdraw: function withdraw(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgWithdraw",
        value: _tx.MsgWithdraw.encode(value).finish()
      };
    },
    exitSale: function exitSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgExitSale",
        value: _tx.MsgExitSale.encode(value).finish()
      };
    },
    finalizeSale: function finalizeSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgFinalizeSale",
        value: _tx.MsgFinalizeSale.encode(value).finish()
      };
    }
  },
  withTypeUrl: {
    createSale: function createSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgCreateSale",
        value: value
      };
    },
    subscribe: function subscribe(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgSubscribe",
        value: value
      };
    },
    withdraw: function withdraw(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgWithdraw",
        value: value
      };
    },
    exitSale: function exitSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgExitSale",
        value: value
      };
    },
    finalizeSale: function finalizeSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgFinalizeSale",
        value: value
      };
    }
  },
  fromPartial: {
    createSale: function createSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgCreateSale",
        value: _tx.MsgCreateSale.fromPartial(value)
      };
    },
    subscribe: function subscribe(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgSubscribe",
        value: _tx.MsgSubscribe.fromPartial(value)
      };
    },
    withdraw: function withdraw(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgWithdraw",
        value: _tx.MsgWithdraw.fromPartial(value)
      };
    },
    exitSale: function exitSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgExitSale",
        value: _tx.MsgExitSale.fromPartial(value)
      };
    },
    finalizeSale: function finalizeSale(value) {
      return {
        typeUrl: "/osmosis.streamswap.v1.MsgFinalizeSale",
        value: _tx.MsgFinalizeSale.fromPartial(value)
      };
    }
  }
};
exports.MessageComposer = MessageComposer;
