// @trezor/blockchain-link-utils@1.0.5/lib/ripple.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformTransaction = exports.transformServerInfo = void 0;
const transformServerInfo = (payload) => ({
    name: 'Ripple',
    shortcut: 'xrp',
    testnet: false,
    version: payload.buildVersion,
    decimals: 6,
    blockHeight: payload.validatedLedger.ledgerVersion,
    blockHash: payload.validatedLedger.hash,
});
exports.transformServerInfo = transformServerInfo;
const BLOCKTIME_OFFSET = 946684800;
const transformTransaction = (descriptor, tx) => {
    const blockTime = typeof tx.date === 'number' && tx.date > 0 ? tx.date + BLOCKTIME_OFFSET : tx.date;
    if (tx.TransactionType !== 'Payment') {
        return {
            type: 'unknown',
            txid: tx.hash,
            amount: '0',
            fee: '0',
            blockTime,
            blockHeight: tx.ledger_index,
            blockHash: tx.hash,
            targets: [],
            tokens: [],
            internalTransfers: [],
            feeRate: undefined,
            details: {
                vin: [],
                vout: [],
                size: 0,
                totalInput: '0',
                totalOutput: '0',
            },
        };
    }
    const type = tx.Account === descriptor ? 'sent' : 'recv';
    const addresses = [tx.Destination];
    const amount = tx.Amount;
    const fee = tx.Fee;
    return {
        type,
        txid: tx.hash,
        blockTime,
        blockHeight: tx.ledger_index,
        blockHash: tx.hash,
        amount,
        fee,
        targets: [
            {
                addresses,
                isAddress: true,
                amount,
                n: 0,
            },
        ],
        tokens: [],
        internalTransfers: [],
        feeRate: undefined,
        details: {
            vin: [],
            vout: [],
            size: 0,
            totalInput: '0',
            totalOutput: '0',
        },
    };
};
exports.transformTransaction = transformTransaction;
