// @shapeshiftoss/metamask-snaps-types@1.0.8/dist/snap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedChainIds = exports.NativeHDWallet = void 0;
var hdwallet_native_1 = require("@shapeshiftoss/hdwallet-native");
Object.defineProperty(exports, "NativeHDWallet", { enumerable: true, get: function () { return hdwallet_native_1.NativeHDWallet; } });
// TODO: Replace local definition with KnownChainIds imported from @shapeshiftoss/types once all chains here are added to KnownChainIds
var SupportedChainIds;
(function (SupportedChainIds) {
    SupportedChainIds["AvalancheMainnet"] = "eip155:43114";
    SupportedChainIds["BinanceMainnet"] = "cosmos:binance-chain-tigris";
    SupportedChainIds["BitcoinCashMainnet"] = "bip122:000000000000000000651ef99cb9fcbe";
    SupportedChainIds["BitcoinMainnet"] = "bip122:000000000019d6689c085ae165831e93";
    SupportedChainIds["CosmosMainnet"] = "cosmos:cosmoshub-4";
    SupportedChainIds["DogecoinMainnet"] = "bip122:00000000001a91e3dace36e2be3bf030";
    SupportedChainIds["EthereumMainnet"] = "eip155:1";
    SupportedChainIds["KavaMainnet"] = "cosmos:kava_2222-10";
    SupportedChainIds["LitecoinMainnet"] = "bip122:12a765e31ffd4059bada1e25190f6e98";
    SupportedChainIds["OsmosisMainnet"] = "cosmos:osmosis-1";
    SupportedChainIds["SecretMainnet"] = "cosmos:secret-4";
    SupportedChainIds["TerraMainnet"] = "cosmos:phoenix-1";
    SupportedChainIds["ThorchainMainnet"] = "cosmos:thorchain-mainnet-v1";
})(SupportedChainIds = exports.SupportedChainIds || (exports.SupportedChainIds = {}));
const CosmosSDKChainIdList = [
    SupportedChainIds.BinanceMainnet,
    SupportedChainIds.CosmosMainnet,
    SupportedChainIds.KavaMainnet,
    SupportedChainIds.OsmosisMainnet,
    SupportedChainIds.SecretMainnet,
    SupportedChainIds.TerraMainnet,
    SupportedChainIds.ThorchainMainnet,
];
const EVMChainIdList = [
    SupportedChainIds.AvalancheMainnet,
    SupportedChainIds.EthereumMainnet,
];
const UTXOChainIdList = [
    SupportedChainIds.BitcoinMainnet,
    SupportedChainIds.BitcoinCashMainnet,
    SupportedChainIds.DogecoinMainnet,
    SupportedChainIds.LitecoinMainnet,
];
