// @pioneer-platform/trezor@3.0.1/dist/index-dea8b114.js
import { a as B, h as M, i as ft, _ as dt, b as At, B as Ge, s as je, c as S, d as ct, e as wt } from "./bignumber-9ccd6dd9.js";
import { c as st, g as kt, L as ze, a as Et } from "./index-e4de2709.js";
import "events";
import "buffer";
import "util";
import "string_decoder";
var ht = { exports: {} };
/**
 * [js-sha3]{@link https://github.com/emn178/js-sha3}
 *
 * @version 0.8.0
 * @author Chen, Yi-Cyuan [emn178@gmail.com]
 * @copyright Chen, Yi-Cyuan 2015-2018
 * @license MIT
 */
(function(t) {
  (function() {
    var i = "input is invalid type", u = "finalize already called", h = typeof window == "object", y = h ? window : {};
    y.JS_SHA3_NO_WINDOW && (h = !1);
    var k = !h && typeof self == "object", d = !y.JS_SHA3_NO_NODE_JS && typeof process == "object" && process.versions && process.versions.node;
    d ? y = st : k && (y = self);
    var A = !y.JS_SHA3_NO_COMMON_JS && !0 && t.exports, E = !y.JS_SHA3_NO_ARRAY_BUFFER && typeof ArrayBuffer < "u", v = "0123456789abcdef".split(""), yt = [31, 7936, 2031616, 520093696], Te = [4, 1024, 262144, 67108864], xt = [1, 256, 65536, 16777216], gt = [6, 1536, 393216, 100663296], w = [0, 8, 16, 24], We = [
      1,
      0,
      32898,
      0,
      32906,
      2147483648,
      2147516416,
      2147483648,
      32907,
      0,
      2147483649,
      0,
      2147516545,
      2147483648,
      32777,
      2147483648,
      138,
      0,
      136,
      0,
      2147516425,
      0,
      2147483658,
      0,
      2147516555,
      0,
      139,
      2147483648,
      32905,
      2147483648,
      32771,
      2147483648,
      32770,
      2147483648,
      128,
      2147483648,
      32778,
      0,
      2147483658,
      2147483648,
      2147516545,
      2147483648,
      32896,
      2147483648,
      2147483649,
      0,
      2147516424,
      2147483648
    ], Ve = [224, 256, 384, 512], De = [128, 256], Ye = ["hex", "buffer", "arrayBuffer", "array", "digest"], Xe = {
      128: 168,
      256: 136
    };
    (y.JS_SHA3_NO_NODE_JS || !Array.isArray) && (Array.isArray = function(e) {
      return Object.prototype.toString.call(e) === "[object Array]";
    }), E && (y.JS_SHA3_NO_ARRAY_BUFFER_IS_VIEW || !ArrayBuffer.isView) && (ArrayBuffer.isView = function(e) {
      return typeof e == "object" && e.buffer && e.buffer.constructor === ArrayBuffer;
    });
    for (var Ze = function(e, r, n) {
      return function(o) {
        return new g(e, r, e).update(o)[n]();
      };
    }, qe = function(e, r, n) {
      return function(o, a) {
        return new g(e, r, a).update(o)[n]();
      };
    }, Qe = function(e, r, n) {
      return function(o, a, c, l) {
        return F["cshake" + e].update(o, a, c, l)[n]();
      };
    }, et = function(e, r, n) {
      return function(o, a, c, l) {
        return F["kmac" + e].update(o, a, c, l)[n]();
      };
    }, N = function(e, r, n, o) {
      for (var a = 0; a < Ye.length; ++a) {
        var c = Ye[a];
        e[c] = r(n, o, c);
      }
      return e;
    }, tt = function(e, r) {
      var n = Ze(e, r, "hex");
      return n.create = function() {
        return new g(e, r, e);
      }, n.update = function(o) {
        return n.create().update(o);
      }, N(n, Ze, e, r);
    }, bt = function(e, r) {
      var n = qe(e, r, "hex");
      return n.create = function(o) {
        return new g(e, r, o);
      }, n.update = function(o, a) {
        return n.create(a).update(o);
      }, N(n, qe, e, r);
    }, vt = function(e, r) {
      var n = Xe[e], o = Qe(e, r, "hex");
      return o.create = function(a, c, l) {
        return !c && !l ? F["shake" + e].create(a) : new g(e, r, a).bytepad([c, l], n);
      }, o.update = function(a, c, l, f) {
        return o.create(c, l, f).update(a);
      }, N(o, Qe, e, r);
    }, mt = function(e, r) {
      var n = Xe[e], o = et(e, r, "hex");
      return o.create = function(a, c, l) {
        return new He(e, r, c).bytepad(["KMAC", l], n).bytepad([a], n);
      }, o.update = function(a, c, l, f) {
        return o.create(a, l, f).update(c);
      }, N(o, et, e, r);
    }, rt = [
      { name: "keccak", padding: xt, bits: Ve, createMethod: tt },
      { name: "sha3", padding: gt, bits: Ve, createMethod: tt },
      { name: "shake", padding: yt, bits: De, createMethod: bt },
      { name: "cshake", padding: Te, bits: De, createMethod: vt },
      { name: "kmac", padding: Te, bits: De, createMethod: mt }
    ], F = {}, P = [], _ = 0; _ < rt.length; ++_)
      for (var C = rt[_], R = C.bits, L = 0; L < R.length; ++L) {
        var Ke = C.name + "_" + R[L];
        if (P.push(Ke), F[Ke] = C.createMethod(R[L], C.padding), C.name !== "sha3") {
          var nt = C.name + R[L];
          P.push(nt), F[nt] = F[Ke];
        }
      }
    function g(e, r, n) {
      this.blocks = [], this.s = [], this.padding = r, this.outputBits = n, this.reset = !0, this.finalized = !1, this.block = 0, this.start = 0, this.blockCount = 1600 - (e << 1) >> 5, this.byteCount = this.blockCount << 2, this.outputBlocks = n >> 5, this.extraBytes = (n & 31) >> 3;
      for (var o = 0; o < 50; ++o)
        this.s[o] = 0;
    }
    g.prototype.update = function(e) {
      if (this.finalized)
        throw new Error(u);
      var r, n = typeof e;
      if (n !== "string") {
        if (n === "object") {
          if (e === null)
            throw new Error(i);
          if (E && e.constructor === ArrayBuffer)
            e = new Uint8Array(e);
          else if (!Array.isArray(e) && (!E || !ArrayBuffer.isView(e)))
            throw new Error(i);
        } else
          throw new Error(i);
        r = !0;
      }
      for (var o = this.blocks, a = this.byteCount, c = e.length, l = this.blockCount, f = 0, b = this.s, p, x; f < c; ) {
        if (this.reset)
          for (this.reset = !1, o[0] = this.block, p = 1; p < l + 1; ++p)
            o[p] = 0;
        if (r)
          for (p = this.start; f < c && p < a; ++f)
            o[p >> 2] |= e[f] << w[p++ & 3];
        else
          for (p = this.start; f < c && p < a; ++f)
            x = e.charCodeAt(f), x < 128 ? o[p >> 2] |= x << w[p++ & 3] : x < 2048 ? (o[p >> 2] |= (192 | x >> 6) << w[p++ & 3], o[p >> 2] |= (128 | x & 63) << w[p++ & 3]) : x < 55296 || x >= 57344 ? (o[p >> 2] |= (224 | x >> 12) << w[p++ & 3], o[p >> 2] |= (128 | x >> 6 & 63) << w[p++ & 3], o[p >> 2] |= (128 | x & 63) << w[p++ & 3]) : (x = 65536 + ((x & 1023) << 10 | e.charCodeAt(++f) & 1023), o[p >> 2] |= (240 | x >> 18) << w[p++ & 3], o[p >> 2] |= (128 | x >> 12 & 63) << w[p++ & 3], o[p >> 2] |= (128 | x >> 6 & 63) << w[p++ & 3], o[p >> 2] |= (128 | x & 63) << w[p++ & 3]);
        if (this.lastByteIndex = p, p >= a) {
          for (this.start = p - a, this.block = o[l], p = 0; p < l; ++p)
            b[p] ^= o[p];
          O(b), this.reset = !0;
        } else
          this.start = p;
      }
      return this;
    }, g.prototype.encode = function(e, r) {
      var n = e & 255, o = 1, a = [n];
      for (e = e >> 8, n = e & 255; n > 0; )
        a.unshift(n), e = e >> 8, n = e & 255, ++o;
      return r ? a.push(o) : a.unshift(o), this.update(a), a.length;
    }, g.prototype.encodeString = function(e) {
      var r, n = typeof e;
      if (n !== "string") {
        if (n === "object") {
          if (e === null)
            throw new Error(i);
          if (E && e.constructor === ArrayBuffer)
            e = new Uint8Array(e);
          else if (!Array.isArray(e) && (!E || !ArrayBuffer.isView(e)))
            throw new Error(i);
        } else
          throw new Error(i);
        r = !0;
      }
      var o = 0, a = e.length;
      if (r)
        o = a;
      else
        for (var c = 0; c < e.length; ++c) {
          var l = e.charCodeAt(c);
          l < 128 ? o += 1 : l < 2048 ? o += 2 : l < 55296 || l >= 57344 ? o += 3 : (l = 65536 + ((l & 1023) << 10 | e.charCodeAt(++c) & 1023), o += 4);
        }
      return o += this.encode(o * 8), this.update(e), o;
    }, g.prototype.bytepad = function(e, r) {
      for (var n = this.encode(r), o = 0; o < e.length; ++o)
        n += this.encodeString(e[o]);
      var a = r - n % r, c = [];
      return c.length = a, this.update(c), this;
    }, g.prototype.finalize = function() {
      if (!this.finalized) {
        this.finalized = !0;
        var e = this.blocks, r = this.lastByteIndex, n = this.blockCount, o = this.s;
        if (e[r >> 2] |= this.padding[r & 3], this.lastByteIndex === this.byteCount)
          for (e[0] = e[n], r = 1; r < n + 1; ++r)
            e[r] = 0;
        for (e[n - 1] |= 2147483648, r = 0; r < n; ++r)
          o[r] ^= e[r];
        O(o);
      }
    }, g.prototype.toString = g.prototype.hex = function() {
      this.finalize();
      for (var e = this.blockCount, r = this.s, n = this.outputBlocks, o = this.extraBytes, a = 0, c = 0, l = "", f; c < n; ) {
        for (a = 0; a < e && c < n; ++a, ++c)
          f = r[a], l += v[f >> 4 & 15] + v[f & 15] + v[f >> 12 & 15] + v[f >> 8 & 15] + v[f >> 20 & 15] + v[f >> 16 & 15] + v[f >> 28 & 15] + v[f >> 24 & 15];
        c % e === 0 && (O(r), a = 0);
      }
      return o && (f = r[a], l += v[f >> 4 & 15] + v[f & 15], o > 1 && (l += v[f >> 12 & 15] + v[f >> 8 & 15]), o > 2 && (l += v[f >> 20 & 15] + v[f >> 16 & 15])), l;
    }, g.prototype.arrayBuffer = function() {
      this.finalize();
      var e = this.blockCount, r = this.s, n = this.outputBlocks, o = this.extraBytes, a = 0, c = 0, l = this.outputBits >> 3, f;
      o ? f = new ArrayBuffer(n + 1 << 2) : f = new ArrayBuffer(l);
      for (var b = new Uint32Array(f); c < n; ) {
        for (a = 0; a < e && c < n; ++a, ++c)
          b[c] = r[a];
        c % e === 0 && O(r);
      }
      return o && (b[a] = r[a], f = f.slice(0, l)), f;
    }, g.prototype.buffer = g.prototype.arrayBuffer, g.prototype.digest = g.prototype.array = function() {
      this.finalize();
      for (var e = this.blockCount, r = this.s, n = this.outputBlocks, o = this.extraBytes, a = 0, c = 0, l = [], f, b; c < n; ) {
        for (a = 0; a < e && c < n; ++a, ++c)
          f = c << 2, b = r[a], l[f] = b & 255, l[f + 1] = b >> 8 & 255, l[f + 2] = b >> 16 & 255, l[f + 3] = b >> 24 & 255;
        c % e === 0 && O(r);
      }
      return o && (f = c << 2, b = r[a], l[f] = b & 255, o > 1 && (l[f + 1] = b >> 8 & 255), o > 2 && (l[f + 2] = b >> 16 & 255)), l;
    };
    function He(e, r, n) {
      g.call(this, e, r, n);
    }
    He.prototype = new g(), He.prototype.finalize = function() {
      return this.encode(this.outputBits, !0), g.prototype.finalize.call(this);
    };
    var O = function(e) {
      var r, n, o, a, c, l, f, b, p, x, z, D, K, H, U, G, j, J, $, T, W, V, Y, X, Z, q, Q, ee, te, re, ne, oe, ie, ae, ue, fe, ce, he, le, pe, ye, xe, ge, be, ve, me, de, Ae, we, se, ke, Ee, Fe, _e, Se, Ce, Be, Pe, Le, Oe, Me, Ie, Ne;
      for (o = 0; o < 48; o += 2)
        a = e[0] ^ e[10] ^ e[20] ^ e[30] ^ e[40], c = e[1] ^ e[11] ^ e[21] ^ e[31] ^ e[41], l = e[2] ^ e[12] ^ e[22] ^ e[32] ^ e[42], f = e[3] ^ e[13] ^ e[23] ^ e[33] ^ e[43], b = e[4] ^ e[14] ^ e[24] ^ e[34] ^ e[44], p = e[5] ^ e[15] ^ e[25] ^ e[35] ^ e[45], x = e[6] ^ e[16] ^ e[26] ^ e[36] ^ e[46], z = e[7] ^ e[17] ^ e[27] ^ e[37] ^ e[47], D = e[8] ^ e[18] ^ e[28] ^ e[38] ^ e[48], K = e[9] ^ e[19] ^ e[29] ^ e[39] ^ e[49], r = D ^ (l << 1 | f >>> 31), n = K ^ (f << 1 | l >>> 31), e[0] ^= r, e[1] ^= n, e[10] ^= r, e[11] ^= n, e[20] ^= r, e[21] ^= n, e[30] ^= r, e[31] ^= n, e[40] ^= r, e[41] ^= n, r = a ^ (b << 1 | p >>> 31), n = c ^ (p << 1 | b >>> 31), e[2] ^= r, e[3] ^= n, e[12] ^= r, e[13] ^= n, e[22] ^= r, e[23] ^= n, e[32] ^= r, e[33] ^= n, e[42] ^= r, e[43] ^= n, r = l ^ (x << 1 | z >>> 31), n = f ^ (z << 1 | x >>> 31), e[4] ^= r, e[5] ^= n, e[14] ^= r, e[15] ^= n, e[24] ^= r, e[25] ^= n, e[34] ^= r, e[35] ^= n, e[44] ^= r, e[45] ^= n, r = b ^ (D << 1 | K >>> 31), n = p ^ (K << 1 | D >>> 31), e[6] ^= r, e[7] ^= n, e[16] ^= r, e[17] ^= n, e[26] ^= r, e[27] ^= n, e[36] ^= r, e[37] ^= n, e[46] ^= r, e[47] ^= n, r = x ^ (a << 1 | c >>> 31), n = z ^ (c << 1 | a >>> 31), e[8] ^= r, e[9] ^= n, e[18] ^= r, e[19] ^= n, e[28] ^= r, e[29] ^= n, e[38] ^= r, e[39] ^= n, e[48] ^= r, e[49] ^= n, H = e[0], U = e[1], me = e[11] << 4 | e[10] >>> 28, de = e[10] << 4 | e[11] >>> 28, ee = e[20] << 3 | e[21] >>> 29, te = e[21] << 3 | e[20] >>> 29, Oe = e[31] << 9 | e[30] >>> 23, Me = e[30] << 9 | e[31] >>> 23, xe = e[40] << 18 | e[41] >>> 14, ge = e[41] << 18 | e[40] >>> 14, ae = e[2] << 1 | e[3] >>> 31, ue = e[3] << 1 | e[2] >>> 31, G = e[13] << 12 | e[12] >>> 20, j = e[12] << 12 | e[13] >>> 20, Ae = e[22] << 10 | e[23] >>> 22, we = e[23] << 10 | e[22] >>> 22, re = e[33] << 13 | e[32] >>> 19, ne = e[32] << 13 | e[33] >>> 19, Ie = e[42] << 2 | e[43] >>> 30, Ne = e[43] << 2 | e[42] >>> 30, _e = e[5] << 30 | e[4] >>> 2, Se = e[4] << 30 | e[5] >>> 2, fe = e[14] << 6 | e[15] >>> 26, ce = e[15] << 6 | e[14] >>> 26, J = e[25] << 11 | e[24] >>> 21, $ = e[24] << 11 | e[25] >>> 21, se = e[34] << 15 | e[35] >>> 17, ke = e[35] << 15 | e[34] >>> 17, oe = e[45] << 29 | e[44] >>> 3, ie = e[44] << 29 | e[45] >>> 3, X = e[6] << 28 | e[7] >>> 4, Z = e[7] << 28 | e[6] >>> 4, Ce = e[17] << 23 | e[16] >>> 9, Be = e[16] << 23 | e[17] >>> 9, he = e[26] << 25 | e[27] >>> 7, le = e[27] << 25 | e[26] >>> 7, T = e[36] << 21 | e[37] >>> 11, W = e[37] << 21 | e[36] >>> 11, Ee = e[47] << 24 | e[46] >>> 8, Fe = e[46] << 24 | e[47] >>> 8, be = e[8] << 27 | e[9] >>> 5, ve = e[9] << 27 | e[8] >>> 5, q = e[18] << 20 | e[19] >>> 12, Q = e[19] << 20 | e[18] >>> 12, Pe = e[29] << 7 | e[28] >>> 25, Le = e[28] << 7 | e[29] >>> 25, pe = e[38] << 8 | e[39] >>> 24, ye = e[39] << 8 | e[38] >>> 24, V = e[48] << 14 | e[49] >>> 18, Y = e[49] << 14 | e[48] >>> 18, e[0] = H ^ ~G & J, e[1] = U ^ ~j & $, e[10] = X ^ ~q & ee, e[11] = Z ^ ~Q & te, e[20] = ae ^ ~fe & he, e[21] = ue ^ ~ce & le, e[30] = be ^ ~me & Ae, e[31] = ve ^ ~de & we, e[40] = _e ^ ~Ce & Pe, e[41] = Se ^ ~Be & Le, e[2] = G ^ ~J & T, e[3] = j ^ ~$ & W, e[12] = q ^ ~ee & re, e[13] = Q ^ ~te & ne, e[22] = fe ^ ~he & pe, e[23] = ce ^ ~le & ye, e[32] = me ^ ~Ae & se, e[33] = de ^ ~we & ke, e[42] = Ce ^ ~Pe & Oe, e[43] = Be ^ ~Le & Me, e[4] = J ^ ~T & V, e[5] = $ ^ ~W & Y, e[14] = ee ^ ~re & oe, e[15] = te ^ ~ne & ie, e[24] = he ^ ~pe & xe, e[25] = le ^ ~ye & ge, e[34] = Ae ^ ~se & Ee, e[35] = we ^ ~ke & Fe, e[44] = Pe ^ ~Oe & Ie, e[45] = Le ^ ~Me & Ne, e[6] = T ^ ~V & H, e[7] = W ^ ~Y & U, e[16] = re ^ ~oe & X, e[17] = ne ^ ~ie & Z, e[26] = pe ^ ~xe & ae, e[27] = ye ^ ~ge & ue, e[36] = se ^ ~Ee & be, e[37] = ke ^ ~Fe & ve, e[46] = Oe ^ ~Ie & _e, e[47] = Me ^ ~Ne & Se, e[8] = V ^ ~H & G, e[9] = Y ^ ~U & j, e[18] = oe ^ ~X & q, e[19] = ie ^ ~Z & Q, e[28] = xe ^ ~ae & fe, e[29] = ge ^ ~ue & ce, e[38] = Ee ^ ~be & me, e[39] = Fe ^ ~ve & de, e[48] = Ie ^ ~_e & Ce, e[49] = Ne ^ ~Se & Be, e[0] ^= We[o], e[1] ^= We[o + 1];
    };
    if (A)
      t.exports = F;
    else
      for (_ = 0; _ < P.length; ++_)
        y[P[_]] = F[P[_]];
  })();
})(ht);
var Ft = ht.exports;
const _t = /* @__PURE__ */ kt(Ft);
function St(t) {
  return "0x" + _t.keccak_256(B(t));
}
const Ct = "rlp/5.7.0", Bt = new ze(Ct);
function ot(t) {
  const i = [];
  for (; t; )
    i.unshift(t & 255), t >>= 8;
  return i;
}
function lt(t) {
  if (Array.isArray(t)) {
    let h = [];
    if (t.forEach(function(k) {
      h = h.concat(lt(k));
    }), h.length <= 55)
      return h.unshift(192 + h.length), h;
    const y = ot(h.length);
    return y.unshift(247 + y.length), y.concat(h);
  }
  ft(t) || Bt.throwArgumentError("RLP object must be BytesLike", "object", t);
  const i = Array.prototype.slice.call(B(t));
  if (i.length === 1 && i[0] <= 127)
    return i;
  if (i.length <= 55)
    return i.unshift(128 + i.length), i;
  const u = ot(i.length);
  return u.unshift(183 + u.length), u.concat(i);
}
function Re(t) {
  return M(lt(t));
}
const Pt = "address/5.7.0", I = new ze(Pt);
function it(t) {
  At(t, 20) || I.throwArgumentError("invalid address", "address", t), t = t.toLowerCase();
  const i = t.substring(2).split(""), u = new Uint8Array(40);
  for (let y = 0; y < 40; y++)
    u[y] = i[y].charCodeAt(0);
  const h = B(St(u));
  for (let y = 0; y < 40; y += 2)
    h[y >> 1] >> 4 >= 8 && (i[y] = i[y].toUpperCase()), (h[y >> 1] & 15) >= 8 && (i[y + 1] = i[y + 1].toUpperCase());
  return "0x" + i.join("");
}
const Lt = 9007199254740991;
function Ot(t) {
  return Math.log10 ? Math.log10(t) : Math.log(t) / Math.LN10;
}
const Je = {};
for (let t = 0; t < 10; t++)
  Je[String(t)] = String(t);
for (let t = 0; t < 26; t++)
  Je[String.fromCharCode(65 + t)] = String(10 + t);
const at = Math.floor(Ot(Lt));
function Mt(t) {
  t = t.toUpperCase(), t = t.substring(4) + t.substring(0, 2) + "00";
  let i = t.split("").map((h) => Je[h]).join("");
  for (; i.length >= at; ) {
    let h = i.substring(0, at);
    i = parseInt(h, 10) % 97 + i.substring(h.length);
  }
  let u = String(98 - parseInt(i, 10) % 97);
  for (; u.length < 2; )
    u = "0" + u;
  return u;
}
function $e(t) {
  let i = null;
  if (typeof t != "string" && I.throwArgumentError("invalid address", "address", t), t.match(/^(0x)?[0-9a-fA-F]{40}$/))
    t.substring(0, 2) !== "0x" && (t = "0x" + t), i = it(t), t.match(/([A-F].*[a-f])|([a-f].*[A-F])/) && i !== t && I.throwArgumentError("bad address checksum", "address", t);
  else if (t.match(/^XE[0-9]{2}[0-9A-Za-z]{30,31}$/)) {
    for (t.substring(2, 4) !== Mt(t) && I.throwArgumentError("bad icap checksum", "address", t), i = dt(t.substring(4)); i.length < 40; )
      i = "0" + i;
    i = it("0x" + i);
  } else
    I.throwArgumentError("invalid address", "address", t);
  return i;
}
const It = "transactions/5.7.0", s = new ze(It);
var ut;
(function(t) {
  t[t.legacy = 0] = "legacy", t[t.eip2930 = 1] = "eip2930", t[t.eip1559 = 2] = "eip1559";
})(ut || (ut = {}));
const Nt = [
  { name: "nonce", maxLength: 32, numeric: !0 },
  { name: "gasPrice", maxLength: 32, numeric: !0 },
  { name: "gasLimit", maxLength: 32, numeric: !0 },
  { name: "to", length: 20 },
  { name: "value", maxLength: 32, numeric: !0 },
  { name: "data" }
], Rt = {
  chainId: !0,
  data: !0,
  gasLimit: !0,
  gasPrice: !0,
  nonce: !0,
  to: !0,
  type: !0,
  value: !0
};
function m(t, i) {
  const u = S(Ge.from(t).toHexString());
  return u.length > 32 && s.throwArgumentError("invalid length for " + i, "transaction:" + i, t), u;
}
function Ue(t, i) {
  return {
    address: $e(t),
    storageKeys: (i || []).map((u, h) => (wt(u) !== 32 && s.throwArgumentError("invalid access list storageKey", `accessList[${t}:${h}]`, u), u.toLowerCase()))
  };
}
function zt(t) {
  if (Array.isArray(t))
    return t.map((u, h) => Array.isArray(u) ? (u.length > 2 && s.throwArgumentError("access list expected to be [ address, storageKeys[] ]", `value[${h}]`, u), Ue(u[0], u[1])) : Ue(u.address, u.storageKeys));
  const i = Object.keys(t).map((u) => {
    const h = t[u].reduce((y, k) => (y[k] = !0, y), {});
    return Ue(u, Object.keys(h).sort());
  });
  return i.sort((u, h) => u.address.localeCompare(h.address)), i;
}
function pt(t) {
  return zt(t).map((i) => [i.address, i.storageKeys]);
}
function Dt(t, i) {
  if (t.gasPrice != null) {
    const h = Ge.from(t.gasPrice), y = Ge.from(t.maxFeePerGas || 0);
    h.eq(y) || s.throwArgumentError("mismatch EIP-1559 gasPrice != maxFeePerGas", "tx", {
      gasPrice: h,
      maxFeePerGas: y
    });
  }
  const u = [
    m(t.chainId || 0, "chainId"),
    m(t.nonce || 0, "nonce"),
    m(t.maxPriorityFeePerGas || 0, "maxPriorityFeePerGas"),
    m(t.maxFeePerGas || 0, "maxFeePerGas"),
    m(t.gasLimit || 0, "gasLimit"),
    t.to != null ? $e(t.to) : "0x",
    m(t.value || 0, "value"),
    t.data || "0x",
    pt(t.accessList || [])
  ];
  if (i) {
    const h = je(i);
    u.push(m(h.recoveryParam, "recoveryParam")), u.push(S(h.r)), u.push(S(h.s));
  }
  return ct(["0x02", Re(u)]);
}
function Kt(t, i) {
  const u = [
    m(t.chainId || 0, "chainId"),
    m(t.nonce || 0, "nonce"),
    m(t.gasPrice || 0, "gasPrice"),
    m(t.gasLimit || 0, "gasLimit"),
    t.to != null ? $e(t.to) : "0x",
    m(t.value || 0, "value"),
    t.data || "0x",
    pt(t.accessList || [])
  ];
  if (i) {
    const h = je(i);
    u.push(m(h.recoveryParam, "recoveryParam")), u.push(S(h.r)), u.push(S(h.s));
  }
  return ct(["0x01", Re(u)]);
}
function Ht(t, i) {
  Et(t, Rt);
  const u = [];
  Nt.forEach(function(d) {
    let A = t[d.name] || [];
    const E = {};
    d.numeric && (E.hexPad = "left"), A = B(M(A, E)), d.length && A.length !== d.length && A.length > 0 && s.throwArgumentError("invalid length for " + d.name, "transaction:" + d.name, A), d.maxLength && (A = S(A), A.length > d.maxLength && s.throwArgumentError("invalid length for " + d.name, "transaction:" + d.name, A)), u.push(M(A));
  });
  let h = 0;
  if (t.chainId != null ? (h = t.chainId, typeof h != "number" && s.throwArgumentError("invalid transaction.chainId", "transaction", t)) : i && !ft(i) && i.v > 28 && (h = Math.floor((i.v - 35) / 2)), h !== 0 && (u.push(M(h)), u.push("0x"), u.push("0x")), !i)
    return Re(u);
  const y = je(i);
  let k = 27 + y.recoveryParam;
  return h !== 0 ? (u.pop(), u.pop(), u.pop(), k += h * 2 + 8, y.v > 28 && y.v !== k && s.throwArgumentError("transaction.chainId/signature.v mismatch", "signature", i)) : y.v !== k && s.throwArgumentError("transaction.chainId/signature.v mismatch", "signature", i), u.push(M(k)), u.push(S(B(y.r))), u.push(S(B(y.s))), Re(u);
}
function Wt(t, i) {
  if (t.type == null || t.type === 0)
    return t.accessList != null && s.throwArgumentError("untyped transactions do not support accessList; include type: 1", "transaction", t), Ht(t, i);
  switch (t.type) {
    case 1:
      return Kt(t, i);
    case 2:
      return Dt(t, i);
  }
  return s.throwError(`unsupported transaction type: ${t.type}`, ze.errors.UNSUPPORTED_OPERATION, {
    operation: "serializeTransaction",
    transactionType: t.type
  });
}
export {
  ut as TransactionTypes,
  zt as accessListify,
  Wt as serialize
};
