// @pioneer-platform/toolbox-cosmos@3.0.1/dist/MsgCompiled-f025c504.js
import { f as et, g as ot, i as rt, j as ft, k as ut, u as ct, p as lt, c as S, n as st } from "./index-d653805f.js";
var V = {}, j = {}, B, U;
function dt() {
  if (U)
    return B;
  U = 1, B = i;
  var f = P();
  function i(n, o) {
    this.lo = n >>> 0, this.hi = o >>> 0;
  }
  var r = i.zero = new i(0, 0);
  r.toNumber = function() {
    return 0;
  }, r.zzEncode = r.zzDecode = function() {
    return this;
  }, r.length = function() {
    return 1;
  };
  var t = i.zeroHash = "\0\0\0\0\0\0\0\0";
  i.fromNumber = function(o) {
    if (o === 0)
      return r;
    var u = o < 0;
    u && (o = -o);
    var l = o >>> 0, h = (o - l) / 4294967296 >>> 0;
    return u && (h = ~h >>> 0, l = ~l >>> 0, ++l > 4294967295 && (l = 0, ++h > 4294967295 && (h = 0))), new i(l, h);
  }, i.from = function(o) {
    if (typeof o == "number")
      return i.fromNumber(o);
    if (f.isString(o))
      if (f.Long)
        o = f.Long.fromString(o);
      else
        return i.fromNumber(parseInt(o, 10));
    return o.low || o.high ? new i(o.low >>> 0, o.high >>> 0) : r;
  }, i.prototype.toNumber = function(o) {
    if (!o && this.hi >>> 31) {
      var u = ~this.lo + 1 >>> 0, l = ~this.hi >>> 0;
      return u || (l = l + 1 >>> 0), -(u + l * 4294967296);
    }
    return this.lo + this.hi * 4294967296;
  }, i.prototype.toLong = function(o) {
    return f.Long ? new f.Long(this.lo | 0, this.hi | 0, !!o) : { low: this.lo | 0, high: this.hi | 0, unsigned: !!o };
  };
  var e = String.prototype.charCodeAt;
  return i.fromHash = function(o) {
    return o === t ? r : new i(
      (e.call(o, 0) | e.call(o, 1) << 8 | e.call(o, 2) << 16 | e.call(o, 3) << 24) >>> 0,
      (e.call(o, 4) | e.call(o, 5) << 8 | e.call(o, 6) << 16 | e.call(o, 7) << 24) >>> 0
    );
  }, i.prototype.toHash = function() {
    return String.fromCharCode(
      this.lo & 255,
      this.lo >>> 8 & 255,
      this.lo >>> 16 & 255,
      this.lo >>> 24,
      this.hi & 255,
      this.hi >>> 8 & 255,
      this.hi >>> 16 & 255,
      this.hi >>> 24
    );
  }, i.prototype.zzEncode = function() {
    var o = this.hi >> 31;
    return this.hi = ((this.hi << 1 | this.lo >>> 31) ^ o) >>> 0, this.lo = (this.lo << 1 ^ o) >>> 0, this;
  }, i.prototype.zzDecode = function() {
    var o = -(this.lo & 1);
    return this.lo = ((this.lo >>> 1 | this.hi << 31) ^ o) >>> 0, this.hi = (this.hi >>> 1 ^ o) >>> 0, this;
  }, i.prototype.length = function() {
    var o = this.lo, u = (this.lo >>> 28 | this.hi << 4) >>> 0, l = this.hi >>> 24;
    return l === 0 ? u === 0 ? o < 16384 ? o < 128 ? 1 : 2 : o < 2097152 ? 3 : 4 : u < 16384 ? u < 128 ? 5 : 6 : u < 2097152 ? 7 : 8 : l < 128 ? 9 : 10;
  }, B;
}
var F;
function P() {
  return F || (F = 1, function(f) {
    var i = f;
    i.asPromise = et, i.base64 = ot, i.EventEmitter = rt, i.float = ft, i.inquire = ut, i.utf8 = ct, i.pool = lt, i.LongBits = dt(), i.isNode = !!(typeof S < "u" && S && S.process && S.process.versions && S.process.versions.node), i.global = i.isNode && S || typeof window < "u" && window || typeof self < "u" && self || S, i.emptyArray = Object.freeze ? Object.freeze([]) : (
      /* istanbul ignore next */
      []
    ), i.emptyObject = Object.freeze ? Object.freeze({}) : (
      /* istanbul ignore next */
      {}
    ), i.isInteger = Number.isInteger || /* istanbul ignore next */
    function(n) {
      return typeof n == "number" && isFinite(n) && Math.floor(n) === n;
    }, i.isString = function(n) {
      return typeof n == "string" || n instanceof String;
    }, i.isObject = function(n) {
      return n && typeof n == "object";
    }, i.isset = /**
     * Checks if a property on a message is considered to be present.
     * @param {Object} obj Plain object or message instance
     * @param {string} prop Property name
     * @returns {boolean} `true` if considered to be present, otherwise `false`
     */
    i.isSet = function(n, o) {
      var u = n[o];
      return u != null && n.hasOwnProperty(o) ? typeof u != "object" || (Array.isArray(u) ? u.length : Object.keys(u).length) > 0 : !1;
    }, i.Buffer = function() {
      try {
        var e = i.inquire("buffer").Buffer;
        return e.prototype.utf8Write ? e : (
          /* istanbul ignore next */
          null
        );
      } catch {
        return null;
      }
    }(), i._Buffer_from = null, i._Buffer_allocUnsafe = null, i.newBuffer = function(n) {
      return typeof n == "number" ? i.Buffer ? i._Buffer_allocUnsafe(n) : new i.Array(n) : i.Buffer ? i._Buffer_from(n) : typeof Uint8Array > "u" ? n : new Uint8Array(n);
    }, i.Array = typeof Uint8Array < "u" ? Uint8Array : Array, i.Long = /* istanbul ignore next */
    i.global.dcodeIO && /* istanbul ignore next */
    i.global.dcodeIO.Long || /* istanbul ignore next */
    i.global.Long || i.inquire("long"), i.key2Re = /^true|false|0|1$/, i.key32Re = /^-?(?:0|[1-9][0-9]*)$/, i.key64Re = /^(?:[\\x00-\\xff]{8}|-?(?:0|[1-9][0-9]*))$/, i.longToHash = function(n) {
      return n ? i.LongBits.from(n).toHash() : i.LongBits.zeroHash;
    }, i.longFromHash = function(n, o) {
      var u = i.LongBits.fromHash(n);
      return i.Long ? i.Long.fromBits(u.lo, u.hi, o) : u.toNumber(!!o);
    };
    function r(e, n, o) {
      for (var u = Object.keys(n), l = 0; l < u.length; ++l)
        (e[u[l]] === void 0 || !o) && (e[u[l]] = n[u[l]]);
      return e;
    }
    i.merge = r, i.lcFirst = function(n) {
      return n.charAt(0).toLowerCase() + n.substring(1);
    };
    function t(e) {
      function n(o, u) {
        if (!(this instanceof n))
          return new n(o, u);
        Object.defineProperty(this, "message", { get: function() {
          return o;
        } }), Error.captureStackTrace ? Error.captureStackTrace(this, n) : Object.defineProperty(this, "stack", { value: new Error().stack || "" }), u && r(this, u);
      }
      return n.prototype = Object.create(Error.prototype, {
        constructor: {
          value: n,
          writable: !0,
          enumerable: !1,
          configurable: !0
        },
        name: {
          get: function() {
            return e;
          },
          set: void 0,
          enumerable: !1,
          // configurable: false would accurately preserve the behavior of
          // the original, but I'm guessing that was not intentional.
          // For an actual error subclass, this property would
          // be configurable.
          configurable: !0
        },
        toString: {
          value: function() {
            return this.name + ": " + this.message;
          },
          writable: !0,
          enumerable: !1,
          configurable: !0
        }
      }), n;
    }
    i.newError = t, i.ProtocolError = t("ProtocolError"), i.oneOfGetter = function(n) {
      for (var o = {}, u = 0; u < n.length; ++u)
        o[n[u]] = 1;
      return function() {
        for (var l = Object.keys(this), h = l.length - 1; h > -1; --h)
          if (o[l[h]] === 1 && this[l[h]] !== void 0 && this[l[h]] !== null)
            return l[h];
      };
    }, i.oneOfSetter = function(n) {
      return function(o) {
        for (var u = 0; u < n.length; ++u)
          n[u] !== o && delete this[n[u]];
      };
    }, i.toJSONOptions = {
      longs: String,
      enums: String,
      bytes: String,
      json: !0
    }, i._configure = function() {
      var e = i.Buffer;
      if (!e) {
        i._Buffer_from = i._Buffer_allocUnsafe = null;
        return;
      }
      i._Buffer_from = e.from !== Uint8Array.from && e.from || /* istanbul ignore next */
      function(o, u) {
        return new e(o, u);
      }, i._Buffer_allocUnsafe = e.allocUnsafe || /* istanbul ignore next */
      function(o) {
        return new e(o);
      };
    };
  }(j)), j;
}
var G = p, b = P(), C, D = b.LongBits, R = b.base64, H = b.utf8;
function v(f, i, r) {
  this.fn = f, this.len = i, this.next = void 0, this.val = r;
}
function T() {
}
function ht(f) {
  this.head = f.head, this.tail = f.tail, this.len = f.len, this.next = f.states;
}
function p() {
  this.len = 0, this.head = new v(T, 0, 0), this.tail = this.head, this.states = null;
}
var Q = function() {
  return b.Buffer ? function() {
    return (p.create = function() {
      return new C();
    })();
  } : function() {
    return new p();
  };
};
p.create = Q();
p.alloc = function(i) {
  return new b.Array(i);
};
b.Array !== Array && (p.alloc = b.pool(p.alloc, b.Array.prototype.subarray));
p.prototype._push = function(i, r, t) {
  return this.tail = this.tail.next = new v(i, r, t), this.len += r, this;
};
function J(f, i, r) {
  i[r] = f & 255;
}
function pt(f, i, r) {
  for (; f > 127; )
    i[r++] = f & 127 | 128, f >>>= 7;
  i[r] = f;
}
function $(f, i) {
  this.len = f, this.next = void 0, this.val = i;
}
$.prototype = Object.create(v.prototype);
$.prototype.fn = pt;
p.prototype.uint32 = function(i) {
  return this.len += (this.tail = this.tail.next = new $(
    (i = i >>> 0) < 128 ? 1 : i < 16384 ? 2 : i < 2097152 ? 3 : i < 268435456 ? 4 : 5,
    i
  )).len, this;
};
p.prototype.int32 = function(i) {
  return i < 0 ? this._push(I, 10, D.fromNumber(i)) : this.uint32(i);
};
p.prototype.sint32 = function(i) {
  return this.uint32((i << 1 ^ i >> 31) >>> 0);
};
function I(f, i, r) {
  for (; f.hi; )
    i[r++] = f.lo & 127 | 128, f.lo = (f.lo >>> 7 | f.hi << 25) >>> 0, f.hi >>>= 7;
  for (; f.lo > 127; )
    i[r++] = f.lo & 127 | 128, f.lo = f.lo >>> 7;
  i[r++] = f.lo;
}
p.prototype.uint64 = function(i) {
  var r = D.from(i);
  return this._push(I, r.length(), r);
};
p.prototype.int64 = p.prototype.uint64;
p.prototype.sint64 = function(i) {
  var r = D.from(i).zzEncode();
  return this._push(I, r.length(), r);
};
p.prototype.bool = function(i) {
  return this._push(J, 1, i ? 1 : 0);
};
function E(f, i, r) {
  i[r] = f & 255, i[r + 1] = f >>> 8 & 255, i[r + 2] = f >>> 16 & 255, i[r + 3] = f >>> 24;
}
p.prototype.fixed32 = function(i) {
  return this._push(E, 4, i >>> 0);
};
p.prototype.sfixed32 = p.prototype.fixed32;
p.prototype.fixed64 = function(i) {
  var r = D.from(i);
  return this._push(E, 4, r.lo)._push(E, 4, r.hi);
};
p.prototype.sfixed64 = p.prototype.fixed64;
p.prototype.float = function(i) {
  return this._push(b.float.writeFloatLE, 4, i);
};
p.prototype.double = function(i) {
  return this._push(b.float.writeDoubleLE, 8, i);
};
var at = b.Array.prototype.set ? function(i, r, t) {
  r.set(i, t);
} : function(i, r, t) {
  for (var e = 0; e < i.length; ++e)
    r[t + e] = i[e];
};
p.prototype.bytes = function(i) {
  var r = i.length >>> 0;
  if (!r)
    return this._push(J, 1, 0);
  if (b.isString(i)) {
    var t = p.alloc(r = R.length(i));
    R.decode(i, t, 0), i = t;
  }
  return this.uint32(r)._push(at, r, i);
};
p.prototype.string = function(i) {
  var r = H.length(i);
  return r ? this.uint32(r)._push(H.write, r, i) : this._push(J, 1, 0);
};
p.prototype.fork = function() {
  return this.states = new ht(this), this.head = this.tail = new v(T, 0, 0), this.len = 0, this;
};
p.prototype.reset = function() {
  return this.states ? (this.head = this.states.head, this.tail = this.states.tail, this.len = this.states.len, this.states = this.states.next) : (this.head = this.tail = new v(T, 0, 0), this.len = 0), this;
};
p.prototype.ldelim = function() {
  var i = this.head, r = this.tail, t = this.len;
  return this.reset().uint32(t), t && (this.tail.next = i.next, this.tail = r, this.len += t), this;
};
p.prototype.finish = function() {
  for (var i = this.head.next, r = this.constructor.alloc(this.len), t = 0; i; )
    i.fn(i.val, r, t), t += i.len, i = i.next;
  return r;
};
p._configure = function(f) {
  C = f, p.create = Q(), C._configure();
};
var yt = A, X = G;
(A.prototype = Object.create(X.prototype)).constructor = A;
var k = P();
function A() {
  X.call(this);
}
A._configure = function() {
  A.alloc = k._Buffer_allocUnsafe, A.writeBytesBuffer = k.Buffer && k.Buffer.prototype instanceof Uint8Array && k.Buffer.prototype.set.name === "set" ? function(i, r, t) {
    r.set(i, t);
  } : function(i, r, t) {
    if (i.copy)
      i.copy(r, t, 0, i.length);
    else
      for (var e = 0; e < i.length; )
        r[t++] = i[e++];
  };
};
A.prototype.bytes = function(i) {
  k.isString(i) && (i = k._Buffer_from(i, "base64"));
  var r = i.length >>> 0;
  return this.uint32(r), r && this._push(A.writeBytesBuffer, r, i), this;
};
function mt(f, i, r) {
  f.length < 40 ? k.utf8.write(f, i, r) : i.utf8Write ? i.utf8Write(f, r) : i.write(f, r);
}
A.prototype.string = function(i) {
  var r = k.Buffer.byteLength(i);
  return this.uint32(r), r && this._push(mt, r, i), this;
};
A._configure();
var Y = a, O = P(), L, Z = O.LongBits, bt = O.utf8;
function w(f, i) {
  return RangeError("index out of range: " + f.pos + " + " + (i || 1) + " > " + f.len);
}
function a(f) {
  this.buf = f, this.pos = 0, this.len = f.length;
}
var W = typeof Uint8Array < "u" ? function(i) {
  if (i instanceof Uint8Array || Array.isArray(i))
    return new a(i);
  throw Error("illegal buffer");
} : function(i) {
  if (Array.isArray(i))
    return new a(i);
  throw Error("illegal buffer");
}, tt = function() {
  return O.Buffer ? function(r) {
    return (a.create = function(e) {
      return O.Buffer.isBuffer(e) ? new L(e) : W(e);
    })(r);
  } : W;
};
a.create = tt();
a.prototype._slice = O.Array.prototype.subarray || /* istanbul ignore next */
O.Array.prototype.slice;
a.prototype.uint32 = function() {
  var i = 4294967295;
  return function() {
    if (i = (this.buf[this.pos] & 127) >>> 0, this.buf[this.pos++] < 128 || (i = (i | (this.buf[this.pos] & 127) << 7) >>> 0, this.buf[this.pos++] < 128) || (i = (i | (this.buf[this.pos] & 127) << 14) >>> 0, this.buf[this.pos++] < 128) || (i = (i | (this.buf[this.pos] & 127) << 21) >>> 0, this.buf[this.pos++] < 128) || (i = (i | (this.buf[this.pos] & 15) << 28) >>> 0, this.buf[this.pos++] < 128))
      return i;
    if ((this.pos += 5) > this.len)
      throw this.pos = this.len, w(this, 10);
    return i;
  };
}();
a.prototype.int32 = function() {
  return this.uint32() | 0;
};
a.prototype.sint32 = function() {
  var i = this.uint32();
  return i >>> 1 ^ -(i & 1) | 0;
};
function N() {
  var f = new Z(0, 0), i = 0;
  if (this.len - this.pos > 4) {
    for (; i < 4; ++i)
      if (f.lo = (f.lo | (this.buf[this.pos] & 127) << i * 7) >>> 0, this.buf[this.pos++] < 128)
        return f;
    if (f.lo = (f.lo | (this.buf[this.pos] & 127) << 28) >>> 0, f.hi = (f.hi | (this.buf[this.pos] & 127) >> 4) >>> 0, this.buf[this.pos++] < 128)
      return f;
    i = 0;
  } else {
    for (; i < 3; ++i) {
      if (this.pos >= this.len)
        throw w(this);
      if (f.lo = (f.lo | (this.buf[this.pos] & 127) << i * 7) >>> 0, this.buf[this.pos++] < 128)
        return f;
    }
    return f.lo = (f.lo | (this.buf[this.pos++] & 127) << i * 7) >>> 0, f;
  }
  if (this.len - this.pos > 4) {
    for (; i < 5; ++i)
      if (f.hi = (f.hi | (this.buf[this.pos] & 127) << i * 7 + 3) >>> 0, this.buf[this.pos++] < 128)
        return f;
  } else
    for (; i < 5; ++i) {
      if (this.pos >= this.len)
        throw w(this);
      if (f.hi = (f.hi | (this.buf[this.pos] & 127) << i * 7 + 3) >>> 0, this.buf[this.pos++] < 128)
        return f;
    }
  throw Error("invalid varint encoding");
}
a.prototype.bool = function() {
  return this.uint32() !== 0;
};
function x(f, i) {
  return (f[i - 4] | f[i - 3] << 8 | f[i - 2] << 16 | f[i - 1] << 24) >>> 0;
}
a.prototype.fixed32 = function() {
  if (this.pos + 4 > this.len)
    throw w(this, 4);
  return x(this.buf, this.pos += 4);
};
a.prototype.sfixed32 = function() {
  if (this.pos + 4 > this.len)
    throw w(this, 4);
  return x(this.buf, this.pos += 4) | 0;
};
function q() {
  if (this.pos + 8 > this.len)
    throw w(this, 8);
  return new Z(x(this.buf, this.pos += 4), x(this.buf, this.pos += 4));
}
a.prototype.float = function() {
  if (this.pos + 4 > this.len)
    throw w(this, 4);
  var i = O.float.readFloatLE(this.buf, this.pos);
  return this.pos += 4, i;
};
a.prototype.double = function() {
  if (this.pos + 8 > this.len)
    throw w(this, 4);
  var i = O.float.readDoubleLE(this.buf, this.pos);
  return this.pos += 8, i;
};
a.prototype.bytes = function() {
  var i = this.uint32(), r = this.pos, t = this.pos + i;
  if (t > this.len)
    throw w(this, i);
  if (this.pos += i, Array.isArray(this.buf))
    return this.buf.slice(r, t);
  if (r === t) {
    var e = O.Buffer;
    return e ? e.alloc(0) : new this.buf.constructor(0);
  }
  return this._slice.call(this.buf, r, t);
};
a.prototype.string = function() {
  var i = this.bytes();
  return bt.read(i, 0, i.length);
};
a.prototype.skip = function(i) {
  if (typeof i == "number") {
    if (this.pos + i > this.len)
      throw w(this, i);
    this.pos += i;
  } else
    do
      if (this.pos >= this.len)
        throw w(this);
    while (this.buf[this.pos++] & 128);
  return this;
};
a.prototype.skipType = function(f) {
  switch (f) {
    case 0:
      this.skip();
      break;
    case 1:
      this.skip(8);
      break;
    case 2:
      this.skip(this.uint32());
      break;
    case 3:
      for (; (f = this.uint32() & 7) !== 4; )
        this.skipType(f);
      break;
    case 5:
      this.skip(4);
      break;
    default:
      throw Error("invalid wire type " + f + " at offset " + this.pos);
  }
  return this;
};
a._configure = function(f) {
  L = f, a.create = tt(), L._configure();
  var i = O.Long ? "toLong" : (
    /* istanbul ignore next */
    "toNumber"
  );
  O.merge(a.prototype, {
    int64: function() {
      return N.call(this)[i](!1);
    },
    uint64: function() {
      return N.call(this)[i](!0);
    },
    sint64: function() {
      return N.call(this).zzDecode()[i](!1);
    },
    fixed64: function() {
      return q.call(this)[i](!0);
    },
    sfixed64: function() {
      return q.call(this)[i](!1);
    }
  });
};
var Ot = g, nt = Y;
(g.prototype = Object.create(nt.prototype)).constructor = g;
var K = P();
function g(f) {
  nt.call(this, f);
}
g._configure = function() {
  K.Buffer && (g.prototype._slice = K.Buffer.prototype.slice);
};
g.prototype.string = function() {
  var i = this.uint32();
  return this.buf.utf8Slice ? this.buf.utf8Slice(this.pos, this.pos = Math.min(this.pos + i, this.len)) : this.buf.toString("utf-8", this.pos, this.pos = Math.min(this.pos + i, this.len));
};
g._configure();
var it = {}, wt = _, M = P();
(_.prototype = Object.create(M.EventEmitter.prototype)).constructor = _;
function _(f, i, r) {
  if (typeof f != "function")
    throw TypeError("rpcImpl must be a function");
  M.EventEmitter.call(this), this.rpcImpl = f, this.requestDelimited = !!i, this.responseDelimited = !!r;
}
_.prototype.rpcCall = function f(i, r, t, e, n) {
  if (!e)
    throw TypeError("request must be specified");
  var o = this;
  if (!n)
    return M.asPromise(f, o, i, r, t, e);
  if (!o.rpcImpl) {
    setTimeout(function() {
      n(Error("already ended"));
    }, 0);
    return;
  }
  try {
    return o.rpcImpl(
      i,
      r[o.requestDelimited ? "encodeDelimited" : "encode"](e).finish(),
      function(l, h) {
        if (l)
          return o.emit("error", l, i), n(l);
        if (h === null) {
          o.end(
            /* endedByRPC */
            !0
          );
          return;
        }
        if (!(h instanceof t))
          try {
            h = t[o.responseDelimited ? "decodeDelimited" : "decode"](h);
          } catch (z) {
            return o.emit("error", z, i), n(z);
          }
        return o.emit("data", h, i), n(null, h);
      }
    );
  } catch (u) {
    o.emit("error", u, i), setTimeout(function() {
      n(u);
    }, 0);
    return;
  }
};
_.prototype.end = function(i) {
  return this.rpcImpl && (i || this.rpcImpl(null, null, null), this.rpcImpl = null, this.emit("end").off()), this;
};
(function(f) {
  var i = f;
  i.Service = wt;
})(it);
var At = {};
(function(f) {
  var i = f;
  i.build = "minimal", i.Writer = G, i.BufferWriter = yt, i.Reader = Y, i.BufferReader = Ot, i.util = P(), i.rpc = it, i.roots = At, i.configure = r;
  function r() {
    i.util._configure(), i.Writer._configure(i.BufferWriter), i.Reader._configure(i.BufferReader);
  }
  r();
})(V);
var kt = V;
const y = /* @__PURE__ */ st(kt), d = y.Reader, m = y.Writer, s = y.util, c = y.roots.default || (y.roots.default = {});
c.common = function() {
  let f = {};
  return f.Asset = function() {
    function i(r) {
      if (r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.chain = "", i.prototype.symbol = "", i.prototype.ticker = "", i.prototype.synth = !1, i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      return e || (e = m.create()), t.chain != null && Object.hasOwnProperty.call(t, "chain") && e.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(t.chain), t.symbol != null && Object.hasOwnProperty.call(t, "symbol") && e.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(t.symbol), t.ticker != null && Object.hasOwnProperty.call(t, "ticker") && e.uint32(
        /* id 3, wireType 2 =*/
        26
      ).string(t.ticker), t.synth != null && Object.hasOwnProperty.call(t, "synth") && e.uint32(
        /* id 4, wireType 0 =*/
        32
      ).bool(t.synth), e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.common.Asset();
      for (; t.pos < n; ) {
        let u = t.uint32();
        switch (u >>> 3) {
          case 1:
            o.chain = t.string();
            break;
          case 2:
            o.symbol = t.string();
            break;
          case 3:
            o.ticker = t.string();
            break;
          case 4:
            o.synth = t.bool();
            break;
          default:
            t.skipType(u & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      return typeof t != "object" || t === null ? "object expected" : t.chain != null && t.hasOwnProperty("chain") && !s.isString(t.chain) ? "chain: string expected" : t.symbol != null && t.hasOwnProperty("symbol") && !s.isString(t.symbol) ? "symbol: string expected" : t.ticker != null && t.hasOwnProperty("ticker") && !s.isString(t.ticker) ? "ticker: string expected" : t.synth != null && t.hasOwnProperty("synth") && typeof t.synth != "boolean" ? "synth: boolean expected" : null;
    }, i.fromObject = function(t) {
      if (t instanceof c.common.Asset)
        return t;
      let e = new c.common.Asset();
      return t.chain != null && (e.chain = String(t.chain)), t.symbol != null && (e.symbol = String(t.symbol)), t.ticker != null && (e.ticker = String(t.ticker)), t.synth != null && (e.synth = !!t.synth), e;
    }, i.toObject = function(t, e) {
      e || (e = {});
      let n = {};
      return e.defaults && (n.chain = "", n.symbol = "", n.ticker = "", n.synth = !1), t.chain != null && t.hasOwnProperty("chain") && (n.chain = t.chain), t.symbol != null && t.hasOwnProperty("symbol") && (n.symbol = t.symbol), t.ticker != null && t.hasOwnProperty("ticker") && (n.ticker = t.ticker), t.synth != null && t.hasOwnProperty("synth") && (n.synth = t.synth), n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f.Coin = function() {
    function i(r) {
      if (r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.asset = null, i.prototype.amount = "", i.prototype.decimals = s.Long ? s.Long.fromBits(0, 0, !1) : 0, i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      return e || (e = m.create()), t.asset != null && Object.hasOwnProperty.call(t, "asset") && c.common.Asset.encode(
        t.asset,
        e.uint32(
          /* id 1, wireType 2 =*/
          10
        ).fork()
      ).ldelim(), t.amount != null && Object.hasOwnProperty.call(t, "amount") && e.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(t.amount), t.decimals != null && Object.hasOwnProperty.call(t, "decimals") && e.uint32(
        /* id 3, wireType 0 =*/
        24
      ).int64(t.decimals), e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.common.Coin();
      for (; t.pos < n; ) {
        let u = t.uint32();
        switch (u >>> 3) {
          case 1:
            o.asset = c.common.Asset.decode(t, t.uint32());
            break;
          case 2:
            o.amount = t.string();
            break;
          case 3:
            o.decimals = t.int64();
            break;
          default:
            t.skipType(u & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      if (typeof t != "object" || t === null)
        return "object expected";
      if (t.asset != null && t.hasOwnProperty("asset")) {
        let e = c.common.Asset.verify(t.asset);
        if (e)
          return "asset." + e;
      }
      return t.amount != null && t.hasOwnProperty("amount") && !s.isString(t.amount) ? "amount: string expected" : t.decimals != null && t.hasOwnProperty("decimals") && !s.isInteger(t.decimals) && !(t.decimals && s.isInteger(t.decimals.low) && s.isInteger(t.decimals.high)) ? "decimals: integer|Long expected" : null;
    }, i.fromObject = function(t) {
      if (t instanceof c.common.Coin)
        return t;
      let e = new c.common.Coin();
      if (t.asset != null) {
        if (typeof t.asset != "object")
          throw TypeError(".common.Coin.asset: object expected");
        e.asset = c.common.Asset.fromObject(t.asset);
      }
      return t.amount != null && (e.amount = String(t.amount)), t.decimals != null && (s.Long ? (e.decimals = s.Long.fromValue(t.decimals)).unsigned = !1 : typeof t.decimals == "string" ? e.decimals = parseInt(t.decimals, 10) : typeof t.decimals == "number" ? e.decimals = t.decimals : typeof t.decimals == "object" && (e.decimals = new s.LongBits(
        t.decimals.low >>> 0,
        t.decimals.high >>> 0
      ).toNumber())), e;
    }, i.toObject = function(t, e) {
      e || (e = {});
      let n = {};
      if (e.defaults)
        if (n.asset = null, n.amount = "", s.Long) {
          let o = new s.Long(0, 0, !1);
          n.decimals = e.longs === String ? o.toString() : e.longs === Number ? o.toNumber() : o;
        } else
          n.decimals = e.longs === String ? "0" : 0;
      return t.asset != null && t.hasOwnProperty("asset") && (n.asset = c.common.Asset.toObject(t.asset, e)), t.amount != null && t.hasOwnProperty("amount") && (n.amount = t.amount), t.decimals != null && t.hasOwnProperty("decimals") && (typeof t.decimals == "number" ? n.decimals = e.longs === String ? String(t.decimals) : t.decimals : n.decimals = e.longs === String ? s.Long.prototype.toString.call(t.decimals) : e.longs === Number ? new s.LongBits(
        t.decimals.low >>> 0,
        t.decimals.high >>> 0
      ).toNumber() : t.decimals), n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f.PubKeySet = function() {
    function i(r) {
      if (r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.secp256k1 = "", i.prototype.ed25519 = "", i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      return e || (e = m.create()), t.secp256k1 != null && Object.hasOwnProperty.call(t, "secp256k1") && e.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(t.secp256k1), t.ed25519 != null && Object.hasOwnProperty.call(t, "ed25519") && e.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(t.ed25519), e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.common.PubKeySet();
      for (; t.pos < n; ) {
        let u = t.uint32();
        switch (u >>> 3) {
          case 1:
            o.secp256k1 = t.string();
            break;
          case 2:
            o.ed25519 = t.string();
            break;
          default:
            t.skipType(u & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      return typeof t != "object" || t === null ? "object expected" : t.secp256k1 != null && t.hasOwnProperty("secp256k1") && !s.isString(t.secp256k1) ? "secp256k1: string expected" : t.ed25519 != null && t.hasOwnProperty("ed25519") && !s.isString(t.ed25519) ? "ed25519: string expected" : null;
    }, i.fromObject = function(t) {
      if (t instanceof c.common.PubKeySet)
        return t;
      let e = new c.common.PubKeySet();
      return t.secp256k1 != null && (e.secp256k1 = String(t.secp256k1)), t.ed25519 != null && (e.ed25519 = String(t.ed25519)), e;
    }, i.toObject = function(t, e) {
      e || (e = {});
      let n = {};
      return e.defaults && (n.secp256k1 = "", n.ed25519 = ""), t.secp256k1 != null && t.hasOwnProperty("secp256k1") && (n.secp256k1 = t.secp256k1), t.ed25519 != null && t.hasOwnProperty("ed25519") && (n.ed25519 = t.ed25519), n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f.Tx = function() {
    function i(r) {
      if (this.coins = [], this.gas = [], r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.id = "", i.prototype.chain = "", i.prototype.fromAddress = "", i.prototype.toAddress = "", i.prototype.coins = s.emptyArray, i.prototype.gas = s.emptyArray, i.prototype.memo = "", i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      var n, o;
      if (e || (e = m.create()), t.id != null && Object.hasOwnProperty.call(t, "id") && e.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(t.id), t.chain != null && Object.hasOwnProperty.call(t, "chain") && e.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(t.chain), t.fromAddress != null && Object.hasOwnProperty.call(t, "fromAddress") && e.uint32(
        /* id 3, wireType 2 =*/
        26
      ).string(t.fromAddress), t.toAddress != null && Object.hasOwnProperty.call(t, "toAddress") && e.uint32(
        /* id 4, wireType 2 =*/
        34
      ).string(t.toAddress), (n = t.coins) != null && n.length)
        for (let u = 0; u < t.coins.length; ++u)
          c.common.Coin.encode(
            t.coins[u],
            e.uint32(
              /* id 5, wireType 2 =*/
              42
            ).fork()
          ).ldelim();
      if ((o = t.gas) != null && o.length)
        for (let u = 0; u < t.gas.length; ++u)
          c.common.Coin.encode(
            t.gas[u],
            e.uint32(
              /* id 6, wireType 2 =*/
              50
            ).fork()
          ).ldelim();
      return t.memo != null && Object.hasOwnProperty.call(t, "memo") && e.uint32(
        /* id 7, wireType 2 =*/
        58
      ).string(t.memo), e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      var u, l;
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.common.Tx();
      for (; t.pos < n; ) {
        let h = t.uint32();
        switch (h >>> 3) {
          case 1:
            o.id = t.string();
            break;
          case 2:
            o.chain = t.string();
            break;
          case 3:
            o.fromAddress = t.string();
            break;
          case 4:
            o.toAddress = t.string();
            break;
          case 5:
            (u = o.coins) != null && u.length || (o.coins = []), o.coins.push(c.common.Coin.decode(t, t.uint32()));
            break;
          case 6:
            (l = o.gas) != null && l.length || (o.gas = []), o.gas.push(c.common.Coin.decode(t, t.uint32()));
            break;
          case 7:
            o.memo = t.string();
            break;
          default:
            t.skipType(h & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      if (typeof t != "object" || t === null)
        return "object expected";
      if (t.id != null && t.hasOwnProperty("id") && !s.isString(t.id))
        return "id: string expected";
      if (t.chain != null && t.hasOwnProperty("chain") && !s.isString(t.chain))
        return "chain: string expected";
      if (t.fromAddress != null && t.hasOwnProperty("fromAddress") && !s.isString(t.fromAddress))
        return "fromAddress: string expected";
      if (t.toAddress != null && t.hasOwnProperty("toAddress") && !s.isString(t.toAddress))
        return "toAddress: string expected";
      if (t.coins != null && t.hasOwnProperty("coins")) {
        if (!Array.isArray(t.coins))
          return "coins: array expected";
        for (let e = 0; e < t.coins.length; ++e) {
          let n = c.common.Coin.verify(t.coins[e]);
          if (n)
            return "coins." + n;
        }
      }
      if (t.gas != null && t.hasOwnProperty("gas")) {
        if (!Array.isArray(t.gas))
          return "gas: array expected";
        for (let e = 0; e < t.gas.length; ++e) {
          let n = c.common.Coin.verify(t.gas[e]);
          if (n)
            return "gas." + n;
        }
      }
      return t.memo != null && t.hasOwnProperty("memo") && !s.isString(t.memo) ? "memo: string expected" : null;
    }, i.fromObject = function(t) {
      if (t instanceof c.common.Tx)
        return t;
      let e = new c.common.Tx();
      if (t.id != null && (e.id = String(t.id)), t.chain != null && (e.chain = String(t.chain)), t.fromAddress != null && (e.fromAddress = String(t.fromAddress)), t.toAddress != null && (e.toAddress = String(t.toAddress)), t.coins) {
        if (!Array.isArray(t.coins))
          throw TypeError(".common.Tx.coins: array expected");
        e.coins = [];
        for (let n = 0; n < t.coins.length; ++n) {
          if (typeof t.coins[n] != "object")
            throw TypeError(".common.Tx.coins: object expected");
          e.coins[n] = c.common.Coin.fromObject(t.coins[n]);
        }
      }
      if (t.gas) {
        if (!Array.isArray(t.gas))
          throw TypeError(".common.Tx.gas: array expected");
        e.gas = [];
        for (let n = 0; n < t.gas.length; ++n) {
          if (typeof t.gas[n] != "object")
            throw TypeError(".common.Tx.gas: object expected");
          e.gas[n] = c.common.Coin.fromObject(t.gas[n]);
        }
      }
      return t.memo != null && (e.memo = String(t.memo)), e;
    }, i.toObject = function(t, e) {
      var o, u;
      e || (e = {});
      let n = {};
      if ((e.arrays || e.defaults) && (n.coins = [], n.gas = []), e.defaults && (n.id = "", n.chain = "", n.fromAddress = "", n.toAddress = "", n.memo = ""), t.id != null && t.hasOwnProperty("id") && (n.id = t.id), t.chain != null && t.hasOwnProperty("chain") && (n.chain = t.chain), t.fromAddress != null && t.hasOwnProperty("fromAddress") && (n.fromAddress = t.fromAddress), t.toAddress != null && t.hasOwnProperty("toAddress") && (n.toAddress = t.toAddress), (o = t.coins) != null && o.length) {
        n.coins = [];
        for (let l = 0; l < t.coins.length; ++l)
          n.coins[l] = c.common.Coin.toObject(t.coins[l], e);
      }
      if ((u = t.gas) != null && u.length) {
        n.gas = [];
        for (let l = 0; l < t.gas.length; ++l)
          n.gas[l] = c.common.Coin.toObject(t.gas[l], e);
      }
      return t.memo != null && t.hasOwnProperty("memo") && (n.memo = t.memo), n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f.Fee = function() {
    function i(r) {
      if (this.coins = [], r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.coins = s.emptyArray, i.prototype.poolDeduct = "", i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      var n;
      if (e || (e = m.create()), (n = t.coins) != null && n.length)
        for (let o = 0; o < t.coins.length; ++o)
          c.common.Coin.encode(
            t.coins[o],
            e.uint32(
              /* id 1, wireType 2 =*/
              10
            ).fork()
          ).ldelim();
      return t.poolDeduct != null && Object.hasOwnProperty.call(t, "poolDeduct") && e.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(t.poolDeduct), e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      var u;
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.common.Fee();
      for (; t.pos < n; ) {
        let l = t.uint32();
        switch (l >>> 3) {
          case 1:
            (u = o.coins) != null && u.length || (o.coins = []), o.coins.push(c.common.Coin.decode(t, t.uint32()));
            break;
          case 2:
            o.poolDeduct = t.string();
            break;
          default:
            t.skipType(l & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      if (typeof t != "object" || t === null)
        return "object expected";
      if (t.coins != null && t.hasOwnProperty("coins")) {
        if (!Array.isArray(t.coins))
          return "coins: array expected";
        for (let e = 0; e < t.coins.length; ++e) {
          let n = c.common.Coin.verify(t.coins[e]);
          if (n)
            return "coins." + n;
        }
      }
      return t.poolDeduct != null && t.hasOwnProperty("poolDeduct") && !s.isString(t.poolDeduct) ? "poolDeduct: string expected" : null;
    }, i.fromObject = function(t) {
      if (t instanceof c.common.Fee)
        return t;
      let e = new c.common.Fee();
      if (t.coins) {
        if (!Array.isArray(t.coins))
          throw TypeError(".common.Fee.coins: array expected");
        e.coins = [];
        for (let n = 0; n < t.coins.length; ++n) {
          if (typeof t.coins[n] != "object")
            throw TypeError(".common.Fee.coins: object expected");
          e.coins[n] = c.common.Coin.fromObject(t.coins[n]);
        }
      }
      return t.poolDeduct != null && (e.poolDeduct = String(t.poolDeduct)), e;
    }, i.toObject = function(t, e) {
      var o;
      e || (e = {});
      let n = {};
      if ((e.arrays || e.defaults) && (n.coins = []), e.defaults && (n.poolDeduct = ""), (o = t.coins) != null && o.length) {
        n.coins = [];
        for (let u = 0; u < t.coins.length; ++u)
          n.coins[u] = c.common.Coin.toObject(t.coins[u], e);
      }
      return t.poolDeduct != null && t.hasOwnProperty("poolDeduct") && (n.poolDeduct = t.poolDeduct), n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f.ProtoUint = function() {
    function i(r) {
      if (r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.value = "", i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      return e || (e = m.create()), t.value != null && Object.hasOwnProperty.call(t, "value") && e.uint32(
        /* id 1, wireType 2 =*/
        10
      ).string(t.value), e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.common.ProtoUint();
      for (; t.pos < n; ) {
        let u = t.uint32();
        switch (u >>> 3) {
          case 1:
            o.value = t.string();
            break;
          default:
            t.skipType(u & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      return typeof t != "object" || t === null ? "object expected" : t.value != null && t.hasOwnProperty("value") && !s.isString(t.value) ? "value: string expected" : null;
    }, i.fromObject = function(t) {
      if (t instanceof c.common.ProtoUint)
        return t;
      let e = new c.common.ProtoUint();
      return t.value != null && (e.value = String(t.value)), e;
    }, i.toObject = function(t, e) {
      e || (e = {});
      let n = {};
      return e.defaults && (n.value = ""), t.value != null && t.hasOwnProperty("value") && (n.value = t.value), n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f;
}();
c.types = function() {
  let f = {};
  return f.MsgDeposit = function() {
    function i(r) {
      if (this.coins = [], r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.coins = s.emptyArray, i.prototype.memo = "", i.prototype.signer = s.newBuffer([]), i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      var n;
      if (e || (e = m.create()), (n = t.coins) != null && n.length)
        for (let o = 0; o < t.coins.length; ++o)
          c.common.Coin.encode(
            t.coins[o],
            e.uint32(
              /* id 1, wireType 2 =*/
              10
            ).fork()
          ).ldelim();
      return t.memo != null && Object.hasOwnProperty.call(t, "memo") && e.uint32(
        /* id 2, wireType 2 =*/
        18
      ).string(t.memo), t.signer != null && Object.hasOwnProperty.call(t, "signer") && e.uint32(
        /* id 3, wireType 2 =*/
        26
      ).bytes(t.signer), e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      var u;
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.types.MsgDeposit();
      for (; t.pos < n; ) {
        let l = t.uint32();
        switch (l >>> 3) {
          case 1:
            (u = o.coins) != null && u.length || (o.coins = []), o.coins.push(c.common.Coin.decode(t, t.uint32()));
            break;
          case 2:
            o.memo = t.string();
            break;
          case 3:
            o.signer = t.bytes();
            break;
          default:
            t.skipType(l & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      if (typeof t != "object" || t === null)
        return "object expected";
      if (t.coins != null && t.hasOwnProperty("coins")) {
        if (!Array.isArray(t.coins))
          return "coins: array expected";
        for (let e = 0; e < t.coins.length; ++e) {
          let n = c.common.Coin.verify(t.coins[e]);
          if (n)
            return "coins." + n;
        }
      }
      return t.memo != null && t.hasOwnProperty("memo") && !s.isString(t.memo) ? "memo: string expected" : t.signer != null && t.hasOwnProperty("signer") && !(t.signer && typeof t.signer.length == "number" || s.isString(t.signer)) ? "signer: buffer expected" : null;
    }, i.fromObject = function(t) {
      if (t instanceof c.types.MsgDeposit)
        return t;
      let e = new c.types.MsgDeposit();
      if (t.coins) {
        if (!Array.isArray(t.coins))
          throw TypeError(".types.MsgDeposit.coins: array expected");
        e.coins = [];
        for (let n = 0; n < t.coins.length; ++n) {
          if (typeof t.coins[n] != "object")
            throw TypeError(".types.MsgDeposit.coins: object expected");
          e.coins[n] = c.common.Coin.fromObject(t.coins[n]);
        }
      }
      return t.memo != null && (e.memo = String(t.memo)), t.signer != null && (typeof t.signer == "string" ? s.base64.decode(
        t.signer,
        e.signer = s.newBuffer(s.base64.length(t.signer)),
        0
      ) : t.signer.length && (e.signer = t.signer)), e;
    }, i.toObject = function(t, e) {
      var o;
      e || (e = {});
      let n = {};
      if ((e.arrays || e.defaults) && (n.coins = []), e.defaults && (n.memo = "", e.bytes === String ? n.signer = "" : (n.signer = [], e.bytes !== Array && (n.signer = s.newBuffer(n.signer)))), (o = t.coins) != null && o.length) {
        n.coins = [];
        for (let u = 0; u < t.coins.length; ++u)
          n.coins[u] = c.common.Coin.toObject(t.coins[u], e);
      }
      return t.memo != null && t.hasOwnProperty("memo") && (n.memo = t.memo), t.signer != null && t.hasOwnProperty("signer") && (n.signer = e.bytes === String ? s.base64.encode(t.signer, 0, t.signer.length) : e.bytes === Array ? Array.prototype.slice.call(t.signer) : t.signer), n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f.MsgSend = function() {
    function i(r) {
      if (this.amount = [], r)
        for (let t = Object.keys(r), e = 0; e < t.length; ++e)
          r[t[e]] != null && (this[t[e]] = r[t[e]]);
    }
    return i.prototype.fromAddress = s.newBuffer([]), i.prototype.toAddress = s.newBuffer([]), i.prototype.amount = s.emptyArray, i.create = function(t) {
      return new i(t);
    }, i.encode = function(t, e) {
      var n;
      if (e || (e = m.create()), t.fromAddress != null && Object.hasOwnProperty.call(t, "fromAddress") && e.uint32(
        /* id 1, wireType 2 =*/
        10
      ).bytes(t.fromAddress), t.toAddress != null && Object.hasOwnProperty.call(t, "toAddress") && e.uint32(
        /* id 2, wireType 2 =*/
        18
      ).bytes(t.toAddress), (n = t.amount) != null && n.length)
        for (let o = 0; o < t.amount.length; ++o)
          c.cosmos.base.v1beta1.Coin.encode(
            t.amount[o],
            e.uint32(
              /* id 3, wireType 2 =*/
              26
            ).fork()
          ).ldelim();
      return e;
    }, i.encodeDelimited = function(t, e) {
      return this.encode(t, e).ldelim();
    }, i.decode = function(t, e) {
      var u;
      t instanceof d || (t = d.create(t));
      let n = e === void 0 ? t.len : t.pos + e, o = new c.types.MsgSend();
      for (; t.pos < n; ) {
        let l = t.uint32();
        switch (l >>> 3) {
          case 1:
            o.fromAddress = t.bytes();
            break;
          case 2:
            o.toAddress = t.bytes();
            break;
          case 3:
            (u = o.amount) != null && u.length || (o.amount = []), o.amount.push(c.cosmos.base.v1beta1.Coin.decode(t, t.uint32()));
            break;
          default:
            t.skipType(l & 7);
            break;
        }
      }
      return o;
    }, i.decodeDelimited = function(t) {
      return t instanceof d || (t = new d(t)), this.decode(t, t.uint32());
    }, i.verify = function(t) {
      if (typeof t != "object" || t === null)
        return "object expected";
      if (t.fromAddress != null && t.hasOwnProperty("fromAddress") && !(t.fromAddress && typeof t.fromAddress.length == "number" || s.isString(t.fromAddress)))
        return "fromAddress: buffer expected";
      if (t.toAddress != null && t.hasOwnProperty("toAddress") && !(t.toAddress && typeof t.toAddress.length == "number" || s.isString(t.toAddress)))
        return "toAddress: buffer expected";
      if (t.amount != null && t.hasOwnProperty("amount")) {
        if (!Array.isArray(t.amount))
          return "amount: array expected";
        for (let e = 0; e < t.amount.length; ++e) {
          let n = c.cosmos.base.v1beta1.Coin.verify(t.amount[e]);
          if (n)
            return "amount." + n;
        }
      }
      return null;
    }, i.fromObject = function(t) {
      if (t instanceof c.types.MsgSend)
        return t;
      let e = new c.types.MsgSend();
      if (t.fromAddress != null && (typeof t.fromAddress == "string" ? s.base64.decode(
        t.fromAddress,
        e.fromAddress = s.newBuffer(s.base64.length(t.fromAddress)),
        0
      ) : t.fromAddress.length && (e.fromAddress = t.fromAddress)), t.toAddress != null && (typeof t.toAddress == "string" ? s.base64.decode(
        t.toAddress,
        e.toAddress = s.newBuffer(s.base64.length(t.toAddress)),
        0
      ) : t.toAddress.length && (e.toAddress = t.toAddress)), t.amount) {
        if (!Array.isArray(t.amount))
          throw TypeError(".types.MsgSend.amount: array expected");
        e.amount = [];
        for (let n = 0; n < t.amount.length; ++n) {
          if (typeof t.amount[n] != "object")
            throw TypeError(".types.MsgSend.amount: object expected");
          e.amount[n] = c.cosmos.base.v1beta1.Coin.fromObject(t.amount[n]);
        }
      }
      return e;
    }, i.toObject = function(t, e) {
      var o;
      e || (e = {});
      let n = {};
      if ((e.arrays || e.defaults) && (n.amount = []), e.defaults && (e.bytes === String ? n.fromAddress = "" : (n.fromAddress = [], e.bytes !== Array && (n.fromAddress = s.newBuffer(n.fromAddress))), e.bytes === String ? n.toAddress = "" : (n.toAddress = [], e.bytes !== Array && (n.toAddress = s.newBuffer(n.toAddress)))), t.fromAddress != null && t.hasOwnProperty("fromAddress") && (n.fromAddress = e.bytes === String ? s.base64.encode(t.fromAddress, 0, t.fromAddress.length) : e.bytes === Array ? Array.prototype.slice.call(t.fromAddress) : t.fromAddress), t.toAddress != null && t.hasOwnProperty("toAddress") && (n.toAddress = e.bytes === String ? s.base64.encode(t.toAddress, 0, t.toAddress.length) : e.bytes === Array ? Array.prototype.slice.call(t.toAddress) : t.toAddress), (o = t.amount) != null && o.length) {
        n.amount = [];
        for (let u = 0; u < t.amount.length; ++u)
          n.amount[u] = c.cosmos.base.v1beta1.Coin.toObject(t.amount[u], e);
      }
      return n;
    }, i.prototype.toJSON = function() {
      return this.constructor.toObject(this, y.util.toJSONOptions);
    }, i;
  }(), f;
}();
c.cosmos = function() {
  let f = {};
  return f.base = function() {
    let i = {};
    return i.v1beta1 = function() {
      let r = {};
      return r.Coin = function() {
        function t(e) {
          if (e)
            for (let n = Object.keys(e), o = 0; o < n.length; ++o)
              e[n[o]] != null && (this[n[o]] = e[n[o]]);
        }
        return t.prototype.denom = "", t.prototype.amount = "", t.create = function(n) {
          return new t(n);
        }, t.encode = function(n, o) {
          return o || (o = m.create()), n.denom != null && Object.hasOwnProperty.call(n, "denom") && o.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(n.denom), n.amount != null && Object.hasOwnProperty.call(n, "amount") && o.uint32(
            /* id 2, wireType 2 =*/
            18
          ).string(n.amount), o;
        }, t.encodeDelimited = function(n, o) {
          return this.encode(n, o).ldelim();
        }, t.decode = function(n, o) {
          n instanceof d || (n = d.create(n));
          let u = o === void 0 ? n.len : n.pos + o, l = new c.cosmos.base.v1beta1.Coin();
          for (; n.pos < u; ) {
            let h = n.uint32();
            switch (h >>> 3) {
              case 1:
                l.denom = n.string();
                break;
              case 2:
                l.amount = n.string();
                break;
              default:
                n.skipType(h & 7);
                break;
            }
          }
          return l;
        }, t.decodeDelimited = function(n) {
          return n instanceof d || (n = new d(n)), this.decode(n, n.uint32());
        }, t.verify = function(n) {
          return typeof n != "object" || n === null ? "object expected" : n.denom != null && n.hasOwnProperty("denom") && !s.isString(n.denom) ? "denom: string expected" : n.amount != null && n.hasOwnProperty("amount") && !s.isString(n.amount) ? "amount: string expected" : null;
        }, t.fromObject = function(n) {
          if (n instanceof c.cosmos.base.v1beta1.Coin)
            return n;
          let o = new c.cosmos.base.v1beta1.Coin();
          return n.denom != null && (o.denom = String(n.denom)), n.amount != null && (o.amount = String(n.amount)), o;
        }, t.toObject = function(n, o) {
          o || (o = {});
          let u = {};
          return o.defaults && (u.denom = "", u.amount = ""), n.denom != null && n.hasOwnProperty("denom") && (u.denom = n.denom), n.amount != null && n.hasOwnProperty("amount") && (u.amount = n.amount), u;
        }, t.prototype.toJSON = function() {
          return this.constructor.toObject(this, y.util.toJSONOptions);
        }, t;
      }(), r.DecCoin = function() {
        function t(e) {
          if (e)
            for (let n = Object.keys(e), o = 0; o < n.length; ++o)
              e[n[o]] != null && (this[n[o]] = e[n[o]]);
        }
        return t.prototype.denom = "", t.prototype.amount = "", t.create = function(n) {
          return new t(n);
        }, t.encode = function(n, o) {
          return o || (o = m.create()), n.denom != null && Object.hasOwnProperty.call(n, "denom") && o.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(n.denom), n.amount != null && Object.hasOwnProperty.call(n, "amount") && o.uint32(
            /* id 2, wireType 2 =*/
            18
          ).string(n.amount), o;
        }, t.encodeDelimited = function(n, o) {
          return this.encode(n, o).ldelim();
        }, t.decode = function(n, o) {
          n instanceof d || (n = d.create(n));
          let u = o === void 0 ? n.len : n.pos + o, l = new c.cosmos.base.v1beta1.DecCoin();
          for (; n.pos < u; ) {
            let h = n.uint32();
            switch (h >>> 3) {
              case 1:
                l.denom = n.string();
                break;
              case 2:
                l.amount = n.string();
                break;
              default:
                n.skipType(h & 7);
                break;
            }
          }
          return l;
        }, t.decodeDelimited = function(n) {
          return n instanceof d || (n = new d(n)), this.decode(n, n.uint32());
        }, t.verify = function(n) {
          return typeof n != "object" || n === null ? "object expected" : n.denom != null && n.hasOwnProperty("denom") && !s.isString(n.denom) ? "denom: string expected" : n.amount != null && n.hasOwnProperty("amount") && !s.isString(n.amount) ? "amount: string expected" : null;
        }, t.fromObject = function(n) {
          if (n instanceof c.cosmos.base.v1beta1.DecCoin)
            return n;
          let o = new c.cosmos.base.v1beta1.DecCoin();
          return n.denom != null && (o.denom = String(n.denom)), n.amount != null && (o.amount = String(n.amount)), o;
        }, t.toObject = function(n, o) {
          o || (o = {});
          let u = {};
          return o.defaults && (u.denom = "", u.amount = ""), n.denom != null && n.hasOwnProperty("denom") && (u.denom = n.denom), n.amount != null && n.hasOwnProperty("amount") && (u.amount = n.amount), u;
        }, t.prototype.toJSON = function() {
          return this.constructor.toObject(this, y.util.toJSONOptions);
        }, t;
      }(), r.IntProto = function() {
        function t(e) {
          if (e)
            for (let n = Object.keys(e), o = 0; o < n.length; ++o)
              e[n[o]] != null && (this[n[o]] = e[n[o]]);
        }
        return t.prototype.int = "", t.create = function(n) {
          return new t(n);
        }, t.encode = function(n, o) {
          return o || (o = m.create()), n.int != null && Object.hasOwnProperty.call(n, "int") && o.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(n.int), o;
        }, t.encodeDelimited = function(n, o) {
          return this.encode(n, o).ldelim();
        }, t.decode = function(n, o) {
          n instanceof d || (n = d.create(n));
          let u = o === void 0 ? n.len : n.pos + o, l = new c.cosmos.base.v1beta1.IntProto();
          for (; n.pos < u; ) {
            let h = n.uint32();
            switch (h >>> 3) {
              case 1:
                l.int = n.string();
                break;
              default:
                n.skipType(h & 7);
                break;
            }
          }
          return l;
        }, t.decodeDelimited = function(n) {
          return n instanceof d || (n = new d(n)), this.decode(n, n.uint32());
        }, t.verify = function(n) {
          return typeof n != "object" || n === null ? "object expected" : n.int != null && n.hasOwnProperty("int") && !s.isString(n.int) ? "int: string expected" : null;
        }, t.fromObject = function(n) {
          if (n instanceof c.cosmos.base.v1beta1.IntProto)
            return n;
          let o = new c.cosmos.base.v1beta1.IntProto();
          return n.int != null && (o.int = String(n.int)), o;
        }, t.toObject = function(n, o) {
          o || (o = {});
          let u = {};
          return o.defaults && (u.int = ""), n.int != null && n.hasOwnProperty("int") && (u.int = n.int), u;
        }, t.prototype.toJSON = function() {
          return this.constructor.toObject(this, y.util.toJSONOptions);
        }, t;
      }(), r.DecProto = function() {
        function t(e) {
          if (e)
            for (let n = Object.keys(e), o = 0; o < n.length; ++o)
              e[n[o]] != null && (this[n[o]] = e[n[o]]);
        }
        return t.prototype.dec = "", t.create = function(n) {
          return new t(n);
        }, t.encode = function(n, o) {
          return o || (o = m.create()), n.dec != null && Object.hasOwnProperty.call(n, "dec") && o.uint32(
            /* id 1, wireType 2 =*/
            10
          ).string(n.dec), o;
        }, t.encodeDelimited = function(n, o) {
          return this.encode(n, o).ldelim();
        }, t.decode = function(n, o) {
          n instanceof d || (n = d.create(n));
          let u = o === void 0 ? n.len : n.pos + o, l = new c.cosmos.base.v1beta1.DecProto();
          for (; n.pos < u; ) {
            let h = n.uint32();
            switch (h >>> 3) {
              case 1:
                l.dec = n.string();
                break;
              default:
                n.skipType(h & 7);
                break;
            }
          }
          return l;
        }, t.decodeDelimited = function(n) {
          return n instanceof d || (n = new d(n)), this.decode(n, n.uint32());
        }, t.verify = function(n) {
          return typeof n != "object" || n === null ? "object expected" : n.dec != null && n.hasOwnProperty("dec") && !s.isString(n.dec) ? "dec: string expected" : null;
        }, t.fromObject = function(n) {
          if (n instanceof c.cosmos.base.v1beta1.DecProto)
            return n;
          let o = new c.cosmos.base.v1beta1.DecProto();
          return n.dec != null && (o.dec = String(n.dec)), o;
        }, t.toObject = function(n, o) {
          o || (o = {});
          let u = {};
          return o.defaults && (u.dec = ""), n.dec != null && n.hasOwnProperty("dec") && (u.dec = n.dec), u;
        }, t.prototype.toJSON = function() {
          return this.constructor.toObject(this, y.util.toJSONOptions);
        }, t;
      }(), r;
    }(), i;
  }(), f;
}();
export {
  c as default
};
