// @shapeshiftoss/unchained-client@10.1.1/dist/evm/parser/erc20.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parser = void 0;
const caip_1 = require("@shapeshiftoss/caip");
const ethers_1 = require("ethers");
const erc20_1 = __importDefault(require("./abi/erc20"));
const utils_1 = require("./utils");
class Parser {
    constructor(args) {
        this.abiInterface = new ethers_1.ethers.utils.Interface(erc20_1.default);
        this.supportedFunctions = {
            approveSigHash: this.abiInterface.getSighash('approve'),
        };
        this.provider = args.provider;
        this.chainId = args.chainId;
    }
    async parse(tx) {
        if (!tx.inputData)
            return;
        const txSigHash = (0, utils_1.getSigHash)(tx.inputData);
        if (!Object.values(this.supportedFunctions).some((hash) => hash === txSigHash))
            return;
        const decoded = this.abiInterface.parseTransaction({ data: tx.inputData });
        // failed to decode input data
        if (!decoded)
            return;
        const data = {
            assetId: (0, caip_1.toAssetId)({
                ...(0, caip_1.fromChainId)(this.chainId),
                assetNamespace: 'erc20',
                assetReference: tx.to,
            }),
            method: decoded.name,
            parser: 'erc20',
        };
        switch (txSigHash) {
            case this.supportedFunctions.approveSigHash:
                return { data: { ...data, value: (decoded === null || decoded === void 0 ? void 0 : decoded.args.amount).toString() } };
            default:
                return { data };
        }
    }
}
exports.Parser = Parser;
