// @trezor/connect@9.1.1/lib/api/cardano/cardanoOutputs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendOutput = exports.transformOutput = void 0;
const paramsValidator_1 = require("../common/paramsValidator");
const cardanoAddressParameters_1 = require("./cardanoAddressParameters");
const cardanoTokenBundle_1 = require("./cardanoTokenBundle");
const cardanoUtils_1 = require("./cardanoUtils");
const transformOutput = (output) => {
    (0, paramsValidator_1.validateParams)(output, [
        { name: 'address', type: 'string' },
        { name: 'amount', type: 'uint', required: true },
        { name: 'tokenBundle', type: 'array', allowEmpty: true },
        { name: 'datumHash', type: 'string' },
        { name: 'format', type: 'number' },
        { name: 'inlineDatum', type: 'string' },
        { name: 'referenceScript', type: 'string' },
    ]);
    const result = {
        output: {
            amount: output.amount,
            asset_groups_count: 0,
            datum_hash: output.datumHash,
            format: output.format,
            inline_datum_size: output.inlineDatum
                ? (0, cardanoUtils_1.hexStringByteLength)(output.inlineDatum)
                : undefined,
            reference_script_size: output.referenceScript
                ? (0, cardanoUtils_1.hexStringByteLength)(output.referenceScript)
                : undefined,
        },
        inlineDatum: output.inlineDatum,
        referenceScript: output.referenceScript,
    };
    if (output.addressParameters) {
        (0, cardanoAddressParameters_1.validateAddressParameters)(output.addressParameters);
        result.output.address_parameters = (0, cardanoAddressParameters_1.addressParametersToProto)(output.addressParameters);
    }
    else {
        result.output.address = output.address;
    }
    if (output.tokenBundle) {
        result.tokenBundle = (0, cardanoTokenBundle_1.tokenBundleToProto)(output.tokenBundle);
        result.output.asset_groups_count = result.tokenBundle.length;
    }
    else {
        result.output.asset_groups_count = 0;
    }
    return result;
};
exports.transformOutput = transformOutput;
const sendOutput = async (typedCall, outputWithData) => {
    const MAX_CHUNK_SIZE = 1024 * 2;
    const { output, tokenBundle, inlineDatum, referenceScript } = outputWithData;
    await typedCall('CardanoTxOutput', 'CardanoTxItemAck', output);
    if (tokenBundle) {
        for (const assetGroup of tokenBundle) {
            await typedCall('CardanoAssetGroup', 'CardanoTxItemAck', {
                policy_id: assetGroup.policyId,
                tokens_count: assetGroup.tokens.length,
            });
            for (const token of assetGroup.tokens) {
                await typedCall('CardanoToken', 'CardanoTxItemAck', token);
            }
        }
    }
    if (inlineDatum) {
        await (0, cardanoUtils_1.sendChunkedHexString)(typedCall, inlineDatum, MAX_CHUNK_SIZE, 'CardanoTxInlineDatumChunk');
    }
    if (referenceScript) {
        await (0, cardanoUtils_1.sendChunkedHexString)(typedCall, referenceScript, MAX_CHUNK_SIZE, 'CardanoTxReferenceScriptChunk');
    }
};
exports.sendOutput = sendOutput;
