// @trezor/connect@9.1.1/lib/api/cardano/cardanoInputs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformReferenceInput = exports.transformCollateralInput = exports.transformInput = void 0;
const paramsValidator_1 = require("../common/paramsValidator");
const pathUtils_1 = require("../../utils/pathUtils");
const transformInput = (input) => {
    (0, paramsValidator_1.validateParams)(input, [
        { name: 'prev_hash', type: 'string', required: true },
        { name: 'prev_index', type: 'number', required: true },
    ]);
    return {
        input: {
            prev_hash: input.prev_hash,
            prev_index: input.prev_index,
        },
        path: input.path ? (0, pathUtils_1.validatePath)(input.path, 5) : undefined,
    };
};
exports.transformInput = transformInput;
const transformCollateralInput = (collateralInput) => {
    (0, paramsValidator_1.validateParams)(collateralInput, [
        { name: 'prev_hash', type: 'string', required: true },
        { name: 'prev_index', type: 'number', required: true },
    ]);
    return {
        collateralInput: {
            prev_hash: collateralInput.prev_hash,
            prev_index: collateralInput.prev_index,
        },
        path: collateralInput.path ? (0, pathUtils_1.validatePath)(collateralInput.path, 5) : undefined,
    };
};
exports.transformCollateralInput = transformCollateralInput;
const transformReferenceInput = (referenceInput) => {
    (0, paramsValidator_1.validateParams)(referenceInput, [
        { name: 'prev_hash', type: 'string', required: true },
        { name: 'prev_index', type: 'number', required: true },
    ]);
    return {
        prev_hash: referenceInput.prev_hash,
        prev_index: referenceInput.prev_index,
    };
};
exports.transformReferenceInput = transformReferenceInput;
