// fiosdk-offline@1.2.21/lib/entities/FioActionParameters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FioVoteProducerActionName = exports.FioVoteProducerActionAccount = exports.FioUnregisterProxyActionName = exports.FioUnregisterProxyActionAccount = exports.FioUnregisterProducerActionName = exports.FioUnregisterProducerActionAccount = exports.FioTransferTokensPubKeyActionName = exports.FioTransferTokensPubKeyActionAccount = exports.FioSubmitFeeRatiosActionName = exports.FioSubmitFeeRatiosActionAccount = exports.FioSubmitFeeMultiplierActionName = exports.FioSubmitFeeMultiplierActionAccount = exports.FioSubmitBundledTransactionActionName = exports.FioSubmitBundledTransactionActionAccount = exports.FioSetFioDomainPublicActionName = exports.FioSetFioDomainPublicActionAccount = exports.FioRenewFioDomainActionName = exports.FioRenewFioDomainActionAccount = exports.FioRenewFioAddressActionName = exports.FioRenewFioAddressActionAccount = exports.FioRejectFundsRequestActionName = exports.FioRejectFundsRequestActionAccount = exports.FioRegisterProxyActionName = exports.FioRegisterProxyActionAccount = exports.FioRegisterProducerActionName = exports.FioRegisterProducerActionAccount = exports.FioRegisterFioDomainActionName = exports.FioRegisterFioDomainActionAccount = exports.FioRegisterFioAddressActionName = exports.FioRegisterFioAddressActionAccount = exports.FioRecordObtDataActionName = exports.FioRecordObtDataActionAccount = exports.FioProxyVoteActionName = exports.FioProxyVoteActionAccount = exports.FioPayTpidRewardsActionName = exports.FioPayTpidRewardsActionAccount = exports.FioNewFundsRequestActionName = exports.FioNewFundsRequestActionAccount = exports.FioClaimBPRewardsActionName = exports.FioClaimBPRewardsActionAccount = exports.FioBurnExpiredActionName = exports.FioBurnExpiredActionAccount = exports.FioAddPubAddressActionName = exports.FioAddPubAddressActionAccount = void 0;
exports.FioAddPubAddressActionAccount = 'fio.address';
exports.FioAddPubAddressActionName = 'addaddress';
exports.FioBurnExpiredActionAccount = 'fio.address';
exports.FioBurnExpiredActionName = 'burnexpired';
exports.FioClaimBPRewardsActionAccount = 'fio.treasury';
exports.FioClaimBPRewardsActionName = 'bpclaim';
exports.FioNewFundsRequestActionAccount = 'fio.reqobt';
exports.FioNewFundsRequestActionName = 'newfundsreq';
exports.FioPayTpidRewardsActionAccount = 'fio.treasury';
exports.FioPayTpidRewardsActionName = 'tpidclaim';
exports.FioProxyVoteActionAccount = 'eosio';
exports.FioProxyVoteActionName = 'voteproxy';
exports.FioRecordObtDataActionAccount = 'fio.reqobt';
exports.FioRecordObtDataActionName = 'recordobt';
exports.FioRegisterFioAddressActionAccount = 'fio.address';
exports.FioRegisterFioAddressActionName = 'regaddress';
exports.FioRegisterFioDomainActionAccount = 'fio.address';
exports.FioRegisterFioDomainActionName = 'regdomain';
exports.FioRegisterProducerActionAccount = 'eosio';
exports.FioRegisterProducerActionName = 'regproducer';
exports.FioRegisterProxyActionAccount = 'eosio';
exports.FioRegisterProxyActionName = 'regproxy';
exports.FioRejectFundsRequestActionAccount = 'fio.reqobt';
exports.FioRejectFundsRequestActionName = 'rejectfndreq';
exports.FioRenewFioAddressActionAccount = 'fio.address';
exports.FioRenewFioAddressActionName = 'renewaddress';
exports.FioRenewFioDomainActionAccount = 'fio.address';
exports.FioRenewFioDomainActionName = 'renewdomain';
exports.FioSetFioDomainPublicActionAccount = 'fio.address';
exports.FioSetFioDomainPublicActionName = 'setdomainpub';
exports.FioSubmitBundledTransactionActionAccount = 'fio.fee';
exports.FioSubmitBundledTransactionActionName = 'bundlevote';
exports.FioSubmitFeeMultiplierActionAccount = 'fio.fee';
exports.FioSubmitFeeMultiplierActionName = 'setfeemult';
exports.FioSubmitFeeRatiosActionAccount = 'fio.fee';
exports.FioSubmitFeeRatiosActionName = 'setfeevote';
exports.FioTransferTokensPubKeyActionAccount = 'fio.token';
exports.FioTransferTokensPubKeyActionName = 'trnsfiopubky';
exports.FioUnregisterProducerActionAccount = 'eosio';
exports.FioUnregisterProducerActionName = 'unregprod';
exports.FioUnregisterProxyActionAccount = 'eosio';
exports.FioUnregisterProxyActionName = 'unregproxy';
exports.FioVoteProducerActionAccount = 'eosio';
exports.FioVoteProducerActionName = 'voteproducer';
