// @keepkey/keepkey-sdk@0.2.52/dist/generated/apis/DebugApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugApi = void 0;
const runtime = __importStar(require("../runtime"));
const models_1 = require("../models");
/**
 *
 */
class DebugApi extends runtime.BaseAPI {
    /**
     * On a DEBUG_LINK bootloader, fills the storage sectors with a dummy data pattern.
     * Fill storage with dummy data
     */
    async debugLinkFillConfigRaw(initOverrides) {
        const queryParameters = {};
        const headerParameters = {};
        const response = await this.request({
            path: `/system/debug/fill-config`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);
        return new runtime.VoidApiResponse(response);
    }
    /**
     * On a DEBUG_LINK bootloader, fills the storage sectors with a dummy data pattern.
     * Fill storage with dummy data
     */
    async debugLinkFillConfig(initOverrides) {
        await this.debugLinkFillConfigRaw(initOverrides);
    }
    /**
     * On DEBUG_LINK firmware, dumps a section of flash memory.
     * Dump flash memory
     */
    async debugLinkFlashDumpRaw(requestParameters, initOverrides) {
        if (requestParameters.DebugLinkFlashDumpRequest === null || requestParameters.DebugLinkFlashDumpRequest === undefined) {
            throw new runtime.RequiredError('DebugLinkFlashDumpRequest', 'Required parameter requestParameters.DebugLinkFlashDumpRequest was null or undefined when calling debugLinkFlashDump.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        const response = await this.request({
            path: `/system/debug/flash-dump`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.DebugLinkFlashDumpRequestToJSON)(requestParameters.DebugLinkFlashDumpRequest),
        }, initOverrides);
        return new runtime.BlobApiResponse(response);
    }
    /**
     * On DEBUG_LINK firmware, dumps a section of flash memory.
     * Dump flash memory
     */
    async debugLinkFlashDump(DebugLinkFlashDumpRequest, initOverrides) {
        const response = await this.debugLinkFlashDumpRaw({ DebugLinkFlashDumpRequest: DebugLinkFlashDumpRequest }, initOverrides);
        return await response.value();
    }
    /**
     * On DEBUG_LINK firmware, dumps various internal sensitive paramters.
     * Dump sensitive device state
     */
    async debugLinkGetStateRaw(initOverrides) {
        const queryParameters = {};
        const headerParameters = {};
        const response = await this.request({
            path: `/system/debug/get-state`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, models_1.DebugLinkGetState200ResponseFromJSON)(jsonValue));
    }
    /**
     * On DEBUG_LINK firmware, dumps various internal sensitive paramters.
     * Dump sensitive device state
     */
    async debugLinkGetState(initOverrides) {
        const response = await this.debugLinkGetStateRaw(initOverrides);
        return await response.value();
    }
}
exports.DebugApi = DebugApi;
