// @keepkey/keepkey-sdk@0.2.52/dist/generated/apis/ETHApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ETHApi = void 0;
const runtime = __importStar(require("../runtime"));
const models_1 = require("../models");
/**
 *
 */
class ETHApi extends runtime.BaseAPI {
    /**
     * Sign an Etherum message
     */
    async ethSignRaw(requestParameters, initOverrides) {
        if (requestParameters.EthSignRequest === null || requestParameters.EthSignRequest === undefined) {
            throw new runtime.RequiredError('EthSignRequest', 'Required parameter requestParameters.EthSignRequest was null or undefined when calling ethSign.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/eth/sign`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.EthSignRequestToJSON)(requestParameters.EthSignRequest),
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, models_1.TypesEthSignatureFromJSON)(jsonValue));
    }
    /**
     * Sign an Etherum message
     */
    async ethSign(EthSignRequest, initOverrides) {
        const response = await this.ethSignRaw({ EthSignRequest: EthSignRequest }, initOverrides);
        return await response.value();
    }
    /**
     * Sign an Ethereum transaction
     */
    async ethSignTransactionRaw(requestParameters, initOverrides) {
        if (requestParameters.EthSignTransactionRequest === null || requestParameters.EthSignTransactionRequest === undefined) {
            throw new runtime.RequiredError('EthSignTransactionRequest', 'Required parameter requestParameters.EthSignTransactionRequest was null or undefined when calling ethSignTransaction.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/eth/sign-transaction`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.EthSignTransactionRequestToJSON)(requestParameters.EthSignTransactionRequest),
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, models_1.EthSignTransaction200ResponseFromJSON)(jsonValue));
    }
    /**
     * Sign an Ethereum transaction
     */
    async ethSignTransaction(EthSignTransactionRequest, initOverrides) {
        const response = await this.ethSignTransactionRaw({ EthSignTransactionRequest: EthSignTransactionRequest }, initOverrides);
        return await response.value();
    }
    /**
     * Sign EIP-712 typed data
     */
    async ethSignTypedDataRaw(requestParameters, initOverrides) {
        if (requestParameters.EthSignTypedDataRequest === null || requestParameters.EthSignTypedDataRequest === undefined) {
            throw new runtime.RequiredError('EthSignTypedDataRequest', 'Required parameter requestParameters.EthSignTypedDataRequest was null or undefined when calling ethSignTypedData.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/eth/sign-typed-data`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.EthSignTypedDataRequestToJSON)(requestParameters.EthSignTypedDataRequest),
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, models_1.TypesEthSignatureFromJSON)(jsonValue));
    }
    /**
     * Sign EIP-712 typed data
     */
    async ethSignTypedData(EthSignTypedDataRequest, initOverrides) {
        const response = await this.ethSignTypedDataRaw({ EthSignTypedDataRequest: EthSignTypedDataRequest }, initOverrides);
        return await response.value();
    }
    /**
     * Verify an Etherum message
     */
    async ethVerifyRaw(requestParameters, initOverrides) {
        if (requestParameters.EthVerifyRequest === null || requestParameters.EthVerifyRequest === undefined) {
            throw new runtime.RequiredError('EthVerifyRequest', 'Required parameter requestParameters.EthVerifyRequest was null or undefined when calling ethVerify.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/eth/verify`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.EthVerifyRequestToJSON)(requestParameters.EthVerifyRequest),
        }, initOverrides);
        return new runtime.TextApiResponse(response);
    }
    /**
     * Verify an Etherum message
     */
    async ethVerify(EthVerifyRequest, initOverrides) {
        const response = await this.ethVerifyRaw({ EthVerifyRequest: EthVerifyRequest }, initOverrides);
        return await response.value();
    }
}
exports.ETHApi = ETHApi;
