// @keepkey/keepkey-sdk@0.2.52/dist/generated/apis/InitializeApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitializeApi = void 0;
const runtime = __importStar(require("../runtime"));
const models_1 = require("../models");
/**
 *
 */
class InitializeApi extends runtime.BaseAPI {
    /**
     * Load a device with externally-derived key material. (Not recommended.)
     * Load device
     */
    async loadDeviceRaw(requestParameters, initOverrides) {
        if (requestParameters.LoadDeviceRequest === null || requestParameters.LoadDeviceRequest === undefined) {
            throw new runtime.RequiredError('LoadDeviceRequest', 'Required parameter requestParameters.LoadDeviceRequest was null or undefined when calling loadDevice.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/system/initialize/load-device`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.LoadDeviceRequestToJSON)(requestParameters.LoadDeviceRequest),
        }, initOverrides);
        return new runtime.VoidApiResponse(response);
    }
    /**
     * Load a device with externally-derived key material. (Not recommended.)
     * Load device
     */
    async loadDevice(LoadDeviceRequest, initOverrides) {
        await this.loadDeviceRaw({ LoadDeviceRequest: LoadDeviceRequest }, initOverrides);
    }
    /**
     * Start safe recovery workflow to restore from a recovery sentence.
     * Recover device
     */
    async recoverDeviceRaw(requestParameters, initOverrides) {
        if (requestParameters.RecoverDeviceRequest === null || requestParameters.RecoverDeviceRequest === undefined) {
            throw new runtime.RequiredError('RecoverDeviceRequest', 'Required parameter requestParameters.RecoverDeviceRequest was null or undefined when calling recoverDevice.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/system/initialize/recover-device`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.RecoverDeviceRequestToJSON)(requestParameters.RecoverDeviceRequest),
        }, initOverrides);
        return new runtime.VoidApiResponse(response);
    }
    /**
     * Start safe recovery workflow to restore from a recovery sentence.
     * Recover device
     */
    async recoverDevice(RecoverDeviceRequest, initOverrides) {
        await this.recoverDeviceRaw({ RecoverDeviceRequest: RecoverDeviceRequest }, initOverrides);
    }
    /**
     * Perform device setup and generate new seed.
     * Set up device
     */
    async resetDeviceRaw(requestParameters, initOverrides) {
        if (requestParameters.ResetDeviceRequest === null || requestParameters.ResetDeviceRequest === undefined) {
            throw new runtime.RequiredError('ResetDeviceRequest', 'Required parameter requestParameters.ResetDeviceRequest was null or undefined when calling resetDevice.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/system/initialize/reset-device`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.ResetDeviceRequestToJSON)(requestParameters.ResetDeviceRequest),
        }, initOverrides);
        return new runtime.VoidApiResponse(response);
    }
    /**
     * Perform device setup and generate new seed.
     * Set up device
     */
    async resetDevice(ResetDeviceRequest, initOverrides) {
        await this.resetDeviceRaw({ ResetDeviceRequest: ResetDeviceRequest }, initOverrides);
    }
}
exports.InitializeApi = InitializeApi;
