// @keepkey/keepkey-sdk@0.2.52/dist/generated/apis/ThorchainApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThorchainApi = void 0;
const runtime = __importStar(require("../runtime"));
const models_1 = require("../models");
/**
 *
 */
class ThorchainApi extends runtime.BaseAPI {
    /**
     * Sign a Cosmos SDK transaction using SIGN_MODE_AMINO_JSON
     */
    async thorchainSignAminoDepositRaw(requestParameters, initOverrides) {
        if (requestParameters.ThorchainSignAminoDepositRequest === null || requestParameters.ThorchainSignAminoDepositRequest === undefined) {
            throw new runtime.RequiredError('ThorchainSignAminoDepositRequest', 'Required parameter requestParameters.ThorchainSignAminoDepositRequest was null or undefined when calling thorchainSignAminoDeposit.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/thorchain/sign-amino-desposit`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.ThorchainSignAminoDepositRequestToJSON)(requestParameters.ThorchainSignAminoDepositRequest),
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, models_1.ThorchainSignAminoTransfer200ResponseFromJSON)(jsonValue));
    }
    /**
     * Sign a Cosmos SDK transaction using SIGN_MODE_AMINO_JSON
     */
    async thorchainSignAminoDeposit(ThorchainSignAminoDepositRequest, initOverrides) {
        const response = await this.thorchainSignAminoDepositRaw({ ThorchainSignAminoDepositRequest: ThorchainSignAminoDepositRequest }, initOverrides);
        return await response.value();
    }
    /**
     * Sign a Cosmos SDK transaction using SIGN_MODE_AMINO_JSON
     */
    async thorchainSignAminoTransferRaw(requestParameters, initOverrides) {
        if (requestParameters.ThorchainSignAminoTransferRequest === null || requestParameters.ThorchainSignAminoTransferRequest === undefined) {
            throw new runtime.RequiredError('ThorchainSignAminoTransferRequest', 'Required parameter requestParameters.ThorchainSignAminoTransferRequest was null or undefined when calling thorchainSignAminoTransfer.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/thorchain/sign-amino-transfer`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.ThorchainSignAminoTransferRequestToJSON)(requestParameters.ThorchainSignAminoTransferRequest),
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, models_1.ThorchainSignAminoTransfer200ResponseFromJSON)(jsonValue));
    }
    /**
     * Sign a Cosmos SDK transaction using SIGN_MODE_AMINO_JSON
     */
    async thorchainSignAminoTransfer(ThorchainSignAminoTransferRequest, initOverrides) {
        const response = await this.thorchainSignAminoTransferRaw({ ThorchainSignAminoTransferRequest: ThorchainSignAminoTransferRequest }, initOverrides);
        return await response.value();
    }
}
exports.ThorchainApi = ThorchainApi;
