// @keepkey/keepkey-sdk@0.2.52/dist/generated/apis/UTXOApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UTXOApi = void 0;
const runtime = __importStar(require("../runtime"));
const models_1 = require("../models");
/**
 *
 */
class UTXOApi extends runtime.BaseAPI {
    /**
     * Sign a UTXO transaction
     */
    async utxoSignTransactionRaw(requestParameters, initOverrides) {
        if (requestParameters.UtxoSignTransactionRequest === null || requestParameters.UtxoSignTransactionRequest === undefined) {
            throw new runtime.RequiredError('UtxoSignTransactionRequest', 'Required parameter requestParameters.UtxoSignTransactionRequest was null or undefined when calling utxoSignTransaction.');
        }
        const queryParameters = {};
        const headerParameters = {};
        headerParameters['Content-Type'] = 'application/json';
        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("apiKey", []);
            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }
        const response = await this.request({
            path: `/utxo/sign-transaction`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: (0, models_1.UtxoSignTransactionRequestToJSON)(requestParameters.UtxoSignTransactionRequest),
        }, initOverrides);
        return new runtime.JSONApiResponse(response, (jsonValue) => (0, models_1.UtxoSignTransaction200ResponseFromJSON)(jsonValue));
    }
    /**
     * Sign a UTXO transaction
     */
    async utxoSignTransaction(UtxoSignTransactionRequest, initOverrides) {
        const response = await this.utxoSignTransactionRaw({ UtxoSignTransactionRequest: UtxoSignTransactionRequest }, initOverrides);
        return await response.value();
    }
}
exports.UTXOApi = UTXOApi;
