// @trezor/transport@1.1.14/lib/transports/abstract.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractTransport = void 0;
const protobuf = __importStar(require("protobufjs/light"));
const utils_1 = require("@trezor/utils");
const typedEventEmitter_1 = require("@trezor/utils/lib/typedEventEmitter");
const result_1 = require("../utils/result");
const ERRORS = __importStar(require("../errors"));
const constants_1 = require("../constants");
class AbstractTransport extends typedEventEmitter_1.TypedEmitter {
    constructor({ messages, signal, logger }) {
        super();
        this.isOutdated = false;
        this.version = '';
        this.stopped = false;
        this.listening = false;
        this.acquiredUnconfirmed = {};
        this.listenPromise = {};
        this.unknownError = (err, expectedErrors) => {
            this.logger.error(this.name, 'unexpected error: ', err);
            return (0, result_1.unknownError)(typeof err !== 'string' ? err : new Error(err), expectedErrors);
        };
        this.createLocalAbortController = () => {
            const localAbortController = new AbortController();
            const abort = () => localAbortController.abort();
            localAbortController.signal.addEventListener('abort', () => {
                this.abortController.signal.removeEventListener('abort', abort);
            });
            this.abortController.signal.addEventListener('abort', abort);
            return { signal: localAbortController.signal, abort };
        };
        this.scheduleAction = (action, params, errors) => {
            const { signal, abort } = this.createLocalAbortController();
            return {
                promise: (0, utils_1.scheduleAction)(action, Object.assign({ signal, timeout: constants_1.ACTION_TIMEOUT }, params))
                    .catch(err => {
                    const expectedErrors = [ERRORS.ABORTED_BY_TIMEOUT, ERRORS.ABORTED_BY_SIGNAL];
                    if (errors) {
                        expectedErrors.push(...errors);
                    }
                    return (0, result_1.unknownError)(err, expectedErrors);
                })
                    .finally(() => {
                    this.abortController.signal.removeEventListener('abort', abort);
                }),
                abort,
            };
        };
        this.descriptors = [];
        this.messages = protobuf.Root.fromJSON(messages);
        this.abortController = new AbortController();
        if (signal) {
            const abort = () => this.abortController.abort();
            this.abortController.signal.addEventListener('abort', () => signal.removeEventListener('abort', abort));
            signal.addEventListener('abort', abort);
        }
        this.logger = logger || {
            debug: (..._args) => { },
            log: (..._args) => { },
            warn: (..._args) => { },
            error: (..._args) => { },
        };
    }
    _getDiff(nextDescriptors) {
        const connected = nextDescriptors.filter(nextDescriptor => !this.descriptors.find(descriptor => descriptor.path === nextDescriptor.path));
        const disconnected = this.descriptors.filter(d => nextDescriptors.find(x => x.path === d.path) === undefined);
        const changedSessions = nextDescriptors.filter(d => {
            const currentDescriptor = this.descriptors.find(x => x.path === d.path);
            if (currentDescriptor) {
                return currentDescriptor.session !== d.session;
            }
            return false;
        });
        const acquired = changedSessions.filter(d => typeof d.session === 'string');
        const acquiredByMyself = acquired.filter(d => d.session === this.acquiredUnconfirmed[d.path]);
        const acquiredElsewhere = acquired.filter(d => d.session !== this.acquiredUnconfirmed[d.path]);
        const released = changedSessions.filter(d => typeof d.session !== 'string');
        const releasedByMyself = released.filter(d => {
            var _a;
            return ((_a = this.descriptors.find(prevD => prevD.path === d.path)) === null || _a === void 0 ? void 0 : _a.session) ===
                this.releasingSession;
        });
        const releasedElsewhere = released.filter(d => {
            var _a;
            return ((_a = this.descriptors.find(prevD => prevD.path === d.path)) === null || _a === void 0 ? void 0 : _a.session) !==
                this.releasingSession;
        });
        const didUpdate = connected.length + disconnected.length + changedSessions.length > 0;
        return {
            connected,
            disconnected,
            changedSessions,
            acquired,
            acquiredByMyself,
            acquiredElsewhere,
            released,
            releasedByMyself,
            releasedElsewhere,
            didUpdate,
            descriptors: nextDescriptors,
        };
    }
    handleDescriptorsChange(nextDescriptors) {
        if (this.stopped) {
            return;
        }
        const diff = this._getDiff(nextDescriptors);
        this.logger.debug('nextDescriptors', nextDescriptors, 'diff', diff);
        if (!diff.didUpdate) {
            return;
        }
        this.descriptors = nextDescriptors;
        Object.keys(this.listenPromise).forEach(path => {
            const descriptor = nextDescriptors.find(device => device.path === path);
            if (!descriptor) {
                return this.listenPromise[path].resolve(this.error({ error: ERRORS.DEVICE_DISCONNECTED_DURING_ACTION }));
            }
            if (this.acquiredUnconfirmed[path]) {
                const reportedNextSession = descriptor.session;
                if (reportedNextSession === this.acquiredUnconfirmed[path]) {
                    this.listenPromise[path].resolve(this.success(this.acquiredUnconfirmed[path]));
                }
                else {
                    this.listenPromise[path].resolve(this.error({ error: ERRORS.SESSION_WRONG_PREVIOUS }));
                }
                delete this.acquiredUnconfirmed[path];
            }
            else if (this.releasingSession) {
                this.listenPromise[path].resolve(this.success('null'));
            }
            else {
                this.listenPromise[path].resolve(this.error({ error: ERRORS.SESSION_WRONG_PREVIOUS }));
            }
        });
        this.emit(constants_1.TRANSPORT.UPDATE, diff);
        this.releasingSession = undefined;
    }
    success(payload) {
        return (0, result_1.success)(payload);
    }
    error(payload) {
        return (0, result_1.error)(payload);
    }
}
exports.AbstractTransport = AbstractTransport;
