// @trezor/transport@1.1.14/lib/transports/nodeusb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeUsbTransport = void 0;
const usb_1 = require("usb");
const abstractUsb_1 = require("./abstractUsb");
const client_1 = require("../sessions/client");
const background_1 = require("../sessions/background");
const usb_2 = require("../interfaces/usb");
class NodeUsbTransport extends abstractUsb_1.AbstractUsbTransport {
    constructor({ messages, logger }) {
        const sessionsBackground = new background_1.SessionsBackground();
        const sessionsClient = new client_1.SessionsClient({
            requestFn: args => sessionsBackground.handleMessage(args),
            registerBackgroundCallbacks: () => { },
        });
        sessionsBackground.on('descriptors', descriptors => {
            sessionsClient.emit('descriptors', descriptors);
        });
        super({
            messages,
            usbInterface: new usb_2.UsbInterface({
                usbInterface: new usb_1.WebUSB({
                    allowAllDevices: true,
                }),
                logger,
            }),
            sessionsClient,
        });
        this.name = 'NodeUsbTransport';
    }
}
exports.NodeUsbTransport = NodeUsbTransport;
