// hive-keychain-commons@1.2.4/lib/logger/transports.js
"use strict";
/* istanbul ignore file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerTransports = void 0;
const winston = require("winston");
const formats_1 = require("./formats");
const DailyRotateFile = require("winston-daily-rotate-file");
const path = require("path");
let fileRotationTransport;
let consoleTransport;
const initTransport = (config) => {
    if (config.file && config.folder) {
        fileRotationTransport = new DailyRotateFile({
            filename: path.join(config.folder, config.file),
            datePattern: 'YYYY-MM-DD',
            maxSize: '20m',
            maxFiles: '7',
            format: winston.format.combine(formats_1.LoggerFormats.logFormat),
        });
    }
    consoleTransport = new winston.transports.Console({
        format: winston.format.combine(formats_1.LoggerFormats.colorFormat, formats_1.LoggerFormats.logFormat),
    });
};
const getConsoleTransport = () => {
    return consoleTransport;
};
const getFileRotationTransport = () => {
    return fileRotationTransport;
};
exports.LoggerTransports = {
    getConsoleTransport,
    getFileRotationTransport,
    initTransport,
};
