// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/ApplySettingsRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplySettingsRequestToJSON = exports.ApplySettingsRequestFromJSONTyped = exports.ApplySettingsRequestFromJSON = exports.instanceOfApplySettingsRequest = void 0;
const runtime_1 = require("../runtime");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the ApplySettingsRequest interface.
 */
function instanceOfApplySettingsRequest(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfApplySettingsRequest = instanceOfApplySettingsRequest;
function ApplySettingsRequestFromJSON(json) {
    return ApplySettingsRequestFromJSONTyped(json, false);
}
exports.ApplySettingsRequestFromJSON = ApplySettingsRequestFromJSON;
function ApplySettingsRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'u2f_counter': !(0, runtime_1.exists)(json, 'u2f_counter') ? undefined : (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['u2f_counter']),
        'use_passphrase': !(0, runtime_1.exists)(json, 'use_passphrase') ? undefined : json['use_passphrase'],
        'language': !(0, runtime_1.exists)(json, 'language') ? undefined : json['language'],
        'label': !(0, runtime_1.exists)(json, 'label') ? undefined : json['label'],
        'auto_lock_delay_ms': !(0, runtime_1.exists)(json, 'auto_lock_delay_ms') ? undefined : (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['auto_lock_delay_ms']),
    };
}
exports.ApplySettingsRequestFromJSONTyped = ApplySettingsRequestFromJSONTyped;
function ApplySettingsRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'u2f_counter': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.u2f_counter),
        'use_passphrase': value.use_passphrase,
        'language': value.language,
        'label': value.label,
        'auto_lock_delay_ms': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.auto_lock_delay_ms),
    };
}
exports.ApplySettingsRequestToJSON = ApplySettingsRequestToJSON;
