// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/BinanceGetAddress200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinanceGetAddress200ResponseToJSON = exports.BinanceGetAddress200ResponseFromJSONTyped = exports.BinanceGetAddress200ResponseFromJSON = exports.instanceOfBinanceGetAddress200Response = void 0;
const TypesBnbAddress_1 = require("./TypesBnbAddress");
/**
 * Check if a given object implements the BinanceGetAddress200Response interface.
 */
function instanceOfBinanceGetAddress200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfBinanceGetAddress200Response = instanceOfBinanceGetAddress200Response;
function BinanceGetAddress200ResponseFromJSON(json) {
    return BinanceGetAddress200ResponseFromJSONTyped(json, false);
}
exports.BinanceGetAddress200ResponseFromJSON = BinanceGetAddress200ResponseFromJSON;
function BinanceGetAddress200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'address': (0, TypesBnbAddress_1.TypesBnbAddressFromJSON)(json['address']),
    };
}
exports.BinanceGetAddress200ResponseFromJSONTyped = BinanceGetAddress200ResponseFromJSONTyped;
function BinanceGetAddress200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'address': (0, TypesBnbAddress_1.TypesBnbAddressToJSON)(value.address),
    };
}
exports.BinanceGetAddress200ResponseToJSON = BinanceGetAddress200ResponseToJSON;
