// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/BinanceGetAddressRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinanceGetAddressRequestToJSON = exports.BinanceGetAddressRequestFromJSONTyped = exports.BinanceGetAddressRequestFromJSON = exports.instanceOfBinanceGetAddressRequest = void 0;
const runtime_1 = require("../runtime");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the BinanceGetAddressRequest interface.
 */
function instanceOfBinanceGetAddressRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "address_n" in value;
    return isInstance;
}
exports.instanceOfBinanceGetAddressRequest = instanceOfBinanceGetAddressRequest;
function BinanceGetAddressRequestFromJSON(json) {
    return BinanceGetAddressRequestFromJSONTyped(json, false);
}
exports.BinanceGetAddressRequestFromJSON = BinanceGetAddressRequestFromJSON;
function BinanceGetAddressRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'show_display': !(0, runtime_1.exists)(json, 'show_display') ? undefined : json['show_display'],
        'address_n': (json['address_n'].map(TypesNumericU32_1.TypesNumericU32FromJSON)),
    };
}
exports.BinanceGetAddressRequestFromJSONTyped = BinanceGetAddressRequestFromJSONTyped;
function BinanceGetAddressRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'show_display': value.show_display,
        'address_n': (value.address_n.map(TypesNumericU32_1.TypesNumericU32ToJSON)),
    };
}
exports.BinanceGetAddressRequestToJSON = BinanceGetAddressRequestToJSON;
