// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/BnbSignTransaction200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BnbSignTransaction200ResponseToJSON = exports.BnbSignTransaction200ResponseFromJSONTyped = exports.BnbSignTransaction200ResponseFromJSON = exports.instanceOfBnbSignTransaction200Response = void 0;
const TypesBnbSignDoc_1 = require("./TypesBnbSignDoc");
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the BnbSignTransaction200Response interface.
 */
function instanceOfBnbSignTransaction200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "txid" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfBnbSignTransaction200Response = instanceOfBnbSignTransaction200Response;
function BnbSignTransaction200ResponseFromJSON(json) {
    return BnbSignTransaction200ResponseFromJSONTyped(json, false);
}
exports.BnbSignTransaction200ResponseFromJSON = BnbSignTransaction200ResponseFromJSON;
function BnbSignTransaction200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesBnbSignDoc_1.TypesBnbSignDocFromJSON)(json['signed']),
        'txid': json['txid'],
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.BnbSignTransaction200ResponseFromJSONTyped = BnbSignTransaction200ResponseFromJSONTyped;
function BnbSignTransaction200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesBnbSignDoc_1.TypesBnbSignDocToJSON)(value.signed),
        'txid': value.txid,
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.BnbSignTransaction200ResponseToJSON = BnbSignTransaction200ResponseToJSON;
