// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/BnbSignTransactionRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BnbSignTransactionRequestToJSON = exports.BnbSignTransactionRequestFromJSONTyped = exports.BnbSignTransactionRequestFromJSON = exports.instanceOfBnbSignTransactionRequest = void 0;
const TypesBnbAddress_1 = require("./TypesBnbAddress");
const TypesBnbSignDoc_1 = require("./TypesBnbSignDoc");
/**
 * Check if a given object implements the BnbSignTransactionRequest interface.
 */
function instanceOfBnbSignTransactionRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfBnbSignTransactionRequest = instanceOfBnbSignTransactionRequest;
function BnbSignTransactionRequestFromJSON(json) {
    return BnbSignTransactionRequestFromJSONTyped(json, false);
}
exports.BnbSignTransactionRequestFromJSON = BnbSignTransactionRequestFromJSON;
function BnbSignTransactionRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesBnbSignDoc_1.TypesBnbSignDocFromJSON)(json['signDoc']),
        'signerAddress': (0, TypesBnbAddress_1.TypesBnbAddressFromJSON)(json['signerAddress']),
    };
}
exports.BnbSignTransactionRequestFromJSONTyped = BnbSignTransactionRequestFromJSONTyped;
function BnbSignTransactionRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesBnbSignDoc_1.TypesBnbSignDocToJSON)(value.signDoc),
        'signerAddress': (0, TypesBnbAddress_1.TypesBnbAddressToJSON)(value.signerAddress),
    };
}
exports.BnbSignTransactionRequestToJSON = BnbSignTransactionRequestToJSON;
