// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/BTCInputScriptType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BTCInputScriptTypeToJSON = exports.BTCInputScriptTypeFromJSONTyped = exports.BTCInputScriptTypeFromJSON = exports.BTCInputScriptType = void 0;
/**
 *
 * @export
 */
exports.BTCInputScriptType = {
    cashaddr: 'cashaddr',
    bech32: 'bech32',
    p2pkh: 'p2pkh',
    p2sh: 'p2sh',
    external: 'external',
    p2wpkh: 'p2wpkh',
    p2sh_p2wpkh: 'p2sh-p2wpkh'
};
function BTCInputScriptTypeFromJSON(json) {
    return BTCInputScriptTypeFromJSONTyped(json, false);
}
exports.BTCInputScriptTypeFromJSON = BTCInputScriptTypeFromJSON;
function BTCInputScriptTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.BTCInputScriptTypeFromJSONTyped = BTCInputScriptTypeFromJSONTyped;
function BTCInputScriptTypeToJSON(value) {
    return value;
}
exports.BTCInputScriptTypeToJSON = BTCInputScriptTypeToJSON;
