// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/ChangePinRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChangePinRequestToJSON = exports.ChangePinRequestFromJSONTyped = exports.ChangePinRequestFromJSON = exports.instanceOfChangePinRequest = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ChangePinRequest interface.
 */
function instanceOfChangePinRequest(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfChangePinRequest = instanceOfChangePinRequest;
function ChangePinRequestFromJSON(json) {
    return ChangePinRequestFromJSONTyped(json, false);
}
exports.ChangePinRequestFromJSON = ChangePinRequestFromJSON;
function ChangePinRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'remove': !(0, runtime_1.exists)(json, 'remove') ? undefined : json['remove'],
    };
}
exports.ChangePinRequestFromJSONTyped = ChangePinRequestFromJSONTyped;
function ChangePinRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'remove': value.remove,
    };
}
exports.ChangePinRequestToJSON = ChangePinRequestToJSON;
