// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CipherKeyValueRequestAllOf.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CipherKeyValueRequestAllOfToJSON = exports.CipherKeyValueRequestAllOfFromJSONTyped = exports.CipherKeyValueRequestAllOfFromJSON = exports.instanceOfCipherKeyValueRequestAllOf = void 0;
const runtime_1 = require("../runtime");
const CipherKeyValueRequestAllOfIv_1 = require("./CipherKeyValueRequestAllOfIv");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the CipherKeyValueRequestAllOf interface.
 */
function instanceOfCipherKeyValueRequestAllOf(value) {
    let isInstance = true;
    isInstance = isInstance && "iv" in value;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "key" in value;
    isInstance = isInstance && "address_n_list" in value;
    return isInstance;
}
exports.instanceOfCipherKeyValueRequestAllOf = instanceOfCipherKeyValueRequestAllOf;
function CipherKeyValueRequestAllOfFromJSON(json) {
    return CipherKeyValueRequestAllOfFromJSONTyped(json, false);
}
exports.CipherKeyValueRequestAllOfFromJSON = CipherKeyValueRequestAllOfFromJSON;
function CipherKeyValueRequestAllOfFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'iv': (0, CipherKeyValueRequestAllOfIv_1.CipherKeyValueRequestAllOfIvFromJSON)(json['iv']),
        'ask_on_decrypt': !(0, runtime_1.exists)(json, 'ask_on_decrypt') ? undefined : json['ask_on_decrypt'],
        'ask_on_encrypt': !(0, runtime_1.exists)(json, 'ask_on_encrypt') ? undefined : json['ask_on_encrypt'],
        'decrypt': !(0, runtime_1.exists)(json, 'decrypt') ? undefined : json['decrypt'],
        'value': json['value'],
        'key': json['key'],
        'address_n_list': (json['address_n_list'].map(TypesNumericU32_1.TypesNumericU32FromJSON)),
    };
}
exports.CipherKeyValueRequestAllOfFromJSONTyped = CipherKeyValueRequestAllOfFromJSONTyped;
function CipherKeyValueRequestAllOfToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'iv': (0, CipherKeyValueRequestAllOfIv_1.CipherKeyValueRequestAllOfIvToJSON)(value.iv),
        'ask_on_decrypt': value.ask_on_decrypt,
        'ask_on_encrypt': value.ask_on_encrypt,
        'decrypt': value.decrypt,
        'value': value.value,
        'key': value.key,
        'address_n_list': (value.address_n_list.map(TypesNumericU32_1.TypesNumericU32ToJSON)),
    };
}
exports.CipherKeyValueRequestAllOfToJSON = CipherKeyValueRequestAllOfToJSON;
