// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosGetAddress200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosGetAddress200ResponseToJSON = exports.CosmosGetAddress200ResponseFromJSONTyped = exports.CosmosGetAddress200ResponseFromJSON = exports.instanceOfCosmosGetAddress200Response = void 0;
/**
 * Check if a given object implements the CosmosGetAddress200Response interface.
 */
function instanceOfCosmosGetAddress200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfCosmosGetAddress200Response = instanceOfCosmosGetAddress200Response;
function CosmosGetAddress200ResponseFromJSON(json) {
    return CosmosGetAddress200ResponseFromJSONTyped(json, false);
}
exports.CosmosGetAddress200ResponseFromJSON = CosmosGetAddress200ResponseFromJSON;
function CosmosGetAddress200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'address': json['address'],
    };
}
exports.CosmosGetAddress200ResponseFromJSONTyped = CosmosGetAddress200ResponseFromJSONTyped;
function CosmosGetAddress200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'address': value.address,
    };
}
exports.CosmosGetAddress200ResponseToJSON = CosmosGetAddress200ResponseToJSON;
