// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoRedelegateRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoRedelegateRequestToJSON = exports.CosmosSignAminoRedelegateRequestFromJSONTyped = exports.CosmosSignAminoRedelegateRequestFromJSON = exports.instanceOfCosmosSignAminoRedelegateRequest = void 0;
const TypesCosmosAminoSignDocBeginReDelegate_1 = require("./TypesCosmosAminoSignDocBeginReDelegate");
/**
 * Check if a given object implements the CosmosSignAminoRedelegateRequest interface.
 */
function instanceOfCosmosSignAminoRedelegateRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoRedelegateRequest = instanceOfCosmosSignAminoRedelegateRequest;
function CosmosSignAminoRedelegateRequestFromJSON(json) {
    return CosmosSignAminoRedelegateRequestFromJSONTyped(json, false);
}
exports.CosmosSignAminoRedelegateRequestFromJSON = CosmosSignAminoRedelegateRequestFromJSON;
function CosmosSignAminoRedelegateRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocBeginReDelegate_1.TypesCosmosAminoSignDocBeginReDelegateFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.CosmosSignAminoRedelegateRequestFromJSONTyped = CosmosSignAminoRedelegateRequestFromJSONTyped;
function CosmosSignAminoRedelegateRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocBeginReDelegate_1.TypesCosmosAminoSignDocBeginReDelegateToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.CosmosSignAminoRedelegateRequestToJSON = CosmosSignAminoRedelegateRequestToJSON;
