// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoDelegate200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoDelegate200ResponseToJSON = exports.CosmosSignAminoDelegate200ResponseFromJSONTyped = exports.CosmosSignAminoDelegate200ResponseFromJSON = exports.instanceOfCosmosSignAminoDelegate200Response = void 0;
const TypesCosmosAminoSignDocDelegate_1 = require("./TypesCosmosAminoSignDocDelegate");
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the CosmosSignAminoDelegate200Response interface.
 */
function instanceOfCosmosSignAminoDelegate200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoDelegate200Response = instanceOfCosmosSignAminoDelegate200Response;
function CosmosSignAminoDelegate200ResponseFromJSON(json) {
    return CosmosSignAminoDelegate200ResponseFromJSONTyped(json, false);
}
exports.CosmosSignAminoDelegate200ResponseFromJSON = CosmosSignAminoDelegate200ResponseFromJSON;
function CosmosSignAminoDelegate200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocDelegate_1.TypesCosmosAminoSignDocDelegateFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.CosmosSignAminoDelegate200ResponseFromJSONTyped = CosmosSignAminoDelegate200ResponseFromJSONTyped;
function CosmosSignAminoDelegate200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocDelegate_1.TypesCosmosAminoSignDocDelegateToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.CosmosSignAminoDelegate200ResponseToJSON = CosmosSignAminoDelegate200ResponseToJSON;
