// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoUndelegateRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoUndelegateRequestToJSON = exports.CosmosSignAminoUndelegateRequestFromJSONTyped = exports.CosmosSignAminoUndelegateRequestFromJSON = exports.instanceOfCosmosSignAminoUndelegateRequest = void 0;
const TypesCosmosAminoSignDocUnDelegate_1 = require("./TypesCosmosAminoSignDocUnDelegate");
/**
 * Check if a given object implements the CosmosSignAminoUndelegateRequest interface.
 */
function instanceOfCosmosSignAminoUndelegateRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoUndelegateRequest = instanceOfCosmosSignAminoUndelegateRequest;
function CosmosSignAminoUndelegateRequestFromJSON(json) {
    return CosmosSignAminoUndelegateRequestFromJSONTyped(json, false);
}
exports.CosmosSignAminoUndelegateRequestFromJSON = CosmosSignAminoUndelegateRequestFromJSON;
function CosmosSignAminoUndelegateRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocUnDelegate_1.TypesCosmosAminoSignDocUnDelegateFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.CosmosSignAminoUndelegateRequestFromJSONTyped = CosmosSignAminoUndelegateRequestFromJSONTyped;
function CosmosSignAminoUndelegateRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocUnDelegate_1.TypesCosmosAminoSignDocUnDelegateToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.CosmosSignAminoUndelegateRequestToJSON = CosmosSignAminoUndelegateRequestToJSON;
