// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoWithdrawDelegatorRewardsAll200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseToJSON = exports.CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSONTyped = exports.CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSON = exports.instanceOfCosmosSignAminoWithdrawDelegatorRewardsAll200Response = void 0;
const TypesCosmosAminoSignDocWithdrawDelegationReward_1 = require("./TypesCosmosAminoSignDocWithdrawDelegationReward");
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the CosmosSignAminoWithdrawDelegatorRewardsAll200Response interface.
 */
function instanceOfCosmosSignAminoWithdrawDelegatorRewardsAll200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoWithdrawDelegatorRewardsAll200Response = instanceOfCosmosSignAminoWithdrawDelegatorRewardsAll200Response;
function CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSON(json) {
    return CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSONTyped(json, false);
}
exports.CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSON = CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSON;
function CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocWithdrawDelegationReward_1.TypesCosmosAminoSignDocWithdrawDelegationRewardFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSONTyped = CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseFromJSONTyped;
function CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocWithdrawDelegationReward_1.TypesCosmosAminoSignDocWithdrawDelegationRewardToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseToJSON = CosmosSignAminoWithdrawDelegatorRewardsAll200ResponseToJSON;
