// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoDelegateRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoDelegateRequestToJSON = exports.CosmosSignAminoDelegateRequestFromJSONTyped = exports.CosmosSignAminoDelegateRequestFromJSON = exports.instanceOfCosmosSignAminoDelegateRequest = void 0;
const TypesCosmosAminoSignDocDelegate_1 = require("./TypesCosmosAminoSignDocDelegate");
/**
 * Check if a given object implements the CosmosSignAminoDelegateRequest interface.
 */
function instanceOfCosmosSignAminoDelegateRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoDelegateRequest = instanceOfCosmosSignAminoDelegateRequest;
function CosmosSignAminoDelegateRequestFromJSON(json) {
    return CosmosSignAminoDelegateRequestFromJSONTyped(json, false);
}
exports.CosmosSignAminoDelegateRequestFromJSON = CosmosSignAminoDelegateRequestFromJSON;
function CosmosSignAminoDelegateRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocDelegate_1.TypesCosmosAminoSignDocDelegateFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.CosmosSignAminoDelegateRequestFromJSONTyped = CosmosSignAminoDelegateRequestFromJSONTyped;
function CosmosSignAminoDelegateRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocDelegate_1.TypesCosmosAminoSignDocDelegateToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.CosmosSignAminoDelegateRequestToJSON = CosmosSignAminoDelegateRequestToJSON;
