// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoRedelegate200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoRedelegate200ResponseToJSON = exports.CosmosSignAminoRedelegate200ResponseFromJSONTyped = exports.CosmosSignAminoRedelegate200ResponseFromJSON = exports.instanceOfCosmosSignAminoRedelegate200Response = void 0;
const TypesCosmosAminoSignDocBeginReDelegate_1 = require("./TypesCosmosAminoSignDocBeginReDelegate");
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the CosmosSignAminoRedelegate200Response interface.
 */
function instanceOfCosmosSignAminoRedelegate200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoRedelegate200Response = instanceOfCosmosSignAminoRedelegate200Response;
function CosmosSignAminoRedelegate200ResponseFromJSON(json) {
    return CosmosSignAminoRedelegate200ResponseFromJSONTyped(json, false);
}
exports.CosmosSignAminoRedelegate200ResponseFromJSON = CosmosSignAminoRedelegate200ResponseFromJSON;
function CosmosSignAminoRedelegate200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocBeginReDelegate_1.TypesCosmosAminoSignDocBeginReDelegateFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.CosmosSignAminoRedelegate200ResponseFromJSONTyped = CosmosSignAminoRedelegate200ResponseFromJSONTyped;
function CosmosSignAminoRedelegate200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocBeginReDelegate_1.TypesCosmosAminoSignDocBeginReDelegateToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.CosmosSignAminoRedelegate200ResponseToJSON = CosmosSignAminoRedelegate200ResponseToJSON;
