// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoIbcTransferRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoIbcTransferRequestToJSON = exports.CosmosSignAminoIbcTransferRequestFromJSONTyped = exports.CosmosSignAminoIbcTransferRequestFromJSON = exports.instanceOfCosmosSignAminoIbcTransferRequest = void 0;
const TypesCosmosAminoSignDocIbcGoTransfer_1 = require("./TypesCosmosAminoSignDocIbcGoTransfer");
/**
 * Check if a given object implements the CosmosSignAminoIbcTransferRequest interface.
 */
function instanceOfCosmosSignAminoIbcTransferRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoIbcTransferRequest = instanceOfCosmosSignAminoIbcTransferRequest;
function CosmosSignAminoIbcTransferRequestFromJSON(json) {
    return CosmosSignAminoIbcTransferRequestFromJSONTyped(json, false);
}
exports.CosmosSignAminoIbcTransferRequestFromJSON = CosmosSignAminoIbcTransferRequestFromJSON;
function CosmosSignAminoIbcTransferRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocIbcGoTransfer_1.TypesCosmosAminoSignDocIbcGoTransferFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.CosmosSignAminoIbcTransferRequestFromJSONTyped = CosmosSignAminoIbcTransferRequestFromJSONTyped;
function CosmosSignAminoIbcTransferRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocIbcGoTransfer_1.TypesCosmosAminoSignDocIbcGoTransferToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.CosmosSignAminoIbcTransferRequestToJSON = CosmosSignAminoIbcTransferRequestToJSON;
