// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoIbcTransfer200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoIbcTransfer200ResponseToJSON = exports.CosmosSignAminoIbcTransfer200ResponseFromJSONTyped = exports.CosmosSignAminoIbcTransfer200ResponseFromJSON = exports.instanceOfCosmosSignAminoIbcTransfer200Response = void 0;
const TypesCosmosAminoSignDocIbcGoTransfer_1 = require("./TypesCosmosAminoSignDocIbcGoTransfer");
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the CosmosSignAminoIbcTransfer200Response interface.
 */
function instanceOfCosmosSignAminoIbcTransfer200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoIbcTransfer200Response = instanceOfCosmosSignAminoIbcTransfer200Response;
function CosmosSignAminoIbcTransfer200ResponseFromJSON(json) {
    return CosmosSignAminoIbcTransfer200ResponseFromJSONTyped(json, false);
}
exports.CosmosSignAminoIbcTransfer200ResponseFromJSON = CosmosSignAminoIbcTransfer200ResponseFromJSON;
function CosmosSignAminoIbcTransfer200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocIbcGoTransfer_1.TypesCosmosAminoSignDocIbcGoTransferFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.CosmosSignAminoIbcTransfer200ResponseFromJSONTyped = CosmosSignAminoIbcTransfer200ResponseFromJSONTyped;
function CosmosSignAminoIbcTransfer200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocIbcGoTransfer_1.TypesCosmosAminoSignDocIbcGoTransferToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.CosmosSignAminoIbcTransfer200ResponseToJSON = CosmosSignAminoIbcTransfer200ResponseToJSON;
