// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoWithdrawDelegatorRewardsAllRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoWithdrawDelegatorRewardsAllRequestToJSON = exports.CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSONTyped = exports.CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSON = exports.instanceOfCosmosSignAminoWithdrawDelegatorRewardsAllRequest = void 0;
const TypesCosmosAminoSignDocWithdrawDelegationReward_1 = require("./TypesCosmosAminoSignDocWithdrawDelegationReward");
/**
 * Check if a given object implements the CosmosSignAminoWithdrawDelegatorRewardsAllRequest interface.
 */
function instanceOfCosmosSignAminoWithdrawDelegatorRewardsAllRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoWithdrawDelegatorRewardsAllRequest = instanceOfCosmosSignAminoWithdrawDelegatorRewardsAllRequest;
function CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSON(json) {
    return CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSONTyped(json, false);
}
exports.CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSON = CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSON;
function CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocWithdrawDelegationReward_1.TypesCosmosAminoSignDocWithdrawDelegationRewardFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSONTyped = CosmosSignAminoWithdrawDelegatorRewardsAllRequestFromJSONTyped;
function CosmosSignAminoWithdrawDelegatorRewardsAllRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDocWithdrawDelegationReward_1.TypesCosmosAminoSignDocWithdrawDelegationRewardToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.CosmosSignAminoWithdrawDelegatorRewardsAllRequestToJSON = CosmosSignAminoWithdrawDelegatorRewardsAllRequestToJSON;
