// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoRequestToJSON = exports.CosmosSignAminoRequestFromJSONTyped = exports.CosmosSignAminoRequestFromJSON = exports.instanceOfCosmosSignAminoRequest = void 0;
const TypesCosmosAminoSignDoc_1 = require("./TypesCosmosAminoSignDoc");
/**
 * Check if a given object implements the CosmosSignAminoRequest interface.
 */
function instanceOfCosmosSignAminoRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoRequest = instanceOfCosmosSignAminoRequest;
function CosmosSignAminoRequestFromJSON(json) {
    return CosmosSignAminoRequestFromJSONTyped(json, false);
}
exports.CosmosSignAminoRequestFromJSON = CosmosSignAminoRequestFromJSON;
function CosmosSignAminoRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDoc_1.TypesCosmosAminoSignDocFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.CosmosSignAminoRequestFromJSONTyped = CosmosSignAminoRequestFromJSONTyped;
function CosmosSignAminoRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesCosmosAminoSignDoc_1.TypesCosmosAminoSignDocToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.CosmosSignAminoRequestToJSON = CosmosSignAminoRequestToJSON;
