// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAminoUndelegate200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAminoUndelegate200ResponseToJSON = exports.CosmosSignAminoUndelegate200ResponseFromJSONTyped = exports.CosmosSignAminoUndelegate200ResponseFromJSON = exports.instanceOfCosmosSignAminoUndelegate200Response = void 0;
const TypesCosmosAminoSignDocUnDelegate_1 = require("./TypesCosmosAminoSignDocUnDelegate");
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the CosmosSignAminoUndelegate200Response interface.
 */
function instanceOfCosmosSignAminoUndelegate200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAminoUndelegate200Response = instanceOfCosmosSignAminoUndelegate200Response;
function CosmosSignAminoUndelegate200ResponseFromJSON(json) {
    return CosmosSignAminoUndelegate200ResponseFromJSONTyped(json, false);
}
exports.CosmosSignAminoUndelegate200ResponseFromJSON = CosmosSignAminoUndelegate200ResponseFromJSON;
function CosmosSignAminoUndelegate200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocUnDelegate_1.TypesCosmosAminoSignDocUnDelegateFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.CosmosSignAminoUndelegate200ResponseFromJSONTyped = CosmosSignAminoUndelegate200ResponseFromJSONTyped;
function CosmosSignAminoUndelegate200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDocUnDelegate_1.TypesCosmosAminoSignDocUnDelegateToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.CosmosSignAminoUndelegate200ResponseToJSON = CosmosSignAminoUndelegate200ResponseToJSON;
