// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/CosmosSignAmino200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosSignAmino200ResponseToJSON = exports.CosmosSignAmino200ResponseFromJSONTyped = exports.CosmosSignAmino200ResponseFromJSON = exports.instanceOfCosmosSignAmino200Response = void 0;
const TypesCosmosAminoSignDoc_1 = require("./TypesCosmosAminoSignDoc");
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the CosmosSignAmino200Response interface.
 */
function instanceOfCosmosSignAmino200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfCosmosSignAmino200Response = instanceOfCosmosSignAmino200Response;
function CosmosSignAmino200ResponseFromJSON(json) {
    return CosmosSignAmino200ResponseFromJSONTyped(json, false);
}
exports.CosmosSignAmino200ResponseFromJSON = CosmosSignAmino200ResponseFromJSON;
function CosmosSignAmino200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDoc_1.TypesCosmosAminoSignDocFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.CosmosSignAmino200ResponseFromJSONTyped = CosmosSignAmino200ResponseFromJSONTyped;
function CosmosSignAmino200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesCosmosAminoSignDoc_1.TypesCosmosAminoSignDocToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.CosmosSignAmino200ResponseToJSON = CosmosSignAmino200ResponseToJSON;
