// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/DebugLinkFlashDumpRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugLinkFlashDumpRequestToJSON = exports.DebugLinkFlashDumpRequestFromJSONTyped = exports.DebugLinkFlashDumpRequestFromJSON = exports.instanceOfDebugLinkFlashDumpRequest = void 0;
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the DebugLinkFlashDumpRequest interface.
 */
function instanceOfDebugLinkFlashDumpRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "length" in value;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfDebugLinkFlashDumpRequest = instanceOfDebugLinkFlashDumpRequest;
function DebugLinkFlashDumpRequestFromJSON(json) {
    return DebugLinkFlashDumpRequestFromJSONTyped(json, false);
}
exports.DebugLinkFlashDumpRequestFromJSON = DebugLinkFlashDumpRequestFromJSON;
function DebugLinkFlashDumpRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'length': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['length']),
        'address': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['address']),
    };
}
exports.DebugLinkFlashDumpRequestFromJSONTyped = DebugLinkFlashDumpRequestFromJSONTyped;
function DebugLinkFlashDumpRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'length': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.length),
        'address': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.address),
    };
}
exports.DebugLinkFlashDumpRequestToJSON = DebugLinkFlashDumpRequestToJSON;
