// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/DebugLinkGetState200ResponseFirmwareHash.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugLinkGetState200ResponseFirmwareHashToJSON = exports.DebugLinkGetState200ResponseFirmwareHashFromJSONTyped = exports.DebugLinkGetState200ResponseFirmwareHashFromJSON = exports.instanceOfDebugLinkGetState200ResponseFirmwareHash = void 0;
/**
 * Check if a given object implements the DebugLinkGetState200ResponseFirmwareHash interface.
 */
function instanceOfDebugLinkGetState200ResponseFirmwareHash(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfDebugLinkGetState200ResponseFirmwareHash = instanceOfDebugLinkGetState200ResponseFirmwareHash;
function DebugLinkGetState200ResponseFirmwareHashFromJSON(json) {
    return DebugLinkGetState200ResponseFirmwareHashFromJSONTyped(json, false);
}
exports.DebugLinkGetState200ResponseFirmwareHashFromJSON = DebugLinkGetState200ResponseFirmwareHashFromJSON;
function DebugLinkGetState200ResponseFirmwareHashFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.DebugLinkGetState200ResponseFirmwareHashFromJSONTyped = DebugLinkGetState200ResponseFirmwareHashFromJSONTyped;
function DebugLinkGetState200ResponseFirmwareHashToJSON(value) {
    return value;
}
exports.DebugLinkGetState200ResponseFirmwareHashToJSON = DebugLinkGetState200ResponseFirmwareHashToJSON;
