// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/DebugLinkGetState200ResponseNode.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugLinkGetState200ResponseNodeToJSON = exports.DebugLinkGetState200ResponseNodeFromJSONTyped = exports.DebugLinkGetState200ResponseNodeFromJSON = exports.instanceOfDebugLinkGetState200ResponseNode = void 0;
const runtime_1 = require("../runtime");
const TypesHexBytesLower_1 = require("./TypesHexBytesLower");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the DebugLinkGetState200ResponseNode interface.
 */
function instanceOfDebugLinkGetState200ResponseNode(value) {
    let isInstance = true;
    isInstance = isInstance && "chain_code" in value;
    isInstance = isInstance && "child_num" in value;
    isInstance = isInstance && "fingerprint" in value;
    isInstance = isInstance && "depth" in value;
    return isInstance;
}
exports.instanceOfDebugLinkGetState200ResponseNode = instanceOfDebugLinkGetState200ResponseNode;
function DebugLinkGetState200ResponseNodeFromJSON(json) {
    return DebugLinkGetState200ResponseNodeFromJSONTyped(json, false);
}
exports.DebugLinkGetState200ResponseNodeFromJSON = DebugLinkGetState200ResponseNodeFromJSON;
function DebugLinkGetState200ResponseNodeFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'public_key': !(0, runtime_1.exists)(json, 'public_key') ? undefined : (0, TypesHexBytesLower_1.TypesHexBytesLowerFromJSON)(json['public_key']),
        'private_key': !(0, runtime_1.exists)(json, 'private_key') ? undefined : (0, TypesHexBytesLower_1.TypesHexBytesLowerFromJSON)(json['private_key']),
        'chain_code': (0, TypesHexBytesLower_1.TypesHexBytesLowerFromJSON)(json['chain_code']),
        'child_num': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['child_num']),
        'fingerprint': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['fingerprint']),
        'depth': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['depth']),
    };
}
exports.DebugLinkGetState200ResponseNodeFromJSONTyped = DebugLinkGetState200ResponseNodeFromJSONTyped;
function DebugLinkGetState200ResponseNodeToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'public_key': (0, TypesHexBytesLower_1.TypesHexBytesLowerToJSON)(value.public_key),
        'private_key': (0, TypesHexBytesLower_1.TypesHexBytesLowerToJSON)(value.private_key),
        'chain_code': (0, TypesHexBytesLower_1.TypesHexBytesLowerToJSON)(value.chain_code),
        'child_num': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.child_num),
        'fingerprint': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.fingerprint),
        'depth': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.depth),
    };
}
exports.DebugLinkGetState200ResponseNodeToJSON = DebugLinkGetState200ResponseNodeToJSON;
