// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/DebugLinkGetState200ResponseStorageHash.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugLinkGetState200ResponseStorageHashToJSON = exports.DebugLinkGetState200ResponseStorageHashFromJSONTyped = exports.DebugLinkGetState200ResponseStorageHashFromJSON = exports.instanceOfDebugLinkGetState200ResponseStorageHash = void 0;
/**
 * Check if a given object implements the DebugLinkGetState200ResponseStorageHash interface.
 */
function instanceOfDebugLinkGetState200ResponseStorageHash(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfDebugLinkGetState200ResponseStorageHash = instanceOfDebugLinkGetState200ResponseStorageHash;
function DebugLinkGetState200ResponseStorageHashFromJSON(json) {
    return DebugLinkGetState200ResponseStorageHashFromJSONTyped(json, false);
}
exports.DebugLinkGetState200ResponseStorageHashFromJSON = DebugLinkGetState200ResponseStorageHashFromJSON;
function DebugLinkGetState200ResponseStorageHashFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.DebugLinkGetState200ResponseStorageHashFromJSONTyped = DebugLinkGetState200ResponseStorageHashFromJSONTyped;
function DebugLinkGetState200ResponseStorageHashToJSON(value) {
    return value;
}
exports.DebugLinkGetState200ResponseStorageHashToJSON = DebugLinkGetState200ResponseStorageHashToJSON;
