// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/DebugLinkGetState200ResponseResetEntropy.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugLinkGetState200ResponseResetEntropyToJSON = exports.DebugLinkGetState200ResponseResetEntropyFromJSONTyped = exports.DebugLinkGetState200ResponseResetEntropyFromJSON = exports.instanceOfDebugLinkGetState200ResponseResetEntropy = void 0;
/**
 * Check if a given object implements the DebugLinkGetState200ResponseResetEntropy interface.
 */
function instanceOfDebugLinkGetState200ResponseResetEntropy(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfDebugLinkGetState200ResponseResetEntropy = instanceOfDebugLinkGetState200ResponseResetEntropy;
function DebugLinkGetState200ResponseResetEntropyFromJSON(json) {
    return DebugLinkGetState200ResponseResetEntropyFromJSONTyped(json, false);
}
exports.DebugLinkGetState200ResponseResetEntropyFromJSON = DebugLinkGetState200ResponseResetEntropyFromJSON;
function DebugLinkGetState200ResponseResetEntropyFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.DebugLinkGetState200ResponseResetEntropyFromJSONTyped = DebugLinkGetState200ResponseResetEntropyFromJSONTyped;
function DebugLinkGetState200ResponseResetEntropyToJSON(value) {
    return value;
}
exports.DebugLinkGetState200ResponseResetEntropyToJSON = DebugLinkGetState200ResponseResetEntropyToJSON;
