// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/DebugLinkGetState200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugLinkGetState200ResponseToJSON = exports.DebugLinkGetState200ResponseFromJSONTyped = exports.DebugLinkGetState200ResponseFromJSON = exports.instanceOfDebugLinkGetState200Response = void 0;
const runtime_1 = require("../runtime");
const DebugLinkGetState200ResponseFirmwareHash_1 = require("./DebugLinkGetState200ResponseFirmwareHash");
const DebugLinkGetState200ResponseNode_1 = require("./DebugLinkGetState200ResponseNode");
const DebugLinkGetState200ResponseResetEntropy_1 = require("./DebugLinkGetState200ResponseResetEntropy");
const DebugLinkGetState200ResponseStorageHash_1 = require("./DebugLinkGetState200ResponseStorageHash");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the DebugLinkGetState200Response interface.
 */
function instanceOfDebugLinkGetState200Response(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfDebugLinkGetState200Response = instanceOfDebugLinkGetState200Response;
function DebugLinkGetState200ResponseFromJSON(json) {
    return DebugLinkGetState200ResponseFromJSONTyped(json, false);
}
exports.DebugLinkGetState200ResponseFromJSON = DebugLinkGetState200ResponseFromJSON;
function DebugLinkGetState200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'storage_hash': !(0, runtime_1.exists)(json, 'storage_hash') ? undefined : (0, DebugLinkGetState200ResponseStorageHash_1.DebugLinkGetState200ResponseStorageHashFromJSON)(json['storage_hash']),
        'firmware_hash': !(0, runtime_1.exists)(json, 'firmware_hash') ? undefined : (0, DebugLinkGetState200ResponseFirmwareHash_1.DebugLinkGetState200ResponseFirmwareHashFromJSON)(json['firmware_hash']),
        'recovery_auto_completed_word': !(0, runtime_1.exists)(json, 'recovery_auto_completed_word') ? undefined : json['recovery_auto_completed_word'],
        'recovery_cipher': !(0, runtime_1.exists)(json, 'recovery_cipher') ? undefined : json['recovery_cipher'],
        'recovery_word_pos': !(0, runtime_1.exists)(json, 'recovery_word_pos') ? undefined : (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['recovery_word_pos']),
        'recovery_fake_word': !(0, runtime_1.exists)(json, 'recovery_fake_word') ? undefined : json['recovery_fake_word'],
        'reset_entropy': !(0, runtime_1.exists)(json, 'reset_entropy') ? undefined : (0, DebugLinkGetState200ResponseResetEntropy_1.DebugLinkGetState200ResponseResetEntropyFromJSON)(json['reset_entropy']),
        'reset_word': !(0, runtime_1.exists)(json, 'reset_word') ? undefined : json['reset_word'],
        'passphrase_protection': !(0, runtime_1.exists)(json, 'passphrase_protection') ? undefined : json['passphrase_protection'],
        'node': !(0, runtime_1.exists)(json, 'node') ? undefined : (0, DebugLinkGetState200ResponseNode_1.DebugLinkGetState200ResponseNodeFromJSON)(json['node']),
        'mnemonic': !(0, runtime_1.exists)(json, 'mnemonic') ? undefined : json['mnemonic'],
        'matrix': !(0, runtime_1.exists)(json, 'matrix') ? undefined : json['matrix'],
        'pin': !(0, runtime_1.exists)(json, 'pin') ? undefined : json['pin'],
        'layout': !(0, runtime_1.exists)(json, 'layout') ? undefined : json['layout'],
    };
}
exports.DebugLinkGetState200ResponseFromJSONTyped = DebugLinkGetState200ResponseFromJSONTyped;
function DebugLinkGetState200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'storage_hash': (0, DebugLinkGetState200ResponseStorageHash_1.DebugLinkGetState200ResponseStorageHashToJSON)(value.storage_hash),
        'firmware_hash': (0, DebugLinkGetState200ResponseFirmwareHash_1.DebugLinkGetState200ResponseFirmwareHashToJSON)(value.firmware_hash),
        'recovery_auto_completed_word': value.recovery_auto_completed_word,
        'recovery_cipher': value.recovery_cipher,
        'recovery_word_pos': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.recovery_word_pos),
        'recovery_fake_word': value.recovery_fake_word,
        'reset_entropy': (0, DebugLinkGetState200ResponseResetEntropy_1.DebugLinkGetState200ResponseResetEntropyToJSON)(value.reset_entropy),
        'reset_word': value.reset_word,
        'passphrase_protection': value.passphrase_protection,
        'node': (0, DebugLinkGetState200ResponseNode_1.DebugLinkGetState200ResponseNodeToJSON)(value.node),
        'mnemonic': value.mnemonic,
        'matrix': value.matrix,
        'pin': value.pin,
        'layout': value.layout,
    };
}
exports.DebugLinkGetState200ResponseToJSON = DebugLinkGetState200ResponseToJSON;
