// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthereumGetAddress200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumGetAddress200ResponseToJSON = exports.EthereumGetAddress200ResponseFromJSONTyped = exports.EthereumGetAddress200ResponseFromJSON = exports.instanceOfEthereumGetAddress200Response = void 0;
/**
 * Check if a given object implements the EthereumGetAddress200Response interface.
 */
function instanceOfEthereumGetAddress200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfEthereumGetAddress200Response = instanceOfEthereumGetAddress200Response;
function EthereumGetAddress200ResponseFromJSON(json) {
    return EthereumGetAddress200ResponseFromJSONTyped(json, false);
}
exports.EthereumGetAddress200ResponseFromJSON = EthereumGetAddress200ResponseFromJSON;
function EthereumGetAddress200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'address': json['address'],
    };
}
exports.EthereumGetAddress200ResponseFromJSONTyped = EthereumGetAddress200ResponseFromJSONTyped;
function EthereumGetAddress200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'address': value.address,
    };
}
exports.EthereumGetAddress200ResponseToJSON = EthereumGetAddress200ResponseToJSON;
