// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignTypedDataRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignTypedDataRequestToJSON = exports.EthSignTypedDataRequestFromJSONTyped = exports.EthSignTypedDataRequestFromJSON = exports.instanceOfEthSignTypedDataRequest = void 0;
/**
 * Check if a given object implements the EthSignTypedDataRequest interface.
 */
function instanceOfEthSignTypedDataRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "typedData" in value;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfEthSignTypedDataRequest = instanceOfEthSignTypedDataRequest;
function EthSignTypedDataRequestFromJSON(json) {
    return EthSignTypedDataRequestFromJSONTyped(json, false);
}
exports.EthSignTypedDataRequestFromJSON = EthSignTypedDataRequestFromJSON;
function EthSignTypedDataRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'typedData': json['typedData'],
        'address': json['address'],
    };
}
exports.EthSignTypedDataRequestFromJSONTyped = EthSignTypedDataRequestFromJSONTyped;
function EthSignTypedDataRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'typedData': value.typedData,
        'address': value.address,
    };
}
exports.EthSignTypedDataRequestToJSON = EthSignTypedDataRequestToJSON;
