// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignRequestToJSON = exports.EthSignRequestFromJSONTyped = exports.EthSignRequestFromJSON = exports.instanceOfEthSignRequest = void 0;
/**
 * Check if a given object implements the EthSignRequest interface.
 */
function instanceOfEthSignRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "address" in value;
    isInstance = isInstance && "message" in value;
    return isInstance;
}
exports.instanceOfEthSignRequest = instanceOfEthSignRequest;
function EthSignRequestFromJSON(json) {
    return EthSignRequestFromJSONTyped(json, false);
}
exports.EthSignRequestFromJSON = EthSignRequestFromJSON;
function EthSignRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'address': json['address'],
        'message': json['message'],
    };
}
exports.EthSignRequestFromJSONTyped = EthSignRequestFromJSONTyped;
function EthSignRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'address': value.address,
        'message': value.message,
    };
}
exports.EthSignRequestToJSON = EthSignRequestToJSON;
