// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignTransactionRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignTransactionRequestToJSON = exports.EthSignTransactionRequestFromJSONTyped = exports.EthSignTransactionRequestFromJSON = exports.instanceOfEthSignTransactionRequest = void 0;
const runtime_1 = require("../runtime");
const EthSignTransactionRequestChainId_1 = require("./EthSignTransactionRequestChainId");
const EthSignTransactionRequestData_1 = require("./EthSignTransactionRequestData");
const EthSignTransactionRequestFrom_1 = require("./EthSignTransactionRequestFrom");
const EthSignTransactionRequestGasPrice_1 = require("./EthSignTransactionRequestGasPrice");
/**
 * Check if a given object implements the EthSignTransactionRequest interface.
 */
function instanceOfEthSignTransactionRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "chainId" in value;
    isInstance = isInstance && "nonce" in value;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "data" in value;
    isInstance = isInstance && "to" in value;
    return isInstance;
}
exports.instanceOfEthSignTransactionRequest = instanceOfEthSignTransactionRequest;
function EthSignTransactionRequestFromJSON(json) {
    return EthSignTransactionRequestFromJSONTyped(json, false);
}
exports.EthSignTransactionRequestFromJSON = EthSignTransactionRequestFromJSON;
function EthSignTransactionRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'gasPrice': !(0, runtime_1.exists)(json, 'gasPrice') ? undefined : (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceFromJSON)(json['gasPrice']),
        'maxPriorityFeePerGas': !(0, runtime_1.exists)(json, 'maxPriorityFeePerGas') ? undefined : (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceFromJSON)(json['maxPriorityFeePerGas']),
        'maxFeePerGas': !(0, runtime_1.exists)(json, 'maxFeePerGas') ? undefined : (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceFromJSON)(json['maxFeePerGas']),
        'chainId': (0, EthSignTransactionRequestChainId_1.EthSignTransactionRequestChainIdFromJSON)(json['chainId']),
        'nonce': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceFromJSON)(json['nonce']),
        'value': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceFromJSON)(json['value']),
        'gas': !(0, runtime_1.exists)(json, 'gas') ? undefined : (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceFromJSON)(json['gas']),
        'data': (0, EthSignTransactionRequestData_1.EthSignTransactionRequestDataFromJSON)(json['data']),
        'to': json['to'],
        'addressNList': !(0, runtime_1.exists)(json, 'addressNList') ? undefined : json['addressNList'],
        'from': !(0, runtime_1.exists)(json, 'from') ? undefined : (0, EthSignTransactionRequestFrom_1.EthSignTransactionRequestFromFromJSON)(json['from']),
    };
}
exports.EthSignTransactionRequestFromJSONTyped = EthSignTransactionRequestFromJSONTyped;
function EthSignTransactionRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'gasPrice': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceToJSON)(value.gasPrice),
        'maxPriorityFeePerGas': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceToJSON)(value.maxPriorityFeePerGas),
        'maxFeePerGas': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceToJSON)(value.maxFeePerGas),
        'chainId': (0, EthSignTransactionRequestChainId_1.EthSignTransactionRequestChainIdToJSON)(value.chainId),
        'nonce': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceToJSON)(value.nonce),
        'value': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceToJSON)(value.value),
        'gas': (0, EthSignTransactionRequestGasPrice_1.EthSignTransactionRequestGasPriceToJSON)(value.gas),
        'data': (0, EthSignTransactionRequestData_1.EthSignTransactionRequestDataToJSON)(value.data),
        'to': value.to,
        'addressNList': value.addressNList,
        'from': (0, EthSignTransactionRequestFrom_1.EthSignTransactionRequestFromToJSON)(value.from),
    };
}
exports.EthSignTransactionRequestToJSON = EthSignTransactionRequestToJSON;
