// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignTransaction200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignTransaction200ResponseToJSON = exports.EthSignTransaction200ResponseFromJSONTyped = exports.EthSignTransaction200ResponseFromJSON = exports.instanceOfEthSignTransaction200Response = void 0;
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the EthSignTransaction200Response interface.
 */
function instanceOfEthSignTransaction200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "s" in value;
    isInstance = isInstance && "r" in value;
    isInstance = isInstance && "v" in value;
    return isInstance;
}
exports.instanceOfEthSignTransaction200Response = instanceOfEthSignTransaction200Response;
function EthSignTransaction200ResponseFromJSON(json) {
    return EthSignTransaction200ResponseFromJSONTyped(json, false);
}
exports.EthSignTransaction200ResponseFromJSON = EthSignTransaction200ResponseFromJSON;
function EthSignTransaction200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'serialized': json['serialized'],
        's': json['s'],
        'r': json['r'],
        'v': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['v']),
    };
}
exports.EthSignTransaction200ResponseFromJSONTyped = EthSignTransaction200ResponseFromJSONTyped;
function EthSignTransaction200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'serialized': value.serialized,
        's': value.s,
        'r': value.r,
        'v': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.v),
    };
}
exports.EthSignTransaction200ResponseToJSON = EthSignTransaction200ResponseToJSON;
