// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignTransactionRequestFrom.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignTransactionRequestFromToJSON = exports.EthSignTransactionRequestFromFromJSONTyped = exports.EthSignTransactionRequestFromFromJSON = exports.instanceOfEthSignTransactionRequestFrom = void 0;
/**
 * Check if a given object implements the EthSignTransactionRequestFrom interface.
 */
function instanceOfEthSignTransactionRequestFrom(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfEthSignTransactionRequestFrom = instanceOfEthSignTransactionRequestFrom;
function EthSignTransactionRequestFromFromJSON(json) {
    return EthSignTransactionRequestFromFromJSONTyped(json, false);
}
exports.EthSignTransactionRequestFromFromJSON = EthSignTransactionRequestFromFromJSON;
function EthSignTransactionRequestFromFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.EthSignTransactionRequestFromFromJSONTyped = EthSignTransactionRequestFromFromJSONTyped;
function EthSignTransactionRequestFromToJSON(value) {
    return value;
}
exports.EthSignTransactionRequestFromToJSON = EthSignTransactionRequestFromToJSON;
