// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignTransactionRequestData.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignTransactionRequestDataToJSON = exports.EthSignTransactionRequestDataFromJSONTyped = exports.EthSignTransactionRequestDataFromJSON = exports.instanceOfEthSignTransactionRequestData = void 0;
/**
 * Check if a given object implements the EthSignTransactionRequestData interface.
 */
function instanceOfEthSignTransactionRequestData(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfEthSignTransactionRequestData = instanceOfEthSignTransactionRequestData;
function EthSignTransactionRequestDataFromJSON(json) {
    return EthSignTransactionRequestDataFromJSONTyped(json, false);
}
exports.EthSignTransactionRequestDataFromJSON = EthSignTransactionRequestDataFromJSON;
function EthSignTransactionRequestDataFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.EthSignTransactionRequestDataFromJSONTyped = EthSignTransactionRequestDataFromJSONTyped;
function EthSignTransactionRequestDataToJSON(value) {
    return value;
}
exports.EthSignTransactionRequestDataToJSON = EthSignTransactionRequestDataToJSON;
