// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignTransactionRequestGasPrice.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignTransactionRequestGasPriceToJSON = exports.EthSignTransactionRequestGasPriceFromJSONTyped = exports.EthSignTransactionRequestGasPriceFromJSON = exports.instanceOfEthSignTransactionRequestGasPrice = void 0;
/**
 * Check if a given object implements the EthSignTransactionRequestGasPrice interface.
 */
function instanceOfEthSignTransactionRequestGasPrice(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfEthSignTransactionRequestGasPrice = instanceOfEthSignTransactionRequestGasPrice;
function EthSignTransactionRequestGasPriceFromJSON(json) {
    return EthSignTransactionRequestGasPriceFromJSONTyped(json, false);
}
exports.EthSignTransactionRequestGasPriceFromJSON = EthSignTransactionRequestGasPriceFromJSON;
function EthSignTransactionRequestGasPriceFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.EthSignTransactionRequestGasPriceFromJSONTyped = EthSignTransactionRequestGasPriceFromJSONTyped;
function EthSignTransactionRequestGasPriceToJSON(value) {
    return value;
}
exports.EthSignTransactionRequestGasPriceToJSON = EthSignTransactionRequestGasPriceToJSON;
