// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthSignTransactionRequestChainId.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSignTransactionRequestChainIdToJSON = exports.EthSignTransactionRequestChainIdFromJSONTyped = exports.EthSignTransactionRequestChainIdFromJSON = exports.instanceOfEthSignTransactionRequestChainId = void 0;
/**
 * Check if a given object implements the EthSignTransactionRequestChainId interface.
 */
function instanceOfEthSignTransactionRequestChainId(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfEthSignTransactionRequestChainId = instanceOfEthSignTransactionRequestChainId;
function EthSignTransactionRequestChainIdFromJSON(json) {
    return EthSignTransactionRequestChainIdFromJSONTyped(json, false);
}
exports.EthSignTransactionRequestChainIdFromJSON = EthSignTransactionRequestChainIdFromJSON;
function EthSignTransactionRequestChainIdFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.EthSignTransactionRequestChainIdFromJSONTyped = EthSignTransactionRequestChainIdFromJSONTyped;
function EthSignTransactionRequestChainIdToJSON(value) {
    return value;
}
exports.EthSignTransactionRequestChainIdToJSON = EthSignTransactionRequestChainIdToJSON;
