// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/EthVerifyRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthVerifyRequestToJSON = exports.EthVerifyRequestFromJSONTyped = exports.EthVerifyRequestFromJSON = exports.instanceOfEthVerifyRequest = void 0;
const TypesEthSignature_1 = require("./TypesEthSignature");
/**
 * Check if a given object implements the EthVerifyRequest interface.
 */
function instanceOfEthVerifyRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signature" in value;
    isInstance = isInstance && "address" in value;
    isInstance = isInstance && "message" in value;
    return isInstance;
}
exports.instanceOfEthVerifyRequest = instanceOfEthVerifyRequest;
function EthVerifyRequestFromJSON(json) {
    return EthVerifyRequestFromJSONTyped(json, false);
}
exports.EthVerifyRequestFromJSON = EthVerifyRequestFromJSON;
function EthVerifyRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signature': (0, TypesEthSignature_1.TypesEthSignatureFromJSON)(json['signature']),
        'address': json['address'],
        'message': json['message'],
    };
}
exports.EthVerifyRequestFromJSONTyped = EthVerifyRequestFromJSONTyped;
function EthVerifyRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signature': (0, TypesEthSignature_1.TypesEthSignatureToJSON)(value.signature),
        'address': value.address,
        'message': value.message,
    };
}
exports.EthVerifyRequestToJSON = EthVerifyRequestToJSON;
