// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/FlashHashRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlashHashRequestToJSON = exports.FlashHashRequestFromJSONTyped = exports.FlashHashRequestFromJSON = exports.instanceOfFlashHashRequest = void 0;
const runtime_1 = require("../runtime");
const TypesHexBytesLower_1 = require("./TypesHexBytesLower");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the FlashHashRequest interface.
 */
function instanceOfFlashHashRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "length" in value;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfFlashHashRequest = instanceOfFlashHashRequest;
function FlashHashRequestFromJSON(json) {
    return FlashHashRequestFromJSONTyped(json, false);
}
exports.FlashHashRequestFromJSON = FlashHashRequestFromJSON;
function FlashHashRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'challenge': !(0, runtime_1.exists)(json, 'challenge') ? undefined : (0, TypesHexBytesLower_1.TypesHexBytesLowerFromJSON)(json['challenge']),
        'length': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['length']),
        'address': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['address']),
    };
}
exports.FlashHashRequestFromJSONTyped = FlashHashRequestFromJSONTyped;
function FlashHashRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'challenge': (0, TypesHexBytesLower_1.TypesHexBytesLowerToJSON)(value.challenge),
        'length': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.length),
        'address': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.address),
    };
}
exports.FlashHashRequestToJSON = FlashHashRequestToJSON;
