// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/FlashWriteRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlashWriteRequestToJSON = exports.FlashWriteRequestFromJSONTyped = exports.FlashWriteRequestFromJSON = exports.instanceOfFlashWriteRequest = void 0;
const runtime_1 = require("../runtime");
const TypesHexBytesLower_1 = require("./TypesHexBytesLower");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the FlashWriteRequest interface.
 */
function instanceOfFlashWriteRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfFlashWriteRequest = instanceOfFlashWriteRequest;
function FlashWriteRequestFromJSON(json) {
    return FlashWriteRequestFromJSONTyped(json, false);
}
exports.FlashWriteRequestFromJSON = FlashWriteRequestFromJSON;
function FlashWriteRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'erase': !(0, runtime_1.exists)(json, 'erase') ? undefined : json['erase'],
        'data': !(0, runtime_1.exists)(json, 'data') ? undefined : (0, TypesHexBytesLower_1.TypesHexBytesLowerFromJSON)(json['data']),
        'address': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['address']),
    };
}
exports.FlashWriteRequestFromJSONTyped = FlashWriteRequestFromJSONTyped;
function FlashWriteRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'erase': value.erase,
        'data': (0, TypesHexBytesLower_1.TypesHexBytesLowerToJSON)(value.data),
        'address': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.address),
    };
}
exports.FlashWriteRequestToJSON = FlashWriteRequestToJSON;
