// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/GetEntropyRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetEntropyRequestToJSON = exports.GetEntropyRequestFromJSONTyped = exports.GetEntropyRequestFromJSON = exports.instanceOfGetEntropyRequest = void 0;
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the GetEntropyRequest interface.
 */
function instanceOfGetEntropyRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "size" in value;
    return isInstance;
}
exports.instanceOfGetEntropyRequest = instanceOfGetEntropyRequest;
function GetEntropyRequestFromJSON(json) {
    return GetEntropyRequestFromJSONTyped(json, false);
}
exports.GetEntropyRequestFromJSON = GetEntropyRequestFromJSON;
function GetEntropyRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'size': (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['size']),
    };
}
exports.GetEntropyRequestFromJSONTyped = GetEntropyRequestFromJSONTyped;
function GetEntropyRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'size': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.size),
    };
}
exports.GetEntropyRequestToJSON = GetEntropyRequestToJSON;
