// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/GetPublicKey200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetPublicKey200ResponseToJSON = exports.GetPublicKey200ResponseFromJSONTyped = exports.GetPublicKey200ResponseFromJSON = exports.instanceOfGetPublicKey200Response = void 0;
/**
 * Check if a given object implements the GetPublicKey200Response interface.
 */
function instanceOfGetPublicKey200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "xpub" in value;
    return isInstance;
}
exports.instanceOfGetPublicKey200Response = instanceOfGetPublicKey200Response;
function GetPublicKey200ResponseFromJSON(json) {
    return GetPublicKey200ResponseFromJSONTyped(json, false);
}
exports.GetPublicKey200ResponseFromJSON = GetPublicKey200ResponseFromJSON;
function GetPublicKey200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'xpub': json['xpub'],
    };
}
exports.GetPublicKey200ResponseFromJSONTyped = GetPublicKey200ResponseFromJSONTyped;
function GetPublicKey200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'xpub': value.xpub,
    };
}
exports.GetPublicKey200ResponseToJSON = GetPublicKey200ResponseToJSON;
