// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/LoadDeviceRequestAllOf1.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadDeviceRequestAllOf1ToJSON = exports.LoadDeviceRequestAllOf1FromJSONTyped = exports.LoadDeviceRequestAllOf1FromJSON = exports.instanceOfLoadDeviceRequestAllOf1 = void 0;
/**
 * Check if a given object implements the LoadDeviceRequestAllOf1 interface.
 */
function instanceOfLoadDeviceRequestAllOf1(value) {
    let isInstance = true;
    isInstance = isInstance && "mnemonic" in value;
    isInstance = isInstance && "xprv" in value;
    return isInstance;
}
exports.instanceOfLoadDeviceRequestAllOf1 = instanceOfLoadDeviceRequestAllOf1;
function LoadDeviceRequestAllOf1FromJSON(json) {
    return LoadDeviceRequestAllOf1FromJSONTyped(json, false);
}
exports.LoadDeviceRequestAllOf1FromJSON = LoadDeviceRequestAllOf1FromJSON;
function LoadDeviceRequestAllOf1FromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'mnemonic': json['mnemonic'],
        'xprv': json['xprv'],
    };
}
exports.LoadDeviceRequestAllOf1FromJSONTyped = LoadDeviceRequestAllOf1FromJSONTyped;
function LoadDeviceRequestAllOf1ToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'mnemonic': value.mnemonic,
        'xprv': value.xprv,
    };
}
exports.LoadDeviceRequestAllOf1ToJSON = LoadDeviceRequestAllOf1ToJSON;
