// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/LoadDeviceRequestAllOf.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadDeviceRequestAllOfToJSON = exports.LoadDeviceRequestAllOfFromJSONTyped = exports.LoadDeviceRequestAllOfFromJSON = exports.instanceOfLoadDeviceRequestAllOf = void 0;
const runtime_1 = require("../runtime");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the LoadDeviceRequestAllOf interface.
 */
function instanceOfLoadDeviceRequestAllOf(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfLoadDeviceRequestAllOf = instanceOfLoadDeviceRequestAllOf;
function LoadDeviceRequestAllOfFromJSON(json) {
    return LoadDeviceRequestAllOfFromJSONTyped(json, false);
}
exports.LoadDeviceRequestAllOfFromJSON = LoadDeviceRequestAllOfFromJSON;
function LoadDeviceRequestAllOfFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'u2f_counter': !(0, runtime_1.exists)(json, 'u2f_counter') ? undefined : (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['u2f_counter']),
        'skip_checksum': !(0, runtime_1.exists)(json, 'skip_checksum') ? undefined : json['skip_checksum'],
        'label': !(0, runtime_1.exists)(json, 'label') ? undefined : json['label'],
        'language': !(0, runtime_1.exists)(json, 'language') ? undefined : json['language'],
        'passphrase_protection': !(0, runtime_1.exists)(json, 'passphrase_protection') ? undefined : json['passphrase_protection'],
        'pin': !(0, runtime_1.exists)(json, 'pin') ? undefined : (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['pin']),
    };
}
exports.LoadDeviceRequestAllOfFromJSONTyped = LoadDeviceRequestAllOfFromJSONTyped;
function LoadDeviceRequestAllOfToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'u2f_counter': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.u2f_counter),
        'skip_checksum': value.skip_checksum,
        'label': value.label,
        'language': value.language,
        'passphrase_protection': value.passphrase_protection,
        'pin': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.pin),
    };
}
exports.LoadDeviceRequestAllOfToJSON = LoadDeviceRequestAllOfToJSON;
