// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/LoadDeviceRequestAllOf1AnyOf.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadDeviceRequestAllOf1AnyOfToJSON = exports.LoadDeviceRequestAllOf1AnyOfFromJSONTyped = exports.LoadDeviceRequestAllOf1AnyOfFromJSON = exports.instanceOfLoadDeviceRequestAllOf1AnyOf = void 0;
/**
 * Check if a given object implements the LoadDeviceRequestAllOf1AnyOf interface.
 */
function instanceOfLoadDeviceRequestAllOf1AnyOf(value) {
    let isInstance = true;
    isInstance = isInstance && "mnemonic" in value;
    return isInstance;
}
exports.instanceOfLoadDeviceRequestAllOf1AnyOf = instanceOfLoadDeviceRequestAllOf1AnyOf;
function LoadDeviceRequestAllOf1AnyOfFromJSON(json) {
    return LoadDeviceRequestAllOf1AnyOfFromJSONTyped(json, false);
}
exports.LoadDeviceRequestAllOf1AnyOfFromJSON = LoadDeviceRequestAllOf1AnyOfFromJSON;
function LoadDeviceRequestAllOf1AnyOfFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'mnemonic': json['mnemonic'],
    };
}
exports.LoadDeviceRequestAllOf1AnyOfFromJSONTyped = LoadDeviceRequestAllOf1AnyOfFromJSONTyped;
function LoadDeviceRequestAllOf1AnyOfToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'mnemonic': value.mnemonic,
    };
}
exports.LoadDeviceRequestAllOf1AnyOfToJSON = LoadDeviceRequestAllOf1AnyOfToJSON;
