// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/OsmoSignAminoSwapRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsmoSignAminoSwapRequestToJSON = exports.OsmoSignAminoSwapRequestFromJSONTyped = exports.OsmoSignAminoSwapRequestFromJSON = exports.instanceOfOsmoSignAminoSwapRequest = void 0;
const TypesOsmosisSignDocSwap_1 = require("./TypesOsmosisSignDocSwap");
/**
 * Check if a given object implements the OsmoSignAminoSwapRequest interface.
 */
function instanceOfOsmoSignAminoSwapRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfOsmoSignAminoSwapRequest = instanceOfOsmoSignAminoSwapRequest;
function OsmoSignAminoSwapRequestFromJSON(json) {
    return OsmoSignAminoSwapRequestFromJSONTyped(json, false);
}
exports.OsmoSignAminoSwapRequestFromJSON = OsmoSignAminoSwapRequestFromJSON;
function OsmoSignAminoSwapRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesOsmosisSignDocSwap_1.TypesOsmosisSignDocSwapFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.OsmoSignAminoSwapRequestFromJSONTyped = OsmoSignAminoSwapRequestFromJSONTyped;
function OsmoSignAminoSwapRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesOsmosisSignDocSwap_1.TypesOsmosisSignDocSwapToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.OsmoSignAminoSwapRequestToJSON = OsmoSignAminoSwapRequestToJSON;
