// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/OsmoSignAminoLpRemoveRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsmoSignAminoLpRemoveRequestToJSON = exports.OsmoSignAminoLpRemoveRequestFromJSONTyped = exports.OsmoSignAminoLpRemoveRequestFromJSON = exports.instanceOfOsmoSignAminoLpRemoveRequest = void 0;
const TypesOsmosisSignDocLPRemove_1 = require("./TypesOsmosisSignDocLPRemove");
/**
 * Check if a given object implements the OsmoSignAminoLpRemoveRequest interface.
 */
function instanceOfOsmoSignAminoLpRemoveRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfOsmoSignAminoLpRemoveRequest = instanceOfOsmoSignAminoLpRemoveRequest;
function OsmoSignAminoLpRemoveRequestFromJSON(json) {
    return OsmoSignAminoLpRemoveRequestFromJSONTyped(json, false);
}
exports.OsmoSignAminoLpRemoveRequestFromJSON = OsmoSignAminoLpRemoveRequestFromJSON;
function OsmoSignAminoLpRemoveRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesOsmosisSignDocLPRemove_1.TypesOsmosisSignDocLPRemoveFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.OsmoSignAminoLpRemoveRequestFromJSONTyped = OsmoSignAminoLpRemoveRequestFromJSONTyped;
function OsmoSignAminoLpRemoveRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesOsmosisSignDocLPRemove_1.TypesOsmosisSignDocLPRemoveToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.OsmoSignAminoLpRemoveRequestToJSON = OsmoSignAminoLpRemoveRequestToJSON;
