// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/OsmoSignAminoSwap200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsmoSignAminoSwap200ResponseToJSON = exports.OsmoSignAminoSwap200ResponseFromJSONTyped = exports.OsmoSignAminoSwap200ResponseFromJSON = exports.instanceOfOsmoSignAminoSwap200Response = void 0;
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
const TypesOsmosisSignDocSwap_1 = require("./TypesOsmosisSignDocSwap");
/**
 * Check if a given object implements the OsmoSignAminoSwap200Response interface.
 */
function instanceOfOsmoSignAminoSwap200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfOsmoSignAminoSwap200Response = instanceOfOsmoSignAminoSwap200Response;
function OsmoSignAminoSwap200ResponseFromJSON(json) {
    return OsmoSignAminoSwap200ResponseFromJSONTyped(json, false);
}
exports.OsmoSignAminoSwap200ResponseFromJSON = OsmoSignAminoSwap200ResponseFromJSON;
function OsmoSignAminoSwap200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesOsmosisSignDocSwap_1.TypesOsmosisSignDocSwapFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.OsmoSignAminoSwap200ResponseFromJSONTyped = OsmoSignAminoSwap200ResponseFromJSONTyped;
function OsmoSignAminoSwap200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesOsmosisSignDocSwap_1.TypesOsmosisSignDocSwapToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.OsmoSignAminoSwap200ResponseToJSON = OsmoSignAminoSwap200ResponseToJSON;
