// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/OsmoSignAminoLpAdd200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsmoSignAminoLpAdd200ResponseToJSON = exports.OsmoSignAminoLpAdd200ResponseFromJSONTyped = exports.OsmoSignAminoLpAdd200ResponseFromJSON = exports.instanceOfOsmoSignAminoLpAdd200Response = void 0;
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
const TypesOsmosisSignDocLPAdd_1 = require("./TypesOsmosisSignDocLPAdd");
/**
 * Check if a given object implements the OsmoSignAminoLpAdd200Response interface.
 */
function instanceOfOsmoSignAminoLpAdd200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfOsmoSignAminoLpAdd200Response = instanceOfOsmoSignAminoLpAdd200Response;
function OsmoSignAminoLpAdd200ResponseFromJSON(json) {
    return OsmoSignAminoLpAdd200ResponseFromJSONTyped(json, false);
}
exports.OsmoSignAminoLpAdd200ResponseFromJSON = OsmoSignAminoLpAdd200ResponseFromJSON;
function OsmoSignAminoLpAdd200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesOsmosisSignDocLPAdd_1.TypesOsmosisSignDocLPAddFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.OsmoSignAminoLpAdd200ResponseFromJSONTyped = OsmoSignAminoLpAdd200ResponseFromJSONTyped;
function OsmoSignAminoLpAdd200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesOsmosisSignDocLPAdd_1.TypesOsmosisSignDocLPAddToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.OsmoSignAminoLpAdd200ResponseToJSON = OsmoSignAminoLpAdd200ResponseToJSON;
