// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/OsmoSignAminoLpAddRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsmoSignAminoLpAddRequestToJSON = exports.OsmoSignAminoLpAddRequestFromJSONTyped = exports.OsmoSignAminoLpAddRequestFromJSON = exports.instanceOfOsmoSignAminoLpAddRequest = void 0;
const TypesOsmosisSignDocLPAdd_1 = require("./TypesOsmosisSignDocLPAdd");
/**
 * Check if a given object implements the OsmoSignAminoLpAddRequest interface.
 */
function instanceOfOsmoSignAminoLpAddRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfOsmoSignAminoLpAddRequest = instanceOfOsmoSignAminoLpAddRequest;
function OsmoSignAminoLpAddRequestFromJSON(json) {
    return OsmoSignAminoLpAddRequestFromJSONTyped(json, false);
}
exports.OsmoSignAminoLpAddRequestFromJSON = OsmoSignAminoLpAddRequestFromJSON;
function OsmoSignAminoLpAddRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesOsmosisSignDocLPAdd_1.TypesOsmosisSignDocLPAddFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.OsmoSignAminoLpAddRequestFromJSONTyped = OsmoSignAminoLpAddRequestFromJSONTyped;
function OsmoSignAminoLpAddRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesOsmosisSignDocLPAdd_1.TypesOsmosisSignDocLPAddToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.OsmoSignAminoLpAddRequestToJSON = OsmoSignAminoLpAddRequestToJSON;
