// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/OsmoSignAminoLpRemove200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsmoSignAminoLpRemove200ResponseToJSON = exports.OsmoSignAminoLpRemove200ResponseFromJSONTyped = exports.OsmoSignAminoLpRemove200ResponseFromJSON = exports.instanceOfOsmoSignAminoLpRemove200Response = void 0;
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
const TypesOsmosisSignDocLPRemove_1 = require("./TypesOsmosisSignDocLPRemove");
/**
 * Check if a given object implements the OsmoSignAminoLpRemove200Response interface.
 */
function instanceOfOsmoSignAminoLpRemove200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "signed" in value;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfOsmoSignAminoLpRemove200Response = instanceOfOsmoSignAminoLpRemove200Response;
function OsmoSignAminoLpRemove200ResponseFromJSON(json) {
    return OsmoSignAminoLpRemove200ResponseFromJSONTyped(json, false);
}
exports.OsmoSignAminoLpRemove200ResponseFromJSON = OsmoSignAminoLpRemove200ResponseFromJSON;
function OsmoSignAminoLpRemove200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signed': (0, TypesOsmosisSignDocLPRemove_1.TypesOsmosisSignDocLPRemoveFromJSON)(json['signed']),
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.OsmoSignAminoLpRemove200ResponseFromJSONTyped = OsmoSignAminoLpRemove200ResponseFromJSONTyped;
function OsmoSignAminoLpRemove200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signed': (0, TypesOsmosisSignDocLPRemove_1.TypesOsmosisSignDocLPRemoveToJSON)(value.signed),
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.OsmoSignAminoLpRemove200ResponseToJSON = OsmoSignAminoLpRemove200ResponseToJSON;
