// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/PairingInfo.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PairingInfoToJSON = exports.PairingInfoFromJSONTyped = exports.PairingInfoFromJSON = exports.instanceOfPairingInfo = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the PairingInfo interface.
 */
function instanceOfPairingInfo(value) {
    let isInstance = true;
    isInstance = isInstance && "imageUrl" in value;
    isInstance = isInstance && "url" in value;
    isInstance = isInstance && "name" in value;
    return isInstance;
}
exports.instanceOfPairingInfo = instanceOfPairingInfo;
function PairingInfoFromJSON(json) {
    return PairingInfoFromJSONTyped(json, false);
}
exports.PairingInfoFromJSON = PairingInfoFromJSON;
function PairingInfoFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'addedOn': !(0, runtime_1.exists)(json, 'addedOn') ? undefined : json['addedOn'],
        'imageUrl': json['imageUrl'],
        'url': json['url'],
        'name': json['name'],
    };
}
exports.PairingInfoFromJSONTyped = PairingInfoFromJSONTyped;
function PairingInfoToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'addedOn': value.addedOn,
        'imageUrl': value.imageUrl,
        'url': value.url,
        'name': value.name,
    };
}
exports.PairingInfoToJSON = PairingInfoToJSON;
