// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/Pair200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pair200ResponseToJSON = exports.Pair200ResponseFromJSONTyped = exports.Pair200ResponseFromJSON = exports.instanceOfPair200Response = void 0;
/**
 * Check if a given object implements the Pair200Response interface.
 */
function instanceOfPair200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "apiKey" in value;
    return isInstance;
}
exports.instanceOfPair200Response = instanceOfPair200Response;
function Pair200ResponseFromJSON(json) {
    return Pair200ResponseFromJSONTyped(json, false);
}
exports.Pair200ResponseFromJSON = Pair200ResponseFromJSON;
function Pair200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'apiKey': json['apiKey'],
    };
}
exports.Pair200ResponseFromJSONTyped = Pair200ResponseFromJSONTyped;
function Pair200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'apiKey': value.apiKey,
    };
}
exports.Pair200ResponseToJSON = Pair200ResponseToJSON;
