// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/Ping200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ping200ResponseToJSON = exports.Ping200ResponseFromJSONTyped = exports.Ping200ResponseFromJSON = exports.instanceOfPing200Response = void 0;
/**
 * Check if a given object implements the Ping200Response interface.
 */
function instanceOfPing200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "message" in value;
    return isInstance;
}
exports.instanceOfPing200Response = instanceOfPing200Response;
function Ping200ResponseFromJSON(json) {
    return Ping200ResponseFromJSONTyped(json, false);
}
exports.Ping200ResponseFromJSON = Ping200ResponseFromJSON;
function Ping200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'message': json['message'],
    };
}
exports.Ping200ResponseFromJSONTyped = Ping200ResponseFromJSONTyped;
function Ping200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'message': value.message,
    };
}
exports.Ping200ResponseToJSON = Ping200ResponseToJSON;
