// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/PingRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PingRequestToJSON = exports.PingRequestFromJSONTyped = exports.PingRequestFromJSON = exports.instanceOfPingRequest = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the PingRequest interface.
 */
function instanceOfPingRequest(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfPingRequest = instanceOfPingRequest;
function PingRequestFromJSON(json) {
    return PingRequestFromJSONTyped(json, false);
}
exports.PingRequestFromJSON = PingRequestFromJSON;
function PingRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'message': !(0, runtime_1.exists)(json, 'message') ? undefined : json['message'],
        'wipe_code_protection': !(0, runtime_1.exists)(json, 'wipe_code_protection') ? undefined : json['wipe_code_protection'],
        'passphrase_protection': !(0, runtime_1.exists)(json, 'passphrase_protection') ? undefined : json['passphrase_protection'],
        'pin_protection': !(0, runtime_1.exists)(json, 'pin_protection') ? undefined : json['pin_protection'],
        'button_protection': !(0, runtime_1.exists)(json, 'button_protection') ? undefined : json['button_protection'],
    };
}
exports.PingRequestFromJSONTyped = PingRequestFromJSONTyped;
function PingRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'message': value.message,
        'wipe_code_protection': value.wipe_code_protection,
        'passphrase_protection': value.passphrase_protection,
        'pin_protection': value.pin_protection,
        'button_protection': value.button_protection,
    };
}
exports.PingRequestToJSON = PingRequestToJSON;
