// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/RecoverDeviceRequestAutoLockDelayMs.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecoverDeviceRequestAutoLockDelayMsToJSON = exports.RecoverDeviceRequestAutoLockDelayMsFromJSONTyped = exports.RecoverDeviceRequestAutoLockDelayMsFromJSON = exports.instanceOfRecoverDeviceRequestAutoLockDelayMs = void 0;
/**
 * Check if a given object implements the RecoverDeviceRequestAutoLockDelayMs interface.
 */
function instanceOfRecoverDeviceRequestAutoLockDelayMs(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfRecoverDeviceRequestAutoLockDelayMs = instanceOfRecoverDeviceRequestAutoLockDelayMs;
function RecoverDeviceRequestAutoLockDelayMsFromJSON(json) {
    return RecoverDeviceRequestAutoLockDelayMsFromJSONTyped(json, false);
}
exports.RecoverDeviceRequestAutoLockDelayMsFromJSON = RecoverDeviceRequestAutoLockDelayMsFromJSON;
function RecoverDeviceRequestAutoLockDelayMsFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.RecoverDeviceRequestAutoLockDelayMsFromJSONTyped = RecoverDeviceRequestAutoLockDelayMsFromJSONTyped;
function RecoverDeviceRequestAutoLockDelayMsToJSON(value) {
    return value;
}
exports.RecoverDeviceRequestAutoLockDelayMsToJSON = RecoverDeviceRequestAutoLockDelayMsToJSON;
