// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/ResetDeviceRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResetDeviceRequestToJSON = exports.ResetDeviceRequestFromJSONTyped = exports.ResetDeviceRequestFromJSON = exports.instanceOfResetDeviceRequest = exports.ResetDeviceRequestStrengthEnum = void 0;
const runtime_1 = require("../runtime");
const RecoverDeviceRequestAutoLockDelayMs_1 = require("./RecoverDeviceRequestAutoLockDelayMs");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * @export
 */
exports.ResetDeviceRequestStrengthEnum = {
    NUMBER_128: 128,
    NUMBER_192: 192,
    NUMBER_256: 256
};
/**
 * Check if a given object implements the ResetDeviceRequest interface.
 */
function instanceOfResetDeviceRequest(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfResetDeviceRequest = instanceOfResetDeviceRequest;
function ResetDeviceRequestFromJSON(json) {
    return ResetDeviceRequestFromJSONTyped(json, false);
}
exports.ResetDeviceRequestFromJSON = ResetDeviceRequestFromJSON;
function ResetDeviceRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'u2f_counter': !(0, runtime_1.exists)(json, 'u2f_counter') ? undefined : (0, TypesNumericU32_1.TypesNumericU32FromJSON)(json['u2f_counter']),
        'auto_lock_delay_ms': !(0, runtime_1.exists)(json, 'auto_lock_delay_ms') ? undefined : (0, RecoverDeviceRequestAutoLockDelayMs_1.RecoverDeviceRequestAutoLockDelayMsFromJSON)(json['auto_lock_delay_ms']),
        'no_backup': !(0, runtime_1.exists)(json, 'no_backup') ? undefined : json['no_backup'],
        'label': !(0, runtime_1.exists)(json, 'label') ? undefined : json['label'],
        'language': !(0, runtime_1.exists)(json, 'language') ? undefined : json['language'],
        'pin_protection': !(0, runtime_1.exists)(json, 'pin_protection') ? undefined : json['pin_protection'],
        'passphrase_protection': !(0, runtime_1.exists)(json, 'passphrase_protection') ? undefined : json['passphrase_protection'],
        'strength': !(0, runtime_1.exists)(json, 'strength') ? undefined : json['strength'],
        'display_random': !(0, runtime_1.exists)(json, 'display_random') ? undefined : json['display_random'],
    };
}
exports.ResetDeviceRequestFromJSONTyped = ResetDeviceRequestFromJSONTyped;
function ResetDeviceRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'u2f_counter': (0, TypesNumericU32_1.TypesNumericU32ToJSON)(value.u2f_counter),
        'auto_lock_delay_ms': (0, RecoverDeviceRequestAutoLockDelayMs_1.RecoverDeviceRequestAutoLockDelayMsToJSON)(value.auto_lock_delay_ms),
        'no_backup': value.no_backup,
        'label': value.label,
        'language': value.language,
        'pin_protection': value.pin_protection,
        'passphrase_protection': value.passphrase_protection,
        'strength': value.strength,
        'display_random': value.display_random,
    };
}
exports.ResetDeviceRequestToJSON = ResetDeviceRequestToJSON;
