// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/Route.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteToJSON = exports.RouteFromJSONTyped = exports.RouteFromJSON = exports.instanceOfRoute = void 0;
/**
 * Check if a given object implements the Route interface.
 */
function instanceOfRoute(value) {
    let isInstance = true;
    isInstance = isInstance && "token_out_denom" in value;
    isInstance = isInstance && "pool_id" in value;
    return isInstance;
}
exports.instanceOfRoute = instanceOfRoute;
function RouteFromJSON(json) {
    return RouteFromJSONTyped(json, false);
}
exports.RouteFromJSON = RouteFromJSON;
function RouteFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'token_out_denom': json['token_out_denom'],
        'pool_id': json['pool_id'],
    };
}
exports.RouteFromJSONTyped = RouteFromJSONTyped;
function RouteToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'token_out_denom': value.token_out_denom,
        'pool_id': value.pool_id,
    };
}
exports.RouteToJSON = RouteToJSON;
