// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/SignIdentityRequestIndex.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignIdentityRequestIndexToJSON = exports.SignIdentityRequestIndexFromJSONTyped = exports.SignIdentityRequestIndexFromJSON = exports.instanceOfSignIdentityRequestIndex = void 0;
/**
 * Check if a given object implements the SignIdentityRequestIndex interface.
 */
function instanceOfSignIdentityRequestIndex(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfSignIdentityRequestIndex = instanceOfSignIdentityRequestIndex;
function SignIdentityRequestIndexFromJSON(json) {
    return SignIdentityRequestIndexFromJSONTyped(json, false);
}
exports.SignIdentityRequestIndexFromJSON = SignIdentityRequestIndexFromJSON;
function SignIdentityRequestIndexFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.SignIdentityRequestIndexFromJSONTyped = SignIdentityRequestIndexFromJSONTyped;
function SignIdentityRequestIndexToJSON(value) {
    return value;
}
exports.SignIdentityRequestIndexToJSON = SignIdentityRequestIndexToJSON;
