// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/SignIdentityRequestChallengeHidden.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignIdentityRequestChallengeHiddenToJSON = exports.SignIdentityRequestChallengeHiddenFromJSONTyped = exports.SignIdentityRequestChallengeHiddenFromJSON = exports.instanceOfSignIdentityRequestChallengeHidden = void 0;
/**
 * Check if a given object implements the SignIdentityRequestChallengeHidden interface.
 */
function instanceOfSignIdentityRequestChallengeHidden(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfSignIdentityRequestChallengeHidden = instanceOfSignIdentityRequestChallengeHidden;
function SignIdentityRequestChallengeHiddenFromJSON(json) {
    return SignIdentityRequestChallengeHiddenFromJSONTyped(json, false);
}
exports.SignIdentityRequestChallengeHiddenFromJSON = SignIdentityRequestChallengeHiddenFromJSON;
function SignIdentityRequestChallengeHiddenFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.SignIdentityRequestChallengeHiddenFromJSONTyped = SignIdentityRequestChallengeHiddenFromJSONTyped;
function SignIdentityRequestChallengeHiddenToJSON(value) {
    return value;
}
exports.SignIdentityRequestChallengeHiddenToJSON = SignIdentityRequestChallengeHiddenToJSON;
