// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxMsgLPAdd.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxMsgLPAddToJSON = exports.StdTxMsgLPAddFromJSONTyped = exports.StdTxMsgLPAddFromJSON = exports.instanceOfStdTxMsgLPAdd = exports.StdTxMsgLPAddTypeEnum = void 0;
const StdTxMsgLPAddValue_1 = require("./StdTxMsgLPAddValue");
/**
 * @export
 */
exports.StdTxMsgLPAddTypeEnum = {
    osmosis_gamm_join_pool: 'osmosis/gamm/join-pool'
};
/**
 * Check if a given object implements the StdTxMsgLPAdd interface.
 */
function instanceOfStdTxMsgLPAdd(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxMsgLPAdd = instanceOfStdTxMsgLPAdd;
function StdTxMsgLPAddFromJSON(json) {
    return StdTxMsgLPAddFromJSONTyped(json, false);
}
exports.StdTxMsgLPAddFromJSON = StdTxMsgLPAddFromJSON;
function StdTxMsgLPAddFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxMsgLPAddValue_1.StdTxMsgLPAddValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxMsgLPAddFromJSONTyped = StdTxMsgLPAddFromJSONTyped;
function StdTxMsgLPAddToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxMsgLPAddValue_1.StdTxMsgLPAddValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxMsgLPAddToJSON = StdTxMsgLPAddToJSON;
