// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxMsgLPAddValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxMsgLPAddValueToJSON = exports.StdTxMsgLPAddValueFromJSONTyped = exports.StdTxMsgLPAddValueFromJSON = exports.instanceOfStdTxMsgLPAddValue = void 0;
const TypesCosmosCoin_1 = require("./TypesCosmosCoin");
/**
 * Check if a given object implements the StdTxMsgLPAddValue interface.
 */
function instanceOfStdTxMsgLPAddValue(value) {
    let isInstance = true;
    isInstance = isInstance && "pool_id" in value;
    isInstance = isInstance && "sender" in value;
    isInstance = isInstance && "share_out_amount" in value;
    isInstance = isInstance && "token_in_maxs" in value;
    return isInstance;
}
exports.instanceOfStdTxMsgLPAddValue = instanceOfStdTxMsgLPAddValue;
function StdTxMsgLPAddValueFromJSON(json) {
    return StdTxMsgLPAddValueFromJSONTyped(json, false);
}
exports.StdTxMsgLPAddValueFromJSON = StdTxMsgLPAddValueFromJSON;
function StdTxMsgLPAddValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'pool_id': json['pool_id'],
        'sender': json['sender'],
        'share_out_amount': json['share_out_amount'],
        'token_in_maxs': (json['token_in_maxs'].map(TypesCosmosCoin_1.TypesCosmosCoinFromJSON)),
    };
}
exports.StdTxMsgLPAddValueFromJSONTyped = StdTxMsgLPAddValueFromJSONTyped;
function StdTxMsgLPAddValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'pool_id': value.pool_id,
        'sender': value.sender,
        'share_out_amount': value.share_out_amount,
        'token_in_maxs': (value.token_in_maxs.map(TypesCosmosCoin_1.TypesCosmosCoinToJSON)),
    };
}
exports.StdTxMsgLPAddValueToJSON = StdTxMsgLPAddValueToJSON;
