// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxTransferValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxTransferValueToJSON = exports.StdTxTransferValueFromJSONTyped = exports.StdTxTransferValueFromJSON = exports.instanceOfStdTxTransferValue = void 0;
const StdTxTransferValueAmountInner_1 = require("./StdTxTransferValueAmountInner");
/**
 * Check if a given object implements the StdTxTransferValue interface.
 */
function instanceOfStdTxTransferValue(value) {
    let isInstance = true;
    isInstance = isInstance && "amount" in value;
    isInstance = isInstance && "from_address" in value;
    isInstance = isInstance && "to_address" in value;
    return isInstance;
}
exports.instanceOfStdTxTransferValue = instanceOfStdTxTransferValue;
function StdTxTransferValueFromJSON(json) {
    return StdTxTransferValueFromJSONTyped(json, false);
}
exports.StdTxTransferValueFromJSON = StdTxTransferValueFromJSON;
function StdTxTransferValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'amount': (json['amount'].map(StdTxTransferValueAmountInner_1.StdTxTransferValueAmountInnerFromJSON)),
        'from_address': json['from_address'],
        'to_address': json['to_address'],
    };
}
exports.StdTxTransferValueFromJSONTyped = StdTxTransferValueFromJSONTyped;
function StdTxTransferValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'amount': (value.amount.map(StdTxTransferValueAmountInner_1.StdTxTransferValueAmountInnerToJSON)),
        'from_address': value.from_address,
        'to_address': value.to_address,
    };
}
exports.StdTxTransferValueToJSON = StdTxTransferValueToJSON;
