// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxDelegateValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxDelegateValueToJSON = exports.StdTxDelegateValueFromJSONTyped = exports.StdTxDelegateValueFromJSON = exports.instanceOfStdTxDelegateValue = void 0;
/**
 * Check if a given object implements the StdTxDelegateValue interface.
 */
function instanceOfStdTxDelegateValue(value) {
    let isInstance = true;
    isInstance = isInstance && "amount" in value;
    isInstance = isInstance && "delegator_address" in value;
    isInstance = isInstance && "validator_address" in value;
    return isInstance;
}
exports.instanceOfStdTxDelegateValue = instanceOfStdTxDelegateValue;
function StdTxDelegateValueFromJSON(json) {
    return StdTxDelegateValueFromJSONTyped(json, false);
}
exports.StdTxDelegateValueFromJSON = StdTxDelegateValueFromJSON;
function StdTxDelegateValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'amount': json['amount'],
        'delegator_address': json['delegator_address'],
        'validator_address': json['validator_address'],
    };
}
exports.StdTxDelegateValueFromJSONTyped = StdTxDelegateValueFromJSONTyped;
function StdTxDelegateValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'amount': value.amount,
        'delegator_address': value.delegator_address,
        'validator_address': value.validator_address,
    };
}
exports.StdTxDelegateValueToJSON = StdTxDelegateValueToJSON;
