// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxWithdrawDelegationReward.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxWithdrawDelegationRewardToJSON = exports.StdTxWithdrawDelegationRewardFromJSONTyped = exports.StdTxWithdrawDelegationRewardFromJSON = exports.instanceOfStdTxWithdrawDelegationReward = exports.StdTxWithdrawDelegationRewardTypeEnum = void 0;
const StdTxWithdrawDelegationRewardValue_1 = require("./StdTxWithdrawDelegationRewardValue");
/**
 * @export
 */
exports.StdTxWithdrawDelegationRewardTypeEnum = {
    cosmos_sdk_MsgWithdrawDelegationReward: 'cosmos-sdk/MsgWithdrawDelegationReward'
};
/**
 * Check if a given object implements the StdTxWithdrawDelegationReward interface.
 */
function instanceOfStdTxWithdrawDelegationReward(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxWithdrawDelegationReward = instanceOfStdTxWithdrawDelegationReward;
function StdTxWithdrawDelegationRewardFromJSON(json) {
    return StdTxWithdrawDelegationRewardFromJSONTyped(json, false);
}
exports.StdTxWithdrawDelegationRewardFromJSON = StdTxWithdrawDelegationRewardFromJSON;
function StdTxWithdrawDelegationRewardFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxWithdrawDelegationRewardValue_1.StdTxWithdrawDelegationRewardValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxWithdrawDelegationRewardFromJSONTyped = StdTxWithdrawDelegationRewardFromJSONTyped;
function StdTxWithdrawDelegationRewardToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxWithdrawDelegationRewardValue_1.StdTxWithdrawDelegationRewardValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxWithdrawDelegationRewardToJSON = StdTxWithdrawDelegationRewardToJSON;
