// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxMsgTransfer.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxMsgTransferToJSON = exports.StdTxMsgTransferFromJSONTyped = exports.StdTxMsgTransferFromJSON = exports.instanceOfStdTxMsgTransfer = exports.StdTxMsgTransferTypeEnum = void 0;
const StdTxMsgTransferValue_1 = require("./StdTxMsgTransferValue");
/**
 * @export
 */
exports.StdTxMsgTransferTypeEnum = {
    cosmos_sdk_MsgTransfer: 'cosmos-sdk/MsgTransfer'
};
/**
 * Check if a given object implements the StdTxMsgTransfer interface.
 */
function instanceOfStdTxMsgTransfer(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxMsgTransfer = instanceOfStdTxMsgTransfer;
function StdTxMsgTransferFromJSON(json) {
    return StdTxMsgTransferFromJSONTyped(json, false);
}
exports.StdTxMsgTransferFromJSON = StdTxMsgTransferFromJSON;
function StdTxMsgTransferFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxMsgTransferValue_1.StdTxMsgTransferValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxMsgTransferFromJSONTyped = StdTxMsgTransferFromJSONTyped;
function StdTxMsgTransferToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxMsgTransferValue_1.StdTxMsgTransferValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxMsgTransferToJSON = StdTxMsgTransferToJSON;
