// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxUnDelegate.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxUnDelegateToJSON = exports.StdTxUnDelegateFromJSONTyped = exports.StdTxUnDelegateFromJSON = exports.instanceOfStdTxUnDelegate = exports.StdTxUnDelegateTypeEnum = void 0;
const StdTxDelegateValue_1 = require("./StdTxDelegateValue");
/**
 * @export
 */
exports.StdTxUnDelegateTypeEnum = {
    cosmos_sdk_MsgUndelegate: 'cosmos-sdk/MsgUndelegate'
};
/**
 * Check if a given object implements the StdTxUnDelegate interface.
 */
function instanceOfStdTxUnDelegate(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxUnDelegate = instanceOfStdTxUnDelegate;
function StdTxUnDelegateFromJSON(json) {
    return StdTxUnDelegateFromJSONTyped(json, false);
}
exports.StdTxUnDelegateFromJSON = StdTxUnDelegateFromJSON;
function StdTxUnDelegateFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxDelegateValue_1.StdTxDelegateValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxUnDelegateFromJSONTyped = StdTxUnDelegateFromJSONTyped;
function StdTxUnDelegateToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxDelegateValue_1.StdTxDelegateValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxUnDelegateToJSON = StdTxUnDelegateToJSON;
