// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxDeposit.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxDepositToJSON = exports.StdTxDepositFromJSONTyped = exports.StdTxDepositFromJSON = exports.instanceOfStdTxDeposit = exports.StdTxDepositTypeEnum = void 0;
const StdTxDepositValue_1 = require("./StdTxDepositValue");
/**
 * @export
 */
exports.StdTxDepositTypeEnum = {
    thorchain_MsgDeposit: 'thorchain/MsgDeposit'
};
/**
 * Check if a given object implements the StdTxDeposit interface.
 */
function instanceOfStdTxDeposit(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxDeposit = instanceOfStdTxDeposit;
function StdTxDepositFromJSON(json) {
    return StdTxDepositFromJSONTyped(json, false);
}
exports.StdTxDepositFromJSON = StdTxDepositFromJSON;
function StdTxDepositFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxDepositValue_1.StdTxDepositValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxDepositFromJSONTyped = StdTxDepositFromJSONTyped;
function StdTxDepositToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxDepositValue_1.StdTxDepositValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxDepositToJSON = StdTxDepositToJSON;
