// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxBeginReDelegateValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxBeginReDelegateValueToJSON = exports.StdTxBeginReDelegateValueFromJSONTyped = exports.StdTxBeginReDelegateValueFromJSON = exports.instanceOfStdTxBeginReDelegateValue = void 0;
/**
 * Check if a given object implements the StdTxBeginReDelegateValue interface.
 */
function instanceOfStdTxBeginReDelegateValue(value) {
    let isInstance = true;
    isInstance = isInstance && "amount" in value;
    isInstance = isInstance && "delegator_address" in value;
    isInstance = isInstance && "validator_dst_address" in value;
    isInstance = isInstance && "validator_src_address" in value;
    return isInstance;
}
exports.instanceOfStdTxBeginReDelegateValue = instanceOfStdTxBeginReDelegateValue;
function StdTxBeginReDelegateValueFromJSON(json) {
    return StdTxBeginReDelegateValueFromJSONTyped(json, false);
}
exports.StdTxBeginReDelegateValueFromJSON = StdTxBeginReDelegateValueFromJSON;
function StdTxBeginReDelegateValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'amount': json['amount'],
        'delegator_address': json['delegator_address'],
        'validator_dst_address': json['validator_dst_address'],
        'validator_src_address': json['validator_src_address'],
    };
}
exports.StdTxBeginReDelegateValueFromJSONTyped = StdTxBeginReDelegateValueFromJSONTyped;
function StdTxBeginReDelegateValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'amount': value.amount,
        'delegator_address': value.delegator_address,
        'validator_dst_address': value.validator_dst_address,
        'validator_src_address': value.validator_src_address,
    };
}
exports.StdTxBeginReDelegateValueToJSON = StdTxBeginReDelegateValueToJSON;
