// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxMsgSwapValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxMsgSwapValueToJSON = exports.StdTxMsgSwapValueFromJSONTyped = exports.StdTxMsgSwapValueFromJSON = exports.instanceOfStdTxMsgSwapValue = void 0;
const Route_1 = require("./Route");
/**
 * Check if a given object implements the StdTxMsgSwapValue interface.
 */
function instanceOfStdTxMsgSwapValue(value) {
    let isInstance = true;
    isInstance = isInstance && "routes" in value;
    isInstance = isInstance && "sender" in value;
    isInstance = isInstance && "token_in" in value;
    isInstance = isInstance && "token_out_min_amount" in value;
    return isInstance;
}
exports.instanceOfStdTxMsgSwapValue = instanceOfStdTxMsgSwapValue;
function StdTxMsgSwapValueFromJSON(json) {
    return StdTxMsgSwapValueFromJSONTyped(json, false);
}
exports.StdTxMsgSwapValueFromJSON = StdTxMsgSwapValueFromJSON;
function StdTxMsgSwapValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'routes': (json['routes'].map(Route_1.RouteFromJSON)),
        'sender': json['sender'],
        'token_in': json['token_in'],
        'token_out_min_amount': json['token_out_min_amount'],
    };
}
exports.StdTxMsgSwapValueFromJSONTyped = StdTxMsgSwapValueFromJSONTyped;
function StdTxMsgSwapValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'routes': (value.routes.map(Route_1.RouteToJSON)),
        'sender': value.sender,
        'token_in': value.token_in,
        'token_out_min_amount': value.token_out_min_amount,
    };
}
exports.StdTxMsgSwapValueToJSON = StdTxMsgSwapValueToJSON;
