// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxBeginReDelegate.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxBeginReDelegateToJSON = exports.StdTxBeginReDelegateFromJSONTyped = exports.StdTxBeginReDelegateFromJSON = exports.instanceOfStdTxBeginReDelegate = exports.StdTxBeginReDelegateTypeEnum = void 0;
const StdTxBeginReDelegateValue_1 = require("./StdTxBeginReDelegateValue");
/**
 * @export
 */
exports.StdTxBeginReDelegateTypeEnum = {
    cosmos_sdk_MsgBeginRedelegate: 'cosmos-sdk/MsgBeginRedelegate'
};
/**
 * Check if a given object implements the StdTxBeginReDelegate interface.
 */
function instanceOfStdTxBeginReDelegate(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxBeginReDelegate = instanceOfStdTxBeginReDelegate;
function StdTxBeginReDelegateFromJSON(json) {
    return StdTxBeginReDelegateFromJSONTyped(json, false);
}
exports.StdTxBeginReDelegateFromJSON = StdTxBeginReDelegateFromJSON;
function StdTxBeginReDelegateFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxBeginReDelegateValue_1.StdTxBeginReDelegateValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxBeginReDelegateFromJSONTyped = StdTxBeginReDelegateFromJSONTyped;
function StdTxBeginReDelegateToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxBeginReDelegateValue_1.StdTxBeginReDelegateValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxBeginReDelegateToJSON = StdTxBeginReDelegateToJSON;
