// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxDepositValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxDepositValueToJSON = exports.StdTxDepositValueFromJSONTyped = exports.StdTxDepositValueFromJSON = exports.instanceOfStdTxDepositValue = void 0;
/**
 * Check if a given object implements the StdTxDepositValue interface.
 */
function instanceOfStdTxDepositValue(value) {
    let isInstance = true;
    isInstance = isInstance && "coins" in value;
    isInstance = isInstance && "memo" in value;
    isInstance = isInstance && "signer" in value;
    return isInstance;
}
exports.instanceOfStdTxDepositValue = instanceOfStdTxDepositValue;
function StdTxDepositValueFromJSON(json) {
    return StdTxDepositValueFromJSONTyped(json, false);
}
exports.StdTxDepositValueFromJSON = StdTxDepositValueFromJSON;
function StdTxDepositValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'coins': json['coins'],
        'memo': json['memo'],
        'signer': json['signer'],
    };
}
exports.StdTxDepositValueFromJSONTyped = StdTxDepositValueFromJSONTyped;
function StdTxDepositValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'coins': value.coins,
        'memo': value.memo,
        'signer': value.signer,
    };
}
exports.StdTxDepositValueToJSON = StdTxDepositValueToJSON;
