// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxMsgLPRemove.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxMsgLPRemoveToJSON = exports.StdTxMsgLPRemoveFromJSONTyped = exports.StdTxMsgLPRemoveFromJSON = exports.instanceOfStdTxMsgLPRemove = exports.StdTxMsgLPRemoveTypeEnum = void 0;
const StdTxMsgLPRemoveValue_1 = require("./StdTxMsgLPRemoveValue");
/**
 * @export
 */
exports.StdTxMsgLPRemoveTypeEnum = {
    osmosis_gamm_exit_pool: 'osmosis/gamm/exit-pool'
};
/**
 * Check if a given object implements the StdTxMsgLPRemove interface.
 */
function instanceOfStdTxMsgLPRemove(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxMsgLPRemove = instanceOfStdTxMsgLPRemove;
function StdTxMsgLPRemoveFromJSON(json) {
    return StdTxMsgLPRemoveFromJSONTyped(json, false);
}
exports.StdTxMsgLPRemoveFromJSON = StdTxMsgLPRemoveFromJSON;
function StdTxMsgLPRemoveFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxMsgLPRemoveValue_1.StdTxMsgLPRemoveValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxMsgLPRemoveFromJSONTyped = StdTxMsgLPRemoveFromJSONTyped;
function StdTxMsgLPRemoveToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxMsgLPRemoveValue_1.StdTxMsgLPRemoveValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxMsgLPRemoveToJSON = StdTxMsgLPRemoveToJSON;
