// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxTransferValueAmountInner.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxTransferValueAmountInnerToJSON = exports.StdTxTransferValueAmountInnerFromJSONTyped = exports.StdTxTransferValueAmountInnerFromJSON = exports.instanceOfStdTxTransferValueAmountInner = void 0;
/**
 * Check if a given object implements the StdTxTransferValueAmountInner interface.
 */
function instanceOfStdTxTransferValueAmountInner(value) {
    let isInstance = true;
    isInstance = isInstance && "denom" in value;
    isInstance = isInstance && "amount" in value;
    return isInstance;
}
exports.instanceOfStdTxTransferValueAmountInner = instanceOfStdTxTransferValueAmountInner;
function StdTxTransferValueAmountInnerFromJSON(json) {
    return StdTxTransferValueAmountInnerFromJSONTyped(json, false);
}
exports.StdTxTransferValueAmountInnerFromJSON = StdTxTransferValueAmountInnerFromJSON;
function StdTxTransferValueAmountInnerFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'denom': json['denom'],
        'amount': json['amount'],
    };
}
exports.StdTxTransferValueAmountInnerFromJSONTyped = StdTxTransferValueAmountInnerFromJSONTyped;
function StdTxTransferValueAmountInnerToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'denom': value.denom,
        'amount': value.amount,
    };
}
exports.StdTxTransferValueAmountInnerToJSON = StdTxTransferValueAmountInnerToJSON;
