// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxMsgLPRemoveValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxMsgLPRemoveValueToJSON = exports.StdTxMsgLPRemoveValueFromJSONTyped = exports.StdTxMsgLPRemoveValueFromJSON = exports.instanceOfStdTxMsgLPRemoveValue = void 0;
const TypesCosmosCoin_1 = require("./TypesCosmosCoin");
/**
 * Check if a given object implements the StdTxMsgLPRemoveValue interface.
 */
function instanceOfStdTxMsgLPRemoveValue(value) {
    let isInstance = true;
    isInstance = isInstance && "pool_id" in value;
    isInstance = isInstance && "sender" in value;
    isInstance = isInstance && "share_in_amount" in value;
    isInstance = isInstance && "token_out_mins" in value;
    return isInstance;
}
exports.instanceOfStdTxMsgLPRemoveValue = instanceOfStdTxMsgLPRemoveValue;
function StdTxMsgLPRemoveValueFromJSON(json) {
    return StdTxMsgLPRemoveValueFromJSONTyped(json, false);
}
exports.StdTxMsgLPRemoveValueFromJSON = StdTxMsgLPRemoveValueFromJSON;
function StdTxMsgLPRemoveValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'pool_id': json['pool_id'],
        'sender': json['sender'],
        'share_in_amount': json['share_in_amount'],
        'token_out_mins': (json['token_out_mins'].map(TypesCosmosCoin_1.TypesCosmosCoinFromJSON)),
    };
}
exports.StdTxMsgLPRemoveValueFromJSONTyped = StdTxMsgLPRemoveValueFromJSONTyped;
function StdTxMsgLPRemoveValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'pool_id': value.pool_id,
        'sender': value.sender,
        'share_in_amount': value.share_in_amount,
        'token_out_mins': (value.token_out_mins.map(TypesCosmosCoin_1.TypesCosmosCoinToJSON)),
    };
}
exports.StdTxMsgLPRemoveValueToJSON = StdTxMsgLPRemoveValueToJSON;
