// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxWithdrawDelegationRewardValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxWithdrawDelegationRewardValueToJSON = exports.StdTxWithdrawDelegationRewardValueFromJSONTyped = exports.StdTxWithdrawDelegationRewardValueFromJSON = exports.instanceOfStdTxWithdrawDelegationRewardValue = void 0;
/**
 * Check if a given object implements the StdTxWithdrawDelegationRewardValue interface.
 */
function instanceOfStdTxWithdrawDelegationRewardValue(value) {
    let isInstance = true;
    isInstance = isInstance && "delegator_address" in value;
    isInstance = isInstance && "validator_address" in value;
    return isInstance;
}
exports.instanceOfStdTxWithdrawDelegationRewardValue = instanceOfStdTxWithdrawDelegationRewardValue;
function StdTxWithdrawDelegationRewardValueFromJSON(json) {
    return StdTxWithdrawDelegationRewardValueFromJSONTyped(json, false);
}
exports.StdTxWithdrawDelegationRewardValueFromJSON = StdTxWithdrawDelegationRewardValueFromJSON;
function StdTxWithdrawDelegationRewardValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'delegator_address': json['delegator_address'],
        'validator_address': json['validator_address'],
    };
}
exports.StdTxWithdrawDelegationRewardValueFromJSONTyped = StdTxWithdrawDelegationRewardValueFromJSONTyped;
function StdTxWithdrawDelegationRewardValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'delegator_address': value.delegator_address,
        'validator_address': value.validator_address,
    };
}
exports.StdTxWithdrawDelegationRewardValueToJSON = StdTxWithdrawDelegationRewardValueToJSON;
