// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxTransfer.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxTransferToJSON = exports.StdTxTransferFromJSONTyped = exports.StdTxTransferFromJSON = exports.instanceOfStdTxTransfer = exports.StdTxTransferTypeEnum = void 0;
const StdTxTransferValue_1 = require("./StdTxTransferValue");
/**
 * @export
 */
exports.StdTxTransferTypeEnum = {
    thorchain_MsgSend: 'thorchain/MsgSend'
};
/**
 * Check if a given object implements the StdTxTransfer interface.
 */
function instanceOfStdTxTransfer(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxTransfer = instanceOfStdTxTransfer;
function StdTxTransferFromJSON(json) {
    return StdTxTransferFromJSONTyped(json, false);
}
exports.StdTxTransferFromJSON = StdTxTransferFromJSON;
function StdTxTransferFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxTransferValue_1.StdTxTransferValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxTransferFromJSONTyped = StdTxTransferFromJSONTyped;
function StdTxTransferToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxTransferValue_1.StdTxTransferValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxTransferToJSON = StdTxTransferToJSON;
