// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/StdTxMsgSwap.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdTxMsgSwapToJSON = exports.StdTxMsgSwapFromJSONTyped = exports.StdTxMsgSwapFromJSON = exports.instanceOfStdTxMsgSwap = exports.StdTxMsgSwapTypeEnum = void 0;
const StdTxMsgSwapValue_1 = require("./StdTxMsgSwapValue");
/**
 * @export
 */
exports.StdTxMsgSwapTypeEnum = {
    osmosis_gamm_swap_exact_amount_in: 'osmosis/gamm/swap-exact-amount-in'
};
/**
 * Check if a given object implements the StdTxMsgSwap interface.
 */
function instanceOfStdTxMsgSwap(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfStdTxMsgSwap = instanceOfStdTxMsgSwap;
function StdTxMsgSwapFromJSON(json) {
    return StdTxMsgSwapFromJSONTyped(json, false);
}
exports.StdTxMsgSwapFromJSON = StdTxMsgSwapFromJSON;
function StdTxMsgSwapFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, StdTxMsgSwapValue_1.StdTxMsgSwapValueFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.StdTxMsgSwapFromJSONTyped = StdTxMsgSwapFromJSONTyped;
function StdTxMsgSwapToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, StdTxMsgSwapValue_1.StdTxMsgSwapValueToJSON)(value.value),
        'type': value.type,
    };
}
exports.StdTxMsgSwapToJSON = StdTxMsgSwapToJSON;
