// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/ThorchainSignAminoTransferRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThorchainSignAminoTransferRequestToJSON = exports.ThorchainSignAminoTransferRequestFromJSONTyped = exports.ThorchainSignAminoTransferRequestFromJSON = exports.instanceOfThorchainSignAminoTransferRequest = void 0;
const TypesThorchainSignDocTransfer_1 = require("./TypesThorchainSignDocTransfer");
/**
 * Check if a given object implements the ThorchainSignAminoTransferRequest interface.
 */
function instanceOfThorchainSignAminoTransferRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfThorchainSignAminoTransferRequest = instanceOfThorchainSignAminoTransferRequest;
function ThorchainSignAminoTransferRequestFromJSON(json) {
    return ThorchainSignAminoTransferRequestFromJSONTyped(json, false);
}
exports.ThorchainSignAminoTransferRequestFromJSON = ThorchainSignAminoTransferRequestFromJSON;
function ThorchainSignAminoTransferRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesThorchainSignDocTransfer_1.TypesThorchainSignDocTransferFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.ThorchainSignAminoTransferRequestFromJSONTyped = ThorchainSignAminoTransferRequestFromJSONTyped;
function ThorchainSignAminoTransferRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesThorchainSignDocTransfer_1.TypesThorchainSignDocTransferToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.ThorchainSignAminoTransferRequestToJSON = ThorchainSignAminoTransferRequestToJSON;
