// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/ThorchainSignAminoDepositRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThorchainSignAminoDepositRequestToJSON = exports.ThorchainSignAminoDepositRequestFromJSONTyped = exports.ThorchainSignAminoDepositRequestFromJSON = exports.instanceOfThorchainSignAminoDepositRequest = void 0;
const TypesThorchainSignDocDeposit_1 = require("./TypesThorchainSignDocDeposit");
/**
 * Check if a given object implements the ThorchainSignAminoDepositRequest interface.
 */
function instanceOfThorchainSignAminoDepositRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "signDoc" in value;
    isInstance = isInstance && "signerAddress" in value;
    return isInstance;
}
exports.instanceOfThorchainSignAminoDepositRequest = instanceOfThorchainSignAminoDepositRequest;
function ThorchainSignAminoDepositRequestFromJSON(json) {
    return ThorchainSignAminoDepositRequestFromJSONTyped(json, false);
}
exports.ThorchainSignAminoDepositRequestFromJSON = ThorchainSignAminoDepositRequestFromJSON;
function ThorchainSignAminoDepositRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'signDoc': (0, TypesThorchainSignDocDeposit_1.TypesThorchainSignDocDepositFromJSON)(json['signDoc']),
        'signerAddress': json['signerAddress'],
    };
}
exports.ThorchainSignAminoDepositRequestFromJSONTyped = ThorchainSignAminoDepositRequestFromJSONTyped;
function ThorchainSignAminoDepositRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'signDoc': (0, TypesThorchainSignDocDeposit_1.TypesThorchainSignDocDepositToJSON)(value.signDoc),
        'signerAddress': value.signerAddress,
    };
}
exports.ThorchainSignAminoDepositRequestToJSON = ThorchainSignAminoDepositRequestToJSON;
