// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/ThorchainSignAminoTransfer200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThorchainSignAminoTransfer200ResponseToJSON = exports.ThorchainSignAminoTransfer200ResponseFromJSONTyped = exports.ThorchainSignAminoTransfer200ResponseFromJSON = exports.instanceOfThorchainSignAminoTransfer200Response = void 0;
const TypesHexSecp256k1Signature_1 = require("./TypesHexSecp256k1Signature");
/**
 * Check if a given object implements the ThorchainSignAminoTransfer200Response interface.
 */
function instanceOfThorchainSignAminoTransfer200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "serialized" in value;
    isInstance = isInstance && "signature" in value;
    return isInstance;
}
exports.instanceOfThorchainSignAminoTransfer200Response = instanceOfThorchainSignAminoTransfer200Response;
function ThorchainSignAminoTransfer200ResponseFromJSON(json) {
    return ThorchainSignAminoTransfer200ResponseFromJSONTyped(json, false);
}
exports.ThorchainSignAminoTransfer200ResponseFromJSON = ThorchainSignAminoTransfer200ResponseFromJSON;
function ThorchainSignAminoTransfer200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'serialized': json['serialized'],
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureFromJSON)(json['signature']),
    };
}
exports.ThorchainSignAminoTransfer200ResponseFromJSONTyped = ThorchainSignAminoTransfer200ResponseFromJSONTyped;
function ThorchainSignAminoTransfer200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'serialized': value.serialized,
        'signature': (0, TypesHexSecp256k1Signature_1.TypesHexSecp256k1SignatureToJSON)(value.signature),
    };
}
exports.ThorchainSignAminoTransfer200ResponseToJSON = ThorchainSignAminoTransfer200ResponseToJSON;
