// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesBnbCoin.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesBnbCoinToJSON = exports.TypesBnbCoinFromJSONTyped = exports.TypesBnbCoinFromJSON = exports.instanceOfTypesBnbCoin = void 0;
const TypesNumericU64_1 = require("./TypesNumericU64");
/**
 * Check if a given object implements the TypesBnbCoin interface.
 */
function instanceOfTypesBnbCoin(value) {
    let isInstance = true;
    isInstance = isInstance && "denom" in value;
    isInstance = isInstance && "amount" in value;
    return isInstance;
}
exports.instanceOfTypesBnbCoin = instanceOfTypesBnbCoin;
function TypesBnbCoinFromJSON(json) {
    return TypesBnbCoinFromJSONTyped(json, false);
}
exports.TypesBnbCoinFromJSON = TypesBnbCoinFromJSON;
function TypesBnbCoinFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'denom': json['denom'],
        'amount': (0, TypesNumericU64_1.TypesNumericU64FromJSON)(json['amount']),
    };
}
exports.TypesBnbCoinFromJSONTyped = TypesBnbCoinFromJSONTyped;
function TypesBnbCoinToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'denom': value.denom,
        'amount': (0, TypesNumericU64_1.TypesNumericU64ToJSON)(value.amount),
    };
}
exports.TypesBnbCoinToJSON = TypesBnbCoinToJSON;
