// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesFeaturesMajorVersion.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesFeaturesMajorVersionToJSON = exports.TypesFeaturesMajorVersionFromJSONTyped = exports.TypesFeaturesMajorVersionFromJSON = exports.instanceOfTypesFeaturesMajorVersion = void 0;
/**
 * Check if a given object implements the TypesFeaturesMajorVersion interface.
 */
function instanceOfTypesFeaturesMajorVersion(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesFeaturesMajorVersion = instanceOfTypesFeaturesMajorVersion;
function TypesFeaturesMajorVersionFromJSON(json) {
    return TypesFeaturesMajorVersionFromJSONTyped(json, false);
}
exports.TypesFeaturesMajorVersionFromJSON = TypesFeaturesMajorVersionFromJSON;
function TypesFeaturesMajorVersionFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesFeaturesMajorVersionFromJSONTyped = TypesFeaturesMajorVersionFromJSONTyped;
function TypesFeaturesMajorVersionToJSON(value) {
    return value;
}
exports.TypesFeaturesMajorVersionToJSON = TypesFeaturesMajorVersionToJSON;
