// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesFeaturesMinorVersion.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesFeaturesMinorVersionToJSON = exports.TypesFeaturesMinorVersionFromJSONTyped = exports.TypesFeaturesMinorVersionFromJSON = exports.instanceOfTypesFeaturesMinorVersion = void 0;
/**
 * Check if a given object implements the TypesFeaturesMinorVersion interface.
 */
function instanceOfTypesFeaturesMinorVersion(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesFeaturesMinorVersion = instanceOfTypesFeaturesMinorVersion;
function TypesFeaturesMinorVersionFromJSON(json) {
    return TypesFeaturesMinorVersionFromJSONTyped(json, false);
}
exports.TypesFeaturesMinorVersionFromJSON = TypesFeaturesMinorVersionFromJSON;
function TypesFeaturesMinorVersionFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesFeaturesMinorVersionFromJSONTyped = TypesFeaturesMinorVersionFromJSONTyped;
function TypesFeaturesMinorVersionToJSON(value) {
    return value;
}
exports.TypesFeaturesMinorVersionToJSON = TypesFeaturesMinorVersionToJSON;
