// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesBnbMsg.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesBnbMsgToJSON = exports.TypesBnbMsgFromJSONTyped = exports.TypesBnbMsgFromJSON = exports.instanceOfTypesBnbMsg = void 0;
const TypesBnbMsgInputOutput_1 = require("./TypesBnbMsgInputOutput");
/**
 * Check if a given object implements the TypesBnbMsg interface.
 */
function instanceOfTypesBnbMsg(value) {
    let isInstance = true;
    isInstance = isInstance && "outputs" in value;
    isInstance = isInstance && "inputs" in value;
    return isInstance;
}
exports.instanceOfTypesBnbMsg = instanceOfTypesBnbMsg;
function TypesBnbMsgFromJSON(json) {
    return TypesBnbMsgFromJSONTyped(json, false);
}
exports.TypesBnbMsgFromJSON = TypesBnbMsgFromJSON;
function TypesBnbMsgFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'outputs': (json['outputs'].map(TypesBnbMsgInputOutput_1.TypesBnbMsgInputOutputFromJSON)),
        'inputs': (json['inputs'].map(TypesBnbMsgInputOutput_1.TypesBnbMsgInputOutputFromJSON)),
    };
}
exports.TypesBnbMsgFromJSONTyped = TypesBnbMsgFromJSONTyped;
function TypesBnbMsgToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'outputs': (value.outputs.map(TypesBnbMsgInputOutput_1.TypesBnbMsgInputOutputToJSON)),
        'inputs': (value.inputs.map(TypesBnbMsgInputOutput_1.TypesBnbMsgInputOutputToJSON)),
    };
}
exports.TypesBnbMsgToJSON = TypesBnbMsgToJSON;
