// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesFeaturesBootloaderHash.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesFeaturesBootloaderHashToJSON = exports.TypesFeaturesBootloaderHashFromJSONTyped = exports.TypesFeaturesBootloaderHashFromJSON = exports.instanceOfTypesFeaturesBootloaderHash = void 0;
/**
 * Check if a given object implements the TypesFeaturesBootloaderHash interface.
 */
function instanceOfTypesFeaturesBootloaderHash(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesFeaturesBootloaderHash = instanceOfTypesFeaturesBootloaderHash;
function TypesFeaturesBootloaderHashFromJSON(json) {
    return TypesFeaturesBootloaderHashFromJSONTyped(json, false);
}
exports.TypesFeaturesBootloaderHashFromJSON = TypesFeaturesBootloaderHashFromJSON;
function TypesFeaturesBootloaderHashFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesFeaturesBootloaderHashFromJSONTyped = TypesFeaturesBootloaderHashFromJSONTyped;
function TypesFeaturesBootloaderHashToJSON(value) {
    return value;
}
exports.TypesFeaturesBootloaderHashToJSON = TypesFeaturesBootloaderHashToJSON;
