// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesCosmosMessagesIbcGoTransferHeight.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesCosmosMessagesIbcGoTransferHeightToJSON = exports.TypesCosmosMessagesIbcGoTransferHeightFromJSONTyped = exports.TypesCosmosMessagesIbcGoTransferHeightFromJSON = exports.instanceOfTypesCosmosMessagesIbcGoTransferHeight = void 0;
const TypesNumericU64_1 = require("./TypesNumericU64");
/**
 * Check if a given object implements the TypesCosmosMessagesIbcGoTransferHeight interface.
 */
function instanceOfTypesCosmosMessagesIbcGoTransferHeight(value) {
    let isInstance = true;
    isInstance = isInstance && "revision_height" in value;
    isInstance = isInstance && "revision_number" in value;
    return isInstance;
}
exports.instanceOfTypesCosmosMessagesIbcGoTransferHeight = instanceOfTypesCosmosMessagesIbcGoTransferHeight;
function TypesCosmosMessagesIbcGoTransferHeightFromJSON(json) {
    return TypesCosmosMessagesIbcGoTransferHeightFromJSONTyped(json, false);
}
exports.TypesCosmosMessagesIbcGoTransferHeightFromJSON = TypesCosmosMessagesIbcGoTransferHeightFromJSON;
function TypesCosmosMessagesIbcGoTransferHeightFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'revision_height': (0, TypesNumericU64_1.TypesNumericU64FromJSON)(json['revision_height']),
        'revision_number': (0, TypesNumericU64_1.TypesNumericU64FromJSON)(json['revision_number']),
    };
}
exports.TypesCosmosMessagesIbcGoTransferHeightFromJSONTyped = TypesCosmosMessagesIbcGoTransferHeightFromJSONTyped;
function TypesCosmosMessagesIbcGoTransferHeightToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'revision_height': (0, TypesNumericU64_1.TypesNumericU64ToJSON)(value.revision_height),
        'revision_number': (0, TypesNumericU64_1.TypesNumericU64ToJSON)(value.revision_number),
    };
}
exports.TypesCosmosMessagesIbcGoTransferHeightToJSON = TypesCosmosMessagesIbcGoTransferHeightToJSON;
