// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesFeaturesFirmwareHash.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesFeaturesFirmwareHashToJSON = exports.TypesFeaturesFirmwareHashFromJSONTyped = exports.TypesFeaturesFirmwareHashFromJSON = exports.instanceOfTypesFeaturesFirmwareHash = void 0;
/**
 * Check if a given object implements the TypesFeaturesFirmwareHash interface.
 */
function instanceOfTypesFeaturesFirmwareHash(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesFeaturesFirmwareHash = instanceOfTypesFeaturesFirmwareHash;
function TypesFeaturesFirmwareHashFromJSON(json) {
    return TypesFeaturesFirmwareHashFromJSONTyped(json, false);
}
exports.TypesFeaturesFirmwareHashFromJSON = TypesFeaturesFirmwareHashFromJSON;
function TypesFeaturesFirmwareHashFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesFeaturesFirmwareHashFromJSONTyped = TypesFeaturesFirmwareHashFromJSONTyped;
function TypesFeaturesFirmwareHashToJSON(value) {
    return value;
}
exports.TypesFeaturesFirmwareHashToJSON = TypesFeaturesFirmwareHashToJSON;
