// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesCosmosAminoSignDocCosmosSdkSendAllOf.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfToJSON = exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSONTyped = exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSON = exports.instanceOfTypesCosmosAminoSignDocCosmosSdkSendAllOf = exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfTypeEnum = void 0;
const TypesCosmosMessagesCosmosSdkSend_1 = require("./TypesCosmosMessagesCosmosSdkSend");
/**
 * @export
 */
exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfTypeEnum = {
    cosmos_sdk_MsgSend: 'cosmos-sdk/MsgSend'
};
/**
 * Check if a given object implements the TypesCosmosAminoSignDocCosmosSdkSendAllOf interface.
 */
function instanceOfTypesCosmosAminoSignDocCosmosSdkSendAllOf(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfTypesCosmosAminoSignDocCosmosSdkSendAllOf = instanceOfTypesCosmosAminoSignDocCosmosSdkSendAllOf;
function TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSON(json) {
    return TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSONTyped(json, false);
}
exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSON = TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSON;
function TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, TypesCosmosMessagesCosmosSdkSend_1.TypesCosmosMessagesCosmosSdkSendFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSONTyped = TypesCosmosAminoSignDocCosmosSdkSendAllOfFromJSONTyped;
function TypesCosmosAminoSignDocCosmosSdkSendAllOfToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, TypesCosmosMessagesCosmosSdkSend_1.TypesCosmosMessagesCosmosSdkSendToJSON)(value.value),
        'type': value.type,
    };
}
exports.TypesCosmosAminoSignDocCosmosSdkSendAllOfToJSON = TypesCosmosAminoSignDocCosmosSdkSendAllOfToJSON;
