// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesBnbSignDoc.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesBnbSignDocToJSON = exports.TypesBnbSignDocFromJSONTyped = exports.TypesBnbSignDocFromJSON = exports.instanceOfTypesBnbSignDoc = void 0;
const TypesBnbMsg_1 = require("./TypesBnbMsg");
/**
 * Check if a given object implements the TypesBnbSignDoc interface.
 */
function instanceOfTypesBnbSignDoc(value) {
    let isInstance = true;
    isInstance = isInstance && "msgs" in value;
    isInstance = isInstance && "memo" in value;
    isInstance = isInstance && "source" in value;
    isInstance = isInstance && "sequence" in value;
    isInstance = isInstance && "chain_id" in value;
    isInstance = isInstance && "account_number" in value;
    return isInstance;
}
exports.instanceOfTypesBnbSignDoc = instanceOfTypesBnbSignDoc;
function TypesBnbSignDocFromJSON(json) {
    return TypesBnbSignDocFromJSONTyped(json, false);
}
exports.TypesBnbSignDocFromJSON = TypesBnbSignDocFromJSON;
function TypesBnbSignDocFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'msgs': (json['msgs'].map(TypesBnbMsg_1.TypesBnbMsgFromJSON)),
        'memo': json['memo'],
        'source': json['source'],
        'sequence': json['sequence'],
        'chain_id': json['chain_id'],
        'account_number': json['account_number'],
    };
}
exports.TypesBnbSignDocFromJSONTyped = TypesBnbSignDocFromJSONTyped;
function TypesBnbSignDocToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'msgs': (value.msgs.map(TypesBnbMsg_1.TypesBnbMsgToJSON)),
        'memo': value.memo,
        'source': value.source,
        'sequence': value.sequence,
        'chain_id': value.chain_id,
        'account_number': value.account_number,
    };
}
exports.TypesBnbSignDocToJSON = TypesBnbSignDocToJSON;
