// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesThorchainSignDocDeposit.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesThorchainSignDocDepositToJSON = exports.TypesThorchainSignDocDepositFromJSONTyped = exports.TypesThorchainSignDocDepositFromJSON = exports.instanceOfTypesThorchainSignDocDeposit = void 0;
const StdTxDeposit_1 = require("./StdTxDeposit");
const TypesThorchainSignDocDepositFee_1 = require("./TypesThorchainSignDocDepositFee");
/**
 * Check if a given object implements the TypesThorchainSignDocDeposit interface.
 */
function instanceOfTypesThorchainSignDocDeposit(value) {
    let isInstance = true;
    isInstance = isInstance && "fee" in value;
    isInstance = isInstance && "msgs" in value;
    isInstance = isInstance && "memo" in value;
    isInstance = isInstance && "sequence" in value;
    isInstance = isInstance && "chain_id" in value;
    isInstance = isInstance && "account_number" in value;
    return isInstance;
}
exports.instanceOfTypesThorchainSignDocDeposit = instanceOfTypesThorchainSignDocDeposit;
function TypesThorchainSignDocDepositFromJSON(json) {
    return TypesThorchainSignDocDepositFromJSONTyped(json, false);
}
exports.TypesThorchainSignDocDepositFromJSON = TypesThorchainSignDocDepositFromJSON;
function TypesThorchainSignDocDepositFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fee': (0, TypesThorchainSignDocDepositFee_1.TypesThorchainSignDocDepositFeeFromJSON)(json['fee']),
        'msgs': (json['msgs'].map(StdTxDeposit_1.StdTxDepositFromJSON)),
        'memo': json['memo'],
        'sequence': json['sequence'],
        'chain_id': json['chain_id'],
        'account_number': json['account_number'],
    };
}
exports.TypesThorchainSignDocDepositFromJSONTyped = TypesThorchainSignDocDepositFromJSONTyped;
function TypesThorchainSignDocDepositToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fee': (0, TypesThorchainSignDocDepositFee_1.TypesThorchainSignDocDepositFeeToJSON)(value.fee),
        'msgs': (value.msgs.map(StdTxDeposit_1.StdTxDepositToJSON)),
        'memo': value.memo,
        'sequence': value.sequence,
        'chain_id': value.chain_id,
        'account_number': value.account_number,
    };
}
exports.TypesThorchainSignDocDepositToJSON = TypesThorchainSignDocDepositToJSON;
