// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesCosmosAminoAny.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesCosmosAminoAnyToJSON = exports.TypesCosmosAminoAnyFromJSONTyped = exports.TypesCosmosAminoAnyFromJSON = exports.instanceOfTypesCosmosAminoAny = void 0;
/**
 * Check if a given object implements the TypesCosmosAminoAny interface.
 */
function instanceOfTypesCosmosAminoAny(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfTypesCosmosAminoAny = instanceOfTypesCosmosAminoAny;
function TypesCosmosAminoAnyFromJSON(json) {
    return TypesCosmosAminoAnyFromJSONTyped(json, false);
}
exports.TypesCosmosAminoAnyFromJSON = TypesCosmosAminoAnyFromJSON;
function TypesCosmosAminoAnyFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': json['value'],
        'type': json['type'],
    };
}
exports.TypesCosmosAminoAnyFromJSONTyped = TypesCosmosAminoAnyFromJSONTyped;
function TypesCosmosAminoAnyToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': value.value,
        'type': value.type,
    };
}
exports.TypesCosmosAminoAnyToJSON = TypesCosmosAminoAnyToJSON;
