// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesCosmosMessagesCosmosSdkSend.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesCosmosMessagesCosmosSdkSendToJSON = exports.TypesCosmosMessagesCosmosSdkSendFromJSONTyped = exports.TypesCosmosMessagesCosmosSdkSendFromJSON = exports.instanceOfTypesCosmosMessagesCosmosSdkSend = void 0;
const TypesCosmosCoin_1 = require("./TypesCosmosCoin");
/**
 * Check if a given object implements the TypesCosmosMessagesCosmosSdkSend interface.
 */
function instanceOfTypesCosmosMessagesCosmosSdkSend(value) {
    let isInstance = true;
    isInstance = isInstance && "amount" in value;
    isInstance = isInstance && "to_address" in value;
    isInstance = isInstance && "from_address" in value;
    return isInstance;
}
exports.instanceOfTypesCosmosMessagesCosmosSdkSend = instanceOfTypesCosmosMessagesCosmosSdkSend;
function TypesCosmosMessagesCosmosSdkSendFromJSON(json) {
    return TypesCosmosMessagesCosmosSdkSendFromJSONTyped(json, false);
}
exports.TypesCosmosMessagesCosmosSdkSendFromJSON = TypesCosmosMessagesCosmosSdkSendFromJSON;
function TypesCosmosMessagesCosmosSdkSendFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'amount': (json['amount'].map(TypesCosmosCoin_1.TypesCosmosCoinFromJSON)),
        'to_address': json['to_address'],
        'from_address': json['from_address'],
    };
}
exports.TypesCosmosMessagesCosmosSdkSendFromJSONTyped = TypesCosmosMessagesCosmosSdkSendFromJSONTyped;
function TypesCosmosMessagesCosmosSdkSendToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'amount': (value.amount.map(TypesCosmosCoin_1.TypesCosmosCoinToJSON)),
        'to_address': value.to_address,
        'from_address': value.from_address,
    };
}
exports.TypesCosmosMessagesCosmosSdkSendToJSON = TypesCosmosMessagesCosmosSdkSendToJSON;
