// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesCosmosCoin.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesCosmosCoinToJSON = exports.TypesCosmosCoinFromJSONTyped = exports.TypesCosmosCoinFromJSON = exports.instanceOfTypesCosmosCoin = void 0;
/**
 * Check if a given object implements the TypesCosmosCoin interface.
 */
function instanceOfTypesCosmosCoin(value) {
    let isInstance = true;
    isInstance = isInstance && "denom" in value;
    isInstance = isInstance && "amount" in value;
    return isInstance;
}
exports.instanceOfTypesCosmosCoin = instanceOfTypesCosmosCoin;
function TypesCosmosCoinFromJSON(json) {
    return TypesCosmosCoinFromJSONTyped(json, false);
}
exports.TypesCosmosCoinFromJSON = TypesCosmosCoinFromJSON;
function TypesCosmosCoinFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'denom': json['denom'],
        'amount': json['amount'],
    };
}
exports.TypesCosmosCoinFromJSONTyped = TypesCosmosCoinFromJSONTyped;
function TypesCosmosCoinToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'denom': value.denom,
        'amount': value.amount,
    };
}
exports.TypesCosmosCoinToJSON = TypesCosmosCoinToJSON;
