// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesBnbMsgInputOutput.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesBnbMsgInputOutputToJSON = exports.TypesBnbMsgInputOutputFromJSONTyped = exports.TypesBnbMsgInputOutputFromJSON = exports.instanceOfTypesBnbMsgInputOutput = void 0;
const TypesBnbAddress_1 = require("./TypesBnbAddress");
const TypesBnbCoin_1 = require("./TypesBnbCoin");
/**
 * Check if a given object implements the TypesBnbMsgInputOutput interface.
 */
function instanceOfTypesBnbMsgInputOutput(value) {
    let isInstance = true;
    isInstance = isInstance && "coins" in value;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfTypesBnbMsgInputOutput = instanceOfTypesBnbMsgInputOutput;
function TypesBnbMsgInputOutputFromJSON(json) {
    return TypesBnbMsgInputOutputFromJSONTyped(json, false);
}
exports.TypesBnbMsgInputOutputFromJSON = TypesBnbMsgInputOutputFromJSON;
function TypesBnbMsgInputOutputFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'coins': (json['coins'].map(TypesBnbCoin_1.TypesBnbCoinFromJSON)),
        'address': (0, TypesBnbAddress_1.TypesBnbAddressFromJSON)(json['address']),
    };
}
exports.TypesBnbMsgInputOutputFromJSONTyped = TypesBnbMsgInputOutputFromJSONTyped;
function TypesBnbMsgInputOutputToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'coins': (value.coins.map(TypesBnbCoin_1.TypesBnbCoinToJSON)),
        'address': (0, TypesBnbAddress_1.TypesBnbAddressToJSON)(value.address),
    };
}
exports.TypesBnbMsgInputOutputToJSON = TypesBnbMsgInputOutputToJSON;
