// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesNumericU64.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesNumericU64ToJSON = exports.TypesNumericU64FromJSONTyped = exports.TypesNumericU64FromJSON = exports.instanceOfTypesNumericU64 = void 0;
/**
 * Check if a given object implements the TypesNumericU64 interface.
 */
function instanceOfTypesNumericU64(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesNumericU64 = instanceOfTypesNumericU64;
function TypesNumericU64FromJSON(json) {
    return TypesNumericU64FromJSONTyped(json, false);
}
exports.TypesNumericU64FromJSON = TypesNumericU64FromJSON;
function TypesNumericU64FromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesNumericU64FromJSONTyped = TypesNumericU64FromJSONTyped;
function TypesNumericU64ToJSON(value) {
    return value;
}
exports.TypesNumericU64ToJSON = TypesNumericU64ToJSON;
