// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesThorchainSignDocDepositFee.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesThorchainSignDocDepositFeeToJSON = exports.TypesThorchainSignDocDepositFeeFromJSONTyped = exports.TypesThorchainSignDocDepositFeeFromJSON = exports.instanceOfTypesThorchainSignDocDepositFee = void 0;
/**
 * Check if a given object implements the TypesThorchainSignDocDepositFee interface.
 */
function instanceOfTypesThorchainSignDocDepositFee(value) {
    let isInstance = true;
    isInstance = isInstance && "gas" in value;
    isInstance = isInstance && "amount" in value;
    return isInstance;
}
exports.instanceOfTypesThorchainSignDocDepositFee = instanceOfTypesThorchainSignDocDepositFee;
function TypesThorchainSignDocDepositFeeFromJSON(json) {
    return TypesThorchainSignDocDepositFeeFromJSONTyped(json, false);
}
exports.TypesThorchainSignDocDepositFeeFromJSON = TypesThorchainSignDocDepositFeeFromJSON;
function TypesThorchainSignDocDepositFeeFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'gas': json['gas'],
        'amount': json['amount'],
    };
}
exports.TypesThorchainSignDocDepositFeeFromJSONTyped = TypesThorchainSignDocDepositFeeFromJSONTyped;
function TypesThorchainSignDocDepositFeeToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'gas': value.gas,
        'amount': value.amount,
    };
}
exports.TypesThorchainSignDocDepositFeeToJSON = TypesThorchainSignDocDepositFeeToJSON;
