// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesNumericU32.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesNumericU32ToJSON = exports.TypesNumericU32FromJSONTyped = exports.TypesNumericU32FromJSON = exports.instanceOfTypesNumericU32 = void 0;
/**
 * Check if a given object implements the TypesNumericU32 interface.
 */
function instanceOfTypesNumericU32(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesNumericU32 = instanceOfTypesNumericU32;
function TypesNumericU32FromJSON(json) {
    return TypesNumericU32FromJSONTyped(json, false);
}
exports.TypesNumericU32FromJSON = TypesNumericU32FromJSON;
function TypesNumericU32FromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesNumericU32FromJSONTyped = TypesNumericU32FromJSONTyped;
function TypesNumericU32ToJSON(value) {
    return value;
}
exports.TypesNumericU32ToJSON = TypesNumericU32ToJSON;
