// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesCosmosAminoSignDocCosmosSdkSend.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesCosmosAminoSignDocCosmosSdkSendToJSON = exports.TypesCosmosAminoSignDocCosmosSdkSendFromJSONTyped = exports.TypesCosmosAminoSignDocCosmosSdkSendFromJSON = exports.instanceOfTypesCosmosAminoSignDocCosmosSdkSend = exports.TypesCosmosAminoSignDocCosmosSdkSendTypeEnum = void 0;
const TypesCosmosMessagesCosmosSdkSend_1 = require("./TypesCosmosMessagesCosmosSdkSend");
/**
 * @export
 */
exports.TypesCosmosAminoSignDocCosmosSdkSendTypeEnum = {
    cosmos_sdk_MsgSend: 'cosmos-sdk/MsgSend'
};
/**
 * Check if a given object implements the TypesCosmosAminoSignDocCosmosSdkSend interface.
 */
function instanceOfTypesCosmosAminoSignDocCosmosSdkSend(value) {
    let isInstance = true;
    isInstance = isInstance && "value" in value;
    isInstance = isInstance && "type" in value;
    return isInstance;
}
exports.instanceOfTypesCosmosAminoSignDocCosmosSdkSend = instanceOfTypesCosmosAminoSignDocCosmosSdkSend;
function TypesCosmosAminoSignDocCosmosSdkSendFromJSON(json) {
    return TypesCosmosAminoSignDocCosmosSdkSendFromJSONTyped(json, false);
}
exports.TypesCosmosAminoSignDocCosmosSdkSendFromJSON = TypesCosmosAminoSignDocCosmosSdkSendFromJSON;
function TypesCosmosAminoSignDocCosmosSdkSendFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'value': (0, TypesCosmosMessagesCosmosSdkSend_1.TypesCosmosMessagesCosmosSdkSendFromJSON)(json['value']),
        'type': json['type'],
    };
}
exports.TypesCosmosAminoSignDocCosmosSdkSendFromJSONTyped = TypesCosmosAminoSignDocCosmosSdkSendFromJSONTyped;
function TypesCosmosAminoSignDocCosmosSdkSendToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'value': (0, TypesCosmosMessagesCosmosSdkSend_1.TypesCosmosMessagesCosmosSdkSendToJSON)(value.value),
        'type': value.type,
    };
}
exports.TypesCosmosAminoSignDocCosmosSdkSendToJSON = TypesCosmosAminoSignDocCosmosSdkSendToJSON;
