// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesOsmosisSignDocLPAdd.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesOsmosisSignDocLPAddToJSON = exports.TypesOsmosisSignDocLPAddFromJSONTyped = exports.TypesOsmosisSignDocLPAddFromJSON = exports.instanceOfTypesOsmosisSignDocLPAdd = void 0;
const StdTxMsgLPAdd_1 = require("./StdTxMsgLPAdd");
const TypesCosmosAminoSignDocFee_1 = require("./TypesCosmosAminoSignDocFee");
/**
 * Check if a given object implements the TypesOsmosisSignDocLPAdd interface.
 */
function instanceOfTypesOsmosisSignDocLPAdd(value) {
    let isInstance = true;
    isInstance = isInstance && "fee" in value;
    isInstance = isInstance && "msgs" in value;
    isInstance = isInstance && "memo" in value;
    isInstance = isInstance && "sequence" in value;
    isInstance = isInstance && "chain_id" in value;
    isInstance = isInstance && "account_number" in value;
    return isInstance;
}
exports.instanceOfTypesOsmosisSignDocLPAdd = instanceOfTypesOsmosisSignDocLPAdd;
function TypesOsmosisSignDocLPAddFromJSON(json) {
    return TypesOsmosisSignDocLPAddFromJSONTyped(json, false);
}
exports.TypesOsmosisSignDocLPAddFromJSON = TypesOsmosisSignDocLPAddFromJSON;
function TypesOsmosisSignDocLPAddFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fee': (0, TypesCosmosAminoSignDocFee_1.TypesCosmosAminoSignDocFeeFromJSON)(json['fee']),
        'msgs': (json['msgs'].map(StdTxMsgLPAdd_1.StdTxMsgLPAddFromJSON)),
        'memo': json['memo'],
        'sequence': json['sequence'],
        'chain_id': json['chain_id'],
        'account_number': json['account_number'],
    };
}
exports.TypesOsmosisSignDocLPAddFromJSONTyped = TypesOsmosisSignDocLPAddFromJSONTyped;
function TypesOsmosisSignDocLPAddToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fee': (0, TypesCosmosAminoSignDocFee_1.TypesCosmosAminoSignDocFeeToJSON)(value.fee),
        'msgs': (value.msgs.map(StdTxMsgLPAdd_1.StdTxMsgLPAddToJSON)),
        'memo': value.memo,
        'sequence': value.sequence,
        'chain_id': value.chain_id,
        'account_number': value.account_number,
    };
}
exports.TypesOsmosisSignDocLPAddToJSON = TypesOsmosisSignDocLPAddToJSON;
