// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesBnbAddress.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesBnbAddressToJSON = exports.TypesBnbAddressFromJSONTyped = exports.TypesBnbAddressFromJSON = exports.instanceOfTypesBnbAddress = void 0;
/**
 * Check if a given object implements the TypesBnbAddress interface.
 */
function instanceOfTypesBnbAddress(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesBnbAddress = instanceOfTypesBnbAddress;
function TypesBnbAddressFromJSON(json) {
    return TypesBnbAddressFromJSONTyped(json, false);
}
exports.TypesBnbAddressFromJSON = TypesBnbAddressFromJSON;
function TypesBnbAddressFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesBnbAddressFromJSONTyped = TypesBnbAddressFromJSONTyped;
function TypesBnbAddressToJSON(value) {
    return value;
}
exports.TypesBnbAddressToJSON = TypesBnbAddressToJSON;
