// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesEthSignature.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesEthSignatureToJSON = exports.TypesEthSignatureFromJSONTyped = exports.TypesEthSignatureFromJSON = exports.instanceOfTypesEthSignature = void 0;
/**
 * Check if a given object implements the TypesEthSignature interface.
 */
function instanceOfTypesEthSignature(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesEthSignature = instanceOfTypesEthSignature;
function TypesEthSignatureFromJSON(json) {
    return TypesEthSignatureFromJSONTyped(json, false);
}
exports.TypesEthSignatureFromJSON = TypesEthSignatureFromJSON;
function TypesEthSignatureFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesEthSignatureFromJSONTyped = TypesEthSignatureFromJSONTyped;
function TypesEthSignatureToJSON(value) {
    return value;
}
exports.TypesEthSignatureToJSON = TypesEthSignatureToJSON;
