// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesFeaturesPatchVersion.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesFeaturesPatchVersionToJSON = exports.TypesFeaturesPatchVersionFromJSONTyped = exports.TypesFeaturesPatchVersionFromJSON = exports.instanceOfTypesFeaturesPatchVersion = void 0;
/**
 * Check if a given object implements the TypesFeaturesPatchVersion interface.
 */
function instanceOfTypesFeaturesPatchVersion(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesFeaturesPatchVersion = instanceOfTypesFeaturesPatchVersion;
function TypesFeaturesPatchVersionFromJSON(json) {
    return TypesFeaturesPatchVersionFromJSONTyped(json, false);
}
exports.TypesFeaturesPatchVersionFromJSON = TypesFeaturesPatchVersionFromJSON;
function TypesFeaturesPatchVersionFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesFeaturesPatchVersionFromJSONTyped = TypesFeaturesPatchVersionFromJSONTyped;
function TypesFeaturesPatchVersionToJSON(value) {
    return value;
}
exports.TypesFeaturesPatchVersionToJSON = TypesFeaturesPatchVersionToJSON;
