// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesHexSecp256k1Signature.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesHexSecp256k1SignatureToJSON = exports.TypesHexSecp256k1SignatureFromJSONTyped = exports.TypesHexSecp256k1SignatureFromJSON = exports.instanceOfTypesHexSecp256k1Signature = void 0;
/**
 * Check if a given object implements the TypesHexSecp256k1Signature interface.
 */
function instanceOfTypesHexSecp256k1Signature(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesHexSecp256k1Signature = instanceOfTypesHexSecp256k1Signature;
function TypesHexSecp256k1SignatureFromJSON(json) {
    return TypesHexSecp256k1SignatureFromJSONTyped(json, false);
}
exports.TypesHexSecp256k1SignatureFromJSON = TypesHexSecp256k1SignatureFromJSON;
function TypesHexSecp256k1SignatureFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesHexSecp256k1SignatureFromJSONTyped = TypesHexSecp256k1SignatureFromJSONTyped;
function TypesHexSecp256k1SignatureToJSON(value) {
    return value;
}
exports.TypesHexSecp256k1SignatureToJSON = TypesHexSecp256k1SignatureToJSON;
