// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesOsmosisSignDocSwap.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesOsmosisSignDocSwapToJSON = exports.TypesOsmosisSignDocSwapFromJSONTyped = exports.TypesOsmosisSignDocSwapFromJSON = exports.instanceOfTypesOsmosisSignDocSwap = void 0;
const StdTxMsgSwap_1 = require("./StdTxMsgSwap");
const TypesCosmosAminoSignDocFee_1 = require("./TypesCosmosAminoSignDocFee");
/**
 * Check if a given object implements the TypesOsmosisSignDocSwap interface.
 */
function instanceOfTypesOsmosisSignDocSwap(value) {
    let isInstance = true;
    isInstance = isInstance && "fee" in value;
    isInstance = isInstance && "msgs" in value;
    isInstance = isInstance && "memo" in value;
    isInstance = isInstance && "sequence" in value;
    isInstance = isInstance && "chain_id" in value;
    isInstance = isInstance && "account_number" in value;
    return isInstance;
}
exports.instanceOfTypesOsmosisSignDocSwap = instanceOfTypesOsmosisSignDocSwap;
function TypesOsmosisSignDocSwapFromJSON(json) {
    return TypesOsmosisSignDocSwapFromJSONTyped(json, false);
}
exports.TypesOsmosisSignDocSwapFromJSON = TypesOsmosisSignDocSwapFromJSON;
function TypesOsmosisSignDocSwapFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fee': (0, TypesCosmosAminoSignDocFee_1.TypesCosmosAminoSignDocFeeFromJSON)(json['fee']),
        'msgs': (json['msgs'].map(StdTxMsgSwap_1.StdTxMsgSwapFromJSON)),
        'memo': json['memo'],
        'sequence': json['sequence'],
        'chain_id': json['chain_id'],
        'account_number': json['account_number'],
    };
}
exports.TypesOsmosisSignDocSwapFromJSONTyped = TypesOsmosisSignDocSwapFromJSONTyped;
function TypesOsmosisSignDocSwapToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fee': (0, TypesCosmosAminoSignDocFee_1.TypesCosmosAminoSignDocFeeToJSON)(value.fee),
        'msgs': (value.msgs.map(StdTxMsgSwap_1.StdTxMsgSwapToJSON)),
        'memo': value.memo,
        'sequence': value.sequence,
        'chain_id': value.chain_id,
        'account_number': value.account_number,
    };
}
exports.TypesOsmosisSignDocSwapToJSON = TypesOsmosisSignDocSwapToJSON;
