// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesNumericUnsignedInteger.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesNumericUnsignedIntegerToJSON = exports.TypesNumericUnsignedIntegerFromJSONTyped = exports.TypesNumericUnsignedIntegerFromJSON = exports.instanceOfTypesNumericUnsignedInteger = void 0;
/**
 * Check if a given object implements the TypesNumericUnsignedInteger interface.
 */
function instanceOfTypesNumericUnsignedInteger(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesNumericUnsignedInteger = instanceOfTypesNumericUnsignedInteger;
function TypesNumericUnsignedIntegerFromJSON(json) {
    return TypesNumericUnsignedIntegerFromJSONTyped(json, false);
}
exports.TypesNumericUnsignedIntegerFromJSON = TypesNumericUnsignedIntegerFromJSON;
function TypesNumericUnsignedIntegerFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesNumericUnsignedIntegerFromJSONTyped = TypesNumericUnsignedIntegerFromJSONTyped;
function TypesNumericUnsignedIntegerToJSON(value) {
    return value;
}
exports.TypesNumericUnsignedIntegerToJSON = TypesNumericUnsignedIntegerToJSON;
