// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesCosmosAminoSignDocBeginReDelegate.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesCosmosAminoSignDocBeginReDelegateToJSON = exports.TypesCosmosAminoSignDocBeginReDelegateFromJSONTyped = exports.TypesCosmosAminoSignDocBeginReDelegateFromJSON = exports.instanceOfTypesCosmosAminoSignDocBeginReDelegate = void 0;
const StdTxBeginReDelegate_1 = require("./StdTxBeginReDelegate");
const TypesCosmosAminoSignDocFee_1 = require("./TypesCosmosAminoSignDocFee");
/**
 * Check if a given object implements the TypesCosmosAminoSignDocBeginReDelegate interface.
 */
function instanceOfTypesCosmosAminoSignDocBeginReDelegate(value) {
    let isInstance = true;
    isInstance = isInstance && "fee" in value;
    isInstance = isInstance && "msgs" in value;
    isInstance = isInstance && "memo" in value;
    isInstance = isInstance && "sequence" in value;
    isInstance = isInstance && "chain_id" in value;
    isInstance = isInstance && "account_number" in value;
    return isInstance;
}
exports.instanceOfTypesCosmosAminoSignDocBeginReDelegate = instanceOfTypesCosmosAminoSignDocBeginReDelegate;
function TypesCosmosAminoSignDocBeginReDelegateFromJSON(json) {
    return TypesCosmosAminoSignDocBeginReDelegateFromJSONTyped(json, false);
}
exports.TypesCosmosAminoSignDocBeginReDelegateFromJSON = TypesCosmosAminoSignDocBeginReDelegateFromJSON;
function TypesCosmosAminoSignDocBeginReDelegateFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fee': (0, TypesCosmosAminoSignDocFee_1.TypesCosmosAminoSignDocFeeFromJSON)(json['fee']),
        'msgs': (json['msgs'].map(StdTxBeginReDelegate_1.StdTxBeginReDelegateFromJSON)),
        'memo': json['memo'],
        'sequence': json['sequence'],
        'chain_id': json['chain_id'],
        'account_number': json['account_number'],
    };
}
exports.TypesCosmosAminoSignDocBeginReDelegateFromJSONTyped = TypesCosmosAminoSignDocBeginReDelegateFromJSONTyped;
function TypesCosmosAminoSignDocBeginReDelegateToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fee': (0, TypesCosmosAminoSignDocFee_1.TypesCosmosAminoSignDocFeeToJSON)(value.fee),
        'msgs': (value.msgs.map(StdTxBeginReDelegate_1.StdTxBeginReDelegateToJSON)),
        'memo': value.memo,
        'sequence': value.sequence,
        'chain_id': value.chain_id,
        'account_number': value.account_number,
    };
}
exports.TypesCosmosAminoSignDocBeginReDelegateToJSON = TypesCosmosAminoSignDocBeginReDelegateToJSON;
