// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesHexBytesLower.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesHexBytesLowerToJSON = exports.TypesHexBytesLowerFromJSONTyped = exports.TypesHexBytesLowerFromJSON = exports.instanceOfTypesHexBytesLower = void 0;
/**
 * Check if a given object implements the TypesHexBytesLower interface.
 */
function instanceOfTypesHexBytesLower(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesHexBytesLower = instanceOfTypesHexBytesLower;
function TypesHexBytesLowerFromJSON(json) {
    return TypesHexBytesLowerFromJSONTyped(json, false);
}
exports.TypesHexBytesLowerFromJSON = TypesHexBytesLowerFromJSON;
function TypesHexBytesLowerFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesHexBytesLowerFromJSONTyped = TypesHexBytesLowerFromJSONTyped;
function TypesHexBytesLowerToJSON(value) {
    return value;
}
exports.TypesHexBytesLowerToJSON = TypesHexBytesLowerToJSON;
