// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesPolicy.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesPolicyToJSON = exports.TypesPolicyFromJSONTyped = exports.TypesPolicyFromJSON = exports.instanceOfTypesPolicy = void 0;
/**
 * Check if a given object implements the TypesPolicy interface.
 */
function instanceOfTypesPolicy(value) {
    let isInstance = true;
    isInstance = isInstance && "enabled" in value;
    isInstance = isInstance && "policy_name" in value;
    return isInstance;
}
exports.instanceOfTypesPolicy = instanceOfTypesPolicy;
function TypesPolicyFromJSON(json) {
    return TypesPolicyFromJSONTyped(json, false);
}
exports.TypesPolicyFromJSON = TypesPolicyFromJSON;
function TypesPolicyFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'enabled': json['enabled'],
        'policy_name': json['policy_name'],
    };
}
exports.TypesPolicyFromJSONTyped = TypesPolicyFromJSONTyped;
function TypesPolicyToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'enabled': value.enabled,
        'policy_name': value.policy_name,
    };
}
exports.TypesPolicyToJSON = TypesPolicyToJSON;
