// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesFeaturesRevision.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesFeaturesRevisionToJSON = exports.TypesFeaturesRevisionFromJSONTyped = exports.TypesFeaturesRevisionFromJSON = exports.instanceOfTypesFeaturesRevision = void 0;
/**
 * Check if a given object implements the TypesFeaturesRevision interface.
 */
function instanceOfTypesFeaturesRevision(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfTypesFeaturesRevision = instanceOfTypesFeaturesRevision;
function TypesFeaturesRevisionFromJSON(json) {
    return TypesFeaturesRevisionFromJSONTyped(json, false);
}
exports.TypesFeaturesRevisionFromJSON = TypesFeaturesRevisionFromJSON;
function TypesFeaturesRevisionFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.TypesFeaturesRevisionFromJSONTyped = TypesFeaturesRevisionFromJSONTyped;
function TypesFeaturesRevisionToJSON(value) {
    return value;
}
exports.TypesFeaturesRevisionToJSON = TypesFeaturesRevisionToJSON;
