// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/TypesThorchainSignDocTransferFee.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesThorchainSignDocTransferFeeToJSON = exports.TypesThorchainSignDocTransferFeeFromJSONTyped = exports.TypesThorchainSignDocTransferFeeFromJSON = exports.instanceOfTypesThorchainSignDocTransferFee = void 0;
const StdTxTransferValueAmountInner_1 = require("./StdTxTransferValueAmountInner");
/**
 * Check if a given object implements the TypesThorchainSignDocTransferFee interface.
 */
function instanceOfTypesThorchainSignDocTransferFee(value) {
    let isInstance = true;
    isInstance = isInstance && "gas" in value;
    isInstance = isInstance && "amount" in value;
    return isInstance;
}
exports.instanceOfTypesThorchainSignDocTransferFee = instanceOfTypesThorchainSignDocTransferFee;
function TypesThorchainSignDocTransferFeeFromJSON(json) {
    return TypesThorchainSignDocTransferFeeFromJSONTyped(json, false);
}
exports.TypesThorchainSignDocTransferFeeFromJSON = TypesThorchainSignDocTransferFeeFromJSON;
function TypesThorchainSignDocTransferFeeFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'gas': json['gas'],
        'amount': (json['amount'].map(StdTxTransferValueAmountInner_1.StdTxTransferValueAmountInnerFromJSON)),
    };
}
exports.TypesThorchainSignDocTransferFeeFromJSONTyped = TypesThorchainSignDocTransferFeeFromJSONTyped;
function TypesThorchainSignDocTransferFeeToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'gas': value.gas,
        'amount': (value.amount.map(StdTxTransferValueAmountInner_1.StdTxTransferValueAmountInnerToJSON)),
    };
}
exports.TypesThorchainSignDocTransferFeeToJSON = TypesThorchainSignDocTransferFeeToJSON;
