// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/UtxoGetAddress200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UtxoGetAddress200ResponseToJSON = exports.UtxoGetAddress200ResponseFromJSONTyped = exports.UtxoGetAddress200ResponseFromJSON = exports.instanceOfUtxoGetAddress200Response = void 0;
/**
 * Check if a given object implements the UtxoGetAddress200Response interface.
 */
function instanceOfUtxoGetAddress200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "address" in value;
    return isInstance;
}
exports.instanceOfUtxoGetAddress200Response = instanceOfUtxoGetAddress200Response;
function UtxoGetAddress200ResponseFromJSON(json) {
    return UtxoGetAddress200ResponseFromJSONTyped(json, false);
}
exports.UtxoGetAddress200ResponseFromJSON = UtxoGetAddress200ResponseFromJSON;
function UtxoGetAddress200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'address': json['address'],
    };
}
exports.UtxoGetAddress200ResponseFromJSONTyped = UtxoGetAddress200ResponseFromJSONTyped;
function UtxoGetAddress200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'address': value.address,
    };
}
exports.UtxoGetAddress200ResponseToJSON = UtxoGetAddress200ResponseToJSON;
