// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/UtxoSignTransaction200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UtxoSignTransaction200ResponseToJSON = exports.UtxoSignTransaction200ResponseFromJSONTyped = exports.UtxoSignTransaction200ResponseFromJSON = exports.instanceOfUtxoSignTransaction200Response = void 0;
const TypesHexBytesLower_1 = require("./TypesHexBytesLower");
/**
 * Check if a given object implements the UtxoSignTransaction200Response interface.
 */
function instanceOfUtxoSignTransaction200Response(value) {
    let isInstance = true;
    isInstance = isInstance && "serializedTx" in value;
    return isInstance;
}
exports.instanceOfUtxoSignTransaction200Response = instanceOfUtxoSignTransaction200Response;
function UtxoSignTransaction200ResponseFromJSON(json) {
    return UtxoSignTransaction200ResponseFromJSONTyped(json, false);
}
exports.UtxoSignTransaction200ResponseFromJSON = UtxoSignTransaction200ResponseFromJSON;
function UtxoSignTransaction200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'serializedTx': (0, TypesHexBytesLower_1.TypesHexBytesLowerFromJSON)(json['serializedTx']),
    };
}
exports.UtxoSignTransaction200ResponseFromJSONTyped = UtxoSignTransaction200ResponseFromJSONTyped;
function UtxoSignTransaction200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'serializedTx': (0, TypesHexBytesLower_1.TypesHexBytesLowerToJSON)(value.serializedTx),
    };
}
exports.UtxoSignTransaction200ResponseToJSON = UtxoSignTransaction200ResponseToJSON;
