// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/UtxoGetAddressRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UtxoGetAddressRequestToJSON = exports.UtxoGetAddressRequestFromJSONTyped = exports.UtxoGetAddressRequestFromJSON = exports.instanceOfUtxoGetAddressRequest = void 0;
const runtime_1 = require("../runtime");
const BTCInputScriptType_1 = require("./BTCInputScriptType");
const TypesNumericU32_1 = require("./TypesNumericU32");
/**
 * Check if a given object implements the UtxoGetAddressRequest interface.
 */
function instanceOfUtxoGetAddressRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "coin" in value;
    isInstance = isInstance && "address_n" in value;
    return isInstance;
}
exports.instanceOfUtxoGetAddressRequest = instanceOfUtxoGetAddressRequest;
function UtxoGetAddressRequestFromJSON(json) {
    return UtxoGetAddressRequestFromJSONTyped(json, false);
}
exports.UtxoGetAddressRequestFromJSON = UtxoGetAddressRequestFromJSON;
function UtxoGetAddressRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'show_display': !(0, runtime_1.exists)(json, 'show_display') ? undefined : json['show_display'],
        'script_type': !(0, runtime_1.exists)(json, 'script_type') ? undefined : (0, BTCInputScriptType_1.BTCInputScriptTypeFromJSON)(json['script_type']),
        'coin': json['coin'],
        'address_n': (json['address_n'].map(TypesNumericU32_1.TypesNumericU32FromJSON)),
    };
}
exports.UtxoGetAddressRequestFromJSONTyped = UtxoGetAddressRequestFromJSONTyped;
function UtxoGetAddressRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'show_display': value.show_display,
        'script_type': (0, BTCInputScriptType_1.BTCInputScriptTypeToJSON)(value.script_type),
        'coin': value.coin,
        'address_n': (value.address_n.map(TypesNumericU32_1.TypesNumericU32ToJSON)),
    };
}
exports.UtxoGetAddressRequestToJSON = UtxoGetAddressRequestToJSON;
