// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/UtxoSignTransactionRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UtxoSignTransactionRequestToJSON = exports.UtxoSignTransactionRequestFromJSONTyped = exports.UtxoSignTransactionRequestFromJSON = exports.instanceOfUtxoSignTransactionRequest = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the UtxoSignTransactionRequest interface.
 */
function instanceOfUtxoSignTransactionRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "outputs" in value;
    isInstance = isInstance && "inputs" in value;
    isInstance = isInstance && "coin" in value;
    return isInstance;
}
exports.instanceOfUtxoSignTransactionRequest = instanceOfUtxoSignTransactionRequest;
function UtxoSignTransactionRequestFromJSON(json) {
    return UtxoSignTransactionRequestFromJSONTyped(json, false);
}
exports.UtxoSignTransactionRequestFromJSON = UtxoSignTransactionRequestFromJSON;
function UtxoSignTransactionRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'opReturnData': !(0, runtime_1.exists)(json, 'opReturnData') ? undefined : json['opReturnData'],
        'vaultAddress': !(0, runtime_1.exists)(json, 'vaultAddress') ? undefined : json['vaultAddress'],
        'outputs': json['outputs'],
        'inputs': json['inputs'],
        'coin': json['coin'],
    };
}
exports.UtxoSignTransactionRequestFromJSONTyped = UtxoSignTransactionRequestFromJSONTyped;
function UtxoSignTransactionRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'opReturnData': value.opReturnData,
        'vaultAddress': value.vaultAddress,
        'outputs': value.outputs,
        'inputs': value.inputs,
        'coin': value.coin,
    };
}
exports.UtxoSignTransactionRequestToJSON = UtxoSignTransactionRequestToJSON;
