// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/XrpSignTransactionRequestPayment.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrpSignTransactionRequestPaymentToJSON = exports.XrpSignTransactionRequestPaymentFromJSONTyped = exports.XrpSignTransactionRequestPaymentFromJSON = exports.instanceOfXrpSignTransactionRequestPayment = void 0;
/**
 * Check if a given object implements the XrpSignTransactionRequestPayment interface.
 */
function instanceOfXrpSignTransactionRequestPayment(value) {
    let isInstance = true;
    isInstance = isInstance && "destinationTag" in value;
    isInstance = isInstance && "destination" in value;
    isInstance = isInstance && "amount" in value;
    return isInstance;
}
exports.instanceOfXrpSignTransactionRequestPayment = instanceOfXrpSignTransactionRequestPayment;
function XrpSignTransactionRequestPaymentFromJSON(json) {
    return XrpSignTransactionRequestPaymentFromJSONTyped(json, false);
}
exports.XrpSignTransactionRequestPaymentFromJSON = XrpSignTransactionRequestPaymentFromJSON;
function XrpSignTransactionRequestPaymentFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'destinationTag': json['destinationTag'],
        'destination': json['destination'],
        'amount': json['amount'],
    };
}
exports.XrpSignTransactionRequestPaymentFromJSONTyped = XrpSignTransactionRequestPaymentFromJSONTyped;
function XrpSignTransactionRequestPaymentToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'destinationTag': value.destinationTag,
        'destination': value.destination,
        'amount': value.amount,
    };
}
exports.XrpSignTransactionRequestPaymentToJSON = XrpSignTransactionRequestPaymentToJSON;
