// @keepkey/keepkey-sdk@0.2.52/dist/generated/models/XrpSignTransactionRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * keepkey-sdk-server
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrpSignTransactionRequestToJSON = exports.XrpSignTransactionRequestFromJSONTyped = exports.XrpSignTransactionRequestFromJSON = exports.instanceOfXrpSignTransactionRequest = void 0;
const runtime_1 = require("../runtime");
const XrpSignTransactionRequestPayment_1 = require("./XrpSignTransactionRequestPayment");
/**
 * Check if a given object implements the XrpSignTransactionRequest interface.
 */
function instanceOfXrpSignTransactionRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "payment" in value;
    isInstance = isInstance && "lastLedgerSequence" in value;
    isInstance = isInstance && "sequence" in value;
    isInstance = isInstance && "tx" in value;
    isInstance = isInstance && "addressNList" in value;
    return isInstance;
}
exports.instanceOfXrpSignTransactionRequest = instanceOfXrpSignTransactionRequest;
function XrpSignTransactionRequestFromJSON(json) {
    return XrpSignTransactionRequestFromJSONTyped(json, false);
}
exports.XrpSignTransactionRequestFromJSON = XrpSignTransactionRequestFromJSON;
function XrpSignTransactionRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'payment': (0, XrpSignTransactionRequestPayment_1.XrpSignTransactionRequestPaymentFromJSON)(json['payment']),
        'lastLedgerSequence': json['lastLedgerSequence'],
        'sequence': json['sequence'],
        'flags': !(0, runtime_1.exists)(json, 'flags') ? undefined : json['flags'],
        'tx': json['tx'],
        'addressNList': json['addressNList'],
    };
}
exports.XrpSignTransactionRequestFromJSONTyped = XrpSignTransactionRequestFromJSONTyped;
function XrpSignTransactionRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'payment': (0, XrpSignTransactionRequestPayment_1.XrpSignTransactionRequestPaymentToJSON)(value.payment),
        'lastLedgerSequence': value.lastLedgerSequence,
        'sequence': value.sequence,
        'flags': value.flags,
        'tx': value.tx,
        'addressNList': value.addressNList,
    };
}
exports.XrpSignTransactionRequestToJSON = XrpSignTransactionRequestToJSON;
