// osmojs@0.37.0/main/codegen/ibc/lcd.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createLCDClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lcd = require("@osmonauts/lcd");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var createLCDClient = /*#__PURE__*/function () {
  var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(_ref) {
    var restEndpoint, requestClient;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            restEndpoint = _ref.restEndpoint;
            requestClient = new _lcd.LCDClient({
              restEndpoint: restEndpoint
            });
            _context.next = 4;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/authz/v1beta1/query.lcd"));
            });

          case 4:
            _context.t0 = _context.sent.LCDQueryClient;
            _context.t1 = {
              requestClient: requestClient
            };
            _context.t2 = new _context.t0(_context.t1);
            _context.t3 = {
              v1beta1: _context.t2
            };
            _context.next = 10;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/bank/v1beta1/query.lcd"));
            });

          case 10:
            _context.t4 = _context.sent.LCDQueryClient;
            _context.t5 = {
              requestClient: requestClient
            };
            _context.t6 = new _context.t4(_context.t5);
            _context.t7 = {
              v1beta1: _context.t6
            };
            _context.next = 16;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/distribution/v1beta1/query.lcd"));
            });

          case 16:
            _context.t8 = _context.sent.LCDQueryClient;
            _context.t9 = {
              requestClient: requestClient
            };
            _context.t10 = new _context.t8(_context.t9);
            _context.t11 = {
              v1beta1: _context.t10
            };
            _context.next = 22;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1/query.lcd"));
            });

          case 22:
            _context.t12 = _context.sent.LCDQueryClient;
            _context.t13 = {
              requestClient: requestClient
            };
            _context.t14 = new _context.t12(_context.t13);
            _context.next = 27;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1beta1/query.lcd"));
            });

          case 27:
            _context.t15 = _context.sent.LCDQueryClient;
            _context.t16 = {
              requestClient: requestClient
            };
            _context.t17 = new _context.t15(_context.t16);
            _context.t18 = {
              v1: _context.t14,
              v1beta1: _context.t17
            };
            _context.next = 33;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/staking/v1beta1/query.lcd"));
            });

          case 33:
            _context.t19 = _context.sent.LCDQueryClient;
            _context.t20 = {
              requestClient: requestClient
            };
            _context.t21 = new _context.t19(_context.t20);
            _context.t22 = {
              v1beta1: _context.t21
            };
            _context.next = 39;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/tx/v1beta1/service.lcd"));
            });

          case 39:
            _context.t23 = _context.sent.LCDQueryClient;
            _context.t24 = {
              requestClient: requestClient
            };
            _context.t25 = new _context.t23(_context.t24);
            _context.t26 = {
              v1beta1: _context.t25
            };
            _context.next = 45;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/upgrade/v1beta1/query.lcd"));
            });

          case 45:
            _context.t27 = _context.sent.LCDQueryClient;
            _context.t28 = {
              requestClient: requestClient
            };
            _context.t29 = new _context.t27(_context.t28);
            _context.t30 = {
              v1beta1: _context.t29
            };
            _context.t31 = {
              authz: _context.t3,
              bank: _context.t7,
              distribution: _context.t11,
              gov: _context.t18,
              staking: _context.t22,
              tx: _context.t26,
              upgrade: _context.t30
            };
            _context.next = 52;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./applications/transfer/v1/query.lcd"));
            });

          case 52:
            _context.t32 = _context.sent.LCDQueryClient;
            _context.t33 = {
              requestClient: requestClient
            };
            _context.t34 = new _context.t32(_context.t33);
            _context.t35 = {
              v1: _context.t34
            };
            _context.t36 = {
              transfer: _context.t35
            };
            _context.next = 59;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./core/channel/v1/query.lcd"));
            });

          case 59:
            _context.t37 = _context.sent.LCDQueryClient;
            _context.t38 = {
              requestClient: requestClient
            };
            _context.t39 = new _context.t37(_context.t38);
            _context.t40 = {
              v1: _context.t39
            };
            _context.next = 65;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./core/client/v1/query.lcd"));
            });

          case 65:
            _context.t41 = _context.sent.LCDQueryClient;
            _context.t42 = {
              requestClient: requestClient
            };
            _context.t43 = new _context.t41(_context.t42);
            _context.t44 = {
              v1: _context.t43
            };
            _context.next = 71;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./core/connection/v1/query.lcd"));
            });

          case 71:
            _context.t45 = _context.sent.LCDQueryClient;
            _context.t46 = {
              requestClient: requestClient
            };
            _context.t47 = new _context.t45(_context.t46);
            _context.t48 = {
              v1: _context.t47
            };
            _context.t49 = {
              channel: _context.t40,
              client: _context.t44,
              connection: _context.t48
            };
            _context.t50 = {
              applications: _context.t36,
              core: _context.t49
            };
            return _context.abrupt("return", {
              cosmos: _context.t31,
              ibc: _context.t50
            });

          case 78:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function createLCDClient(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.createLCDClient = createLCDClient;
