// osmojs@0.37.0/main/codegen/osmosis/superfluid/tx.rpc.msg.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MsgClientImpl = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));

var _tx = require("./tx");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MsgClientImpl = /*#__PURE__*/function () {
  function MsgClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, MsgClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.superfluidDelegate = this.superfluidDelegate.bind(this);
    this.superfluidUndelegate = this.superfluidUndelegate.bind(this);
    this.superfluidUnbondLock = this.superfluidUnbondLock.bind(this);
    this.lockAndSuperfluidDelegate = this.lockAndSuperfluidDelegate.bind(this);
    this.unPoolWhitelistedPool = this.unPoolWhitelistedPool.bind(this);
  }

  (0, _createClass2["default"])(MsgClientImpl, [{
    key: "superfluidDelegate",
    value: function superfluidDelegate(request) {
      var data = _tx.MsgSuperfluidDelegate.encode(request).finish();

      var promise = this.rpc.request("osmosis.superfluid.Msg", "SuperfluidDelegate", data);
      return promise.then(function (data) {
        return _tx.MsgSuperfluidDelegateResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "superfluidUndelegate",
    value: function superfluidUndelegate(request) {
      var data = _tx.MsgSuperfluidUndelegate.encode(request).finish();

      var promise = this.rpc.request("osmosis.superfluid.Msg", "SuperfluidUndelegate", data);
      return promise.then(function (data) {
        return _tx.MsgSuperfluidUndelegateResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "superfluidUnbondLock",
    value: function superfluidUnbondLock(request) {
      var data = _tx.MsgSuperfluidUnbondLock.encode(request).finish();

      var promise = this.rpc.request("osmosis.superfluid.Msg", "SuperfluidUnbondLock", data);
      return promise.then(function (data) {
        return _tx.MsgSuperfluidUnbondLockResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "lockAndSuperfluidDelegate",
    value: function lockAndSuperfluidDelegate(request) {
      var data = _tx.MsgLockAndSuperfluidDelegate.encode(request).finish();

      var promise = this.rpc.request("osmosis.superfluid.Msg", "LockAndSuperfluidDelegate", data);
      return promise.then(function (data) {
        return _tx.MsgLockAndSuperfluidDelegateResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "unPoolWhitelistedPool",
    value: function unPoolWhitelistedPool(request) {
      var data = _tx.MsgUnPoolWhitelistedPool.encode(request).finish();

      var promise = this.rpc.request("osmosis.superfluid.Msg", "UnPoolWhitelistedPool", data);
      return promise.then(function (data) {
        return _tx.MsgUnPoolWhitelistedPoolResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return MsgClientImpl;
}();

exports.MsgClientImpl = MsgClientImpl;
