// @metamask/rpc-methods@1.0.2/dist/restricted/getBip32PublicKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBip32PublicKeyImplementation = exports.getBip32PublicKeyBuilder = exports.Bip32PublicKeyArgsStruct = void 0;
const key_tree_1 = require("@metamask/key-tree");
const permission_controller_1 = require("@metamask/permission-controller");
const snaps_utils_1 = require("@metamask/snaps-utils");
const utils_1 = require("@metamask/utils");
const eth_rpc_errors_1 = require("eth-rpc-errors");
const superstruct_1 = require("superstruct");
const targetName = 'snap_getBip32PublicKey';
exports.Bip32PublicKeyArgsStruct = (0, snaps_utils_1.bip32entropy)((0, superstruct_1.object)({
    path: snaps_utils_1.Bip32PathStruct,
    curve: (0, superstruct_1.enums)(['ed225519', 'secp256k1']),
    compressed: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
}));
/**
 * The specification builder for the `snap_getBip32PublicKey` permission.
 * `snap_getBip32PublicKey` lets the Snap retrieve public keys for a particular
 * BIP-32 node.
 *
 * @param options - The specification builder options.
 * @param options.methodHooks - The RPC method hooks needed by the method implementation.
 * @returns The specification for the `snap_getBip32PublicKey` permission.
 */
const specificationBuilder = ({ methodHooks }) => {
    return {
        permissionType: permission_controller_1.PermissionType.RestrictedMethod,
        targetName,
        allowedCaveats: [snaps_utils_1.SnapCaveatType.PermittedDerivationPaths],
        methodImplementation: getBip32PublicKeyImplementation(methodHooks),
        validator: ({ caveats }) => {
            if (caveats?.length !== 1 ||
                caveats[0].type !== snaps_utils_1.SnapCaveatType.PermittedDerivationPaths) {
                throw eth_rpc_errors_1.ethErrors.rpc.invalidParams({
                    message: `Expected a single "${snaps_utils_1.SnapCaveatType.PermittedDerivationPaths}" caveat.`,
                });
            }
        },
        subjectTypes: [permission_controller_1.SubjectType.Snap],
    };
};
const methodHooks = {
    getMnemonic: true,
    getUnlockPromise: true,
};
exports.getBip32PublicKeyBuilder = Object.freeze({
    targetName,
    specificationBuilder,
    methodHooks,
});
/**
 * Builds the method implementation for `snap_getBip32PublicKey`.
 *
 * @param hooks - The RPC method hooks.
 * @param hooks.getMnemonic - A function to retrieve the Secret Recovery Phrase of the user.
 * @param hooks.getUnlockPromise - A function that resolves once the MetaMask extension is unlocked
 * and prompts the user to unlock their MetaMask if it is locked.
 * @returns The method implementation which returns a public key.
 * @throws If the params are invalid.
 */
function getBip32PublicKeyImplementation({ getMnemonic, getUnlockPromise, }) {
    return async function getBip32PublicKey(args) {
        await getUnlockPromise(true);
        (0, utils_1.assertStruct)(args.params, exports.Bip32PublicKeyArgsStruct, 'Invalid BIP-32 public key params', eth_rpc_errors_1.ethErrors.rpc.invalidParams);
        const { params } = args;
        const node = await key_tree_1.SLIP10Node.fromDerivationPath({
            curve: params.curve,
            derivationPath: [
                await getMnemonic(),
                ...params.path
                    .slice(1)
                    .map((index) => `bip32:${index}`),
            ],
        });
        if (params.compressed) {
            return node.compressedPublicKey;
        }
        return node.publicKey;
    };
}
exports.getBip32PublicKeyImplementation = getBip32PublicKeyImplementation;
