// @metamask/rpc-methods@1.0.2/dist/restricted/notify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getValidatedParams = exports.getImplementation = exports.notifyBuilder = exports.specificationBuilder = exports.NotificationType = void 0;
const permission_controller_1 = require("@metamask/permission-controller");
const utils_1 = require("@metamask/utils");
const eth_rpc_errors_1 = require("eth-rpc-errors");
const methodName = 'snap_notify';
// TODO: Move all the types to a shared place when implementing more
//  notifications.
var NotificationType;
(function (NotificationType) {
    NotificationType["InApp"] = "inApp";
    NotificationType["Native"] = "native";
})(NotificationType = exports.NotificationType || (exports.NotificationType = {}));
/**
 * The specification builder for the `snap_notify` permission.
 * `snap_notify` allows snaps to send multiple types of notifications to its users.
 *
 * @param options - The specification builder options.
 * @param options.allowedCaveats - The optional allowed caveats for the permission.
 * @param options.methodHooks - The RPC method hooks needed by the method implementation.
 * @returns The specification for the `snap_notify` permission.
 */
const specificationBuilder = ({ allowedCaveats = null, methodHooks }) => {
    return {
        permissionType: permission_controller_1.PermissionType.RestrictedMethod,
        targetName: methodName,
        allowedCaveats,
        methodImplementation: getImplementation(methodHooks),
        subjectTypes: [permission_controller_1.SubjectType.Snap],
    };
};
exports.specificationBuilder = specificationBuilder;
const methodHooks = {
    showNativeNotification: true,
    showInAppNotification: true,
};
exports.notifyBuilder = Object.freeze({
    targetName: methodName,
    specificationBuilder: exports.specificationBuilder,
    methodHooks,
});
/**
 * Builds the method implementation for `snap_notify`.
 *
 * @param hooks - The RPC method hooks.
 * @param hooks.showNativeNotification - A function that shows a native browser notification.
 * @param hooks.showInAppNotification - A function that shows a notification in the MetaMask UI.
 * @returns The method implementation which returns `null` on success.
 * @throws If the params are invalid.
 */
function getImplementation({ showNativeNotification, showInAppNotification, }) {
    return async function implementation(args) {
        const { params, context: { origin }, } = args;
        const validatedParams = getValidatedParams(params);
        switch (validatedParams.type) {
            case NotificationType.Native:
                return await showNativeNotification(origin, validatedParams);
            case NotificationType.InApp:
                return await showInAppNotification(origin, validatedParams);
            default:
                throw eth_rpc_errors_1.ethErrors.rpc.invalidParams({
                    message: 'Must specify a valid notification "type".',
                });
        }
    };
}
exports.getImplementation = getImplementation;
/**
 * Validates the notify method `params` and returns them cast to the correct
 * type. Throws if validation fails.
 *
 * @param params - The unvalidated params object from the method request.
 * @returns The validated method parameter object.
 */
function getValidatedParams(params) {
    if (!(0, utils_1.isObject)(params)) {
        throw eth_rpc_errors_1.ethErrors.rpc.invalidParams({
            message: 'Expected params to be a single object.',
        });
    }
    const { type, message } = params;
    if (!type ||
        typeof type !== 'string' ||
        !Object.values(NotificationType).includes(type)) {
        throw eth_rpc_errors_1.ethErrors.rpc.invalidParams({
            message: 'Must specify a valid notification "type".',
        });
    }
    // Set to the max message length on a Mac notification for now.
    if (!message || typeof message !== 'string' || message.length >= 50) {
        throw eth_rpc_errors_1.ethErrors.rpc.invalidParams({
            message: 'Must specify a non-empty string "message" less than 50 characters long.',
        });
    }
    return params;
}
exports.getValidatedParams = getValidatedParams;
