// @coinmasters/toolbox-cosmos@4.0.4/dist/index-7bcca4ca.js
var aspromise = asPromise;
function asPromise(u, c) {
  for (var n = new Array(arguments.length - 1), i = 0, t = 2, h = !0; t < arguments.length; )
    n[i++] = arguments[t++];
  return new Promise(function(a, r) {
    n[i] = function(l) {
      if (h)
        if (h = !1, l)
          r(l);
        else {
          for (var o = new Array(arguments.length - 1), E = 0; E < o.length; )
            o[E++] = arguments[E];
          a.apply(null, o);
        }
    };
    try {
      u.apply(c || null, n);
    } catch (e) {
      h && (h = !1, r(e));
    }
  });
}
var base64 = {};
(function(u) {
  var c = u;
  c.length = function(a) {
    var r = a.length;
    if (!r)
      return 0;
    for (var e = 0; --r % 4 > 1 && a.charAt(r) === "="; )
      ++e;
    return Math.ceil(a.length * 3) / 4 - e;
  };
  for (var n = new Array(64), i = new Array(123), t = 0; t < 64; )
    i[n[t] = t < 26 ? t + 65 : t < 52 ? t + 71 : t < 62 ? t - 4 : t - 59 | 43] = t++;
  c.encode = function(a, r, e) {
    for (var l = null, o = [], E = 0, y = 0, g; r < e; ) {
      var v = a[r++];
      switch (y) {
        case 0:
          o[E++] = n[v >> 2], g = (v & 3) << 4, y = 1;
          break;
        case 1:
          o[E++] = n[g | v >> 4], g = (v & 15) << 2, y = 2;
          break;
        case 2:
          o[E++] = n[g | v >> 6], o[E++] = n[v & 63], y = 0;
          break;
      }
      E > 8191 && ((l || (l = [])).push(String.fromCharCode.apply(String, o)), E = 0);
    }
    return y && (o[E++] = n[g], o[E++] = 61, y === 1 && (o[E++] = 61)), l ? (E && l.push(String.fromCharCode.apply(String, o.slice(0, E))), l.join("")) : String.fromCharCode.apply(String, o.slice(0, E));
  };
  var h = "invalid encoding";
  c.decode = function(a, r, e) {
    for (var l = e, o = 0, E, y = 0; y < a.length; ) {
      var g = a.charCodeAt(y++);
      if (g === 61 && o > 1)
        break;
      if ((g = i[g]) === void 0)
        throw Error(h);
      switch (o) {
        case 0:
          E = g, o = 1;
          break;
        case 1:
          r[e++] = E << 2 | (g & 48) >> 4, E = g, o = 2;
          break;
        case 2:
          r[e++] = (E & 15) << 4 | (g & 60) >> 2, E = g, o = 3;
          break;
        case 3:
          r[e++] = (E & 3) << 6 | g, o = 0;
          break;
      }
    }
    if (o === 1)
      throw Error(h);
    return e - l;
  }, c.test = function(a) {
    return /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(a);
  };
})(base64);
var eventemitter = EventEmitter;
function EventEmitter() {
  this._listeners = {};
}
EventEmitter.prototype.on = function(c, n, i) {
  return (this._listeners[c] || (this._listeners[c] = [])).push({
    fn: n,
    ctx: i || this
  }), this;
};
EventEmitter.prototype.off = function(c, n) {
  if (c === void 0)
    this._listeners = {};
  else if (n === void 0)
    this._listeners[c] = [];
  else
    for (var i = this._listeners[c], t = 0; t < i.length; )
      i[t].fn === n ? i.splice(t, 1) : ++t;
  return this;
};
EventEmitter.prototype.emit = function(c) {
  var n = this._listeners[c];
  if (n) {
    for (var i = [], t = 1; t < arguments.length; )
      i.push(arguments[t++]);
    for (t = 0; t < n.length; )
      n[t].fn.apply(n[t++].ctx, i);
  }
  return this;
};
var float = factory(factory);
function factory(u) {
  return typeof Float32Array < "u" ? function() {
    var c = new Float32Array([-0]), n = new Uint8Array(c.buffer), i = n[3] === 128;
    function t(r, e, l) {
      c[0] = r, e[l] = n[0], e[l + 1] = n[1], e[l + 2] = n[2], e[l + 3] = n[3];
    }
    function h(r, e, l) {
      c[0] = r, e[l] = n[3], e[l + 1] = n[2], e[l + 2] = n[1], e[l + 3] = n[0];
    }
    u.writeFloatLE = i ? t : h, u.writeFloatBE = i ? h : t;
    function d(r, e) {
      return n[0] = r[e], n[1] = r[e + 1], n[2] = r[e + 2], n[3] = r[e + 3], c[0];
    }
    function a(r, e) {
      return n[3] = r[e], n[2] = r[e + 1], n[1] = r[e + 2], n[0] = r[e + 3], c[0];
    }
    u.readFloatLE = i ? d : a, u.readFloatBE = i ? a : d;
  }() : function() {
    function c(i, t, h, d) {
      var a = t < 0 ? 1 : 0;
      if (a && (t = -t), t === 0)
        i(1 / t > 0 ? (
          /* positive */
          0
        ) : (
          /* negative 0 */
          2147483648
        ), h, d);
      else if (isNaN(t))
        i(2143289344, h, d);
      else if (t > 34028234663852886e22)
        i((a << 31 | 2139095040) >>> 0, h, d);
      else if (t < 11754943508222875e-54)
        i((a << 31 | Math.round(t / 1401298464324817e-60)) >>> 0, h, d);
      else {
        var r = Math.floor(Math.log(t) / Math.LN2), e = Math.round(t * Math.pow(2, -r) * 8388608) & 8388607;
        i((a << 31 | r + 127 << 23 | e) >>> 0, h, d);
      }
    }
    u.writeFloatLE = c.bind(null, writeUintLE), u.writeFloatBE = c.bind(null, writeUintBE);
    function n(i, t, h) {
      var d = i(t, h), a = (d >> 31) * 2 + 1, r = d >>> 23 & 255, e = d & 8388607;
      return r === 255 ? e ? NaN : a * (1 / 0) : r === 0 ? a * 1401298464324817e-60 * e : a * Math.pow(2, r - 150) * (e + 8388608);
    }
    u.readFloatLE = n.bind(null, readUintLE), u.readFloatBE = n.bind(null, readUintBE);
  }(), typeof Float64Array < "u" ? function() {
    var c = new Float64Array([-0]), n = new Uint8Array(c.buffer), i = n[7] === 128;
    function t(r, e, l) {
      c[0] = r, e[l] = n[0], e[l + 1] = n[1], e[l + 2] = n[2], e[l + 3] = n[3], e[l + 4] = n[4], e[l + 5] = n[5], e[l + 6] = n[6], e[l + 7] = n[7];
    }
    function h(r, e, l) {
      c[0] = r, e[l] = n[7], e[l + 1] = n[6], e[l + 2] = n[5], e[l + 3] = n[4], e[l + 4] = n[3], e[l + 5] = n[2], e[l + 6] = n[1], e[l + 7] = n[0];
    }
    u.writeDoubleLE = i ? t : h, u.writeDoubleBE = i ? h : t;
    function d(r, e) {
      return n[0] = r[e], n[1] = r[e + 1], n[2] = r[e + 2], n[3] = r[e + 3], n[4] = r[e + 4], n[5] = r[e + 5], n[6] = r[e + 6], n[7] = r[e + 7], c[0];
    }
    function a(r, e) {
      return n[7] = r[e], n[6] = r[e + 1], n[5] = r[e + 2], n[4] = r[e + 3], n[3] = r[e + 4], n[2] = r[e + 5], n[1] = r[e + 6], n[0] = r[e + 7], c[0];
    }
    u.readDoubleLE = i ? d : a, u.readDoubleBE = i ? a : d;
  }() : function() {
    function c(i, t, h, d, a, r) {
      var e = d < 0 ? 1 : 0;
      if (e && (d = -d), d === 0)
        i(0, a, r + t), i(1 / d > 0 ? (
          /* positive */
          0
        ) : (
          /* negative 0 */
          2147483648
        ), a, r + h);
      else if (isNaN(d))
        i(0, a, r + t), i(2146959360, a, r + h);
      else if (d > 17976931348623157e292)
        i(0, a, r + t), i((e << 31 | 2146435072) >>> 0, a, r + h);
      else {
        var l;
        if (d < 22250738585072014e-324)
          l = d / 5e-324, i(l >>> 0, a, r + t), i((e << 31 | l / 4294967296) >>> 0, a, r + h);
        else {
          var o = Math.floor(Math.log(d) / Math.LN2);
          o === 1024 && (o = 1023), l = d * Math.pow(2, -o), i(l * 4503599627370496 >>> 0, a, r + t), i((e << 31 | o + 1023 << 20 | l * 1048576 & 1048575) >>> 0, a, r + h);
        }
      }
    }
    u.writeDoubleLE = c.bind(null, writeUintLE, 0, 4), u.writeDoubleBE = c.bind(null, writeUintBE, 4, 0);
    function n(i, t, h, d, a) {
      var r = i(d, a + t), e = i(d, a + h), l = (e >> 31) * 2 + 1, o = e >>> 20 & 2047, E = 4294967296 * (e & 1048575) + r;
      return o === 2047 ? E ? NaN : l * (1 / 0) : o === 0 ? l * 5e-324 * E : l * Math.pow(2, o - 1075) * (E + 4503599627370496);
    }
    u.readDoubleLE = n.bind(null, readUintLE, 0, 4), u.readDoubleBE = n.bind(null, readUintBE, 4, 0);
  }(), u;
}
function writeUintLE(u, c, n) {
  c[n] = u & 255, c[n + 1] = u >>> 8 & 255, c[n + 2] = u >>> 16 & 255, c[n + 3] = u >>> 24;
}
function writeUintBE(u, c, n) {
  c[n] = u >>> 24, c[n + 1] = u >>> 16 & 255, c[n + 2] = u >>> 8 & 255, c[n + 3] = u & 255;
}
function readUintLE(u, c) {
  return (u[c] | u[c + 1] << 8 | u[c + 2] << 16 | u[c + 3] << 24) >>> 0;
}
function readUintBE(u, c) {
  return (u[c] << 24 | u[c + 1] << 16 | u[c + 2] << 8 | u[c + 3]) >>> 0;
}
var inquire_1 = inquire;
function inquire(moduleName) {
  try {
    var mod = eval("quire".replace(/^/, "re"))(moduleName);
    if (mod && (mod.length || Object.keys(mod).length))
      return mod;
  } catch (u) {
  }
  return null;
}
var utf8 = {};
(function(u) {
  var c = u;
  c.length = function(i) {
    for (var t = 0, h = 0, d = 0; d < i.length; ++d)
      h = i.charCodeAt(d), h < 128 ? t += 1 : h < 2048 ? t += 2 : (h & 64512) === 55296 && (i.charCodeAt(d + 1) & 64512) === 56320 ? (++d, t += 4) : t += 3;
    return t;
  }, c.read = function(i, t, h) {
    var d = h - t;
    if (d < 1)
      return "";
    for (var a = null, r = [], e = 0, l; t < h; )
      l = i[t++], l < 128 ? r[e++] = l : l > 191 && l < 224 ? r[e++] = (l & 31) << 6 | i[t++] & 63 : l > 239 && l < 365 ? (l = ((l & 7) << 18 | (i[t++] & 63) << 12 | (i[t++] & 63) << 6 | i[t++] & 63) - 65536, r[e++] = 55296 + (l >> 10), r[e++] = 56320 + (l & 1023)) : r[e++] = (l & 15) << 12 | (i[t++] & 63) << 6 | i[t++] & 63, e > 8191 && ((a || (a = [])).push(String.fromCharCode.apply(String, r)), e = 0);
    return a ? (e && a.push(String.fromCharCode.apply(String, r.slice(0, e))), a.join("")) : String.fromCharCode.apply(String, r.slice(0, e));
  }, c.write = function(i, t, h) {
    for (var d = h, a, r, e = 0; e < i.length; ++e)
      a = i.charCodeAt(e), a < 128 ? t[h++] = a : a < 2048 ? (t[h++] = a >> 6 | 192, t[h++] = a & 63 | 128) : (a & 64512) === 55296 && ((r = i.charCodeAt(e + 1)) & 64512) === 56320 ? (a = 65536 + ((a & 1023) << 10) + (r & 1023), ++e, t[h++] = a >> 18 | 240, t[h++] = a >> 12 & 63 | 128, t[h++] = a >> 6 & 63 | 128, t[h++] = a & 63 | 128) : (t[h++] = a >> 12 | 224, t[h++] = a >> 6 & 63 | 128, t[h++] = a & 63 | 128);
    return h - d;
  };
})(utf8);
var pool_1 = pool;
function pool(u, c, n) {
  var i = n || 8192, t = i >>> 1, h = null, d = i;
  return function(r) {
    if (r < 1 || r > t)
      return u(r);
    d + r > i && (h = u(i), d = 0);
    var e = c.call(h, d, d += r);
    return d & 7 && (d = (d | 7) + 1), e;
  };
}
export {
  aspromise as a,
  base64 as b,
  eventemitter as e,
  float as f,
  inquire_1 as i,
  pool_1 as p,
  utf8 as u
};
