// hive-keychain-commons@1.2.4/lib/swaps/swap.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepHistoryStatus = exports.SwapStatus = exports.Provider = exports.SwapStepType = void 0;
var SwapStepType;
(function (SwapStepType) {
    SwapStepType["DEPOSIT_TO_HIVE_ENGINE"] = "DEPOSIT_TO_HIVE_ENGINE";
    SwapStepType["WITHDRAWAL_FROM_HIVE_ENGINE"] = "WITHDRAWAL_FROM_HIVE_ENGINE";
    SwapStepType["SWAP_TOKEN"] = "SWAP_TOKEN";
    SwapStepType["BUY_ON_HIVE_ENGINE_MARKET"] = "BUY_ON_HIVE_ENGINE_MARKET";
    SwapStepType["SELL_ON_HIVE_ENGINE_MARKET"] = "SELL_ON_HIVE_ENGINE_MARKET";
    SwapStepType["BUY_ON_MARKET"] = "BUY_ON_MARKET";
    SwapStepType["SELL_ON_MARKET"] = "SELL_ON_MARKET";
})(SwapStepType = exports.SwapStepType || (exports.SwapStepType = {}));
var Provider;
(function (Provider) {
    Provider["HIVE_INTERNAL_MARKET"] = "HIVE_INTERNAL_MARKET";
    Provider["BEESWAP"] = "BEESWAP";
    Provider["HIVE_PAY"] = "HIVE_PAY";
    Provider["DISCOUNTED_BRIDGE"] = "DISCOUNTED_BRIDGE";
    Provider["LEODEX"] = "LEODEX";
    Provider["HIVE_ENGINE"] = "HIVE_ENGINE";
    Provider["LIQUIDITY_POOL"] = "LIQUIDITY_POOL";
    Provider["HIVE_ENGINE_INTERNAL_MARKET"] = "HIVE_ENGINE_INTERNAL_MARKET";
    Provider["NULL"] = "NULL";
})(Provider = exports.Provider || (exports.Provider = {}));
var SwapStatus;
(function (SwapStatus) {
    SwapStatus["PENDING"] = "PENDING";
    SwapStatus["STARTED"] = "STARTED";
    SwapStatus["CANCELED_DUE_TO_ERROR"] = "CANCELED_DUE_TO_ERROR";
    SwapStatus["REFUNDED_SLIPPAGE"] = "REFUND_SLIPPAGE";
    SwapStatus["COMPLETED"] = "COMPLETED";
    SwapStatus["FUNDS_RETURNED"] = "FUNDS_RETURNED";
})(SwapStatus = exports.SwapStatus || (exports.SwapStatus = {}));
var StepHistoryStatus;
(function (StepHistoryStatus) {
    StepHistoryStatus["PENDING"] = "pending";
    StepHistoryStatus["SUCCESS"] = "success";
    StepHistoryStatus["FAILED"] = "failed";
})(StepHistoryStatus = exports.StepHistoryStatus || (exports.StepHistoryStatus = {}));
