// @pioneer-platform/xdefi@3.0.1/dist/index.es-24d5c8d6.js
var Pe = /^-?(?:\d+(?:\.\d*)?|\.\d+)(?:e[+-]?\d+)?$/i, Se = Math.ceil, Y = Math.floor, V = "[BigNumber Error] ", we = V + "Number primitive has more than 15 significant digits: ", W = 1e14, T = 14, Re = 9007199254740991, Ie = [1, 10, 100, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8, 1e9, 1e10, 1e11, 1e12, 1e13], Z = 1e7, M = 1e9;
function Be(n) {
  var h, A, D, _ = N.prototype = { constructor: N, toString: null, valueOf: null }, B = new N(1), g = 20, w = 4, G = -7, x = 21, ee = -1e7, q = 1e7, ie = !1, se = 1, j = 0, he = {
    prefix: "",
    groupSize: 3,
    secondaryGroupSize: 0,
    groupSeparator: ",",
    decimalSeparator: ".",
    fractionGroupSize: 0,
    fractionGroupSeparator: " ",
    // non-breaking space
    suffix: ""
  }, ne = "0123456789abcdefghijklmnopqrstuvwxyz", Ee = !0;
  function N(e, i) {
    var r, l, s, c, a, t, o, u, f = this;
    if (!(f instanceof N))
      return new N(e, i);
    if (i == null) {
      if (e && e._isBigNumber === !0) {
        f.s = e.s, !e.c || e.e > q ? f.c = f.e = null : e.e < ee ? f.c = [f.e = 0] : (f.e = e.e, f.c = e.c.slice());
        return;
      }
      if ((t = typeof e == "number") && e * 0 == 0) {
        if (f.s = 1 / e < 0 ? (e = -e, -1) : 1, e === ~~e) {
          for (c = 0, a = e; a >= 10; a /= 10, c++)
            ;
          c > q ? f.c = f.e = null : (f.e = c, f.c = [e]);
          return;
        }
        u = String(e);
      } else {
        if (!Pe.test(u = String(e)))
          return D(f, u, t);
        f.s = u.charCodeAt(0) == 45 ? (u = u.slice(1), -1) : 1;
      }
      (c = u.indexOf(".")) > -1 && (u = u.replace(".", "")), (a = u.search(/e/i)) > 0 ? (c < 0 && (c = a), c += +u.slice(a + 1), u = u.substring(0, a)) : c < 0 && (c = u.length);
    } else {
      if (U(i, 2, ne.length, "Base"), i == 10 && Ee)
        return f = new N(e), F(f, g + f.e + 1, w);
      if (u = String(e), t = typeof e == "number") {
        if (e * 0 != 0)
          return D(f, u, t, i);
        if (f.s = 1 / e < 0 ? (u = u.slice(1), -1) : 1, N.DEBUG && u.replace(/^0\.0*|\./, "").length > 15)
          throw Error(we + e);
      } else
        f.s = u.charCodeAt(0) === 45 ? (u = u.slice(1), -1) : 1;
      for (r = ne.slice(0, i), c = a = 0, o = u.length; a < o; a++)
        if (r.indexOf(l = u.charAt(a)) < 0) {
          if (l == ".") {
            if (a > c) {
              c = o;
              continue;
            }
          } else if (!s && (u == u.toUpperCase() && (u = u.toLowerCase()) || u == u.toLowerCase() && (u = u.toUpperCase()))) {
            s = !0, a = -1, c = 0;
            continue;
          }
          return D(f, String(e), t, i);
        }
      t = !1, u = A(u, i, 10, f.s), (c = u.indexOf(".")) > -1 ? u = u.replace(".", "") : c = u.length;
    }
    for (a = 0; u.charCodeAt(a) === 48; a++)
      ;
    for (o = u.length; u.charCodeAt(--o) === 48; )
      ;
    if (u = u.slice(a, ++o)) {
      if (o -= a, t && N.DEBUG && o > 15 && (e > Re || e !== Y(e)))
        throw Error(we + f.s * e);
      if ((c = c - a - 1) > q)
        f.c = f.e = null;
      else if (c < ee)
        f.c = [f.e = 0];
      else {
        if (f.e = c, f.c = [], a = (c + 1) % T, c < 0 && (a += T), a < o) {
          for (a && f.c.push(+u.slice(0, a)), o -= T; a < o; )
            f.c.push(+u.slice(a, a += T));
          a = T - (u = u.slice(a)).length;
        } else
          a -= o;
        for (; a--; u += "0")
          ;
        f.c.push(+u);
      }
    } else
      f.c = [f.e = 0];
  }
  N.clone = Be, N.ROUND_UP = 0, N.ROUND_DOWN = 1, N.ROUND_CEIL = 2, N.ROUND_FLOOR = 3, N.ROUND_HALF_UP = 4, N.ROUND_HALF_DOWN = 5, N.ROUND_HALF_EVEN = 6, N.ROUND_HALF_CEIL = 7, N.ROUND_HALF_FLOOR = 8, N.EUCLID = 9, N.config = N.set = function(e) {
    var i, r;
    if (e != null)
      if (typeof e == "object") {
        if (e.hasOwnProperty(i = "DECIMAL_PLACES") && (r = e[i], U(r, 0, M, i), g = r), e.hasOwnProperty(i = "ROUNDING_MODE") && (r = e[i], U(r, 0, 8, i), w = r), e.hasOwnProperty(i = "EXPONENTIAL_AT") && (r = e[i], r && r.pop ? (U(r[0], -M, 0, i), U(r[1], 0, M, i), G = r[0], x = r[1]) : (U(r, -M, M, i), G = -(x = r < 0 ? -r : r))), e.hasOwnProperty(i = "RANGE"))
          if (r = e[i], r && r.pop)
            U(r[0], -M, -1, i), U(r[1], 1, M, i), ee = r[0], q = r[1];
          else if (U(r, -M, M, i), r)
            ee = -(q = r < 0 ? -r : r);
          else
            throw Error(V + i + " cannot be zero: " + r);
        if (e.hasOwnProperty(i = "CRYPTO"))
          if (r = e[i], r === !!r)
            if (r)
              if (typeof crypto < "u" && crypto && (crypto.getRandomValues || crypto.randomBytes))
                ie = r;
              else
                throw ie = !r, Error(V + "crypto unavailable");
            else
              ie = r;
          else
            throw Error(V + i + " not true or false: " + r);
        if (e.hasOwnProperty(i = "MODULO_MODE") && (r = e[i], U(r, 0, 9, i), se = r), e.hasOwnProperty(i = "POW_PRECISION") && (r = e[i], U(r, 0, M, i), j = r), e.hasOwnProperty(i = "FORMAT"))
          if (r = e[i], typeof r == "object")
            he = r;
          else
            throw Error(V + i + " not an object: " + r);
        if (e.hasOwnProperty(i = "ALPHABET"))
          if (r = e[i], typeof r == "string" && !/^.?$|[+\-.\s]|(.).*\1/.test(r))
            Ee = r.slice(0, 10) == "0123456789", ne = r;
          else
            throw Error(V + i + " invalid: " + r);
      } else
        throw Error(V + "Object expected: " + e);
    return {
      DECIMAL_PLACES: g,
      ROUNDING_MODE: w,
      EXPONENTIAL_AT: [G, x],
      RANGE: [ee, q],
      CRYPTO: ie,
      MODULO_MODE: se,
      POW_PRECISION: j,
      FORMAT: he,
      ALPHABET: ne
    };
  }, N.isBigNumber = function(e) {
    if (!e || e._isBigNumber !== !0)
      return !1;
    if (!N.DEBUG)
      return !0;
    var i, r, l = e.c, s = e.e, c = e.s;
    e:
      if ({}.toString.call(l) == "[object Array]") {
        if ((c === 1 || c === -1) && s >= -M && s <= M && s === Y(s)) {
          if (l[0] === 0) {
            if (s === 0 && l.length === 1)
              return !0;
            break e;
          }
          if (i = (s + 1) % T, i < 1 && (i += T), String(l[0]).length == i) {
            for (i = 0; i < l.length; i++)
              if (r = l[i], r < 0 || r >= W || r !== Y(r))
                break e;
            if (r !== 0)
              return !0;
          }
        }
      } else if (l === null && s === null && (c === null || c === 1 || c === -1))
        return !0;
    throw Error(V + "Invalid BigNumber: " + e);
  }, N.maximum = N.max = function() {
    return me(arguments, -1);
  }, N.minimum = N.min = function() {
    return me(arguments, 1);
  }, N.random = function() {
    var e = 9007199254740992, i = Math.random() * e & 2097151 ? function() {
      return Y(Math.random() * e);
    } : function() {
      return (Math.random() * 1073741824 | 0) * 8388608 + (Math.random() * 8388608 | 0);
    };
    return function(r) {
      var l, s, c, a, t, o = 0, u = [], f = new N(B);
      if (r == null ? r = g : U(r, 0, M), a = Se(r / T), ie)
        if (crypto.getRandomValues) {
          for (l = crypto.getRandomValues(new Uint32Array(a *= 2)); o < a; )
            t = l[o] * 131072 + (l[o + 1] >>> 11), t >= 9e15 ? (s = crypto.getRandomValues(new Uint32Array(2)), l[o] = s[0], l[o + 1] = s[1]) : (u.push(t % 1e14), o += 2);
          o = a / 2;
        } else if (crypto.randomBytes) {
          for (l = crypto.randomBytes(a *= 7); o < a; )
            t = (l[o] & 31) * 281474976710656 + l[o + 1] * 1099511627776 + l[o + 2] * 4294967296 + l[o + 3] * 16777216 + (l[o + 4] << 16) + (l[o + 5] << 8) + l[o + 6], t >= 9e15 ? crypto.randomBytes(7).copy(l, o) : (u.push(t % 1e14), o += 7);
          o = a / 7;
        } else
          throw ie = !1, Error(V + "crypto unavailable");
      if (!ie)
        for (; o < a; )
          t = i(), t < 9e15 && (u[o++] = t % 1e14);
      for (a = u[--o], r %= T, a && r && (t = Ie[T - r], u[o] = Y(a / t) * t); u[o] === 0; u.pop(), o--)
        ;
      if (o < 0)
        u = [c = 0];
      else {
        for (c = -1; u[0] === 0; u.splice(0, 1), c -= T)
          ;
        for (o = 1, t = u[0]; t >= 10; t /= 10, o++)
          ;
        o < T && (c -= T - o);
      }
      return f.e = c, f.c = u, f;
    };
  }(), N.sum = function() {
    for (var e = 1, i = arguments, r = new N(i[0]); e < i.length; )
      r = r.plus(i[e++]);
    return r;
  }, A = function() {
    var e = "0123456789";
    function i(r, l, s, c) {
      for (var a, t = [0], o, u = 0, f = r.length; u < f; ) {
        for (o = t.length; o--; t[o] *= l)
          ;
        for (t[0] += c.indexOf(r.charAt(u++)), a = 0; a < t.length; a++)
          t[a] > s - 1 && (t[a + 1] == null && (t[a + 1] = 0), t[a + 1] += t[a] / s | 0, t[a] %= s);
      }
      return t.reverse();
    }
    return function(r, l, s, c, a) {
      var t, o, u, f, E, O, R, C, P = r.indexOf("."), H = g, p = w;
      for (P >= 0 && (f = j, j = 0, r = r.replace(".", ""), C = new N(l), O = C.pow(r.length - P), j = f, C.c = i(
        J(k(O.c), O.e, "0"),
        10,
        s,
        e
      ), C.e = C.c.length), R = i(r, l, s, a ? (t = ne, e) : (t = e, ne)), u = f = R.length; R[--f] == 0; R.pop())
        ;
      if (!R[0])
        return t.charAt(0);
      if (P < 0 ? --u : (O.c = R, O.e = u, O.s = c, O = h(O, C, H, p, s), R = O.c, E = O.r, u = O.e), o = u + H + 1, P = R[o], f = s / 2, E = E || o < 0 || R[o + 1] != null, E = p < 4 ? (P != null || E) && (p == 0 || p == (O.s < 0 ? 3 : 2)) : P > f || P == f && (p == 4 || E || p == 6 && R[o - 1] & 1 || p == (O.s < 0 ? 8 : 7)), o < 1 || !R[0])
        r = E ? J(t.charAt(1), -H, t.charAt(0)) : t.charAt(0);
      else {
        if (R.length = o, E)
          for (--s; ++R[--o] > s; )
            R[o] = 0, o || (++u, R = [1].concat(R));
        for (f = R.length; !R[--f]; )
          ;
        for (P = 0, r = ""; P <= f; r += t.charAt(R[P++]))
          ;
        r = J(r, u, t.charAt(0));
      }
      return r;
    };
  }(), h = function() {
    function e(l, s, c) {
      var a, t, o, u, f = 0, E = l.length, O = s % Z, R = s / Z | 0;
      for (l = l.slice(); E--; )
        o = l[E] % Z, u = l[E] / Z | 0, a = R * o + u * O, t = O * o + a % Z * Z + f, f = (t / c | 0) + (a / Z | 0) + R * u, l[E] = t % c;
      return f && (l = [f].concat(l)), l;
    }
    function i(l, s, c, a) {
      var t, o;
      if (c != a)
        o = c > a ? 1 : -1;
      else
        for (t = o = 0; t < c; t++)
          if (l[t] != s[t]) {
            o = l[t] > s[t] ? 1 : -1;
            break;
          }
      return o;
    }
    function r(l, s, c, a) {
      for (var t = 0; c--; )
        l[c] -= t, t = l[c] < s[c] ? 1 : 0, l[c] = t * a + l[c] - s[c];
      for (; !l[0] && l.length > 1; l.splice(0, 1))
        ;
    }
    return function(l, s, c, a, t) {
      var o, u, f, E, O, R, C, P, H, p, m, d, fe, Ae, Oe, z, oe, y = l.s == s.s ? 1 : -1, v = l.c, L = s.c;
      if (!v || !v[0] || !L || !L[0])
        return new N(
          // Return NaN if either NaN, or both Infinity or 0.
          !l.s || !s.s || (v ? L && v[0] == L[0] : !L) ? NaN : (
            // Return ±0 if x is ±0 or y is ±Infinity, or return ±Infinity as y is ±0.
            v && v[0] == 0 || !L ? y * 0 : y / 0
          )
        );
      for (P = new N(y), H = P.c = [], u = l.e - s.e, y = c + u + 1, t || (t = W, u = $(l.e / T) - $(s.e / T), y = y / T | 0), f = 0; L[f] == (v[f] || 0); f++)
        ;
      if (L[f] > (v[f] || 0) && u--, y < 0)
        H.push(1), E = !0;
      else {
        for (Ae = v.length, z = L.length, f = 0, y += 2, O = Y(t / (L[0] + 1)), O > 1 && (L = e(L, O, t), v = e(v, O, t), z = L.length, Ae = v.length), fe = z, p = v.slice(0, z), m = p.length; m < z; p[m++] = 0)
          ;
        oe = L.slice(), oe = [0].concat(oe), Oe = L[0], L[1] >= t / 2 && Oe++;
        do {
          if (O = 0, o = i(L, p, z, m), o < 0) {
            if (d = p[0], z != m && (d = d * t + (p[1] || 0)), O = Y(d / Oe), O > 1)
              for (O >= t && (O = t - 1), R = e(L, O, t), C = R.length, m = p.length; i(R, p, C, m) == 1; )
                O--, r(R, z < C ? oe : L, C, t), C = R.length, o = 1;
            else
              O == 0 && (o = O = 1), R = L.slice(), C = R.length;
            if (C < m && (R = [0].concat(R)), r(p, R, m, t), m = p.length, o == -1)
              for (; i(L, p, z, m) < 1; )
                O++, r(p, z < m ? oe : L, m, t), m = p.length;
          } else
            o === 0 && (O++, p = [0]);
          H[f++] = O, p[0] ? p[m++] = v[fe] || 0 : (p = [v[fe]], m = 1);
        } while ((fe++ < Ae || p[0] != null) && y--);
        E = p[0] != null, H[0] || H.splice(0, 1);
      }
      if (t == W) {
        for (f = 1, y = H[0]; y >= 10; y /= 10, f++)
          ;
        F(P, c + (P.e = f + u * T - 1) + 1, a, E);
      } else
        P.e = u, P.r = +E;
      return P;
    };
  }();
  function _e(e, i, r, l) {
    var s, c, a, t, o;
    if (r == null ? r = w : U(r, 0, 8), !e.c)
      return e.toString();
    if (s = e.c[0], a = e.e, i == null)
      o = k(e.c), o = l == 1 || l == 2 && (a <= G || a >= x) ? ue(o, a) : J(o, a, "0");
    else if (e = F(new N(e), i, r), c = e.e, o = k(e.c), t = o.length, l == 1 || l == 2 && (i <= c || c <= G)) {
      for (; t < i; o += "0", t++)
        ;
      o = ue(o, c);
    } else if (i -= a, o = J(o, c, "0"), c + 1 > t) {
      if (--i > 0)
        for (o += "."; i--; o += "0")
          ;
    } else if (i += c - t, i > 0)
      for (c + 1 == t && (o += "."); i--; o += "0")
        ;
    return e.s < 0 && s ? "-" + o : o;
  }
  function me(e, i) {
    for (var r, l, s = 1, c = new N(e[0]); s < e.length; s++)
      l = new N(e[s]), (!l.s || (r = re(c, l)) === i || r === 0 && c.s === i) && (c = l);
    return c;
  }
  function Ne(e, i, r) {
    for (var l = 1, s = i.length; !i[--s]; i.pop())
      ;
    for (s = i[0]; s >= 10; s /= 10, l++)
      ;
    return (r = l + r * T - 1) > q ? e.c = e.e = null : r < ee ? e.c = [e.e = 0] : (e.e = r, e.c = i), e;
  }
  D = function() {
    var e = /^(-?)0([xbo])(?=\w[\w.]*$)/i, i = /^([^.]+)\.$/, r = /^\.([^.]+)$/, l = /^-?(Infinity|NaN)$/, s = /^\s*\+(?=[\w.])|^\s+|\s+$/g;
    return function(c, a, t, o) {
      var u, f = t ? a : a.replace(s, "");
      if (l.test(f))
        c.s = isNaN(f) ? null : f < 0 ? -1 : 1;
      else {
        if (!t && (f = f.replace(e, function(E, O, R) {
          return u = (R = R.toLowerCase()) == "x" ? 16 : R == "b" ? 2 : 8, !o || o == u ? O : E;
        }), o && (u = o, f = f.replace(i, "$1").replace(r, "0.$1")), a != f))
          return new N(f, u);
        if (N.DEBUG)
          throw Error(V + "Not a" + (o ? " base " + o : "") + " number: " + a);
        c.s = null;
      }
      c.c = c.e = null;
    };
  }();
  function F(e, i, r, l) {
    var s, c, a, t, o, u, f, E = e.c, O = Ie;
    if (E) {
      e: {
        for (s = 1, t = E[0]; t >= 10; t /= 10, s++)
          ;
        if (c = i - s, c < 0)
          c += T, a = i, o = E[u = 0], f = Y(o / O[s - a - 1] % 10);
        else if (u = Se((c + 1) / T), u >= E.length)
          if (l) {
            for (; E.length <= u; E.push(0))
              ;
            o = f = 0, s = 1, c %= T, a = c - T + 1;
          } else
            break e;
        else {
          for (o = t = E[u], s = 1; t >= 10; t /= 10, s++)
            ;
          c %= T, a = c - T + s, f = a < 0 ? 0 : Y(o / O[s - a - 1] % 10);
        }
        if (l = l || i < 0 || // Are there any non-zero digits after the rounding digit?
        // The expression  n % pows10[d - j - 1]  returns all digits of n to the right
        // of the digit at j, e.g. if n is 908714 and j is 2, the expression gives 714.
        E[u + 1] != null || (a < 0 ? o : o % O[s - a - 1]), l = r < 4 ? (f || l) && (r == 0 || r == (e.s < 0 ? 3 : 2)) : f > 5 || f == 5 && (r == 4 || l || r == 6 && // Check whether the digit to the left of the rounding digit is odd.
        (c > 0 ? a > 0 ? o / O[s - a] : 0 : E[u - 1]) % 10 & 1 || r == (e.s < 0 ? 8 : 7)), i < 1 || !E[0])
          return E.length = 0, l ? (i -= e.e + 1, E[0] = O[(T - i % T) % T], e.e = -i || 0) : E[0] = e.e = 0, e;
        if (c == 0 ? (E.length = u, t = 1, u--) : (E.length = u + 1, t = O[T - c], E[u] = a > 0 ? Y(o / O[s - a] % O[a]) * t : 0), l)
          for (; ; )
            if (u == 0) {
              for (c = 1, a = E[0]; a >= 10; a /= 10, c++)
                ;
              for (a = E[0] += t, t = 1; a >= 10; a /= 10, t++)
                ;
              c != t && (e.e++, E[0] == W && (E[0] = 1));
              break;
            } else {
              if (E[u] += t, E[u] != W)
                break;
              E[u--] = 0, t = 1;
            }
        for (c = E.length; E[--c] === 0; E.pop())
          ;
      }
      e.e > q ? e.c = e.e = null : e.e < ee && (e.c = [e.e = 0]);
    }
    return e;
  }
  function K(e) {
    var i, r = e.e;
    return r === null ? e.toString() : (i = k(e.c), i = r <= G || r >= x ? ue(i, r) : J(i, r, "0"), e.s < 0 ? "-" + i : i);
  }
  return _.absoluteValue = _.abs = function() {
    var e = new N(this);
    return e.s < 0 && (e.s = 1), e;
  }, _.comparedTo = function(e, i) {
    return re(this, new N(e, i));
  }, _.decimalPlaces = _.dp = function(e, i) {
    var r, l, s, c = this;
    if (e != null)
      return U(e, 0, M), i == null ? i = w : U(i, 0, 8), F(new N(c), e + c.e + 1, i);
    if (!(r = c.c))
      return null;
    if (l = ((s = r.length - 1) - $(this.e / T)) * T, s = r[s])
      for (; s % 10 == 0; s /= 10, l--)
        ;
    return l < 0 && (l = 0), l;
  }, _.dividedBy = _.div = function(e, i) {
    return h(this, new N(e, i), g, w);
  }, _.dividedToIntegerBy = _.idiv = function(e, i) {
    return h(this, new N(e, i), 0, 1);
  }, _.exponentiatedBy = _.pow = function(e, i) {
    var r, l, s, c, a, t, o, u, f, E = this;
    if (e = new N(e), e.c && !e.isInteger())
      throw Error(V + "Exponent not an integer: " + K(e));
    if (i != null && (i = new N(i)), t = e.e > 14, !E.c || !E.c[0] || E.c[0] == 1 && !E.e && E.c.length == 1 || !e.c || !e.c[0])
      return f = new N(Math.pow(+K(E), t ? e.s * (2 - le(e)) : +K(e))), i ? f.mod(i) : f;
    if (o = e.s < 0, i) {
      if (i.c ? !i.c[0] : !i.s)
        return new N(NaN);
      l = !o && E.isInteger() && i.isInteger(), l && (E = E.mod(i));
    } else {
      if (e.e > 9 && (E.e > 0 || E.e < -1 || (E.e == 0 ? E.c[0] > 1 || t && E.c[1] >= 24e7 : E.c[0] < 8e13 || t && E.c[0] <= 9999975e7)))
        return c = E.s < 0 && le(e) ? -0 : 0, E.e > -1 && (c = 1 / c), new N(o ? 1 / c : c);
      j && (c = Se(j / T + 2));
    }
    for (t ? (r = new N(0.5), o && (e.s = 1), u = le(e)) : (s = Math.abs(+K(e)), u = s % 2), f = new N(B); ; ) {
      if (u) {
        if (f = f.times(E), !f.c)
          break;
        c ? f.c.length > c && (f.c.length = c) : l && (f = f.mod(i));
      }
      if (s) {
        if (s = Y(s / 2), s === 0)
          break;
        u = s % 2;
      } else if (e = e.times(r), F(e, e.e + 1, 1), e.e > 14)
        u = le(e);
      else {
        if (s = +K(e), s === 0)
          break;
        u = s % 2;
      }
      E = E.times(E), c ? E.c && E.c.length > c && (E.c.length = c) : l && (E = E.mod(i));
    }
    return l ? f : (o && (f = B.div(f)), i ? f.mod(i) : c ? F(f, j, w, a) : f);
  }, _.integerValue = function(e) {
    var i = new N(this);
    return e == null ? e = w : U(e, 0, 8), F(i, i.e + 1, e);
  }, _.isEqualTo = _.eq = function(e, i) {
    return re(this, new N(e, i)) === 0;
  }, _.isFinite = function() {
    return !!this.c;
  }, _.isGreaterThan = _.gt = function(e, i) {
    return re(this, new N(e, i)) > 0;
  }, _.isGreaterThanOrEqualTo = _.gte = function(e, i) {
    return (i = re(this, new N(e, i))) === 1 || i === 0;
  }, _.isInteger = function() {
    return !!this.c && $(this.e / T) > this.c.length - 2;
  }, _.isLessThan = _.lt = function(e, i) {
    return re(this, new N(e, i)) < 0;
  }, _.isLessThanOrEqualTo = _.lte = function(e, i) {
    return (i = re(this, new N(e, i))) === -1 || i === 0;
  }, _.isNaN = function() {
    return !this.s;
  }, _.isNegative = function() {
    return this.s < 0;
  }, _.isPositive = function() {
    return this.s > 0;
  }, _.isZero = function() {
    return !!this.c && this.c[0] == 0;
  }, _.minus = function(e, i) {
    var r, l, s, c, a = this, t = a.s;
    if (e = new N(e, i), i = e.s, !t || !i)
      return new N(NaN);
    if (t != i)
      return e.s = -i, a.plus(e);
    var o = a.e / T, u = e.e / T, f = a.c, E = e.c;
    if (!o || !u) {
      if (!f || !E)
        return f ? (e.s = -i, e) : new N(E ? a : NaN);
      if (!f[0] || !E[0])
        return E[0] ? (e.s = -i, e) : new N(f[0] ? a : (
          // IEEE 754 (2008) 6.3: n - n = -0 when rounding to -Infinity
          w == 3 ? -0 : 0
        ));
    }
    if (o = $(o), u = $(u), f = f.slice(), t = o - u) {
      for ((c = t < 0) ? (t = -t, s = f) : (u = o, s = E), s.reverse(), i = t; i--; s.push(0))
        ;
      s.reverse();
    } else
      for (l = (c = (t = f.length) < (i = E.length)) ? t : i, t = i = 0; i < l; i++)
        if (f[i] != E[i]) {
          c = f[i] < E[i];
          break;
        }
    if (c && (s = f, f = E, E = s, e.s = -e.s), i = (l = E.length) - (r = f.length), i > 0)
      for (; i--; f[r++] = 0)
        ;
    for (i = W - 1; l > t; ) {
      if (f[--l] < E[l]) {
        for (r = l; r && !f[--r]; f[r] = i)
          ;
        --f[r], f[l] += W;
      }
      f[l] -= E[l];
    }
    for (; f[0] == 0; f.splice(0, 1), --u)
      ;
    return f[0] ? Ne(e, f, u) : (e.s = w == 3 ? -1 : 1, e.c = [e.e = 0], e);
  }, _.modulo = _.mod = function(e, i) {
    var r, l, s = this;
    return e = new N(e, i), !s.c || !e.s || e.c && !e.c[0] ? new N(NaN) : !e.c || s.c && !s.c[0] ? new N(s) : (se == 9 ? (l = e.s, e.s = 1, r = h(s, e, 0, 3), e.s = l, r.s *= l) : r = h(s, e, 0, se), e = s.minus(r.times(e)), !e.c[0] && se == 1 && (e.s = s.s), e);
  }, _.multipliedBy = _.times = function(e, i) {
    var r, l, s, c, a, t, o, u, f, E, O, R, C, P, H, p = this, m = p.c, d = (e = new N(e, i)).c;
    if (!m || !d || !m[0] || !d[0])
      return !p.s || !e.s || m && !m[0] && !d || d && !d[0] && !m ? e.c = e.e = e.s = null : (e.s *= p.s, !m || !d ? e.c = e.e = null : (e.c = [0], e.e = 0)), e;
    for (l = $(p.e / T) + $(e.e / T), e.s *= p.s, o = m.length, E = d.length, o < E && (C = m, m = d, d = C, s = o, o = E, E = s), s = o + E, C = []; s--; C.push(0))
      ;
    for (P = W, H = Z, s = E; --s >= 0; ) {
      for (r = 0, O = d[s] % H, R = d[s] / H | 0, a = o, c = s + a; c > s; )
        u = m[--a] % H, f = m[a] / H | 0, t = R * u + f * O, u = O * u + t % H * H + C[c] + r, r = (u / P | 0) + (t / H | 0) + R * f, C[c--] = u % P;
      C[c] = r;
    }
    return r ? ++l : C.splice(0, 1), Ne(e, C, l);
  }, _.negated = function() {
    var e = new N(this);
    return e.s = -e.s || null, e;
  }, _.plus = function(e, i) {
    var r, l = this, s = l.s;
    if (e = new N(e, i), i = e.s, !s || !i)
      return new N(NaN);
    if (s != i)
      return e.s = -i, l.minus(e);
    var c = l.e / T, a = e.e / T, t = l.c, o = e.c;
    if (!c || !a) {
      if (!t || !o)
        return new N(s / 0);
      if (!t[0] || !o[0])
        return o[0] ? e : new N(t[0] ? l : s * 0);
    }
    if (c = $(c), a = $(a), t = t.slice(), s = c - a) {
      for (s > 0 ? (a = c, r = o) : (s = -s, r = t), r.reverse(); s--; r.push(0))
        ;
      r.reverse();
    }
    for (s = t.length, i = o.length, s - i < 0 && (r = o, o = t, t = r, i = s), s = 0; i; )
      s = (t[--i] = t[i] + o[i] + s) / W | 0, t[i] = W === t[i] ? 0 : t[i] % W;
    return s && (t = [s].concat(t), ++a), Ne(e, t, a);
  }, _.precision = _.sd = function(e, i) {
    var r, l, s, c = this;
    if (e != null && e !== !!e)
      return U(e, 1, M), i == null ? i = w : U(i, 0, 8), F(new N(c), e, i);
    if (!(r = c.c))
      return null;
    if (s = r.length - 1, l = s * T + 1, s = r[s]) {
      for (; s % 10 == 0; s /= 10, l--)
        ;
      for (s = r[0]; s >= 10; s /= 10, l++)
        ;
    }
    return e && c.e + 1 > l && (l = c.e + 1), l;
  }, _.shiftedBy = function(e) {
    return U(e, -Re, Re), this.times("1e" + e);
  }, _.squareRoot = _.sqrt = function() {
    var e, i, r, l, s, c = this, a = c.c, t = c.s, o = c.e, u = g + 4, f = new N("0.5");
    if (t !== 1 || !a || !a[0])
      return new N(!t || t < 0 && (!a || a[0]) ? NaN : a ? c : 1 / 0);
    if (t = Math.sqrt(+K(c)), t == 0 || t == 1 / 0 ? (i = k(a), (i.length + o) % 2 == 0 && (i += "0"), t = Math.sqrt(+i), o = $((o + 1) / 2) - (o < 0 || o % 2), t == 1 / 0 ? i = "5e" + o : (i = t.toExponential(), i = i.slice(0, i.indexOf("e") + 1) + o), r = new N(i)) : r = new N(t + ""), r.c[0]) {
      for (o = r.e, t = o + u, t < 3 && (t = 0); ; )
        if (s = r, r = f.times(s.plus(h(c, s, u, 1))), k(s.c).slice(0, t) === (i = k(r.c)).slice(0, t))
          if (r.e < o && --t, i = i.slice(t - 3, t + 1), i == "9999" || !l && i == "4999") {
            if (!l && (F(s, s.e + g + 2, 0), s.times(s).eq(c))) {
              r = s;
              break;
            }
            u += 4, t += 4, l = 1;
          } else {
            (!+i || !+i.slice(1) && i.charAt(0) == "5") && (F(r, r.e + g + 2, 1), e = !r.times(r).eq(c));
            break;
          }
    }
    return F(r, r.e + g + 1, w, e);
  }, _.toExponential = function(e, i) {
    return e != null && (U(e, 0, M), e++), _e(this, e, i, 1);
  }, _.toFixed = function(e, i) {
    return e != null && (U(e, 0, M), e = e + this.e + 1), _e(this, e, i);
  }, _.toFormat = function(e, i, r) {
    var l, s = this;
    if (r == null)
      e != null && i && typeof i == "object" ? (r = i, i = null) : e && typeof e == "object" ? (r = e, e = i = null) : r = he;
    else if (typeof r != "object")
      throw Error(V + "Argument not an object: " + r);
    if (l = s.toFixed(e, i), s.c) {
      var c, a = l.split("."), t = +r.groupSize, o = +r.secondaryGroupSize, u = r.groupSeparator || "", f = a[0], E = a[1], O = s.s < 0, R = O ? f.slice(1) : f, C = R.length;
      if (o && (c = t, t = o, o = c, C -= c), t > 0 && C > 0) {
        for (c = C % t || t, f = R.substr(0, c); c < C; c += t)
          f += u + R.substr(c, t);
        o > 0 && (f += u + R.slice(c)), O && (f = "-" + f);
      }
      l = E ? f + (r.decimalSeparator || "") + ((o = +r.fractionGroupSize) ? E.replace(
        new RegExp("\\d{" + o + "}\\B", "g"),
        "$&" + (r.fractionGroupSeparator || "")
      ) : E) : f;
    }
    return (r.prefix || "") + l + (r.suffix || "");
  }, _.toFraction = function(e) {
    var i, r, l, s, c, a, t, o, u, f, E, O, R = this, C = R.c;
    if (e != null && (t = new N(e), !t.isInteger() && (t.c || t.s !== 1) || t.lt(B)))
      throw Error(V + "Argument " + (t.isInteger() ? "out of range: " : "not an integer: ") + K(t));
    if (!C)
      return new N(R);
    for (i = new N(B), u = r = new N(B), l = o = new N(B), O = k(C), c = i.e = O.length - R.e - 1, i.c[0] = Ie[(a = c % T) < 0 ? T + a : a], e = !e || t.comparedTo(i) > 0 ? c > 0 ? i : u : t, a = q, q = 1 / 0, t = new N(O), o.c[0] = 0; f = h(t, i, 0, 1), s = r.plus(f.times(l)), s.comparedTo(e) != 1; )
      r = l, l = s, u = o.plus(f.times(s = u)), o = s, i = t.minus(f.times(s = i)), t = s;
    return s = h(e.minus(r), l, 0, 1), o = o.plus(s.times(u)), r = r.plus(s.times(l)), o.s = u.s = R.s, c = c * 2, E = h(u, l, c, w).minus(R).abs().comparedTo(
      h(o, r, c, w).minus(R).abs()
    ) < 1 ? [u, l] : [o, r], q = a, E;
  }, _.toNumber = function() {
    return +K(this);
  }, _.toPrecision = function(e, i) {
    return e != null && U(e, 1, M), _e(this, e, i, 2);
  }, _.toString = function(e) {
    var i, r = this, l = r.s, s = r.e;
    return s === null ? l ? (i = "Infinity", l < 0 && (i = "-" + i)) : i = "NaN" : (e == null ? i = s <= G || s >= x ? ue(k(r.c), s) : J(k(r.c), s, "0") : e === 10 && Ee ? (r = F(new N(r), g + s + 1, w), i = J(k(r.c), r.e, "0")) : (U(e, 2, ne.length, "Base"), i = A(J(k(r.c), s, "0"), 10, e, l, !0)), l < 0 && r.c[0] && (i = "-" + i)), i;
  }, _.valueOf = _.toJSON = function() {
    return K(this);
  }, _._isBigNumber = !0, _[Symbol.toStringTag] = "BigNumber", _[Symbol.for("nodejs.util.inspect.custom")] = _.valueOf, n != null && N.set(n), N;
}
function $(n) {
  var h = n | 0;
  return n > 0 || n === h ? h : h - 1;
}
function k(n) {
  for (var h, A, D = 1, _ = n.length, B = n[0] + ""; D < _; ) {
    for (h = n[D++] + "", A = T - h.length; A--; h = "0" + h)
      ;
    B += h;
  }
  for (_ = B.length; B.charCodeAt(--_) === 48; )
    ;
  return B.slice(0, _ + 1 || 1);
}
function re(n, h) {
  var A, D, _ = n.c, B = h.c, g = n.s, w = h.s, G = n.e, x = h.e;
  if (!g || !w)
    return null;
  if (A = _ && !_[0], D = B && !B[0], A || D)
    return A ? D ? 0 : -w : g;
  if (g != w)
    return g;
  if (A = g < 0, D = G == x, !_ || !B)
    return D ? 0 : !_ ^ A ? 1 : -1;
  if (!D)
    return G > x ^ A ? 1 : -1;
  for (w = (G = _.length) < (x = B.length) ? G : x, g = 0; g < w; g++)
    if (_[g] != B[g])
      return _[g] > B[g] ^ A ? 1 : -1;
  return G == x ? 0 : G > x ^ A ? 1 : -1;
}
function U(n, h, A, D) {
  if (n < h || n > A || n !== Y(n))
    throw Error(V + (D || "Argument") + (typeof n == "number" ? n < h || n > A ? " out of range: " : " not an integer: " : " not a primitive number: ") + String(n));
}
function le(n) {
  var h = n.c.length - 1;
  return $(n.e / T) == h && n.c[h] % 2 != 0;
}
function ue(n, h) {
  return (n.length > 1 ? n.charAt(0) + "." + n.slice(1) : n) + (h < 0 ? "e" : "e+") + h;
}
function J(n, h, A) {
  var D, _;
  if (h < 0) {
    for (_ = A + "."; ++h; _ += A)
      ;
    n = _ + n;
  } else if (D = n.length, ++h > D) {
    for (_ = A, h -= D; --h; _ += A)
      ;
    n += _;
  } else
    h < D && (n = n.slice(0, h) + "." + n.slice(h));
  return n;
}
var Te = Be(), Le = Object.defineProperty, He = (n, h, A) => h in n ? Le(n, h, { enumerable: !0, configurable: !0, writable: !0, value: A }) : n[h] = A, b = (n, h, A) => (He(n, typeof h != "symbol" ? h + "" : h, A), A), te = /* @__PURE__ */ ((n) => (n.THOR = "THOR", n.RUNE = "RUNE", n.ATOM = "ATOM", n.MUON = "MUON", n.USDC = "USDC", n))(te || {}), I = /* @__PURE__ */ ((n) => (n.INVALID_INPUT_PARAMETERS = "1000", n.UNKNOWN_PROVIDERS = "1001", n.CANNOT_FIND_INBOUND_ADDRESS = "1002", n.NO_INBOUND_ADDRESSES = "1003", n.CHAIN_HALTED_OR_UNSUPPORTED = "1004", n.MISSING_INPUT_PARAMETER = "1005", n.INVALID_TYPE_GENERIC = "1100", n.INVALID_NUMBER_STRING = "1101", n.INVALID_NUMBER = "1102", n.INVALID_BOOLEAN = "1103", n.INVALID_OBJECT = "1104", n.INVALID_ARRAY = "1105", n.SELL_AMOUNT_MUST_BE_POSITIVE_INTEGER = "2000", n.SELL_BUY_ASSETS_ARE_THE_SAME = "2001", n.MISSING_SOURCE_ADDRESS_FOR_SYNTH = "2002", n.AFF_ADDRESS_AND_BPS_OR_NEITHER = "2003", n.AFF_ADDRESS_TOO_LONG = "2004", n.AFF_BPS_INTEGER_0_100 = "2005", n.SOURCE_ADDRESS_INVALID_FOR_SELL_CHAIN = "2006", n.DESTINATION_ADDRESS_INVALID_FOR_BUY_CHAIN = "2007", n.PREFERRED_PROFVIDER_NOT_SUPPORTED = "2008", n.DESTINATION_ADDRESS_SMART_CONTRACT = "2009", n.BUY_AMOUNT_MUST_BE_POSITIVE_INTEGER = "2010", n.SOURCE_ADDRESS_SMART_CONTRACT = "2011", n.INVALID_PROVIDER = "2100", n.MISSING_CROSS_CHAIN_PROVIDER = "2101", n.MISSING_AVAX_PROVIDER = "2102", n.MISSING_BSC_PROVIDER = "2103", n.MISSING_ETH_PROVIDER = "2104", n.INVALID_PROVIDER_FOR_SWAP_OUT = "2105", n.MISSING_ARB_PROVIDER = "2106", n.INVALID_CHAIN = "2200", n.INVALID_ASSET = "2201", n.INVALID_ASSET_IDENTIFIER = "2202", n.UNSUPPORTED_CHAIN = "2204", n.UNSUPPORTED_ASSET = "2203", n.UNSUPPORTED_ASSET_FOR_SWAPOUT = "2205", n.THORNODE_QUOTE_GENERIC_ERROR = "3000", n.NOT_ENOUGH_SYNTH_BALANCE = "3001", n.SYNTH_MINTING_CAP_REACHED = "3002", n.INVALID_QUOTE_MODE = "4000", n.NO_QUOTES = "4001", n.SERVICE_UNAVAILABLE_GENERIC = "5000", n.MISSING_GAS_DATA_GENERIC = "5100", n.MISSING_TOKEN_INFO_GENERIC = "5200", n.CANT_FIND_TOKEN_LIST = "5201", n.NO_PRICE = "5202", n.PRICE_IS_STALE = "5203", n.ADDRESS_NOT_WHITELISTED = "6000", n.ADDRESS_ALREADY_CLAIMED = "6001", n.TEMPORARY_ERROR = "9999", n))(I || {});
I.INVALID_INPUT_PARAMETERS + "", I.UNKNOWN_PROVIDERS + "", I.CANNOT_FIND_INBOUND_ADDRESS + "", I.NO_INBOUND_ADDRESSES + "", I.CHAIN_HALTED_OR_UNSUPPORTED + "", I.MISSING_INPUT_PARAMETER + "", I.INVALID_TYPE_GENERIC + "", I.INVALID_NUMBER_STRING + "", I.INVALID_NUMBER + "", I.INVALID_BOOLEAN + "", I.INVALID_OBJECT + "", I.INVALID_ARRAY + "", I.SELL_AMOUNT_MUST_BE_POSITIVE_INTEGER + "", I.SELL_BUY_ASSETS_ARE_THE_SAME + "", I.MISSING_SOURCE_ADDRESS_FOR_SYNTH + "", I.AFF_ADDRESS_AND_BPS_OR_NEITHER + "", I.AFF_ADDRESS_TOO_LONG + "", I.AFF_BPS_INTEGER_0_100 + "", I.SOURCE_ADDRESS_INVALID_FOR_SELL_CHAIN + "", I.DESTINATION_ADDRESS_INVALID_FOR_BUY_CHAIN + "", I.PREFERRED_PROFVIDER_NOT_SUPPORTED + "", I.DESTINATION_ADDRESS_SMART_CONTRACT + "", I.BUY_AMOUNT_MUST_BE_POSITIVE_INTEGER + "", I.INVALID_PROVIDER + "", I.MISSING_CROSS_CHAIN_PROVIDER + "", I.MISSING_AVAX_PROVIDER + "", I.MISSING_BSC_PROVIDER + "", I.MISSING_ETH_PROVIDER + "", I.MISSING_ARB_PROVIDER + "", I.INVALID_PROVIDER_FOR_SWAP_OUT + "", I.INVALID_CHAIN + "", I.INVALID_ASSET + "", I.UNSUPPORTED_CHAIN + "", I.UNSUPPORTED_ASSET + "", I.UNSUPPORTED_ASSET_FOR_SWAPOUT + "", I.THORNODE_QUOTE_GENERIC_ERROR + "", I.NOT_ENOUGH_SYNTH_BALANCE + "", I.SYNTH_MINTING_CAP_REACHED + "", I.INVALID_QUOTE_MODE + "", I.NO_QUOTES + "", I.SERVICE_UNAVAILABLE_GENERIC + "", I.MISSING_GAS_DATA_GENERIC + "", I.MISSING_TOKEN_INFO_GENERIC + "", I.CANT_FIND_TOKEN_LIST + "", I.NO_PRICE + "", I.PRICE_IS_STALE + "", I.ADDRESS_NOT_WHITELISTED + "", I.ADDRESS_ALREADY_CLAIMED + "";
var S = /* @__PURE__ */ ((n) => (n.Arbitrum = "ARB", n.Avalanche = "AVAX", n.Binance = "BNB", n.BinanceSmartChain = "BSC", n.Bitcoin = "BTC", n.BitcoinCash = "BCH", n.Cosmos = "GAIA", n.Dogecoin = "DOGE", n.Ethereum = "ETH", n.Litecoin = "LTC", n.Optimism = "OP", n.Polygon = "MATIC", n.THORChain = "THOR", n))(S || {}), ce = /* @__PURE__ */ ((n) => (n[n.ARB = 18] = "ARB", n[n.AVAX = 18] = "AVAX", n[n.BCH = 8] = "BCH", n[n.BNB = 8] = "BNB", n[n.BSC = 18] = "BSC", n[n.BTC = 8] = "BTC", n[n.DOGE = 8] = "DOGE", n[n.ETH = 18] = "ETH", n[n.GAIA = 6] = "GAIA", n[n.LTC = 8] = "LTC", n[n.MATIC = 18] = "MATIC", n[n.OP = 18] = "OP", n[n.THOR = 8] = "THOR", n))(ce || {}), Ce = /* @__PURE__ */ ((n) => (n.Arbitrum = "42161", n.ArbitrumHex = "0xa4b1", n.Avalanche = "43114", n.AvalancheHex = "0xa86a", n.Binance = "Binance-Chain-Tigris", n.BinanceHex = "", n.BinanceSmartChain = "56", n.BinanceSmartChainHex = "0x38", n.Bitcoin = "bitcoin", n.BitcoinHex = "", n.BitcoinCash = "bitcoincash", n.BitcoinCashHex = "", n.Cosmos = "cosmoshub-4", n.CosmosHex = "", n.Dogecoin = "dogecoin", n.DogecoinHex = "", n.Ethereum = "1", n.EthereumHex = "0x1", n.Litecoin = "litecoin", n.LitecoinHex = "", n.Optimism = "10", n.OptimismHex = "0xa", n.Polygon = "137", n.PolygonHex = "0x89", n.THORChain = "thorchain-mainnet-v1", n.THORChainHex = "", n.THORChainStagenet = "thorchain-stagenet-v2", n))(Ce || {}), Ue = /* @__PURE__ */ ((n) => (n.Arbitrum = "https://arb1.arbitrum.io/rpc", n.Avalanche = "https://node-router.thorswap.net/avalanche-c", n.Binance = "", n.BinanceSmartChain = "https://bsc-dataseed.binance.org", n.Bitcoin = "https://node-router.thorswap.net/bitcoin", n.BitcoinCash = "https://node-router.thorswap.net/bitcoin-cash", n.Cosmos = "https://node-router.thorswap.net/cosmos/rpc", n.Dogecoin = "https://node-router.thorswap.net/dogecoin", n.Ethereum = "https://node-router.thorswap.net/ethereum", n.Litecoin = "https://node-router.thorswap.net/litecoin", n.Optimism = "https://mainnet.optimism.io", n.Polygon = "https://polygon-rpc.com", n.THORChain = "https://rpc.thorswap.net", n.THORChainStagenet = "https://stagenet-rpc.ninerealms.com", n))(Ue || {});
const ae = Object.values(S), Me = Object.keys(S), ge = ae.reduce(
  (n, h) => {
    const A = Me.find((D) => S[D] === h);
    return A && (n[h] = A), n;
  },
  {}
);
ae.reduce(
  (n, h) => (n[h] = Ce[ge[h]], n),
  {}
);
ae.reduce(
  (n, h) => (n[h] = Ue[ge[h]], n),
  {}
);
ae.reduce(
  (n, h) => (n[h] = Ce[`${ge[h]}Hex`], n),
  {}
);
const de = (n, h, A = !1) => {
  if (A)
    return "Synth";
  switch (n) {
    case S.Bitcoin:
    case S.BitcoinCash:
    case S.Dogecoin:
    case S.Litecoin:
    case S.THORChain:
      return "Native";
    case S.Cosmos:
      return h === "ATOM" ? "Native" : "GAIA";
    case S.Binance:
      return h === S.Binance ? "Native" : "BEP2";
    case S.BinanceSmartChain:
      return h === S.Binance ? "Native" : "BEP20";
    case S.Ethereum:
      return h === S.Ethereum ? "Native" : "ERC20";
    case S.Avalanche:
      return h === S.Avalanche ? "Native" : "AVAX";
    case S.Polygon:
      return h === S.Polygon ? "Native" : "POLYGON";
    case S.Arbitrum:
      return [S.Ethereum, S.Arbitrum].includes(h) ? "Native" : "ARBITRUM";
    case S.Optimism:
      return [S.Ethereum, S.Optimism].includes(h) ? "Native" : "OPTIMISM";
    default:
      return n;
  }
};
Te.ROUND_DOWN, Te.ROUND_HALF_UP, Te.ROUND_UP;
const ve = (n, h) => n === S.Bitcoin ? "Bitcoin" : n === S.Dogecoin ? "Dogecoin" : n === S.Litecoin ? "Litecoin" : n === S.BitcoinCash ? "Bitcoin Cash" : n === S.Ethereum && h === "ETH" ? "Ethereum" : h, De = class pe {
  constructor(h, A, D = !1, _) {
    b(this, "chain"), b(this, "symbol"), b(this, "ticker"), b(this, "type"), b(this, "network"), b(this, "name"), b(this, "decimal"), b(this, "isSynth", !1), b(this, "L1Chain"), b(this, "setDecimal", (B) => {
      this.decimal = B || ce[this.chain] || ce.THOR;
    }), this.chain = h, this.symbol = A.toUpperCase(), this.ticker = _ || A.toUpperCase().split("-")[0], this.type = de(h, this.ticker, D), this.name = D ? `Synth ${this.ticker}` : this.ticker, this.network = ve(h, this.ticker), this.decimal = D ? ce.THOR : ce[h], this.isSynth = D, this.L1Chain = D ? S.THORChain : h;
  }
  static fromAssetString(h) {
    var A;
    if (!h)
      return null;
    const D = h.includes("/"), [_, ...B] = h.split(D ? "/" : "."), g = B.join("."), w = (A = g == null ? void 0 : g.split("-")) == null ? void 0 : A[0];
    return _ && g && w ? new pe(_, g, D) : null;
  }
  getAssetObj() {
    if (this.isSynth) {
      const h = `${this.chain.toLowerCase()}/${this.symbol.toLowerCase()}`;
      return {
        chain: S.THORChain,
        symbol: h,
        ticker: h
      };
    }
    return { chain: this.chain, symbol: this.symbol, ticker: this.ticker };
  }
  /**
   * convert asset entity to string
   * @returns L1 asset -> btc.btc, Synth asset -> btc/btc
   */
  toString() {
    return `${this.chain}${this.isSynth ? "/" : "."}${this.symbol}`;
  }
  toURLEncoded() {
    return `${this.isSynth ? "THOR." : ""}${this.chain}.${this.symbol}`;
  }
  // full compare chain, symbol, synth
  eq(h) {
    return this.chain === h.chain && this.symbol.toUpperCase() === h.symbol.toUpperCase() && this.ticker.toUpperCase() === h.ticker.toUpperCase() && this.isSynth === h.isSynth;
  }
  // compare chain, symbol but not synth
  shallowEq(h) {
    return this.chain === h.chain && this.symbol.toUpperCase() === h.symbol.toUpperCase() && this.ticker.toUpperCase() === h.ticker.toUpperCase();
  }
  isRUNE() {
    return this.eq(xe(S.THORChain));
  }
  identifier() {
    return `${this.chain}.${this.symbol}`;
  }
  toSynth() {
    return this.isSynth ? this : new pe(S.THORChain, `${this.chain}/${this.symbol}`, !0);
  }
};
b(De, "decodeFromURL", (n) => {
  var h, A, D;
  let _ = n.toUpperCase();
  return _.startsWith("THOR.") && ((h = _.split("THOR.")) == null ? void 0 : h[1]) !== "RUNE" && (_ = (D = (A = _.split("THOR.")) == null ? void 0 : A[1]) == null ? void 0 : D.replace(".", "/")), De.fromAssetString(_);
});
let Q = De;
const Ge = "THOR-0xa5f2211B9b8170F694421f2046281775E8468044", Ve = "VTHOR-0x815C23eCA83261b6Ec689b60Cc4a58b54BC24D8D", X = {}, xe = (n, h = !1) => {
  if (X[n])
    return X[n];
  switch (n) {
    case S.Avalanche:
    case S.Binance:
    case S.Bitcoin:
    case S.BitcoinCash:
    case S.Dogecoin:
    case S.Ethereum:
    case S.Litecoin: {
      const A = new Q(n, n, h);
      return X[n] = A, A;
    }
    case S.Optimism:
    case S.Arbitrum: {
      const A = new Q(n, S.Ethereum, h);
      return A.setDecimal(18), X[n] = A, A;
    }
    case S.BinanceSmartChain: {
      const A = new Q(n, S.Binance, h);
      return A.setDecimal(18), X[n] = A, A;
    }
    case S.Polygon: {
      const A = new Q(n, n, h);
      return A.setDecimal(18), X[n] = A, A;
    }
    case S.Cosmos: {
      const A = new Q(n, te.ATOM, h, te.ATOM);
      return X[n] = A, A;
    }
    case S.THORChain: {
      const A = new Q(S.THORChain, te.RUNE, h, te.RUNE);
      return X[n] = A, A;
    }
    case "USD": {
      const A = new Q(S.THORChain, "USD-USD", h, "USD-USD");
      return X[n] = A, A;
    }
    case "ETH_THOR": {
      const A = new Q(S.Ethereum, Ge, h);
      return A.setDecimal(18), X[n] = A, A;
    }
    case "ETH_VTHOR": {
      const A = new Q(S.Ethereum, Ve, h);
      return A.setDecimal(18), X[n] = A, A;
    }
    default:
      return new Q(S.THORChain, te.RUNE, h, te.RUNE);
  }
};
export {
  Q as AssetEntity,
  de as getAssetType,
  ve as getNetworkName,
  xe as getSignatureAssetFor
};
