// @trezor/transport@1.1.14/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeUsbTransport = exports.WebUsbTransport = exports.BridgeTransport = exports.Messages = exports.Transport = exports.TRANSPORT = exports.TREZOR_USB_DESCRIPTORS = exports.TRANSPORT_ERROR = void 0;
const protobuf = __importStar(require("protobufjs/light"));
const long_1 = __importDefault(require("long"));
exports.TRANSPORT_ERROR = __importStar(require("./errors"));
protobuf.util.Long = long_1.default;
protobuf.configure();
var constants_1 = require("./constants");
Object.defineProperty(exports, "TREZOR_USB_DESCRIPTORS", { enumerable: true, get: function () { return constants_1.TREZOR_USB_DESCRIPTORS; } });
Object.defineProperty(exports, "TRANSPORT", { enumerable: true, get: function () { return constants_1.TRANSPORT; } });
var abstract_1 = require("./transports/abstract");
Object.defineProperty(exports, "Transport", { enumerable: true, get: function () { return abstract_1.AbstractTransport; } });
exports.Messages = __importStar(require("./types/messages"));
var bridge_1 = require("./transports/bridge");
Object.defineProperty(exports, "BridgeTransport", { enumerable: true, get: function () { return bridge_1.BridgeTransport; } });
var webusb_1 = require("./transports/webusb");
Object.defineProperty(exports, "WebUsbTransport", { enumerable: true, get: function () { return webusb_1.WebUsbTransport; } });
var nodeusb_1 = require("./transports/nodeusb");
Object.defineProperty(exports, "NodeUsbTransport", { enumerable: true, get: function () { return nodeusb_1.NodeUsbTransport; } });
