// bip84@0.2.7/src/constants.js
module.exports = {
  NETWORKS: {
    bch: {
      mainnet: {
        messagePrefix: '\x19Bitcoin Cash Signed Message:\n',
        bech32: 'bch',
        bip44: 145,
        bip32: {
          private: 0x0488ade4,
          public: 0x0488b21e
        },
        pubKeyHash: 0x00,
        scriptHash: 0x05,
        wif: 0x80
      },
      testnet: {
        messagePrefix: '\x19Litecoin Signed Message:\n',
        bech32: 'bchtest',
        bip44: 1,
        bip32: {
          private: 0x04358394,
          public: 0x043587cf
        },
        pubKeyHash: 0x6f,
        scriptHash: 0xc4,
        wif: 0xef
      }
    },
    ltc: {
      mainnet: {
        messagePrefix: '\x19Litecoin Signed Message:\n',
        bech32: 'ltc',
        bip44: 2,
        bip32: {
          private: 0x01b26792, // 0x019d9cfe
          public: 0x01b26ef6 // 0x019da462
        },
        pubKeyHash: 0x30,
        scriptHash: 0x32,
        wif: 0xb0
      },
      testnet: {
        messagePrefix: '\x19Litecoin Signed Message:\n',
        bech32: 'tltc',
        bip44: 1,
        bip32: {
          private: 0x04358394, // 0x0436ef7d
          public: 0x043587cf // 0x0436f6e1
        },
        pubKeyHash: 0x6f,
        scriptHash: 0xc4,
        wif: 0xef
      }
    }
  }
}
