// @metamask/key-tree@7.1.1/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BIP_32_HARDENED_OFFSET = exports.BIP_39_PATH_REGEX = exports.SLIP_10_PATH_REGEX = exports.BIP_32_PATH_REGEX = exports.UNPREFIXED_BIP_32_PATH_REGEX = exports.UNPREFIXED_PATH_REGEX = exports.BIP44PurposeNodeToken = exports.MAX_BIP_32_INDEX = exports.MAX_UNHARDENED_BIP_32_INDEX = exports.MAX_BIP_44_DEPTH = exports.MIN_BIP_44_DEPTH = exports.BYTES_KEY_LENGTH = void 0;
exports.BYTES_KEY_LENGTH = 32;
exports.MIN_BIP_44_DEPTH = 0;
exports.MAX_BIP_44_DEPTH = 5;
exports.MAX_UNHARDENED_BIP_32_INDEX = 0x7fffffff; // 2^31 - 1
exports.MAX_BIP_32_INDEX = 0xffffffff; // 2^32 - 1
exports.BIP44PurposeNodeToken = `bip32:44'`;
exports.UNPREFIXED_PATH_REGEX = /^\d+$/u;
/**
 * e.g.
 * -  0
 * -  0'
 */
exports.UNPREFIXED_BIP_32_PATH_REGEX = /^(?<index>\d+)'?$/u;
/**
 * e.g.
 * -  bip32:0
 * -  bip32:0'
 */
exports.BIP_32_PATH_REGEX = /^bip32:\d+'?$/u;
/**
 * e.g.
 * -  slip10:0
 * -  slip10:0'
 */
exports.SLIP_10_PATH_REGEX = /^slip10:\d+'?$/u;
/**
 * bip39:<SPACE_DELMITED_SEED_PHRASE>
 *
 * The seed phrase must consist of 12 <= 24 words.
 */
exports.BIP_39_PATH_REGEX = /^bip39:([a-z]+){1}( [a-z]+){11,23}$/u;
exports.BIP_32_HARDENED_OFFSET = 0x80000000;
