// osmojs@0.37.0/main/codegen/osmosis/txfees/v1beta1/query.rpc.Query.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));

var _stargate = require("@cosmjs/stargate");

var _query = require("./query");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var QueryClientImpl = /*#__PURE__*/function () {
  function QueryClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, QueryClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.feeTokens = this.feeTokens.bind(this);
    this.denomSpotPrice = this.denomSpotPrice.bind(this);
    this.denomPoolId = this.denomPoolId.bind(this);
    this.baseDenom = this.baseDenom.bind(this);
  }

  (0, _createClass2["default"])(QueryClientImpl, [{
    key: "feeTokens",
    value: function feeTokens() {
      var request = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var data = _query.QueryFeeTokensRequest.encode(request).finish();

      var promise = this.rpc.request("osmosis.txfees.v1beta1.Query", "FeeTokens", data);
      return promise.then(function (data) {
        return _query.QueryFeeTokensResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "denomSpotPrice",
    value: function denomSpotPrice(request) {
      var data = _query.QueryDenomSpotPriceRequest.encode(request).finish();

      var promise = this.rpc.request("osmosis.txfees.v1beta1.Query", "DenomSpotPrice", data);
      return promise.then(function (data) {
        return _query.QueryDenomSpotPriceResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "denomPoolId",
    value: function denomPoolId(request) {
      var data = _query.QueryDenomPoolIdRequest.encode(request).finish();

      var promise = this.rpc.request("osmosis.txfees.v1beta1.Query", "DenomPoolId", data);
      return promise.then(function (data) {
        return _query.QueryDenomPoolIdResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "baseDenom",
    value: function baseDenom() {
      var request = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var data = _query.QueryBaseDenomRequest.encode(request).finish();

      var promise = this.rpc.request("osmosis.txfees.v1beta1.Query", "BaseDenom", data);
      return promise.then(function (data) {
        return _query.QueryBaseDenomResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return QueryClientImpl;
}();

exports.QueryClientImpl = QueryClientImpl;

var createRpcQueryExtension = function createRpcQueryExtension(base) {
  var rpc = (0, _stargate.createProtobufRpcClient)(base);
  var queryService = new QueryClientImpl(rpc);
  return {
    feeTokens: function feeTokens(request) {
      return queryService.feeTokens(request);
    },
    denomSpotPrice: function denomSpotPrice(request) {
      return queryService.denomSpotPrice(request);
    },
    denomPoolId: function denomPoolId(request) {
      return queryService.denomPoolId(request);
    },
    baseDenom: function baseDenom(request) {
      return queryService.baseDenom(request);
    }
  };
};

exports.createRpcQueryExtension = createRpcQueryExtension;
