// @bithighlander/hdwallet-native@1.17.4/dist/terra.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const core = __importStar(require("@bithighlander/hdwallet-core"));
const native = __importStar(require("./native"));
const MNEMONIC = "all all all all all all all all all all all all";
const mswMock = require("mswMock")().startServer();
afterEach(() => expect(mswMock).not.toHaveBeenCalled());
const untouchable = require("untouchableMock");
describe("NativeTerraWalletInfo", () => {
    const info = native.info();
    it("should return some static metadata", () => __awaiter(void 0, void 0, void 0, function* () {
        yield expect(untouchable.call(info, "terraSupportsNetwork")).resolves.toBe(true);
        yield expect(untouchable.call(info, "terraSupportsSecureTransfer")).resolves.toBe(false);
    }));
    it("should return the correct account paths", () => __awaiter(void 0, void 0, void 0, function* () {
        const paths = info.terraGetAccountPaths({ accountIdx: 0 });
        expect(paths).toMatchObject([{ addressNList: core.bip32ToAddressNList("m/44'/330'/0'/0/0") }]);
    }));
    it("does not support getting the next account path", () => __awaiter(void 0, void 0, void 0, function* () {
        expect(untouchable.call(info, "terraNextAccountPath", {})).toBe(undefined);
    }));
});
describe("NativeTerraWallet", () => {
    let wallet;
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        wallet = native.create({ deviceId: "native" });
        yield wallet.loadDevice({ mnemonic: MNEMONIC });
        yield expect(wallet.initialize()).resolves.toBe(true);
    }));
    it("should generate a correct terra address", () => __awaiter(void 0, void 0, void 0, function* () {
        yield expect(wallet.terraGetAddress({ addressNList: core.bip32ToAddressNList("m/44'/330'/0'/0/0") })).resolves.toBe("terra1f95csal3u6cyyj23ept3x7ap3u247npf8u2yhz");
    }));
    it("should generate another correct terra address", () => __awaiter(void 0, void 0, void 0, function* () {
        yield expect(wallet.terraGetAddress({ addressNList: core.bip32ToAddressNList("m/44'/330'/1337'/123/4") })).resolves.toBe("terra153l3gzmg5xlr8aldndpcg7achjejre04azdf9q");
    }));
    it("does not support signing transactions", () => __awaiter(void 0, void 0, void 0, function* () {
        const signed = yield wallet.terraSignTx({
            addressNList: core.bip32ToAddressNList("m/44'/330'/0'/0/0"),
            tx: {
                msg: [{ type: "foo", value: "bar" }],
                fee: {
                    amount: [{ denom: "foo", amount: "bar" }],
                    gas: "baz",
                },
                signatures: null,
                memo: "foobar",
            },
            chain_id: "foobar",
            account_number: "foo",
            sequence: "bar",
        });
        expect(signed.signatures.length).toBe(1);
        expect(signed.signatures[0].pub_key.value).toMatchInlineSnapshot(`"A6cAUgKWL3/P3nQY+j2fMUBaAW/QC/FGmQzTJ4nqXo0E"`);
        expect(signed.signatures[0].signature).toMatchInlineSnapshot(`"zUPR10sr2QwRa10fcb3z/KC6/mWuLq0iff5ImhylIJpqU1RSg49Jxbmvp07D3sWuY0fE5mcSdMWQXWJFw2zsWQ=="`);
    }));
});
