// @bithighlander/hdwallet-native@1.17.4/dist/terra.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MixinNativeTerraWallet = exports.MixinNativeTerraWalletInfo = void 0;
const core = __importStar(require("@bithighlander/hdwallet-core"));
const bech32 = __importStar(require("bech32"));
const crypto_js_1 = __importDefault(require("crypto-js"));
const txBuilder = __importStar(require("tendermint-tx-builder"));
const util_1 = __importDefault(require("./util"));
const Isolation = __importStar(require("./crypto/isolation"));
function MixinNativeTerraWalletInfo(Base) {
    return class MixinNativeTerraWalletInfo extends Base {
        constructor() {
            super(...arguments);
            this._supportsTerraInfo = true;
        }
        terraSupportsNetwork() {
            return __awaiter(this, void 0, void 0, function* () {
                return true;
            });
        }
        terraSupportsSecureTransfer() {
            return __awaiter(this, void 0, void 0, function* () {
                return false;
            });
        }
        terraGetAccountPaths(msg) {
            const slip44 = core.slip44ByCoin("Terra");
            return [
                {
                    addressNList: [0x80000000 + 44, 0x80000000 + slip44, 0x80000000 + msg.accountIdx, 0, 0],
                },
            ];
        }
        terraNextAccountPath(msg) {
            // Only support one account for now (like portis).
            return undefined;
        }
    };
}
exports.MixinNativeTerraWalletInfo = MixinNativeTerraWalletInfo;
function MixinNativeTerraWallet(Base) {
    var _MixinNativeTerraWallet_seed, _a;
    return _a = class MixinNativeTerraWallet extends Base {
            constructor() {
                super(...arguments);
                this._supportsTerra = true;
                _MixinNativeTerraWallet_seed.set(this, void 0);
            }
            terraInitializeWallet(seed) {
                return __awaiter(this, void 0, void 0, function* () {
                    __classPrivateFieldSet(this, _MixinNativeTerraWallet_seed, seed, "f");
                });
            }
            terraWipe() {
                __classPrivateFieldSet(this, _MixinNativeTerraWallet_seed, undefined, "f");
            }
            terraBech32ify(address, prefix) {
                const words = bech32.toWords(address);
                return bech32.encode(prefix, words);
            }
            createTerraAddress(publicKey) {
                const message = crypto_js_1.default.SHA256(crypto_js_1.default.enc.Hex.parse(publicKey));
                const hash = crypto_js_1.default.RIPEMD160(message).toString();
                const address = Buffer.from(hash, `hex`);
                return this.terraBech32ify(address, `terra`);
            }
            terraGetAddress(msg) {
                return __awaiter(this, void 0, void 0, function* () {
                    return this.needsMnemonic(!!__classPrivateFieldGet(this, _MixinNativeTerraWallet_seed, "f"), () => __awaiter(this, void 0, void 0, function* () {
                        return this.createTerraAddress(util_1.default.getKeyPair(__classPrivateFieldGet(this, _MixinNativeTerraWallet_seed, "f"), msg.addressNList, "terra").publicKey.toString("hex"));
                    }));
                });
            }
            terraSignTx(msg) {
                return __awaiter(this, void 0, void 0, function* () {
                    return this.needsMnemonic(!!__classPrivateFieldGet(this, _MixinNativeTerraWallet_seed, "f"), () => __awaiter(this, void 0, void 0, function* () {
                        const keyPair = util_1.default.getKeyPair(core.mustBeDefined(__classPrivateFieldGet(this, _MixinNativeTerraWallet_seed, "f")), msg.addressNList, "terra");
                        const adapter = new Isolation.Adapters.Cosmos(keyPair);
                        const result = yield txBuilder.sign(msg.tx, adapter, msg.sequence, msg.account_number, "terra");
                        return txBuilder.createSignedTx(msg.tx, result);
                    }));
                });
            }
        },
        _MixinNativeTerraWallet_seed = new WeakMap(),
        _a;
}
exports.MixinNativeTerraWallet = MixinNativeTerraWallet;
