// @trezor/connect@9.1.1/lib/api/blockchainDisconnect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const constants_1 = require("../constants");
const BlockchainLink_1 = require("../backend/BlockchainLink");
const coinInfo_1 = require("../data/coinInfo");
class BlockchainDisconnect extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = [];
        this.info = '';
        this.useDevice = false;
        this.useUi = false;
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [{ name: 'coin', type: 'string', required: true }]);
        const coinInfo = (0, coinInfo_1.getCoinInfo)(payload.coin);
        if (!coinInfo) {
            throw constants_1.ERRORS.TypedError('Method_UnknownCoin');
        }
        (0, BlockchainLink_1.isBackendSupported)(coinInfo);
        this.params = {
            coinInfo,
        };
    }
    run() {
        const backend = (0, BlockchainLink_1.findBackend)(this.params.coinInfo.shortcut);
        if (backend) {
            backend.disconnect();
        }
        return Promise.resolve({ disconnected: true });
    }
}
exports.default = BlockchainDisconnect;
