// @trezor/connect@9.1.1/lib/api/ethereumSignMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const pathUtils_1 = require("../utils/pathUtils");
const coinInfo_1 = require("../data/coinInfo");
const ethereumUtils_1 = require("../utils/ethereumUtils");
const formatUtils_1 = require("../utils/formatUtils");
const ethereumDefinitions_1 = require("./ethereum/ethereumDefinitions");
class EthereumSignMessage extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = ['read', 'write'];
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'path', required: true },
            { name: 'message', type: 'string', required: true },
            { name: 'hex', type: 'boolean' },
        ]);
        const path = (0, pathUtils_1.validatePath)(payload.path, 3);
        const network = (0, coinInfo_1.getEthereumNetwork)(path);
        this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, network, this.firmwareRange);
        this.info = (0, ethereumUtils_1.getNetworkLabel)('Sign #NETWORK message', network);
        const messageHex = payload.hex
            ? (0, formatUtils_1.messageToHex)(payload.message)
            : Buffer.from(payload.message, 'utf8').toString('hex');
        this.params = {
            address_n: path,
            message: messageHex,
        };
    }
    async run() {
        const cmd = this.device.getCommands();
        const { address_n, message } = this.params;
        const network = (0, coinInfo_1.getEthereumNetwork)(address_n);
        const slip44 = (0, pathUtils_1.getSlip44ByPath)(address_n);
        const definitions = await (0, ethereumDefinitions_1.getEthereumDefinitions)({
            chainId: network === null || network === void 0 ? void 0 : network.chainId,
            slip44,
        });
        const definitionParams = {
            ...(definitions.encoded_network && {
                encoded_network: definitions.encoded_network,
            }),
        };
        const response = await cmd.typedCall('EthereumSignMessage', 'EthereumMessageSignature', {
            ...definitionParams,
            address_n,
            message,
        });
        return response.message;
    }
}
exports.default = EthereumSignMessage;
