// @trezor/connect@9.1.1/lib/api/ethereumVerifyMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const formatUtils_1 = require("../utils/formatUtils");
class EthereumVerifyMessage extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = ['read', 'write'];
        this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, null, this.firmwareRange);
        this.info = 'Verify message';
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'address', type: 'string', required: true },
            { name: 'signature', type: 'string', required: true },
            { name: 'message', type: 'string', required: true },
            { name: 'hex', type: 'boolean' },
        ]);
        const messageHex = payload.hex
            ? (0, formatUtils_1.messageToHex)(payload.message)
            : Buffer.from(payload.message, 'utf8').toString('hex');
        this.params = {
            address: (0, formatUtils_1.stripHexPrefix)(payload.address),
            signature: (0, formatUtils_1.stripHexPrefix)(payload.signature),
            message: messageHex,
        };
    }
    async run() {
        const cmd = this.device.getCommands();
        const response = await cmd.typedCall('EthereumVerifyMessage', 'Success', this.params);
        return response.message;
    }
}
exports.default = EthereumVerifyMessage;
