// @trezor/connect@9.1.1/lib/api/ethereumSignTypedData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const pathUtils_1 = require("../utils/pathUtils");
const coinInfo_1 = require("../data/coinInfo");
const ethereumUtils_1 = require("../utils/ethereumUtils");
const constants_1 = require("../constants");
const ethereumSignTypedData_1 = require("./ethereum/ethereumSignTypedData");
const formatUtils_1 = require("../utils/formatUtils");
const ethereumDefinitions_1 = require("./ethereum/ethereumDefinitions");
class EthereumSignTypedData extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = ['read', 'write'];
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'path', required: true },
            { name: 'metamask_v4_compat', type: 'boolean', required: true },
            { name: 'data', type: 'object', required: true },
            { name: 'domain_separator_hash', type: 'string' },
            { name: 'message_hash', type: 'string' },
        ]);
        const path = (0, pathUtils_1.validatePath)(payload.path, 3);
        const network = (0, coinInfo_1.getEthereumNetwork)(path);
        this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, network, this.firmwareRange);
        this.info = (0, ethereumUtils_1.getNetworkLabel)('Sign #NETWORK typed data', network);
        this.params = {
            address_n: path,
            metamask_v4_compat: payload.metamask_v4_compat,
            data: payload.data,
        };
        if (payload.domain_separator_hash) {
            this.params = {
                ...this.params,
                domain_separator_hash: (0, formatUtils_1.messageToHex)(payload.domain_separator_hash),
            };
            if (payload.message_hash) {
                this.params = {
                    ...this.params,
                    message_hash: (0, formatUtils_1.messageToHex)(payload.message_hash),
                };
            }
            else if (this.params.data.primaryType !== 'EIP712Domain') {
                throw constants_1.ERRORS.TypedError('Method_InvalidParameter', 'message_hash should only be empty when data.primaryType=EIP712Domain');
            }
        }
        if (this.params.data.primaryType === 'EIP712Domain') {
            this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)('eip712-domain-only', network, this.firmwareRange);
            if ('message_hash' in this.params) {
                throw constants_1.ERRORS.TypedError('Method_InvalidParameter', 'message_hash should be empty when data.primaryType=EIP712Domain');
            }
        }
    }
    async run() {
        const cmd = this.device.getCommands();
        const { address_n } = this.params;
        const network = (0, coinInfo_1.getEthereumNetwork)(address_n);
        const slip44 = (0, pathUtils_1.getSlip44ByPath)(address_n);
        const definitions = await (0, ethereumDefinitions_1.getEthereumDefinitions)({
            chainId: network === null || network === void 0 ? void 0 : network.chainId,
            slip44,
        });
        if (this.device.features.model === '1') {
            (0, paramsValidator_1.validateParams)(this.params, [
                { name: 'domain_separator_hash', type: 'string', required: true },
                { name: 'message_hash', type: 'string' },
            ]);
            const { domain_separator_hash, message_hash } = this.params;
            const definitionParams = {
                ...(definitions.encoded_network && {
                    encoded_network: definitions.encoded_network,
                }),
            };
            const response = await cmd.typedCall('EthereumSignTypedHash', 'EthereumTypedDataSignature', {
                address_n,
                domain_separator_hash: domain_separator_hash,
                message_hash,
                ...definitionParams,
            });
            const { address, signature } = response.message;
            return {
                address,
                signature: `0x${signature}`,
            };
        }
        const { data, metamask_v4_compat } = this.params;
        const { types, primaryType, domain, message } = data;
        let response = await cmd.typedCall('EthereumSignTypedData', [
            'EthereumTypedDataStructRequest',
            'EthereumTypedDataValueRequest',
            'EthereumTypedDataSignature',
        ], {
            address_n,
            primary_type: primaryType,
            metamask_v4_compat,
            definitions,
        });
        while (response.type === 'EthereumTypedDataStructRequest') {
            const { name: typeDefinitionName } = response.message;
            const typeDefinition = types[typeDefinitionName];
            if (typeDefinition === undefined) {
                throw constants_1.ERRORS.TypedError('Runtime', `Type ${typeDefinitionName} was not defined in types object`);
            }
            const dataStruckAck = {
                members: typeDefinition.map(({ name, type: typeName }) => ({
                    name,
                    type: (0, ethereumSignTypedData_1.getFieldType)(typeName, types),
                })),
            };
            response = await cmd.typedCall('EthereumTypedDataStructAck', [
                'EthereumTypedDataStructRequest',
                'EthereumTypedDataValueRequest',
                'EthereumTypedDataSignature',
            ], dataStruckAck);
        }
        while (response.type === 'EthereumTypedDataValueRequest') {
            const { member_path } = response.message;
            let memberData;
            let memberTypeName;
            const [rootIndex, ...nestedMemberPath] = member_path;
            switch (rootIndex) {
                case 0:
                    memberData = domain;
                    memberTypeName = 'EIP712Domain';
                    break;
                case 1:
                    memberData = message;
                    memberTypeName = primaryType;
                    break;
                default:
                    throw constants_1.ERRORS.TypedError('Runtime', 'Root index can only be 0 or 1');
            }
            for (const index of nestedMemberPath) {
                if (Array.isArray(memberData)) {
                    memberTypeName = (0, ethereumSignTypedData_1.parseArrayType)(memberTypeName).entryTypeName;
                    memberData = memberData[index];
                }
                else if (typeof memberData === 'object' && memberData !== null) {
                    const memberTypeDefinition = types[memberTypeName][index];
                    memberTypeName = memberTypeDefinition.type;
                    memberData = memberData[memberTypeDefinition.name];
                }
                else {
                }
            }
            let encodedData;
            if (Array.isArray(memberData)) {
                encodedData = (0, ethereumSignTypedData_1.encodeData)('uint16', memberData.length);
            }
            else {
                encodedData = (0, ethereumSignTypedData_1.encodeData)(memberTypeName, memberData);
            }
            response = await cmd.typedCall('EthereumTypedDataValueAck', ['EthereumTypedDataValueRequest', 'EthereumTypedDataSignature'], {
                value: encodedData,
            });
        }
        const { address, signature } = response.message;
        return {
            address,
            signature: `0x${signature}`,
        };
    }
}
exports.default = EthereumSignTypedData;
