// @trezor/connect@9.1.1/lib/api/firmwareUpdate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const randombytes_1 = tslib_1.__importDefault(require("randombytes"));
const AbstractMethod_1 = require("../core/AbstractMethod");
const constants_1 = require("../constants");
const events_1 = require("../events");
const firmware_1 = require("./firmware");
const paramsValidator_1 = require("./common/paramsValidator");
const firmwareInfo_1 = require("../data/firmwareInfo");
class FirmwareUpdate extends AbstractMethod_1.AbstractMethod {
    init() {
        this.useEmptyPassphrase = true;
        this.requiredPermissions = ['management'];
        this.allowDeviceMode = [events_1.UI.BOOTLOADER, events_1.UI.INITIALIZE];
        this.requireDeviceMode = [events_1.UI.BOOTLOADER];
        this.useDeviceState = false;
        this.skipFirmwareCheck = true;
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'version', type: 'array' },
            { name: 'btcOnly', type: 'boolean' },
            { name: 'baseUrl', type: 'string' },
            { name: 'binary', type: 'array-buffer' },
            { name: 'intermediaryVersion', type: 'number' },
        ]);
        if ('version' in payload) {
            this.params = {
                version: payload.version,
                btcOnly: payload.btcOnly,
                baseUrl: payload.baseUrl || 'https://data.trezor.io',
                intermediaryVersion: payload.intermediaryVersion,
            };
        }
        if ('binary' in payload) {
            this.params = {
                ...this.params,
                binary: payload.binary,
            };
        }
    }
    async confirmation() {
        await this.getPopupPromise().promise;
        const uiPromise = this.createUiPromise(events_1.UI.RECEIVE_CONFIRMATION);
        this.postMessage((0, events_1.createUiMessage)(events_1.UI.REQUEST_CONFIRMATION, {
            view: 'device-management',
            customConfirmButton: {
                className: 'wipe',
                label: 'Proceed',
            },
            label: 'Do you want to update firmware? Never do this without your recovery card.',
        }));
        const uiResp = await uiPromise.promise;
        return uiResp.payload;
    }
    async run() {
        const { device, params } = this;
        let binary;
        try {
            if (params.binary) {
                binary = params.binary;
            }
            else {
                binary = await (0, firmware_1.getBinary)({
                    features: device.features,
                    releases: (0, firmwareInfo_1.getReleases)(device.features.major_version),
                    version: params.version,
                    btcOnly: params.btcOnly,
                    baseUrl: params.baseUrl,
                    intermediaryVersion: params.intermediaryVersion,
                });
            }
        }
        catch (err) {
            throw constants_1.ERRORS.TypedError('Method_FirmwareUpdate_DownloadFailed', `Failed to download firmware binary ${err.message}`);
        }
        const stripped = (0, firmware_1.stripFwHeaders)(binary);
        await (0, firmware_1.uploadFirmware)(this.device.getCommands().typedCall.bind(this.device.getCommands()), this.postMessage, device, { payload: (0, firmware_1.shouldStripFwHeaders)(device.features) ? stripped : binary });
        return (0, firmware_1.calculateFirmwareHash)(device.features.major_version, stripped, (0, randombytes_1.default)(32));
    }
}
exports.default = FirmwareUpdate;
