// @trezor/connect@9.1.1/lib/api/getAddress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const pathUtils_1 = require("../utils/pathUtils");
const coinInfo_1 = require("../data/coinInfo");
const constants_1 = require("../constants");
const events_1 = require("../events");
class GetAddress extends AbstractMethod_1.AbstractMethod {
    constructor() {
        super(...arguments);
        this.progress = 0;
    }
    init() {
        this.requiredPermissions = ['read'];
        this.hasBundle = !!this.payload.bundle;
        const payload = !this.payload.bundle
            ? { ...this.payload, bundle: [this.payload] }
            : this.payload;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'bundle', type: 'array' },
            { name: 'useEventListener', type: 'boolean' },
        ]);
        this.params = payload.bundle.map(batch => {
            (0, paramsValidator_1.validateParams)(batch, [
                { name: 'path', required: true },
                { name: 'coin', type: 'string' },
                { name: 'address', type: 'string' },
                { name: 'showOnTrezor', type: 'boolean' },
                { name: 'multisig', type: 'object' },
                { name: 'scriptType', type: 'string' },
                { name: 'unlockPath', type: 'object' },
            ]);
            if (batch.unlockPath) {
                (0, paramsValidator_1.validateParams)(batch.unlockPath, [
                    { name: 'address_n', required: true, type: 'array' },
                    { name: 'mac', required: true, type: 'string' },
                ]);
            }
            const path = (0, pathUtils_1.validatePath)(batch.path, 1);
            let coinInfo;
            if (batch.coin) {
                coinInfo = (0, coinInfo_1.getBitcoinNetwork)(batch.coin);
            }
            if (coinInfo && !batch.crossChain) {
                (0, paramsValidator_1.validateCoinPath)(path, coinInfo);
            }
            else if (!coinInfo) {
                coinInfo = (0, coinInfo_1.getBitcoinNetwork)(path);
            }
            if (!coinInfo) {
                throw constants_1.ERRORS.TypedError('Method_UnknownCoin');
            }
            else if (coinInfo) {
                this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, coinInfo, this.firmwareRange);
            }
            coinInfo = (0, coinInfo_1.fixCoinInfoNetwork)(coinInfo, path);
            return {
                address_n: path,
                address: batch.address,
                show_display: typeof batch.showOnTrezor === 'boolean' ? batch.showOnTrezor : true,
                multisig: batch.multisig,
                script_type: batch.scriptType,
                coinInfo,
                unlockPath: batch.unlockPath,
            };
        });
        const useEventListener = payload.useEventListener &&
            this.params.length === 1 &&
            typeof this.params[0].address === 'string' &&
            this.params[0].show_display;
        this.confirmed = useEventListener;
        this.useUi = !useEventListener;
        if (this.params.length === 1) {
            this.info = (0, pathUtils_1.getLabel)('Export #NETWORK address', this.params[0].coinInfo);
        }
        else {
            const requestedNetworks = this.params.map(b => b.coinInfo);
            const uniqNetworks = (0, coinInfo_1.getUniqueNetworks)(requestedNetworks);
            if (uniqNetworks.length === 1 && uniqNetworks[0]) {
                this.info = (0, pathUtils_1.getLabel)('Export multiple #NETWORK addresses', uniqNetworks[0]);
            }
            else {
                this.info = 'Export multiple addresses';
            }
        }
    }
    getButtonRequestData(code) {
        if (code === 'ButtonRequest_Address') {
            return {
                type: 'address',
                serializedPath: (0, pathUtils_1.getSerializedPath)(this.params[this.progress].address_n),
                address: this.params[this.progress].address || 'not-set',
            };
        }
    }
    async confirmation() {
        if (this.confirmed)
            return true;
        await this.getPopupPromise().promise;
        const uiPromise = this.createUiPromise(events_1.UI.RECEIVE_CONFIRMATION);
        this.postMessage((0, events_1.createUiMessage)(events_1.UI.REQUEST_CONFIRMATION, {
            view: 'export-address',
            label: this.info,
        }));
        const uiResp = await uiPromise.promise;
        this.confirmed = uiResp.payload;
        return this.confirmed;
    }
    async noBackupConfirmation() {
        await this.getPopupPromise().promise;
        const uiPromise = this.createUiPromise(events_1.UI.RECEIVE_CONFIRMATION);
        this.postMessage((0, events_1.createUiMessage)(events_1.UI.REQUEST_CONFIRMATION, {
            view: 'no-backup',
        }));
        const uiResp = await uiPromise.promise;
        return uiResp.payload;
    }
    async _call({ address_n, show_display, multisig, script_type, coinInfo, unlockPath }) {
        const cmd = this.device.getCommands();
        if (unlockPath) {
            await cmd.unlockPath(unlockPath);
        }
        return cmd.getAddress({
            address_n,
            show_display,
            multisig,
            script_type,
        }, coinInfo);
    }
    async run() {
        const responses = [];
        for (let i = 0; i < this.params.length; i++) {
            const batch = this.params[i];
            if (batch.show_display) {
                const silent = await this._call({
                    ...batch,
                    show_display: false,
                });
                if (typeof batch.address === 'string') {
                    if (batch.address !== silent.address) {
                        throw constants_1.ERRORS.TypedError('Method_AddressNotMatch');
                    }
                }
                else {
                    batch.address = silent.address;
                }
            }
            const response = await this._call(batch);
            responses.push(response);
            if (this.hasBundle) {
                this.postMessage((0, events_1.createUiMessage)(events_1.UI.BUNDLE_PROGRESS, {
                    progress: i,
                    response,
                }));
            }
            this.progress++;
        }
        return this.hasBundle ? responses : responses[0];
    }
}
exports.default = GetAddress;
