// @trezor/connect@9.1.1/lib/api/getFirmwareHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const events_1 = require("../events");
const paramsValidator_1 = require("./common/paramsValidator");
class GetFirmwareHash extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = ['management'];
        this.useEmptyPassphrase = true;
        this.useDeviceState = false;
        this.allowDeviceMode = [events_1.UI.INITIALIZE];
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [{ name: 'challenge', type: 'string' }]);
        this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, null, {
            1: { min: '1.11.1', max: '0' },
            2: { min: '2.5.1', max: '0' },
        });
        this.params = {
            challenge: payload.challenge,
        };
    }
    async run() {
        const cmd = this.device.getCommands();
        const response = await cmd.typedCall('GetFirmwareHash', 'FirmwareHash', this.params);
        return response.message;
    }
}
exports.default = GetFirmwareHash;
