// @trezor/connect@9.1.1/lib/api/getPublicKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const pathUtils_1 = require("../utils/pathUtils");
const events_1 = require("../events");
const coinInfo_1 = require("../data/coinInfo");
const accountUtils_1 = require("../utils/accountUtils");
class GetPublicKey extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = ['read'];
        this.info = 'Export public key';
        this.hasBundle = !!this.payload.bundle;
        const payload = !this.payload.bundle
            ? { ...this.payload, bundle: [this.payload] }
            : this.payload;
        (0, paramsValidator_1.validateParams)(payload, [{ name: 'bundle', type: 'array' }]);
        this.params = payload.bundle.map(batch => {
            (0, paramsValidator_1.validateParams)(batch, [
                { name: 'path', required: true },
                { name: 'coin', type: 'string' },
                { name: 'crossChain', type: 'boolean' },
                { name: 'showOnTrezor', type: 'boolean' },
                { name: 'scriptType', type: ['string', 'number'] },
                { name: 'ignoreXpubMagic', type: 'boolean' },
                { name: 'ecdsaCurveName', type: 'boolean' },
                { name: 'unlockPath', type: 'object' },
            ]);
            if (batch.unlockPath) {
                (0, paramsValidator_1.validateParams)(batch.unlockPath, [
                    { name: 'address_n', required: true, type: 'array' },
                    { name: 'mac', required: true, type: 'string' },
                ]);
            }
            let coinInfo;
            if (batch.coin) {
                coinInfo = (0, coinInfo_1.getBitcoinNetwork)(batch.coin);
            }
            const address_n = (0, pathUtils_1.validatePath)(batch.path, coinInfo ? 3 : 0);
            if (coinInfo && !batch.crossChain) {
                (0, paramsValidator_1.validateCoinPath)(address_n, coinInfo);
            }
            else if (!coinInfo) {
                coinInfo = (0, coinInfo_1.getBitcoinNetwork)(address_n);
            }
            if (coinInfo) {
                this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, coinInfo, this.firmwareRange);
            }
            return {
                address_n,
                coin_name: coinInfo === null || coinInfo === void 0 ? void 0 : coinInfo.name,
                show_display: batch.showOnTrezor,
                script_type: batch.scriptType,
                ignore_xpub_magic: batch.ignoreXpubMagic,
                ecdsa_curve_name: batch.ecdsaCurveName,
                coinInfo,
                unlockPath: batch.unlockPath,
            };
        });
    }
    async confirmation() {
        if (this.confirmed)
            return true;
        await this.getPopupPromise().promise;
        const uiPromise = this.createUiPromise(events_1.UI.RECEIVE_CONFIRMATION);
        let label;
        if (this.params.length > 1) {
            label = 'Export multiple public keys';
        }
        else {
            label = (0, accountUtils_1.getPublicKeyLabel)(this.params[0].address_n, this.params[0].coinInfo);
        }
        this.postMessage((0, events_1.createUiMessage)(events_1.UI.REQUEST_CONFIRMATION, {
            view: 'export-xpub',
            label,
        }));
        const uiResp = await uiPromise.promise;
        this.confirmed = uiResp.payload;
        return this.confirmed;
    }
    async noBackupConfirmation() {
        await this.getPopupPromise().promise;
        const uiPromise = this.createUiPromise(events_1.UI.RECEIVE_CONFIRMATION);
        this.postMessage((0, events_1.createUiMessage)(events_1.UI.REQUEST_CONFIRMATION, {
            view: 'get-public-key-no-backup',
        }));
        const uiResp = await uiPromise.promise;
        return uiResp.payload;
    }
    async run() {
        const responses = [];
        const cmd = this.device.getCommands();
        for (let i = 0; i < this.params.length; i++) {
            const { coinInfo, unlockPath, ...batch } = this.params[i];
            const response = await cmd.getHDNode(batch, { coinInfo, unlockPath });
            responses.push(response);
            if (this.hasBundle) {
                this.postMessage((0, events_1.createUiMessage)(events_1.UI.BUNDLE_PROGRESS, {
                    progress: i,
                    response,
                }));
            }
        }
        return this.hasBundle ? responses : responses[0];
    }
}
exports.default = GetPublicKey;
