// @trezor/connect@9.1.1/lib/api/pushTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const coinInfo_1 = require("../data/coinInfo");
const constants_1 = require("../constants");
const BlockchainLink_1 = require("../backend/BlockchainLink");
class PushTransaction extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = [];
        this.useUi = false;
        this.useDevice = false;
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'tx', type: 'string', required: true },
            { name: 'coin', type: 'string', required: true },
        ]);
        const coinInfo = (0, coinInfo_1.getCoinInfo)(payload.coin);
        if (!coinInfo) {
            throw constants_1.ERRORS.TypedError('Method_UnknownCoin');
        }
        (0, BlockchainLink_1.isBackendSupported)(coinInfo);
        if (coinInfo.type === 'bitcoin' && !/^[0-9A-Fa-f]*$/.test(payload.tx)) {
            throw constants_1.ERRORS.TypedError('Method_InvalidParameter', 'Transaction must be hexadecimal');
        }
        this.params = {
            tx: payload.tx,
            coinInfo,
        };
    }
    async run() {
        const backend = await (0, BlockchainLink_1.initBlockchain)(this.params.coinInfo, this.postMessage);
        const txid = await backend.pushTransaction(this.params.tx);
        return {
            txid,
        };
    }
}
exports.default = PushTransaction;
