// @trezor/connect@9.1.1/lib/api/tezosSignTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const AbstractMethod_1 = require("../core/AbstractMethod");
const paramsValidator_1 = require("./common/paramsValidator");
const coinInfo_1 = require("../data/coinInfo");
const pathUtils_1 = require("../utils/pathUtils");
const helper = tslib_1.__importStar(require("./tezos/tezosSignTx"));
class TezosSignTransaction extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = ['read', 'write'];
        this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, (0, coinInfo_1.getMiscNetwork)('Tezos'), this.firmwareRange);
        this.info = 'Sign Tezos transaction';
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'path', required: true },
            { name: 'branch', type: 'string', required: true },
            { name: 'operation', required: true },
        ]);
        const path = (0, pathUtils_1.validatePath)(payload.path, 3);
        this.params = helper.createTx(path, payload.branch, payload.operation);
    }
    async run() {
        const cmd = this.device.getCommands();
        const response = await cmd.typedCall('TezosSignTx', 'TezosSignedTx', this.params);
        return response.message;
    }
}
exports.default = TezosSignTransaction;
