// @trezor/connect@9.1.1/lib/api/unlockPath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractMethod_1 = require("../core/AbstractMethod");
const pathUtils_1 = require("../utils/pathUtils");
const paramsValidator_1 = require("./common/paramsValidator");
class UnlockPath extends AbstractMethod_1.AbstractMethod {
    init() {
        this.requiredPermissions = ['read'];
        this.skipFinalReload = true;
        this.firmwareRange = (0, paramsValidator_1.getFirmwareRange)(this.name, undefined, this.firmwareRange);
        const { payload } = this;
        (0, paramsValidator_1.validateParams)(payload, [
            { name: 'path', required: true },
            { name: 'mac', type: 'string' },
        ]);
        const path = (0, pathUtils_1.validatePath)(payload.path, 1);
        this.params = {
            address_n: path,
            mac: payload.mac,
        };
    }
    async run() {
        const cmd = this.device.getCommands();
        const { message } = await cmd.unlockPath(this.params);
        return {
            address_n: this.params.address_n,
            mac: message.mac,
        };
    }
}
exports.default = UnlockPath;
