// @trezor/utxo-lib@1.0.9/lib/script/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPS = exports.signature = exports.number = exports.isCanonicalScriptSignature = exports.isDefinedHashType = exports.isCanonicalPubKey = exports.toStack = exports.fromASM = exports.toASM = exports.decompile = exports.compile = exports.isPushOnly = void 0;
const bip66 = require("bip66");
const pushdata = require("pushdata-bitcoin");
const typeforce = require("typeforce");
const ecc = require("tiny-secp256k1");
const scriptNumber = require("./scriptNumber");
const scriptSignature = require("./scriptSignature");
const ops_1 = require("./ops");
const types = require("../types");
const OP_INT_BASE = ops_1.OPS.OP_RESERVED;
function isOPInt(value) {
    return (types.Number(value) &&
        (value === ops_1.OPS.OP_0 ||
            (value >= ops_1.OPS.OP_1 && value <= ops_1.OPS.OP_16) ||
            value === ops_1.OPS.OP_1NEGATE));
}
function isPushOnlyChunk(value) {
    return types.Buffer(value) || isOPInt(value);
}
function isPushOnly(value) {
    return types.Array(value) && value.every(isPushOnlyChunk);
}
exports.isPushOnly = isPushOnly;
function asMinimalOP(buffer) {
    if (buffer.length === 0)
        return ops_1.OPS.OP_0;
    if (buffer.length !== 1)
        return;
    if (buffer[0] >= 1 && buffer[0] <= 16)
        return OP_INT_BASE + buffer[0];
    if (buffer[0] === 0x81)
        return ops_1.OPS.OP_1NEGATE;
}
function compile(chunks) {
    if (types.Buffer(chunks))
        return chunks;
    typeforce(types.Array, chunks);
    const bufferSize = chunks.reduce((accum, chunk) => {
        if (types.Buffer(chunk)) {
            if (chunk.length === 1 && asMinimalOP(chunk) !== undefined) {
                return accum + 1;
            }
            return accum + pushdata.encodingLength(chunk.length) + chunk.length;
        }
        return accum + 1;
    }, 0.0);
    const buffer = Buffer.allocUnsafe(bufferSize);
    let offset = 0;
    chunks.forEach(chunk => {
        if (types.Buffer(chunk)) {
            const opcode = asMinimalOP(chunk);
            if (opcode !== undefined) {
                buffer.writeUInt8(opcode, offset);
                offset += 1;
                return;
            }
            offset += pushdata.encode(buffer, chunk.length, offset);
            chunk.copy(buffer, offset);
            offset += chunk.length;
        }
        else {
            buffer.writeUInt8(chunk, offset);
            offset += 1;
        }
    });
    if (offset !== buffer.length)
        throw new Error('Could not decode chunks');
    return buffer;
}
exports.compile = compile;
function decompile(buffer) {
    if (types.Array(buffer))
        return buffer;
    typeforce(types.Buffer, buffer);
    const chunks = [];
    let i = 0;
    while (i < buffer.length) {
        const opcode = buffer[i];
        if (opcode > ops_1.OPS.OP_0 && opcode <= ops_1.OPS.OP_PUSHDATA4) {
            const d = pushdata.decode(buffer, i);
            if (d === null)
                return [];
            i += d.size;
            if (i + d.number > buffer.length)
                return [];
            const data = buffer.slice(i, i + d.number);
            i += d.number;
            const op = asMinimalOP(data);
            if (op !== undefined) {
                chunks.push(op);
            }
            else {
                chunks.push(data);
            }
        }
        else {
            chunks.push(opcode);
            i += 1;
        }
    }
    return chunks;
}
exports.decompile = decompile;
function toASM(chunks) {
    if (types.Buffer(chunks)) {
        chunks = decompile(chunks);
    }
    return chunks
        .map(chunk => {
        if (types.Buffer(chunk)) {
            const op = asMinimalOP(chunk);
            if (op === undefined)
                return chunk.toString('hex');
            chunk = op;
        }
        return ops_1.REVERSE_OPS[chunk];
    })
        .join(' ');
}
exports.toASM = toASM;
function fromASM(asm) {
    typeforce(types.String, asm);
    return compile(asm.split(' ').map(chunkStr => {
        if (ops_1.OPS[chunkStr] !== undefined)
            return ops_1.OPS[chunkStr];
        typeforce(types.Hex, chunkStr);
        return Buffer.from(chunkStr, 'hex');
    }));
}
exports.fromASM = fromASM;
function toStack(chunks0) {
    const chunks = decompile(chunks0);
    typeforce(isPushOnly, chunks);
    return chunks === null || chunks === void 0 ? void 0 : chunks.map(op => {
        if (types.Buffer(op))
            return op;
        if (op === ops_1.OPS.OP_0)
            return Buffer.allocUnsafe(0);
        return scriptNumber.encode(op - OP_INT_BASE);
    });
}
exports.toStack = toStack;
function isCanonicalPubKey(buffer) {
    return ecc.isPoint(buffer);
}
exports.isCanonicalPubKey = isCanonicalPubKey;
function isDefinedHashType(hashType) {
    const hashTypeMod = hashType & ~0x80;
    return hashTypeMod > 0x00 && hashTypeMod < 0x04;
}
exports.isDefinedHashType = isDefinedHashType;
function isCanonicalScriptSignature(buffer) {
    if (!types.Buffer(buffer))
        return false;
    if (!isDefinedHashType(buffer[buffer.length - 1]))
        return false;
    return bip66.check(buffer.slice(0, -1));
}
exports.isCanonicalScriptSignature = isCanonicalScriptSignature;
exports.number = scriptNumber;
exports.signature = scriptSignature;
var ops_2 = require("./ops");
Object.defineProperty(exports, "OPS", { enumerable: true, get: function () { return ops_2.OPS; } });
